{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "035a8d8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import openai\n",
    "import config\n",
    "import os \n",
    "import json\n",
    "from langchain.llms import OpenAI\n",
    "from langchain.prompts import PromptTemplate\n",
    "from langchain.document_loaders import JSONLoader, DirectoryLoader, UnstructuredHTMLLoader\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "from langchain.schema import HumanMessage, SystemMessage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "d3f63367",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(page_content='\\n\\n\\n\\n\\nSupported Browsers - KBase Documentation\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\nKBase DocumentationSearch⌃KKBase DocumentationKBase Terms & ConditionsGetting StartedSigning Up and Signing InSupported BrowsersNarrative Quick StartNarrative Interface User GuideFAQsWorking with DataUsing AppsRunning Common WorkflowsTroubleshootingDeveloping AppsExternal LinksKBase Narrative InterfaceKBase web siteKBase App Catalogkbase.usPowered By GitBookSupported BrowsersThe KBase website and Narrative Interface should work on most modern web browsers, but we primarily focus on supporting Chrome.Supported BrowsersThe following desktop web browsers have been tested against the KBase web site and Narrative user interface. If you experience any issue using these browsers with our site please report it through the Help Board. OSSupported BrowsersMacChromeFirefoxSafari 10 (macOS Sierra) and upWindowsFirefoxMicrosoft Edge - not recommendedLinuxChromeFirefoxAd blockers may prevent KBase from working in your web browserSome ad-blockers and browser extensions may cause errors within the Narrative Interface. We are working to identify issues with commonly used browser extensions, but you may want to disable browser extensions if you encounter issues within KBase.\\n\\nMobileKBase is currently not supported on mobile browsers. The KBase website and services should work with tablet browsers on iPad and Android devices in landscape orientation, but we do not currently test on these platforms so there may be compatibility issues. Watch for future updates as KBase goes responsive and mobile.Installation of Supported BrowsersChromeChrome is the best browser to use for KBase. The Chrome browser is produced by Google, Inc., and is freely available. Chrome operates on a rolling release, meaning that it is updated frequently and it is not easy to track specific versions. Our best advice is to make sure you are using the most recent version available. On Mac and Windows this is built into the Chrome application itself.\\u200bDownload Chrome\\u200bFirefoxFirefox, produced by the non-profit Mozilla Corporation, is a freely available, open source browser for Mac, Windows, Linux, and other systems. Like Chrome, Firefox is updated frequently, so it is not feasible for us to recommend a specific (or even minimum) version. Our best advice is to make sure you are using the most recent version available. Recent versions of Firefox will automatically update themselves on Mac and Windows.\\u200bDownload Firefox\\u200bSafariSafari is distributed with Mac OS X. (We do not support the Windows version of Safari.) Specific versions of Mac OS X support specific versions of Safari – see this Wikipedia entry for details.Microsoft Edge\\u200bMicrosoft Edge is the new web browser for Windows 10; it replaces Internet Explorer. You should be able to use KBase on Edge, but this has not been extensively tested.Unsupported BrowsersKBase is not supported on Internet Explorer. It should work on Microsoft Edge, but it has not been extensively tested. For the best chance of success on Edge, be sure you are using the most recent version.Older versions of supported browsers are also not supported. Please ensure that your browser and system are up-to-date before reporting a problem with a supported browser.Other modern browsers, such as Opera, may work with KBase, but as we do not test with them we cannot make any assurances about their operation.PreviousAuthentication UpdateNextNarrative Quick StartLast modified 1yr agoCopy linkOn this pageSupported BrowsersMobileInstallation of Supported BrowsersChromeFirefoxSafariMicrosoft EdgeUnsupported Browsers\\n\\n\\n\\n\\n', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_browsers.html', 'title': 'Supported Browsers - KBase Documentation'})]"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain.document_loaders import BSHTMLLoader\n",
    "loader = BSHTMLLoader(\"/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_browsers.html\")\n",
    "data = loader.load()\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "20227c3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Supported Browsers - KBase Documentation\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "KBase DocumentationSearch⌃KKBase DocumentationKBase Terms & ConditionsGetting StartedSigning Up and Signing InSupported BrowsersNarrative Quick StartNarrative Interface User GuideFAQsWorking with DataUsing AppsRunning Common WorkflowsTroubleshootingDeveloping AppsExternal LinksKBase Narrative InterfaceKBase web siteKBase App Catalogkbase.usPowered By GitBookSupported BrowsersThe KBase website and Narrative Interface should work on most modern web browsers, but we primarily focus on supporting Chrome.Supported BrowsersThe following desktop web browsers have been tested against the KBase web site and Narrative user interface. If you experience any issue using these browsers with our site please report it through the Help Board. OSSupported BrowsersMacChromeFirefoxSafari 10 (macOS Sierra) and upWindowsFirefoxMicrosoft Edge - not recommendedLinuxChromeFirefoxAd blockers may prevent KBase from working in your web browserSome ad-blockers and browser extensions may cause errors within the Narrative Interface. We are working to identify issues with commonly used browser extensions, but you may want to disable browser extensions if you encounter issues within KBase.\n",
      "\n",
      "MobileKBase is currently not supported on mobile browsers. The KBase website and services should work with tablet browsers on iPad and Android devices in landscape orientation, but we do not currently test on these platforms so there may be compatibility issues. Watch for future updates as KBase goes responsive and mobile.Installation of Supported BrowsersChromeChrome is the best browser to use for KBase. The Chrome browser is produced by Google, Inc., and is freely available. Chrome operates on a rolling release, meaning that it is updated frequently and it is not easy to track specific versions. Our best advice is to make sure you are using the most recent version available. On Mac and Windows this is built into the Chrome application itself.​Download Chrome​FirefoxFirefox, produced by the non-profit Mozilla Corporation, is a freely available, open source browser for Mac, Windows, Linux, and other systems. Like Chrome, Firefox is updated frequently, so it is not feasible for us to recommend a specific (or even minimum) version. Our best advice is to make sure you are using the most recent version available. Recent versions of Firefox will automatically update themselves on Mac and Windows.​Download Firefox​SafariSafari is distributed with Mac OS X. (We do not support the Windows version of Safari.) Specific versions of Mac OS X support specific versions of Safari – see this Wikipedia entry for details.Microsoft Edge​Microsoft Edge is the new web browser for Windows 10; it replaces Internet Explorer. You should be able to use KBase on Edge, but this has not been extensively tested.Unsupported BrowsersKBase is not supported on Internet Explorer. It should work on Microsoft Edge, but it has not been extensively tested. For the best chance of success on Edge, be sure you are using the most recent version.Older versions of supported browsers are also not supported. Please ensure that your browser and system are up-to-date before reporting a problem with a supported browser.Other modern browsers, such as Opera, may work with KBase, but as we do not test with them we cannot make any assurances about their operation.PreviousAuthentication UpdateNextNarrative Quick StartLast modified 1yr agoCopy linkOn this pageSupported BrowsersMobileInstallation of Supported BrowsersChromeFirefoxSafariMicrosoft EdgeUnsupported Browsers\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(data[0].page_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "71c4a9c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = \"sk-jaadX..................Y5cAR58M\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cebf6779",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"OPENAI_API_KEY\"] = key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "4a1c4a90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "92"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "html_dir_loader = DirectoryLoader('/scratch/ac.pgupta/convLLM/kbasedocs', glob=\"**/[!.]*.html\", loader_cls=UnstructuredHTMLLoader)\n",
    "data_html = html_dir_loader.load()\n",
    "len(data_html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "25d94804",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(page_content='KBase Documentation\\n\\nHere you will find guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.\\n\\nHow to cite KBase:\\n\\nArkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163\\n\\nFunding:\\n\\nThis work is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase (KBase) funded by the\\n\\nU.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\\n\\nunder Award Numbers DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.\\n\\nNextKBase Terms & Conditions\\n\\nLast modified \\n\\n2yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/index.html'}),\n",
       " Document(page_content='Getting Started\\n\\nAn introduction to KBase and guide on how to get started.\\n\\nWhat you can find in this section:\\n\\n1.\\u200bSigning Up and Signing In\\u200b\\n\\n2.\\u200bKBase Supported Browsers\\u200b\\n\\n3.\\u200bA Quick Start Guide to KBase\\u200b\\n\\n4.\\u200bKBase Narrative Interface User Guide\\u200b\\n\\n5.\\u200bFAQs\\u200b\\n\\nPreviousKBase Terms & Conditions\\n\\nNextSigning Up and Signing In\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started.html'}),\n",
       " Document(page_content=\"Signing Up and Signing In\\n\\nSign up and get started using KBase\\n\\nSigning Up\\n\\nTo create a KBase account, you will need to use one of our supported identity providers. Sign up for a free KBase account using your existing Google,\\n\\nORCiD \\u200b\\n\\n, or\\n\\nGlobus\\n\\nidentity. For more detailed guidance, see the\\n\\nStep-by-Step Signup Guide\\n\\nAfter you link your Google, ORCiD, or Globus account, you will be able to sign in and use the KBase Narrative Interface to perform your computational systems biology analyses.\\n\\nKBase is an ORCiD\\n\\nMember Organization\\n\\n. Sign in or link your\\n\\n\\u200bto KBase\\n\\nTerms and Conditions\\n\\nKBase is sponsored by the U.S. Department of Energy and is free for all to use (even if you are not in the United States). When you link an identity to KBase, you will need to accept our\\n\\nTerms & Conditions\\n\\n, including our\\n\\nUser Agreement\\n\\n. You may also want to review our\\n\\nPrivacy Policy\\n\\nand\\n\\nData Policy\\n\\n. Users shall follow the\\n\\nKBase Code of Conduct\\n\\nSign up for a free KBase account\\n\\nSigning In\\n\\nTo sign in, click the Sign In button on the KBase website or KBase Narrative Interface. Select the identity provider you chose when you signed up for your KBase account, and use the credentials you would normally use when signing into the that provider's web site. You can\\n\\nlink multiple accounts to sign into KBase\\n\\nHow do I set or reset my password?\\n\\nIf you signed up for KBase and your password does not seem to work, you may reset your password at any time with the identity provider who manages it. KBase itself does not manage passwords of its own, so all password management must be done through your identity provider.\\n\\nIf the last time you logged into your KBase account was prior to June 2017, see the\\n\\nauthentication update page\\n\\nPreviousGetting Started\\n\\nNextStep-by-Step Signup Guide\\n\\nLast modified \\n\\n7mo ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up.html'}),\n",
       " Document(page_content=\"Step-by-Step Signup Guide\\n\\nThis page provides step-by-step instructions for signing up to use KBase. You will use your existing account at Google or\\n\\nGlobus\\n\\nor\\n\\nORCiD\\n\\n, and it will be linked to your new KBase account.\\n\\n1. From the\\n\\nKBase Narrative login\\n\\n, click the “Sign Up” button.\\n\\n2. Use your existing Google,\\n\\nORCiD\\n\\n, or\\n\\nGlobus\\n\\naccount to sign up for KBase.\\n\\nWhat if I don't have a Google, ORCiD, or Globus account?\\n\\nIt’s easy and free to get a new account with one of our identity providers.\\n\\nGet a Google account\\n\\nGet an ORCiD account\\n\\nGet a Globus ID\\n\\nWe are also considering adding support for other identity providers. If there’s one you’d like us to add, please\\n\\ncontact us\\n\\n3. If want to use your Google account, click on the “Sign in with Google” button; you will be prompted to choose a Google account (if you have more than one) and sign in to it (if you aren’t already signed in).\\n\\n4. The KBase account creation form will prompt you to enter some required information, which includes choosing a username and identifying your organization and department.\\n\\n5. You must now agree to the\\n\\nKBase User Agreement\\n\\nand\\n\\nKBase Data Policy\\n\\nNote: If you already had a KBase account, these documents have not changed since you last agreed to them. This is a reconfirmation required by the new authentication process.\\n\\nRead the policies and select the checkboxes that say “I have read and agree to this policy”\\n\\nOnce you’ve checked both boxes for User Agreement and Data Policy, you will be able to click the blue button at the bottom that says “Create KBase Account”.\\n\\n6. You’re done – you now have a KBase account!\\n\\nClick “Continue” and you’ll see your KBase landing page. You may want to take a look at the\\n\\nNarrative Quick Start\\n\\nor\\n\\nother documentation\\n\\nbefore you plunge headfirst into KBase.\\n\\nIf you encounter any difficulties or have any questions, please feel free to\\n\\ncontact us\\n\\nPreviousSigning Up and Signing In\\n\\nNextLinking KBase to ORCiD\\n\\nLast modified \\n\\n7mo ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up_step-by-step.html'}),\n",
       " Document(page_content='Linking KBase to ORCiD\\n\\nIf you have an\\n\\nORCiD\\n\\naccount, you can sign into KBase using your\\n\\nORCiD \\u200b\\n\\nand link it to your KBase account.\\n\\nKBase is an\\n\\nORCiD Member Organization.\\n\\nIf you are already a KBase User, first sign into KBase through your Globus or Google accounts. If you are not already signed in to KBase, go to the\\n\\nsign-in page\\n\\n, click the Sign In button, and click either “Sign in with Google”  or “Sign in with Globus” buttons and go through the normal sign-in process until you get to Narratives.\\n\\nOn the left sidebar menu, click the “Account” icon, and then choose the tab labeled “Linked Sign-In Accounts”.\\n\\nAt the bottom of the tab, choose \"\\n\\n\\u200b\\u200b\\n\\nORCiD” from the pulldown list of identity providers, and then click the “Link” button.\\n\\nThe next screen will prompt you to Sign in using your ORCiD account to link.\\n\\nAfter you click that Link button, you will be returned to the Linked Sign-in Accounts screen in your Account Manager, where you will see your newly linked ORCiD account.\\n\\nYou can now sign-in to KBase using your ORCiD ID and\\n\\nintegrate DOIs\\n\\nWe hope you appreciate the convenience of being able to sign in to KBase with your ORCiD credentials. If you encounter any difficulties or have any questions, please feel free to\\n\\ncontact us\\n\\nPreviousStep-by-Step Signup Guide\\n\\nNextLinking Accounts\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up_linking-orcid.html'}),\n",
       " Document(page_content='Linking Accounts\\n\\nLinking Accounts can be helpful for data transfers, ease of signing in, and sharing data. You can link your KBase Account to Google, Globus, and ORCiD.\\n\\nTo do this, first sign into KBase and go through the normal sign-in process.\\n\\nOn the left sidebar menu, click the “Account” icon, and then choose the tab labeled “Linked Sign-In Accounts”.\\n\\nAt the bottom of the tab, choose an account type from the pulldown list of identity providers, and then click the “Link” button.\\n\\nThe next screen will prompt you to Sign in using your to the chosen account type to link.\\n\\nAfter you click that Link button, you will be returned to the Linked Sign-in Accounts screen, where you will see your newly linked account.\\n\\nExample\\n\\nIf you have a Google (gmail) account, you can link it to your KBase account.\\n\\nTo do this, you must first sign in to KBase with your current KBase account login. If you are not already signed in to KBase, go to the\\n\\nsign-in page\\n\\n, click the Sign In button. Go through the normal sign-in process.\\n\\nOn the left sidebar of your menu, click the “Account” icon, and then choose the tab labeled “Linked Sign-In Accounts”.\\n\\nAt the bottom of the tab, choose “Google” from the pulldown list of identity providers.\\n\\nThen click the “Link” button. The next screen will prompt you to choose a Google account to link (some people have more than one).\\n\\nYou are now ready to link your Google account to your KBase account.\\n\\nAfter you click that Link button, you will be returned to the Linked Sign-in Accounts screen, where you will see your newly linked Google account.\\n\\nThe next time you sign in with your Google credentials\\n\\nNow when you want to sign in to KBase, you can also use the “Sign in with Google” button.\\n\\nThis will bring you straight to the “Choose an account” Google page:\\n\\nAfter you choose your linked Google account, you’ll be signed in to KBase! (Note that if you had signed out of that Google account, you will be directed to sign into it again before you login to KBase.)\\n\\nWe hope you appreciate the convenience of being able to sign-in to KBase with your Google credentials. If you encounter any difficulties or have any questions, please feel free to\\n\\ncontact us\\n\\nPreviousLinking KBase to ORCiD\\n\\nNextAuthentication Update\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up_linking-accounts.html'}),\n",
       " Document(page_content='Authentication Update\\n\\nKBase has simplified the sign up and sign in process for new users. You will be able to sign into KBase with a Google, Globus, or ORCiD once accounts have been linked.\\n\\nNew users: creating a new KBase account\\n\\nYou can ignore the information on this page. For details about setting up a new account, see our\\n\\nstep-by-step signup guide\\n\\nTransitioning an existing KBase account\\n\\nIf you already had a KBase account prior to June 9, 2017, then the first time you want to sign into KBase after the authentication change, you will need to go through a few steps to convert your existing KBase (through Globus) account to the new system. After going through the conversion, you’ll be able to sign in to KBase with your existing account or your Google account, if you choose to link your KBase account. To see what you’ll need to do, continue reading below.\\n\\nWhen you go to\\n\\nKBase\\n\\nto sign in, you will see this:\\n\\nClick the Sign In button below “Use KBase” and sign-in options will appear\\n\\nSign in using Globus. Globus is the home of the existing KBase user accounts, and your Globus ID is your KBase ID.\\n\\nWhen you click “\\n\\nSign in with Globus\\n\\n”, a page to Log in with your Globus credentials will open:\\n\\nClick Continue. You will reach another page at Globus, requesting your Globus username and password (remember, this is the same as your existing KBase account).\\n\\nEnter your usual KBase username and password. Your username is not your email address – it should not contain an “@” sign. If you forgot your password, use the “Forgot password?” link at the lower right.\\n\\nAfter entering your username and password, click the “Log In” button. You will be taken to a page where KBase requests access to your email address and Globus identities. Click the blue “Allow” button. We won’t share your email address or personal information outside of KBase.\\n\\nYou must now agree to the KBase User Agreement and KBase Data Policy. (The User Agreement and Data Policy have not changed since you last agreed to them. This is merely a reconfirmation required by the new authentication process.)\\n\\nRead the\\n\\nKBase User Agreement\\n\\nand\\n\\nKBase Data Policy\\n\\nand check the checkboxes that say “I have read and agree to this policy”\\n\\nOnce you’ve checked both of the checkboxes, you will be able to click the blue button at the bottom that says “Continue to the KBase account [username]”.\\n\\nOnce you click that Continue button, you’re done – you should see the KBase User Interface landing page.\\n\\nNote – the system changes mean that your profile may look a bit bare-bones. Please fill in the information in your profile when you get a chance, and choose or upload a new avatar if you like.\\n\\nThe next time you sign in with your Globus account\\n\\nThe first time you sign into KBase after the authentication change, you will need to go through all the steps described above. Subsequently, if you choose to sign-in using your Google credentials, the process will look like this:\\n\\n1. Go to\\n\\nnarrative.kbase.us\\n\\n2. Click the Sign In button below “Use KBase” and more options will appear:\\n\\n3. Click “Sign in with Globus”.\\n\\nMost of the time, you will not need to sign into Globus (as it will remember you unless you sign out of KBase), so you will see a screen like this:\\n\\n4. Click the blue button at the bottom that says “Sign in to KBase account [username]” and you will be at your KBase landing page!\\n\\nNote that if you had signed out completely from Globus after the last time you used KBase, you might have to go through two extra steps to sign-in to Globus. Click \"Continue\" and then proceed to enter Globus Log in information.\\n\\nIf you encounter any difficulties or have any questions, please feel free to\\n\\ncontact us\\n\\nPreviousLinking Accounts\\n\\nNextSupported Browsers\\n\\nLast modified \\n\\n7mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up_auth-update.html'}),\n",
       " Document(page_content='Supported Browsers\\n\\nThe KBase website and Narrative Interface should work on most modern web browsers, but we primarily focus on supporting Chrome.\\n\\nSupported Browsers\\n\\nThe following desktop web browsers have been tested against the KBase web site and Narrative user interface. If you experience any issue using these browsers with our site please report it through the\\n\\nHelp Board\\n\\nOS Supported Browsers Mac Chrome Firefox Safari 10 (macOS Sierra) and up Windows Firefox Microsoft Edge - not recommended Linux Chrome Firefox\\n\\nAd blockers may prevent KBase from working in your web browser\\n\\nSome ad-blockers and browser extensions may cause errors within the Narrative Interface. We are working to identify issues with commonly used browser extensions, but you may want to disable browser extensions if you encounter issues within KBase.\\n\\nMobile\\n\\nKBase is currently not supported on mobile browsers. The KBase website and services should work with tablet browsers on iPad and Android devices in landscape orientation, but we do not currently test on these platforms so there may be compatibility issues. Watch for future updates as KBase goes responsive and mobile.\\n\\nInstallation of Supported Browsers\\n\\nChrome\\n\\nChrome is the best browser to use for KBase. The Chrome browser is produced by Google, Inc., and is freely available. Chrome operates on a rolling release, meaning that it is updated frequently and it is not easy to track specific versions. Our best advice is to make sure you are using the most recent version available. On Mac and Windows this is built into the Chrome application itself.\\n\\n\\u200bDownload Chrome\\u200b\\n\\nFirefox\\n\\nFirefox, produced by the non-profit Mozilla Corporation, is a freely available, open source browser for Mac, Windows, Linux, and other systems. Like Chrome, Firefox is updated frequently, so it is not feasible for us to recommend a specific (or even minimum) version. Our best advice is to make sure you are using the most recent version available. Recent versions of Firefox will automatically update themselves on Mac and Windows.\\n\\n\\u200bDownload Firefox\\u200b\\n\\nSafari\\n\\nSafari is distributed with Mac OS X. (We do not support the Windows version of Safari.) Specific versions of Mac OS X support specific versions of Safari – see this\\n\\nWikipedia entry\\n\\nfor details.\\n\\nMicrosoft Edge\\n\\nMicrosoft Edge\\n\\nis the new web browser for Windows 10; it replaces Internet Explorer. You should be able to use KBase on Edge, but this has not been extensively tested.\\n\\nUnsupported Browsers\\n\\nKBase is not supported on Internet Explorer. It should work on Microsoft Edge, but it has not been extensively tested. For the best chance of success on Edge, be sure you are using the most recent version.\\n\\nOlder versions of supported browsers are also not supported. Please ensure that your browser and system are up-to-date before reporting a problem with a supported browser.\\n\\nOther modern browsers, such as Opera, may work with KBase, but as we do not test with them we cannot make any assurances about their operation.\\n\\nPreviousAuthentication Update\\n\\nNextNarrative Quick Start\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_browsers.html'}),\n",
       " Document(page_content=\"Narrative Quick Start\\n\\nA fast, one-page introduction to KBase's narrative interface\\n\\nNarrative Quick Start Guide\\n\\nWhat is a Narrative?\\n\\nIn KBase, you can create shareable, reproducible workflows called Narratives that include data, analysis steps, results, visualizations, and commentary. This Quick Start provides an overview of how to create and use the features of Narratives.\\n\\nStep 1. Sign up for a KBase user account\\n\\nTo begin using KBase, you will first need to\\n\\nSign up for a KBase Account\\n\\n. You can now use your existing Google or Globus username and password to get a free KBase account.\\n\\nStep 2. Sign in to KBase\\n\\nAfter you sign in, you will be taken to your Narratives (details of which can be found\\n\\nhere\\n\\n). From Narratives, you can open your existing Narratives, access others that have been shared with you, and create new Narratives. These Narratives can be accessed at\\n\\nnarrative.kbase.us\\n\\nIf you are a new user, your Narratives dash will be mostly empty, since you haven’t created any Narratives yet.\\n\\nStep 3. Create a new Narrative\\n\\nClick the “+ New Narrative” button in Narratives to open a new untitled Narrative. A “Welcome to the Narrative Interface” box offers information and links to documentation for adding content to your Narrative. You can collapse or delete this box using the “…” dropdown menu in the top right corner.\\n\\nTry the Narrative tour!\\n\\nSelect the “Narrative Tour” from the Help menu. This new feature walks you through the user interface, pointing out various useful aspects of it. We recommend taking the tour even if you’ve used KBase before, as it will help familiarize you with the current version.\\n\\nStep 4. Find data to analyze\\n\\nTo load data to your Narrative, click the “Add Data” button in the Data Panel, which can be found under the Analyze tab near the top left of the Narrative. The\\n\\nData Browser\\n\\nwill slide out with tabs that show several data sources.\\n\\nYou can choose a dataset already in KBase by selecting any of the first four tabs. The Example tab, for instance, contains sample data from KBase’s reference data collection that can be used to try out the apps.\\n\\nThe Import tab allows you to upload your own data. KBase currently supports upload of a variety of data types. Please see the\\n\\nData Upload/Download Guide\\n\\nfor more information.\\n\\nNote: Any data you upload to KBase is private unless you choose to share it.\\n\\nStep 5. Add data to your Narrative\\n\\nWhen you hover over a\\n\\ndata object\\n\\nin the Data Browser, a blue “< Add” button will appear to its left. Click the button to add the data to your Narrative.\\n\\nNotice that the data object now appears in the Data Panel. Once you finish adding data, exit the Data Browser by clicking either the “Close” button at the bottom left of the browser window or the arrow at the top right of the Data Panel.\\n\\nDon’t forget to periodically save your Narrative by clicking the “Save” button at the top right of the interface.\\n\\nStep 6. Choose an App\\n\\nOnce you have added data to your Narrative, you can analyze it using one or more of the KBase apps listed in the App Panel below your data.\\n\\nThere are options for filtering by category, name, input type, and more. You can designate apps you like as favorites using the star icon.\\n\\nTo add an app to your Narrative from the App Panel, click on its name or icon. You can get more information about an app by hovering over it until “…” appears and clicking on it.\\n\\nStep 7. Run the App\\n\\nFill in all the required fields in your App. Note that some app fields are “smart” and can recognize which data in your Narrative are valid input for those fields. These smart fields have a pulldown list of all the data objects from your Data Panel that you can choose from as input.\\n\\nWhen all required fields have been filled out, click the green “Run” button near the top left of the app cell to start the analysis.\\n\\nWhile the analysis is running, you will see the progress in the Job Status tab of the app box. When it finishes, the results will show up in the Result tab.\\n\\nPlease see\\n\\nthe KBase Apps section of the Narrative Interface User Guide\\n\\nfor more details on running apps.\\n\\nStep 8. Share your Narrative\\n\\nAll Narratives that you create are private by default. You can\\n\\nmake a Narrative public or share it\\n\\nwith specific collaborators by clicking the “share” button near the top right of your Narrative.\\n\\nNext steps and additional resources\\n\\nWith these basic steps, you now can begin using the Narrative Interface to create Narratives and analyze data. For a more in-depth introduction to any of the topics mentioned here, see the\\n\\nNarrative User Guide\\n\\nPreviousSupported Browsers\\n\\nNextNarrative Interface User Guide\\n\\nLast modified \\n\\n1yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_quick-start.html'}),\n",
       " Document(page_content=\"Narrative Interface User Guide\\n\\nA more leisurely introduction to KBase's Narrative interface\\n\\nWhat is in here:\\n\\n1.\\u200bAbout\\u200b\\n\\n2.\\u200bAccessing the Narrative\\u200b\\n\\n3.\\u200bTour the Narrative\\u200b\\n\\n4.\\u200bNarratives Navigator \\u200b\\n\\n5.\\u200bThe Job Browser\\u200b\\n\\n6.\\u200bCreate a new Narrative\\u200b\\n\\n7.\\u200bExploring data in KBase\\u200b\\n\\n8.\\u200bAdd data to a Narrative\\u200b\\n\\n9.\\u200bBrowsing analysis tools - KBase Apps\\u200b\\n\\n10.\\u200bAnalyze data using Apps\\u200b\\n\\n11.\\u200bRevising the Narrative\\u200b\\n\\n12.\\u200bFormatting Markdown in the Narrative\\u200b\\n\\n13.\\u200bShare a Narrative\\u200b\\n\\n14.\\u200bAccess and copy Narratives\\u200b\\n\\n15.\\u200bCreate an Org\\u200b\\n\\nPreviousNarrative Quick Start\\n\\nNextAbout This Guide\\n\\nLast modified \\n\\n1yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative.html'}),\n",
       " Document(page_content='About This Guide\\n\\nThis is the user guide for the Narrative Interface, the graphical user interface for accessing KBase’s functionality and data. If you are looking for a very quick (one-page) introduction to the Narrative Interface, please see the\\n\\nNarrative Quick Start\\n\\nWhat is the Narrative Interface?\\n\\nThe Narrative Interface enables researchers to design and carry out computational experiments while creating interactive, reproducible records of the data, computational steps, and thought processes underpinning their results. These records can be shared with collaborators and published as “active papers,” called Narratives, which allow others to repeat the computational experiment and alter parameters or input data to produce different or improved results. The Narrative Interface is built on top of the\\n\\nJupyter Notebook\\n\\nframework.\\n\\nPreviousNarrative Interface User Guide\\n\\nNextAccess the Narrative Interface\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_about.html'}),\n",
       " Document(page_content='Access the Narrative Interface\\n\\nThe Narrative Interface is available at\\n\\nnarrative.kbase.us\\n\\n. Once you sign in, you will see the Narratives Navigator.\\n\\nThere are several ways to enter the Narrative Interface from KBase:\\n\\nUse the menu at the top left of the landing page and select “Narrative Interface”.\\n\\nOpen a public Narrative or one that has been shared with you\\n\\nStart a new Narrative by clicking the “+ New Narrative” button\\n\\nMake sure you have the latest version\\n\\nNew versions of the Narrative Interface are released periodically. In most cases, KBase will automatically update to the latest version. However, if you had an older version already running, you will see a green button near the top right of your Narrative that says “New Version Available.”\\n\\nClick this button to update and reload the interface. You also may need to do a hard refresh (shift-reload) in your browser window.\\n\\nPreviousAbout This Guide\\n\\nNextTour the Narrative Interface\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_access.html'}),\n",
       " Document(page_content='Tour the Narrative Interface\\n\\nWhen you open a new Narrative, a “Welcome” cell will appear at the top, offering quick tips for getting started and links for accessing help. You can keep the Welcome cell, collapse it by clicking the \"-\" button in the top right corner of the cell or delete it by selecting \"Delete cell\" from the \"...\" dropdown menu.\\n\\nThe Welcome Cell\\n\\nWhen you first create a Narrative, you may notice the Markdown cell that says \"Welcome to the Narrative\". This is a quick link to this documentation site and shares information on any service status for kbase.us.\\n\\nWelcome Cell\\n\\nBefore adding data and analysis steps to your Narrative, take some time to explore a few high-level features of the Narrative Interface.\\n\\nManaging Narratives and accessing resources\\n\\n1. Top menu\\n\\nNarratives Navigator lets you access other Narratives from a Narrative.\\n\\nAbout the Narrative opens a window describing Narrative Interface properties (including version number) and provides a link to notes about the latest KBase release.\\n\\nShutdown and Restart closes and relaunches the Narrative Interface (which can be useful if your content fails to load or the Narrative appears frozen).\\n\\nThe top-level menu (accessed by clicking the three small horizontal lines on the upper left of the window) offers additional possibilities:\\n\\nNarrative Interface goes to the main Narrative Interface, which allows you to create, edit, run, and share Narratives.\\n\\nNew Narrative opens a new Narrative in a new window.\\n\\nJGI Search goes to the Data Search tab that can also be reached by the Search icon in the side bar.\\n\\nBiochem Search goes to the Compound and Reaction search tabs.\\n\\nAbout opens a page that describes the KBase User Interface and connects to the KBase GitHub.\\n\\nKBase Services Status goes to a page that describes the current version of the Narrative Interface.\\n\\nContact KBase lets you contact the KBase Help Desk with your question, bug report, or feature suggestion.\\n\\nSupport (\"?\" ) brings you to the Narrative User Guide.\\n\\nIn the tabs below Narratives, the top-level menu will look like this:\\n\\nKBase Services Status directs to information on service versions and performance.\\n\\nKBase UI Plugins lists user interface plugins and their version info and links to GitHub.\\n\\nDeveloper links to developer tools and the configuration editor.\\n\\nPublic Search can be used to locate data and filter by data types.\\n\\nAbout directs to information about the KBase User Interface and current version.\\n\\nKBase (home) will open a new window to kbase.us.\\n\\nDocs will open a new window to the KBase Documentation site at docs.kbase.us.\\n\\nSupport will open www.kbase.us/support to access the Help Board or contact the KBase team.\\n\\nIn the next section, we’ll discuss how to create a new Narrative.\\n\\n2. KBase logo\\n\\nLinks to the KBase home page,\\n\\nkbase.us\\n\\n3. Narrative title and creator\\n\\nClick in this space to name (or rename) your Narrative.\\n\\n4. Toggle view-only mode\\n\\nIn\\n\\nview-only mode\\n\\n, you can look at the Narrative but not run or change it.\\n\\n5. Narrative controls\\n\\nHelp menu (“?”) — links to documentation as well as to the Narrative Tour.\\n\\nKernel menu (intended for developers; do not use).\\n\\nShare button — make your Narrative public or give specific KBase users the ability to view, edit, or share it further.\\n\\nSave icon — Be sure to save your work often.\\n\\nUser icon — sign in/out or access your user profile.\\n\\n6. Narrative tabs\\n\\nAnalyze tab — Consists of the Data Panel (see #8) and the Apps Panel below it (see #11).\\n\\nNarratives tab — Allows you to create a new Narrative, copy an existing Narrative, and see a list of other Narratives created by or shared with you.\\n\\nAdding data and apps\\n\\n7. Data Panel\\n\\nYou can add data to your Narrative by searching KBase’s reference collection, importing datasets from external sources, or generating new data from KBase analyses. All data objects from these activities will be listed in this panel, which is empty when you create or open a new Narrative.\\n\\n8. Data Panel controls\\n\\nUse these icons to search, filter, sort, and refresh the data in your panel. The right arrow will open the Data Browser (see #10).\\n\\n9. Data Object\\n\\nEach data object in the panel can be expanded to see more information about the data and several options for downloading, viewing data provenance, and more. The “\\n\\nAdd Data to your Narrative\\n\\n” section describes how to get more information about each data object.\\n\\n10. Add Data\\n\\nThis opens the Data browser, described in detail in the “\\n\\nExplore Data\\n\\n” and “\\n\\nAdd Data to your Narrative\\n\\n” sections, provides options for working with your own data or data from KBase’s reference collection.\\n\\n11. Apps Panel\\n\\nThis panel lists KBase analysis functions. Clicking the app name or icon beside it will add an analysis cell to your Narrative (see “\\n\\nAnalyze Your Data Using KBase Apps\\n\\n” for details).\\n\\n12. App Controls\\n\\nUse these icons to search and refresh the list of apps, as well as expose apps that are available but still in beta. The right arrow will open the\\n\\nApp Catalog\\n\\n, which provides advanced options for browsing apps; designating them as favorites; and filtering them based on analysis type, popularity, and more. For additional information, see\\n\\nBrowse KBase Analysis Tools\\n\\nRunning analyses and adding commentary\\n\\n13. Markdown Cell\\n\\nUse\\n\\nMarkdown language\\n\\nto add rich text commentary and graphics to your Narrative.\\n\\n14. App Cell\\n\\nThere are several types of cells, including those displaying app inputs and outputs, text and commentary, and code cells for writing custom scripts. The cell controls are at the top right of each cell.\\n\\n15. Markdown and Code Cell buttons\\n\\nClick to add these types of cells to your Narrative. Stay tuned for more documentation on how to use code cells in the future!\\n\\nPreviousAccess the Narrative Interface\\n\\nNextNarratives Navigator\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_tour.html'}),\n",
       " Document(page_content='Narratives Navigator\\n\\nWelcome to Narratives! You may notice there are a few changes when you sign in to narrative.kbase.us. Let\\'s get you up to speed with a tour.\\n\\nOverview gif of selecting tabs and sorting\\n\\nThere are five tabs for exploring and locating existing Narratives: My Narratives, Shared with Me, Tutorials, and Public.\\n\\nThe most recently updated Narratives appear first, but can be searched or sorted using the dropdown menu - Recently updated, Least recently updated, Recently created, Oldest, Lexographic (A-Z), Reverse Lexicographic (Z-A).\\n\\nA list of available Narratives will be on the left panel and information on the selected Narrative on the right panel. Here you can see an overview of the Data within the Narrative and a Preview.  Icons under a Narrative name indicate which apps the Narrative includes, how many markdown and code cells it contains, the number of times the Narrative has been shared, whether there are any running jobs, total running time for all completed jobs, and more.\\n\\nClick on the selected Narrative name to open it.\\n\\nToggle between Data and Preview or open a Narrative\\n\\nClick on the Gear icon next to the Narrative title for management options to: Manage SharingCopy this NarrativeRenameLink to OrganizationDelete\\n\\nYou can use the Search Box at the top of each panel to find a particular Narrative within that category.\\n\\nClick the green + New Narrative button at the top to create a new Narrative.\\n\\nClick on the + New Narrative button to create a new Narrative\\n\\nNote that Public Narratives have not necessarily been reviewed or tested by the KBase team; they are simply Narratives that have been made public by their owners.\\n\\nWhere can I go from here?\\n\\nThe Narratives Navigator offers many launch points. To the left of the\\n\\npanel you will see seven icons allowing you to toggle between the\\n\\nand\\n\\ntabs.\\n\\nNarratives\\n\\nTakes you to your accessible Narratives through the Narrative Navigator and features described above.\\n\\nOrgs\\n\\nTakes you to a list of the Organizations within KBase, along with descriptions and statistics for the Organizations, including Narratives and Apps.\\n\\nThe default shows \"My Orgs,\" the Organizations you belong to as a member.\\n\\nYou can view and search all Organizations by selecting the toggle button on the right hand side of the page for \"All Orgs.\"\\n\\nUnder \"All Orgs\" you can search and select different Organizations and request to join by clicking on the \"Join this Organization\" button.\\n\\nCatalog\\n\\nYou can use the App Catalog to browse, search for, and see the specifications of each app KBase has to offer.\\n\\nSearch\\n\\nThe Data Search tab gives you the ability to search JGI data (Beta) as well as KBase user and reference data.\\n\\nIt allows you to search for data present in your own Narratives, those shared with you, and those made public by their owners.\\n\\nIt also enables you to search for reads and assemblies contained in the Joint Genome Institute (JGI) Genomes Online Database (GOLD) and add them directly to your Narratives.\\n\\nJobs\\n\\nThe Job Browser allows you to view apps that you have run recently, see their progress, search for specific runs, and more.\\n\\nYou can filter the runs by Finished, Queued, Running, Success, and Error\\n\\nQuery results also display basic information surround the runs such as the Narrative they can be found in, App ID, Submission Time, Queue Time, Run Time, and Status\\n\\nAccount\\n\\nIn the Account Manager you have access to edit your profile information.\\n\\nFrom this tab you can also get to your Profile Page which gives an overview of your Narratives and your Collaborators.\\n\\nYou can access the user profile of any of your Collaborators by simply clicking on the name under the Your Collaborators panel while in your Profile Page.\\n\\nFeeds\\n\\nTakes you to the Notification Feeds page to view notifications from KBase.\\n\\nPreviousTour the Narrative Interface\\n\\nNextCreate a Narrative\\n\\nLast modified \\n\\n7mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_narratives.html'}),\n",
       " Document(page_content='Create a Narrative\\n\\nCreating a new Narrative\\n\\nYou can create a new Narrative from your\\n\\nNarratives\\n\\nby clicking the “+ New Narrative” button, which will take you to a new Narrative Interface.\\n\\nIf you are already in the Narrative Interface, you can start a new Narrative by selecting the Narratives tab (next to Analyze) and clicking the button labeled “+ New Narrative.”\\n\\nAdding a New Narrative\\n\\nYour Narrative will open in a new browser tab and contain a default “Welcome to the Narrative” cell that provides brief instructions for using the Narrative Interface. You can retain this cell or delete it with the “Delete cell” option in the “…” dropdown menu in the upper right corner of the cell.\\n\\nYou can name the Narrative now by clicking on “Untitled” and entering a new name.\\n\\nNote: The new Narrative will not appear in your Narratives list until it is named something other than ‘Untitled’ and you have clicked the Save button to save it.\\n\\nWe hope that after working on your Narrative, you will want to\\n\\nshare it\\n\\nwith other users. Any Narrative you create is private until you choose to share it.\\n\\nIn the sections that follow, we will explain how to\\n\\nexplore data\\n\\nadd data to your Narrative\\n\\n, and use\\n\\napps\\n\\nto analyze it.\\n\\nCopy and run Narratives created by others\\n\\nIf you don’t feel quite ready to create your own Narrative, you can copy and run a Narrative that someone else has shared with you. Please see the\\n\\nAccess and Copy Narratives\\n\\nsection for more information.\\n\\nPreviousNarratives Navigator\\n\\nNextExplore Data\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_create.html'}),\n",
       " Document(page_content='Explore Data\\n\\nThere are several ways to find data and add it to your Narrative. You can select data available within KBase,\\n\\nupload your own files, or import datasets from external resources\\n\\nsuch as the DOE Joint Genome Institute (JGI) or NCBI. This section will describe how to explore data already in KBase. Instructions for adding data to your Narrative and importing external files are provided in the next section “\\n\\nAdd Data to Your Narrative\\n\\n.”\\n\\nThe Data Browser\\n\\nThere are several ways to explore the\\n\\nwide range of data available in KBase\\n\\nThe Data Browser slide-out is accessed by clicking the right-facing arrow in the menu or the red rectangular \"Add Data\" button in the Data Panel.\\n\\nThe Data Browser will slide out, with tabs that show several data sources that will be discussed in more detail later:\\n\\nMy Data — data you have already loaded or analyzed in another Narrative\\n\\nShared With Me — data included in Narratives that have been shared with you\\n\\nPublic — data in KBase that is accessible to everyone\\n\\nExample — example datasets that can be used as inputs to apps and methods\\n\\nImport — mechanism allowing you to import your own data (of supported types) to your Narrative\\n\\nFind a public reference genome available under the Public tab to examine information and metadata about that genome.\\n\\nClicking the Public tab displays a list of data objects available in the KBase reference collection.\\n\\nThe data-type selector at the top left of the browser window defaults to NCBI RefSeq Genomes. You can change which type of data is displayed under the Public tab by using this pulldown menu.\\n\\nYou can also use the “Search/Filter data...\" field in the Data Browser to find data objects whose names include the text you’ve typed in the search box. (Note — searches are not case-sensitive). Enter “Vibrio brasiliensis” in the search field.\\n\\nIf you hover over a data object in the Data Browser panel, three small images appear: an\\n\\nin a blue rectangle to the left of the object name and\\n\\nicon and a\\n\\nicon to the right. These latter two icons open a\\n\\npage and a\\n\\npage, respectively.\\n\\nGo ahead and click the \"< Add\" button to add the Vibrio brasiliensis LMG 20546 genome to your Narrative. Notice that this genome now appears in your Data Panel.\\n\\nInformation in the Data Panel\\n\\nThe Data Panel shows all the data that you’ve added to your Narrative. (The\\n\\nnext\\n\\nsection\\n\\nof this guide discusses in more detail how to add data to your Narrative.)\\n\\nYou should have at least one object (Vibrio brasiliensis LMG 20546) in your Data Panel. As you add or generate more data during the course of your analyses, the number of objects in this panel will increase. You can search, sort, or filter the list using the icons in the Data Panel header.\\n\\nYou can access more details about a particular data object by hovering over the object and clicking the ”...” that appears in the right-hand side of the data object.\\n\\nThe expanded view of the data object reveals icons that let you examine or manage the data.\\n\\nThe arrow in icon for \"Show Apps with this as input\" shows the apps (in the App Panel) that can use this type of data for input.\\n\\nThe arrow out icon for the \"Show Apps with this as output\" button that shows the apps (in the App Panel) that will generate this type of data as a data object.\\n\\nThe binoculars icon is the \"Explore data\" button that open a Data Landing page for the data object.\\n\\nThe page with three lines and folded upper right corner icon is the \"View associated report\" button and will open the Data View report page for the data object.\\n\\nThe counter clockwise circle with clock hands is the \"View history to revert changes\" icon allows you to see and revert to previous versions of the data.\\n\\nThe graph-like icon is the \"View data provenance and relationships\" button that opens the Provenance page (described below).\\n\\nThe download icon is the \"Export/Download data\" button that lets you download a data object to your local computer. (Note — This capability is still in development; most data objects currently can be downloaded only in the JSON format. In the future, you will be able to choose from a variety of common formats such as GenBank.)\\n\\nThe “A” icon is the \"Rename data\" button that lets you rename your data object. Use with caution. If a Narrative was already using the old object name in analyses, they might stop working. Object names can contain only letters, digits, and underscores; no spaces or other special characters are allowed.\\n\\nThe trash can icon is the \"Delete data\" button that lets you delete a data object from your Narrative. (If the data came from a public or shared data source, only the copy within your Narrative will be deleted.)\\n\\nData viewers\\n\\nMany types of data in KBase have viewers that allow you to learn more about the data. These viewers can be accessed two ways from the Data Panel:\\n\\n1.Click the name of a data object, and its viewer will be added to your Narrative.\\n\\n2.Drag the object from the Data Panel and drop it into the main part of your Narrative.\\n\\nBelow is the genome viewer for the Vibrio brasiliensis LMG 20546 genome that is in the Data Panel.\\n\\nNotice that this genome viewer has tabs for an overview (including GC content, taxonomy information, size, and more) and a list of contigs and genes. Each contig and gene entry in these lists is clickable, opening either a contig browser or a tab with expanded information about the gene.\\n\\nSorting table entries in the viewer\\n\\nYou can sort the table entries by clicking on a column header to sort by that field (e.g., Length). Clicking the same column header again will reverse the sort order. For example, the screenshots below show the table sorted in descending order by Contig name and then in descending order by Length.\\n\\nYou can also sort by more than one column at a time by clicking one column header and then Shift-clicking other column headers. For example, here we have sorted in ascending order by contig length, and then (by shift-clicking the Genes column header) in ascending or descending order by number of genes. Notice how the two rows with length=253 (the bottom two items) have switched places.\\n\\nThere are many types of viewers in KBase in addition to the Genome viewer discussed here. Different viewers may have different options to explore.\\n\\nTo remove a viewer from your Narrative, click the trashcan icon in the top right of the viewer cell. You can always re-add the viewer; removing it from your Narrative doesn’t delete the data object itself.\\n\\nData Landing and Provenance pages\\n\\nData Landing (also known as Data Summary) and Provenance pages are ways to find out even more information about a data object. You can access these pages using the icons that appear when you hover over a data object in the Data Browser or click on one of the objects in your Data Panel. The binoculars icon opens a new Data Landing page about that particular data object, while the graph-like icon opens a Provenance page.\\n\\nData Landing pages (which are still in development) provide both known and contextual information about a data object, allowing users to examine various particulars about the data and, eventually, compare it to other data objects. Depending on the type of data and where it came from, different sorts of information might be presented. For example, for a genome already in KBase (such as the one we loaded earlier,\\n\\nVibrio brasiliensis LMG 20546\\n\\n), you will see:\\n\\nA data object summary\\n\\nA data provenance and reference network (collapsed by default; open it by clicking the > on the left)\\n\\nA genome overview panel that lists information about the genome, including its biological domain, DNA length, etc.\\n\\nA description of the species (including photo, if available)\\n\\nPublications that mention this genome\\n\\nThe taxonomic lineage\\n\\nA species tree (with the option to create a tree if one doesn’t yet exist for this species)\\n\\nA contig browser\\n\\nFunctional categories\\n\\nand more\\n\\nProvenance pages are one way to facilitate reproducibility and transparency of scientific results, two key principles of KBase design. All data in KBase is versioned, and older versions of a data object can be accessed from the Provenance page. This page records and illustrates how data is derived and modified in KBase, including how it entered the system, whether it was produced through analysis of other objects, who generated the data, and when. You also can identify the original “owners” of the data, allowing you to contact them or see their shared analyses.\\n\\nThe image below shows the Provenance page for a flux balance analysis model that was generated in KBase.\\n\\nIn the\\n\\nnext section\\n\\nof this guide, we will discuss in more detail how to add data to your Narrative.\\n\\nPreviousCreate a Narrative\\n\\nNextAdd Data to Your Narrative\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_explore-data.html'}),\n",
       " Document(page_content='Add Data to Your Narrative\\n\\nNow that you are familiar with ways to find and explore data in KBase, you can select or upload data to analyze. The Data Panel in a Narrative shows the data objects that are currently available in that particular Narrative.\\n\\nFrom the Data Panel, you can access the data slide-out, which allows you to search for data of interest and add it to your Narrative. In the Data Panel, click the \"Add Data\" button, the “+” button, or the arrow at the upper right of the panel to access the Data Browser slide-out.\\n\\nData Privacy\\nAny data that you upload to KBase is kept private unless you explicitly choose to share it. You can share any of your Narratives (including their associated data) with one or more specific users, or make it publicly available to all KBase users. Please see the\\n\\nSharing\\n\\npage for more information about how to do that. The\\n\\nTerms and Conditions page\\n\\ndescribes the KBase data policy.\\n\\nThe first four tabs of the\\n\\n, and\\n\\n) let you search data that is already in KBase. The\\n\\ntab lets you import data from your computer to your Narrative so that you can analyze it in KBase.\\n\\nThe My Data tab shows data objects that you have added to your project. You may need to refresh this tab to see your most recently added data.\\n\\nThe Shared With Me and Public tabs display datasets that others have loaded and made accessible to you (or to everyone). Data within each group is searchable and can be filtered. Since there are a large number of public datasets, you may wish to filter them by data type (using the pulldown selector on the left) or narrow the list by searching (in the “Search data” text box) for specific text in the data objects’ names.\\n\\nThe Example tab shows datasets that have been pre-loaded for use with particular apps. These can be handy for trying out the Narrative Interface.\\n\\nThe Import tab allows you to upload your own datasets for analysis.\\n\\nData available in KBase\\n\\nIf you hover your cursor over any data object under the first four tabs, options will appear allowing you to add that object to the Narrative or find out more about it.\\n\\nIn the\\n\\nprevious section\\n\\n, we described the process of adding a genome to your Narrative from the public data in KBase. Now let’s check out the different data types available under the Example tab. The icon to the left of each data object represents its data type.\\n\\nAs described in the previous section, the blue \"< Add\" button next to these icons lets you add the data object to your Narrative. You can add more more data to your Narrative from the\\n\\n, and\\n\\ntabs to try out with various KBase Apps.\\n\\nUploading data from external sources\\n\\nThe Import tab lets you drag & drop data from your computer into your Staging Area to import into your Narrative, where you can then analyze it using KBase’s analysis apps.\\n\\nTo upload data from your computer (or a\\n\\nGlobus endpoint\\n\\nor URL), choose the rightmost tab of the Data Browser to open the Import tab.\\n\\nYou can then click the “?” icon just below the drop zone to the right to launch a short interactive tour that shows the different parts of this user interface.\\n\\nGetting data from your computer to your KBase Narrative is a three-step process:\\n\\n1.Drag and drop the data file(s) from your computer to the new Import tab to upload them to your Staging Area.\\n\\n2.Choose a format for importing data from your Staging Area into your Narrative.\\n\\n3.Run the Import app that is created.\\n\\nWhat is a Staging Area?\\nYour Staging Area is a temporary holding place for your uploaded data files. It is private to you–no one else can see the data in your Staging Area. The files in your Staging Area stay there and can be seen from any of your Narratives. When you’re ready to add data to a Narrative, you can choose a data type from the pulldown menu next to a data file in your staging area in order to add it to your Narrative as an object of that type–see below for instructions.\\n\\nWhy have a Staging Area?\\nIt’s more robust and extensible this way. Unlike our previous importer, the staging upload can handle large data files without timing out. The user interface is more intuitive: you can drag and drop one or more data files into your Staging Area, or even whole directories or zip files. Finally, the new importer is easier for our developers to extend to new data types.\\n\\nWhat\\'s the difference between \\'Upload\\' and \\'Import\\'?\\nIn this documentation, we will use “Upload” to refer to getting data from your computer into your Staging Area and “Import” to refer to the process of converting a data file from your Staging Area into a data object in your Narrative that you can analyze with any of our analysis apps. Sometimes we use the term “Import” to cover the Upload+Import process.\\n\\nDrag & Drop data files into your Staging Area\\n\\nDrag & Drop Limitations\\nThe drag & drop from your local computer works for many files, but there is a size limit that depends on your computer and browser. Some users have reported problems around 20GB. For larger files, use the\\n\\nGlobus Online transfer\\n\\nFind the file(s) you want to import into your KBase account, and drag them into the drop zone (the rectangular area surrounded by a dashed line). You can select multiple files from your computer and drag them all at once. (In the example below, the user is dragging two files into the dashed area.) You can also select a folder of data files and drag the folder into the Staging Area drop zone.\\n\\nIf you don’t like using drag & drop, you can instead click in the upload area to open a file chooser and select a file from your computer to upload.\\n\\nWhile files are being uploaded to your Staging Area, you’ll see a green progress bar.\\n\\nWhen the file is done uploading, you will see it appear in the list of files in your Staging Area. If you don’t see your file, try clicking the refresh button.\\n\\nStaging Area\\n\\nBy default, these are sorted by age, with the most recently uploaded file at the top. To sort the list by other fields, such as name or size, click a column header.\\n\\n90-day lifetime for files in your Staging Area\\nYour Staging Area is meant to be a temporary holding area for data you want to import into your KBase account. After adding files to your Staging Area, be sure to import them into your Narrative soon, as files in the Staging Area are automatically removed after 90 days. Data objects imported into a Narrative will exist indefinitely.\\n\\nTransferring data from a Globus endpoint\\n\\nGlobus is a data management and file transfer system that can facilitate bulk transfer of data (either large data files or a large number of files) between two endpoints. The endpoints that apply here are KBase, JGI, and your local computer. The KBase endpoint is called “KBase Bulk Share,” and JGI has their own way to link to Globus. To do any transfer using Globus, you will\\n\\nneed a Globus account\\n\\nSee\\n\\nTransferring Data with Globus\\n\\nfor more documentation on using Globus.\\n\\nUploading data from JGI\\nIf you are a JGI user, you can transfer public genome reads and assemblies (as well as your private data and annotated genomes) from JGI to your KBase account—see\\n\\nthe JGI data transfer page\\n\\nfor instructions.\\n\\nUploading data from a URL\\n\\nBelow the link to Globus, another link says “Click here to use an App to upload from a public URL” (for example, a GenBank ftp URL, or a Dropbox or Google Drive URL that is publicly accessible).\\n\\nClicking this link adds the “Upload File to Staging from Web” App to your Narrative:\\n\\nThere are also several apps that import specific file types (single- or paired-end reads or SRA files) from a URL directly to your Narrative, bypassing your Staging Area. These are available from the Apps Panel and the\\n\\nApp Catalog\\n\\nImporting files from the Staging Area to the KBase Narrative\\n\\nImport Basics\\n\\nThe files in your Staging Area are ready to import into your Narrative as KBase data objects that can be used in your analyses. To import a file from your Staging Area, choose a format (data type) from the pulldown menu to the right of the file’s age. (You can find out more about KBase data types and accepted formats in the\\n\\nUpload/Download Guide\\n\\n.) Then click the import icon to the right of the format menu.\\n\\nWhen you click the import icon, the Data Browser slides shut and an Import App cell for each data type or a Bulk Import App cell will be created in your Narrative, with the appropriate parameters filled in. For example, here’s an import app created by choosing “GenBank” as the import format:\\n\\nImporting different types of data\\nThis example shows how to import GenBank data. Please see the\\n\\nUpload/Download Guide\\n\\nfor detailed instructions for other supported data types and importing multiple files.\\n\\nIf the GenBank file came from a different source, use the pulldown menu to select it. You can change the output object name, if desired, and then click the \"Run\" button to start the import. When the import is done, you should see the message “Finished with success” near the top of the app cell, and some information about the app run.\\n\\nIf you look at your Data Panel, you should see the new data object created by the import.\\n\\nYou can now use this data object as input into the relevant KBase apps. If you want to see which apps accept a particular data type as input, you can click the “…” menu in the data object cell that appears when you hover over it, and then use the “Show Apps with this as input” icon to filter the apps in the Apps Panel.\\n\\nWhat if my import fails?\\nSometimes an import doesn’t work. One of the most common causes of failure is attempting to import a file that’s the wrong data type, or not the expected format for that data type. For example, the screenshot below shows what a user got when they tried to import a GenBank file as Media.\\n\\nYou can also look at\\n\\ncommon import errors and their meanings.\\n\\nIf the importer objected to something in your file, check the\\n\\nData Upload/Download guide\\n\\nfor details about the relevant format.\\n\\nIn some cases, the cause of an import error will not be obvious. If you can’t figure out why your import isn’t working, please contact us (via the\\n\\nHelp Board\\n\\n) for help. Note, however, that no one besides you has access to your Staging Area, so we will not be able to see the files you uploaded to your Staging Area. You may need to attach your input file to your Help Board ticket in order for us to diagnose the problem.\\n\\nBulk Import for Multiple files\\n\\nMultiple files can be imported together in a single Import App as part of\\n\\nthe bulk import feature\\n\\n. When you select the file type from \"Import As...,\" you will be able to check multiple files in the staging area and import them all by clicking \"Import Selected.\"\\n\\nBulk Import Staging Area. You can select the appropriate type for multiple files, then \"Import Selected\" to import them simultaneously.\\n\\nWhen you import selected, there will be a new cell type which contains a tab for each different file type you selected from the staging area. You will fill out the same parameters for each file type by clicking the file type in the \"Data Types\" column on the left; these are the same parameters as single imports.\\n\\nBulk Import Cell. File types can be selected under the data type column to set the parameters for each type.\\n\\nSupported file types include:\\n\\nAssembly - FASTA\\n\\nSRA Reads\\n\\nFASTQ Reads Interleaved - FASTQ Interleaved reads\\n\\nFASTQ Reads Noninterleaved - FASTQ paired-end reads or FASTQ single-end reads library\\n\\nGenBank Genome - GenBank\\n\\nGFF Metagenome - FASTA and GFF3\\n\\nGetting more information about files in your Staging Area\\n\\nThe list of files in your Staging Area includes their name, size, and age (from when they were uploaded). If you have many files in your Staging Area, use the Search box to locate specific files.\\n\\nCompressed or zipped files have a little double-arrow icon next to the filename. You can click this icon to unpack these files. For how to upload and import file types, follow\\n\\nthis guide\\n\\nFor more information about a file in your staging area, click the arrow to the left of the filename to open a tab like this:\\n\\nYou can click “First 10 lines” or “Last 10 lines” to see that portion of the file:\\n\\nOpening the information about a file in your Staging Area also reveals a trash can icon that allows you to remove the file from your staging area.\\n\\nYou will be asked to confirm that you want to delete the file. This action is not reversible.\\n\\nNote that if you had already imported the file to your Narrative as a data object, that object won’t go away when you delete the file in your Staging Area. If you want to delete a data object, you can do that in your Data Panel.\\n\\nIn progress\\nKBase’s import functionality and user interface are under active development. We\\n\\nwelcome your bug reports and suggestions\\n\\nfor future upload functionality.\\n\\nNow what?\\n\\nOnce you have added data–your own data or reference data that is already in KBase–to your Narrative, you will be ready for the exciting part: analyzing it! The next two sections describe how to\\n\\nchoose\\n\\nand\\n\\nrun an app\\n\\nto analyze your data.\\n\\nPreviousExplore Data\\n\\nNextBrowse KBase Analysis Tools\\n\\nLast modified \\n\\n7mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_add-data.html'}),\n",
       " Document(page_content='Browse KBase Analysis Tools\\n\\nA Narrative can include any number of analysis steps that are added by selecting one or more apps from the Apps Panel directly below your data. Apps are grouped by category. Click the arrow to the left of a category name to see the individual apps in that category.\\n\\nYou can browse KBase Apps by (1) scrolling through the list in the Apps Panel; (2) searching for an app by name; or (3) using the App Catalog (click the right arrow in the top right corner of the Apps Panel to open it) to further explore, filter, and designate apps as favorites.\\n\\nSearch for Apps in the Apps Panel\\n\\nClick the search icon (magnifying glass) at the top of the panel to begin a text search for apps. Once you enter a term, the list will show only the categories that have apps whose names include the text you typed. As before, click the right-arrow to the left of a category name to see matching apps in that category. (To see the complete list again, click the \"x\" to the right of the search box.)\\n\\nHovering over an app in the list will show the more “…” icon. Clicking the \"...\" icon will open an app details page featuring an expanded description of the app, along with technical information about its inputs, outputs, parameters, and more. Links to additional documentation such as tutorials or FAQs also may be provided.\\n\\nAccess the App Catalog\\n\\nAll available apps can be found in the App Catalog, which is accessible three ways:\\n\\nExternal App Catalog (no user account or sign-in required for browsing).\\n\\nFrom KBase by clicking on the Catalog icon in the Menu\\n\\nFrom the Narrative Interface, by clicking the right arrow at the top of the Apps Panel.\\n\\nThe App Catalog will slide out to show available apps. (The “View App Catalog” link at the top opens the App Catalog in a separate web browser window.)\\n\\nAdd your own app!\\n\\nThe App Catalog is growing to contain not only the apps generated by KBase staff, but also those created and contributed by outside developers. KBase’s\\n\\nSoftware Development Toolkit (SDK)\\n\\n, provides a mechanism for users to add their own open-source, open-license tools to the system as new KBase apps.\\n\\nSort and Filter Apps\\n\\n\\u200b\\u200b\\n\\nFrom the App Catalog, you can search for apps by name and use several options to organize and filter how they are displayed (the default sort is by category). Click the “Organize by” drop-down menu to see these options.\\n\\nMy Favorites — Lists your favorited apps (those you’ve clicked the star icon to “favorite”) at the top of the catalog.\\n\\nFavorites Count and Run Count — Suggests some of the most popular and frequently used apps in KBase.\\n\\nName — Filters apps by name (alphabetically or reverse alphabetically).\\n\\nCategory (default sort) — Displays apps in groups by analysis type including annotation, assembly, communities, comparative genomics, expression, metabolic modeling, and more.\\n\\nModule — Sorts apps based on which SDK module they belong to.\\n\\nDeveloper — Displays apps alphabetically by developers’ KBase usernames, which are linked to their profile pages.\\n\\u200b\\n\\nInput Types and Output Types —Groups apps based on the type of data that they take as input or generate as output. Each data type is linked to a specification page that provides some technical details, including version, description, structure, and more. This sort can be useful for determining the data you must add to your Narrative before using a specific app. The Example tab of the Data Browser provides access to various data types to help you get started.\\n\\nExplore Individual Apps\\n\\nSimilar to the App Panel, the App Catalog lets you (1) see a short description of each app (by hovering over the “i”–see second screenshot at right), (2) access the app detail page using the “more” link, and (3) add an app to your favorites by clicking the star.\\n\\nThe number beside the star icon indicates how many other users have selected the app as a favorite, while the next number shows how many times the app has been run.\\n\\nYou can access the app module (a group of related apps) by clicking its hyperlink (in this example, “fba_tools”). Also, some apps will link to the KBase profile pages of their developers (in this case, chenry).\\n\\nFeeling adventurous? Go beta\\n\\nTo try out apps still in development, click the little “R” in the Apps Panel, changing it to a “B”. This will show apps that are in Beta (i.e., still being tested and debugged). Click the “B” again to go back to “R” (show only released apps).\\n\\nNow that you know how to find an app of interest, the\\n\\nnext section\\n\\nwill discuss how to add it to your Narrative so you can begin analyzing data.\\n\\nPreviousAdd Data to Your Narrative\\n\\nNextAnalyze Data Using KBase Apps\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_add-apps.html'}),\n",
       " Document(page_content='Analyze Data Using KBase Apps\\n\\nOnce exploring available\\n\\nKBase Apps\\n\\nand determining which ones work with the data objects in your Narrative, you can start analyzing your data.\\n\\nAdd an App to your Narrative\\n\\nTo add a KBase analysis app, find the app of interest and click its name or the icon to the left of the app name. A box (called a cell) containing the chosen app will appear in the main Narrative panel.\\n\\nA few things to notice about the app cell:\\n\\nApp Cells have three tabs\\n\\n1.Configure (where you set the parameters; it’s what you see when you add a new app cell)\\n\\n2.Info\\n\\n3.Job Status\\n\\n4.Result\\n\\nThe last two tabs are filled in once you run the app; these are discussed later.\\n\\nThe up and down arrows let you move the cell up or down in your Narrative.\\n\\nThe App “…” dropdown menu offers several options for showing the code that will execute the app; getting more info (in a separate browser window) about the app; and deleting the cell. These are discussed more in the section called “Revise Your Narrative“.\\n\\nThe \"-/+\" in the App Cell menu is the collapse/expand cell option. When the App Cell is expanded, the farthest right option will be a \"-\" to collapse the cell. When the App Cell is collapsed, the farthest right option will be a \"+\" to expand the cell.  You can also double click on an App Cell to expand/collapse the app.\\n\\nEvery app has some parameters (fields) that must be filled in before you can run the app.\\n\\nFill in parameters\\n\\nAfter you add an app to your Narrative, the required parameters must be filled in before you can run it. For most apps, you will need to select the input data object(s). Other parameters may also need to be set. A red bar to the right of a field indicates that it is a required field and you have not yet entered a valid value in it. Other indicators, such as banners for errors and warnings, may appear. Read the message and hover over any icons to reveal hints. Once you have filled or corrected the field, the indicator should disappear.\\n\\nSome app fields are “smart” and know which data in your Narrative is valid for that field. These “smart” fields have a pulldown list of data objects that you can choose from. (Remember, only data (of the appropriate type) that you have already added to this particular Narrative will be shown in that list. You can access your data from other Narratives via the My Data and Shared with me tabs in the\\n\\nData Panel\\n\\n.)\\n\\nSome fields are also required, but they will be pre-filled with default values. You can change their values if you chose, or leave them. Additionally, some apps have optional “advanced options” that you can reveal by clicking on the “advanced options” link at the bottom of the cell.\\n\\nThe green \"Run\" button is enabled when all required fields are filled in.\\n\\nSave Your Work\\n\\nAs you add to your Narrative, be sure to save your work frequently, using the Save button at the top right of the screen.\\n\\nRun the App\\n\\nWhen you have filled in all the required fields, the\\n\\nbutton on the left side of the app will be enabled; click it to start running the app. Once clicked, the green\\n\\nbutton will turn to a red\\n\\nbutton. When the analysis starts running, you will see information in the\\n\\ntab\\n\\nDepending on the analysis, the analysis might finish within a few seconds or could take hours to run. You can save your Narrative and come back to it later–the analysis will continue running even if you don’t have the Narrative open.\\n\\nExamine and download results\\n\\nWhen an app finishes running, you will see a summary in the Results tab, as well as an output cell below your app cell. Also, if the analysis generates a new data object, that object will be added to your Data Panel.\\n\\nIn the example, the “Annotate Microbial Contigs” app, once run, produced an output cell that provides information about the newly annotated Shewanella genome object.\\n\\nThis output cell has three tabs: Overview (displayed above), Browse Contigs, which lists the large assembled pieces of a genome, and Browse Features, which allows browsing of all the annotated genes.\\n\\nDifferent apps create different types of output cells when they run, depending on which type of data object is output by the app. The genome output cell shown above is an example of a Data Viewer. Data Viewers are described in the\\n\\nExplore Data\\n\\nsection of this guide.\\n\\nIn addition to creating a new output cell, the app we ran created a new data object and added it to our Narrative. The newly annotated genome object now appears at the top of the Data Panel.\\n\\nRemember that you can hover over an object in the Data Panel and click the “...\" that will appear to see more information and options, including an option (see image) for downloading the object in GenBank or JSON format (see the\\n\\nDownload Guide\\n\\nfor more information). In the future, you will be able to choose from a wider variety of common formats, allowing you to download results to use with other tools and send to colleagues.\\n\\nConduct further analyses\\n\\nOnce you have reviewed your results, you can use your newly generated data in additional analysis steps. Remember, you can click on the data object in the Data Panel to see which KBase Apps work with that data type.\\n\\nFor example, our annotated genome can be used in a number of analyses because it now includes the standard KBase annotations. If you click on the genome in the Data Panel, you see several apps that apply. Among them are the\\n\\nInsert Genomes into Species Tree App\\n\\nthat constructs a phylogenetic tree allowing you to see species closely related to your genome. You also can use the\\n\\nBuild Metabolic Model App\\n\\nto draft a metabolic model from the annotated genome.\\n\\nRe-run Apps with the same or different parameters\\n\\nApp input cells that have already been run have a “Reset” button that allows you to redo the analysis, with the option to change any of the parameter settings (including, perhaps, the input data) before rerunning. Rerunning an App will overwrite the information in the App cells, but will not overwrite any data objects in your Narrative unless you use the same output object name again.\\n\\nReport errors\\n\\nIf a job fails, you will see an error message in the output cell and/or in the Result tab. If you need help figuring out what went wrong, please\\n\\ncontact us\\n\\nwith details about what you were doing when the error occurred and what the error message said. Including the contents of the Job Status tab and the URL of your Narrative will help us debug the problem. We are here to help and appreciate your feedback and error reports!\\n\\nPreviousBrowse KBase Analysis Tools\\n\\nNextJob Browser\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_analyze-data.html'}),\n",
       " Document(page_content='Job Browser\\n\\nThe Job Browser is one of the buttons on the left sidebar menu in KBase. In addition to watching jobs within the Narratives, the Job Browser allows you to monitor and manage your jobs. By default, it shows all jobs submitted within the last 48 hours. You can change the timeframe to an hour, a week, a month, or a custom range of dates.\\n\\nThe check boxes allow you to filter the jobs by\\n\\n, and\\n\\nYou can check as many filters you want. Clicking one of the filters or changing the timeframe will update the page. Otherwise, the table does not auto-refresh. The two arrows in a circle to the right and below the timeframe will refresh the results in the table.\\n\\nThe results display basic information for the jobs such as the Narrative they are located within, App ID, Submission Time, Queue Time, Run Time, and Status. The name of the Narrative is hot-linked, and clicking on it will open a new tab with the Narrative. The App ID is hot-linked and clicking on it will open a new tab with the catalog page for the app.\\n\\nJob Status\\n\\nThe Job Status Cell has the status, an icon for the log, and a red cancel icon. Click the log icon to expand the table and show the contents of the log. Click the icon again to collapse the table. Canceling the job will prompt you to make sure you didn’t click it accidentally. Jobs that have been canceled may continue to show up as “Queued” or “Running” until they clear the system.\\n\\nJob Log\\n\\nThe Job ID and worker node are easily located and the log can be scrolled through within the pop-out. Job Logs can be downloaded in CSV, TSV, JSON, and TEXT formats using the download (downward-facing arrow into tray) icon.\\n\\nPreviousAnalyze Data Using KBase Apps\\n\\nNextRevise Your Narrative\\n\\nLast modified \\n\\n2yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_job-browser.html'}),\n",
       " Document(page_content='Revise Your Narrative\\n\\nYou can browse a list of your Narratives in the Narratives tab:\\n\\nYour Narratives are shown in the Narratives tab in reverse chronological order, with the most recently updated Narrative first. After your personal Narratives, more Narratives owned by others and shared with you are listed.\\n\\nView Narrative history and revert to earlier versions\\n\\nKBase stores a history of everything you’ve done to your Narrative, so you are able to revert changes and go back to an earlier version. When you hover your cursor over one of your Narratives in the Narratives panel, three icons will appear.\\n\\nThe first of these icons (counterclockwise encircling arrow) lets you View Narrative history to revert changes or access prior saved versions of the Narrative. Here you can choose an earlier version of a Narrative to revert to from the saved versions.\\n\\nAdd text to your Narrative\\n\\nAdding text cells to your Narrative to explain what you were doing and what the results might mean will help others understand your computational experiments, as well as helping to remind you what you were doing. You can add formatted text to your Narrative in “Markdown Cells”. Markdown is a markup language (one example is HTML). It’s pretty easy to learn but does require knowing some syntax. A good\\n\\nMarkdown CheatSheet\\n\\ncan be found on GitHub.\\n\\nYou can add Markdown cells to a Narrative using the translucent blue buttons at the bottom right of the main Narrative panel.\\n\\nThe button on the right is the “Add Markdown Cell” button and will add a Markdown cell under the currently selected Narrative cell.\\n\\nNew Markdown cells contain the placeholder text “[Empty Markdown / LaTeX cell]” (LaTeX is a formatting language). To edit a Markdown cell, double-click it to enter edit mode. When you’re done, press Shift-Return to exit the cell.\\n\\nThe \">_\" button will Add a Code Cell, which is beyond the scope of this user guide. TIP: To use a code cell press Shift-Return within the cell to execute the code.\\n\\nCell commands: move, delete, etc.\\n\\nVia the arrows and menu items in a Narrative cell, you can move the cell up or down, collapse or expand it, show the code that runs the analysis, get more info about the app, or delete the app cell. The cell menu can be opened by clicking the “…” icon in the upper right corner of a cell.\\n\\nRename or delete a Narrative\\n\\nTo rename a Narrative, click on its name at the top of the Narrative window. A window will pop up to let you change the name.\\n\\nIf you decide to delete a Narrative that you own, click the trashcan icon. Delete with caution, because currently there is no way to get your deleted Narrative back.\\n\\nSave your work!\\n\\nAlthough your Narrative is periodically autosaved, it is always a good idea to save your work frequently, using the Save button at the top right of the screen.\\n\\nOnce you are happy with your Narrative, you will probably want to share it with others. The section\\n\\nShare Narratives\\n\\ndescribes how to do that.\\n\\nPreviousJob Browser\\n\\nNextFormatting Markdown Cells\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_revise.html'}),\n",
       " Document(page_content='Formatting Markdown Cells\\n\\nTo edit the appearance in Markdown cells for publication, teaching, and to add navigation within the Narrative, use a combination of Markdown and HTML.\\n\\nAdd formatted text to your Narrative using\\n\\nMarkdown cells\\n\\nto explain your process, interpret results, and communicate to others for sharing and publication and more.\\n\\nMarkdown cells use markup languages of Markdown and HTML. Both are pretty easy to learn but does require knowing some syntax. Not every formatting option will work in the Narrative Interface.\\n\\n\\u200bNarrative Formatting Cheat Sheet to copy and save\\n\\nResources for Markdown syntax\\n\\n\\u200bMarkdown Cheat Sheet\\u200b\\n\\n\\u200bMarkdown Guide\\u200b\\n\\nResources for HTML syntax\\n\\n\\u200bHTML Cheat Sheet\\u200b\\n\\n\\u200bHTML Guide\\u200b\\n\\n\\u200bHTML Color Codes\\u200b\\n\\nPreviousRevise Your Narrative\\n\\nNextShare Narratives\\n\\nLast modified \\n\\n2yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_markdown.html'}),\n",
       " Document(page_content='Share Narratives\\n\\nHow to share your KBase Narrative workspaces with individuals and Orgs.\\n\\nNarratives are\\n\\n(multiple people on your project can work on a Narrative together),\\n\\n(you can share your Narrative with specific people or with everyone),\\n\\n(a finished Narrative is similar to a research paper, and can be cited as a URL), and\\n\\n(you can repeat or even change another scientist’s computational experiment if they publish it as a Narrative). We believe that you will ultimately help both yourself and other researchers by sharing your Narratives so that others can see what you were thinking, what you did, and what you concluded from your analyses. However,\\n\\nA quick 4-min Tutorial on how to share Narratives.\\n\\nHow to Share a Narrative\\n\\nYou can share a Narrative that you own with any set of people you choose, or with all KBase Users. To share a Narrative, click the “share” button near the top right and select the \"Manage Sharing\" option in the menu.\\n\\nYou will now see the Change Share Settings pop-up window. Here, you can see the people you have already shared with (right now, only you).\\n\\nIf you want to share your Narrative with everyone, you can click the “make public” link near the top of the sharing panel.\\n\\nUsers tab\\n\\nIf you only want to share it with a few people, you can type their name or username into the “share with” box. Once you have typed something that matches a name in the KBase User Database, the person’s full name will appear. Click the down arrow to the right of the “share with” box to assign permissions to that user–you can allow them only to view your Narrative, to edit it, or to edit and share it.\\n\\nFor example, suppose you want to share your Narrative with kbasehelp so that KBase staff can help you troubleshoot it. You can start typing any part of the username or real name and a list of users that match will appear:\\n\\nThe default permission level for sharing is “View only”, which lets the user you’re sharing with see the Narrative but not let them edit it or share it with others. If you want to share your Narrative with a collaborator and give them more privileges, click the down-arrow to see the permission levels and possibly choose a different one.\\n\\nYou can add multiple users to share with:\\n\\nClick the “Apply” button to add the chosen user(s) to the sharing list for your Narrative. The usernames and permission levels for the users you shared with will appear at the bottom of the Share panel.\\n\\nOrgs tab\\n\\nTo share your Narrative with a specific Organization, select the Orgs tab. Here you can add a single or multiple Organizations to share with a group of users.\\n\\nClick the “Apply” button to add the chosen Organization(s) to the sharing list for your Narrative. The Organizations and permission levels for the users you shared with will appear at the bottom of the Share panel.\\n\\nWhen you’re done sharing, click the \"x\" icon to close the Share panel.\\n\\nEditing access permissions\\n\\nYou can change the access permissions at any time by clicking the Share button again. The up/down arrows next to each user’s permissions let you change the permissions.\\n\\nNote: Unless you alert another user that you have shared a Narrative with them, they will not notice a newly shared Narrative until they see your shared Narrative in their “Narratives that have been shared with you” tab or click on your name in her collaborator list.\\n\\nIf you are working on a shared Narrative, and a collaborator is working on it at the same time, watch for a red box at the top of the screen that says “Narrative updated in another session.” You can update to the latest version by refreshing the page, but note that any changes that you made to the Narrative will be lost, and if you try to save your changes, you might overwrite your collaborator’s changes.\\n\\nPublishing a Static Narrative\\n\\nYou can share you Narrative in publication ready, uneditable format that is accessible outside of KBase by creating a static Narrative. Once you are ready to share the Narrative, there are two steps.\\n\\n1. Click on the share icon in the menu and click on Manage Sharing in the dropdown menu. In the pop-up window click \\'make public?\\' to change the privacy settings. Click the \\'x\\' icon to close the pop-up window.\\n\\nGif of opening the Manage Sharing pop-up window\\n\\n2. Return and click on the \\'share\\' icon, go to \\'Manage Static Narratives in the   dropdown menu. The \\'Static Narratives\\' pop-up window will open and show the status if any versions of this Narrative have been made static. To create a static Narrative of the current version, click on the blue \\'Create static narrative\\' button.\\n\\nGif of creating a Static Narrative\\n\\n3. Access the static Narrative and share the link by clicking on the \\'Existing static Narrative\\' link within the pop-up window.\\n\\nImage of Static Narratives pop-up window.\\n\\nCreate or modify your user profile\\n\\nEvery KBase User has a user profile. Filling in your KBase user profile helps you become part of the social web that KBase is building, making it easier to find and collaborate with other scientists and share data and Narratives. You are not required to complete your user profile in order to use KBase, but we encourage everyone to do so.\\n\\nTo see and edit your user profile, go to the Menu sidebar on the left and click the “Account” button on the left side. Your profile page will open, looking something like the screenshot below. You will be able to enter or change information including your organization, department, job title, research interests, avatar, and more. Required fields are marked with a red \"*.\" You must fill in all required fields in order to save your profile (using the Save button on the upper right of the page).\\n\\nBelow the \"Save\" button is a small preview showing what your profile page looks like to others. You can click the “Open Your Profile Page” link to see a bigger version, along with a list of your Narratives and collaborators (other users with whom you have shared Narratives, or who have shared Narratives with you) and a search box for searching for other users by name or username. As you start typing in the box, you will see a list of users who have that text in their first or last name or their username. Click one of the matches to see that user’s profile.\\n\\nThe Account tab lets you see basic information about your KBase account. The only field that can be edited there is your preferred email address. Note that other users cannot see your email address; it will be used only by KBase staff to contact you occasionally with important information.\\n\\nThe Linked Sign-In Accounts tab lets you see and\\n\\nmanage\\n\\nwhich external (Globus or Google or ORCiD) accounts are linked to your KBase account.\\n\\nThe User Profile editor allows you to edit your position, organization, and location. You can also input your Research Interests, Research or Personal Statement. It also shows which KBase Organizations you belong to and any Affiliations.\\n\\nPreviousFormatting Markdown Cells\\n\\nNextLinking Static Narratives to ORCID\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_share.html'}),\n",
       " Document(page_content=\"Linking Static Narratives to ORCID\\n\\nHow to add KBase static Narrative DOIs to your ORCID account.\\n\\nAdapted from the article\\n\\nDOE OSTI Search & Link Wizard Added to ORCID by  Stephanie Gerics and Carly Robinson\\n\\nAdd Works using the DOE / OSTI Search & Link Wizard\\n\\n1.Sign in to ORCID\\n\\n2.Scroll down to the Works section of your ORCID record and hover over the '+Add works' in the menu and select 'Search & link'\\n\\n3. Under 'Link Works', select DOE / OSTI from the list of member organizations.\\n\\n4. While using the DOE / OSTI Search & link wizard for the first time, you will need to Authorize access to approve DOE / OSTI to make changes to your ORCID Works.\\n\\nOnce DOE / OSTI is authorized, the OSTI.GOV page will appear with the green ORCID iD icon next to the user’s ORCID iD name, as shown below. The ORCID Account Integration page will display\\n\\n, the user’s\\n\\nand directions for adding works to the user’s ORCID record using OSTI.GOV’s search. Browse DOE / OSTI search results to identify records authored, coauthored, or contributed by the user\\n\\n5. Once you find records in the search results to add to your ORCID record, select 'Add to ORCID Works.'\\n\\n6. A pop-up text box will appear and confirm that the user wants to add this record to ORCID Works. Select 'Add to my ORCID Works'.\\n\\n7. In the search results, works added by the user to their ORCID record will now display 'In your ORCID Works' at the bottom right of the popup.\\n\\n8. To view your list of works added to your ORCID record, navigate back to the 'ORCID Account Integration' page. Select the user name next to the green ORCID iD icon at the top of the page.\\n\\n9. Click on the ORCID Works in OSTI.GOV to see the works integrated with your account.\\n\\nPreviousShare Narratives\\n\\nNextAccess and Copy Narratives\\n\\nLast modified \\n\\n1yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_link-doi.html'}),\n",
       " Document(page_content='Access and Copy Narratives\\n\\nView Narratives created by or shared with you\\n\\nThe Narratives tab has two sections that list Narratives that you have created (My Narratives) and, below that list, those that have been shared with you by other users. The Shared With Me list includes only those Narratives that were explicitly shared with you, not those that are viewable by everyone (public Narratives). The public Narratives can be accessed from the Narratives dash.\\n\\nIf you click on one of the Narratives in your Narratives panel, it will open in a new web browser window. If it’s a Narrative that you own, you will be able to run it or edit it. However, if it’s one that someone else owns and shared with you without write permission, you won’t be able to edit or run it–it will open in View-only mode.\\n\\nView-only mode\\n\\nThe key difference you will notice when a Narrative opens in View-only mode is that there is no Analyze panel on the left side–you won’t see any data objects or apps. This is because you can’t modify or run a Narrative that was shared with you with only View permission.\\n\\nWhy can’t you run a Narrative owned by someone else? Because, as we’ve seen, running an app usually creates new data objects in your Narrative, which requires write permission. If you want to run a Narrative owned by someone else who has not given you write permission, you will need to copy it to your account.\\n\\nThe blue box with a small white arrow on the top left of a View-only Narrative opens the Data Panel (so you can see the data objects in this Narrative) and the Narratives list (so that you can go to another Narrative).\\n\\nYou can switch a Narrative that you own to view-only mode by clicking the pencil icon under the title of the Narrative.\\n\\nIf you ever want to print out a Narrative, switch it to view-only mode first, so it doesn’t waste the left side of the page with empty space.\\n\\nCopy a Narrative\\n\\nTo copy a Narrative that has been shared with you, so that you can run or change it, you can use the blue “Copy This Narrative” button at the top of the Narratives panel (or, in View-only mode, at the top right of the screen). This will copy the Narrative that you are currently viewing (which is marked with a green dot in the list of Narratives). You can choose a name for the copy (or leave the default name, which is the original name with “Copy” at the end).\\n\\nWhen you click Copy, the Narrative and its associated data objects are copied to your account, and you should see the new copy appear at the top of “My Narratives”.\\n\\nCopy of a copy warning\\n\\nPlease be aware that if you copy a copy of a Narrative, some things may not work right in the copy of a copy. We are working on resolving this problem.\\n\\nYou can copy any Narrative by hovering your cursor over its name in your Narratives panel to reveal the icons and clicking the Copy icon (the one that looks like two sheets of paper).\\n\\nThe copy is now yours–you can open it, run it, edit it, or even share it with others (if the owner of the original Narrative gave you sharing privileges). After you’ve copied a Narrative, you can run all the steps to reproduce the computational experiment, or try changing the parameters or input data to alter or perhaps even improve the results.\\n\\nCopying Narratives is the key to leveraging KBase’s commitment to reproducibility and reusability. By making it easy to build on previous work and go through rapid cycles of analysis, KBase aims to accelerate the pace of systems biology research.\\n\\nPreviousLinking Static Narratives to ORCID\\n\\nNextOrganizations\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_access-and-copy.html'}),\n",
       " Document(page_content='Organizations\\n\\nHow to access, search, and create KBase Organizations for labs, collaborating researchers, courses, and more.\\n\\nWhat is a KBase Org?\\n\\nIn science, people are members of labs, teams, groups, collaborations, and projects that work together with shared data and analyses. In KBase, organizations are a way for teams of scientists to share their data and associated analyses that are in the Narratives they create with each other as a group. Organization members can see information about the team and a list of the Narratives associated with the organization to which can request access. View-only access is granted upon request while all other access to the Narrative is granted by the Narrative owner. KBase users can be members of more than one organization and Narratives with the associated data might also be added to more than one organization.\\n\\n7 minute video tutorial on KBase Organizations\\n\\nCreating an Organization\\n\\nOn the Orgs page, click \"Create Organization\" button on the right side. This will open the Create Your Organization page. Input the information and Save.\\n\\nTo create an Organization, you will need:\\n\\nName ⏤ the displayed Organization Name\\n\\nID ⏤ the unique URL for the Organization; can only contain lower case letters (a-z), numeric digits (0-9) and the dash \"-\"\\n\\nLogo URL (optional) ⏤ include the link to a publicly available image\\n\\nHome Page URL (optional) ⏤ input your lab or group\\'s website URL to link and share\\n\\nHidden? ⏤ check for the Organization to be Hidden - will be visible only for members of this organization or leave unchecked to remain Visible - will be visible to all KBase users.\\n\\nResearch Interests ⏤ list your Organization\\'s research interests here\\n\\nDescription ⏤ provide a description of your Organization\\n\\nEditing an Organization\\n\\nThere are options to edit and share an organization after it has been created.\\n\\nJoining an Organization\\n\\nIf an invitation appears in your feeds, click on the ID of the organization. A window will pop up with options to Accept or Reject the invitation. Click on your preference.\\n\\nWithout invitation, search for the organization and click on its name. Click on the blue button (upper right) to ‘Join Organization’. You cancel this request if needed. A request to join will be sent to the owner and group administrators. An acceptance or rejection will be added to your Feeds. Note: Private organizations will not appear.\\n\\nIf you know the organization URL (starts with https://Narrative.kbase.us/#org/) go to the webpage and request membership.\\n\\nViewing Narratives in an Org\\n\\nIf you haven\\'t opened the Narrative before, there will be a \"Click for Access/View Access\" buttonClick the \"Click for View Access\" buttonYou will be able to open and view the Narrative when the \"No Access\" line disappearsTo open a Narrative, click on the Narrative name\\n\\nAccess a Narrative by clicking the \"Click for View Access\" button to be able to view the Narrative.\\n\\nThere are several icons that may appear in the Narrative cell:globe   - Narrative is shared publicly with all KBase userspencil  - Narrative is shared with you with edit accessopen lock  - Narrative is shared with you with access to edit and shareclosed lock  - privateeye  - View-Only accesscrown  - you are the owner\\n\\nAdministering an Organization\\n\\nAdd Narratives by either the \"+ Add a Narrative\" button or share Narratives with an Org.\\n\\nInvite members by clicking the invite button on the right above the list of members.In the Search box ‘Select User to Invite’, type in the user’s name or KBase ID. A list of users will appear below the search box. Click on the user to invite and the name will appear on the right with more information that can be used to confirm it is the correct user. If correct, click on the blue ‘Send Invitation’ button.\\n\\nOn the right side of the organization page is a list of organization members. Above the list of members is a tab called ‘Requests’. There is an \"Inbox\" for a list of Users requesting to join and an \"Outbox\" of owner/administrator invitations for pending invitations for users to join that the owner or administrator made. Requests can be accepted or rejected by clicking the respective buttons. Invitations can be cancelled at any time, but will also disappear after a length of time.\\n\\nOwners and administrators can can promote members to administrator. In the membership list, click on the ‘...’ to the right of the member’s name and click on ‘Promote to Admin’.\\n\\nMembers can be removed by click on the ‘...’ to the right of the member’s name in the membership list and click on ‘Remove Member’. The user is not notified that they have been removed.\\n\\nThere is no mechanism to delete an organization. You can un-associate all Narratives, remove all members and set the group to “Private.”\\n\\nPreviousAccess and Copy Narratives\\n\\nNextFAQs\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_orgs.html'}),\n",
       " Document(page_content='FAQs\\n\\nFrequently Asked Questions about KBase.\\n\\nIs KBase free to use?\\n\\nYes! KBase is completely free to use and is supported by the Department of Energy.\\n\\nI\\'m getting an error. What do I do?\\n\\nYou can check\\n\\nthe job log\\n\\nand refer to our\\n\\ntroubleshooting guide\\n\\n. If you do not see a reason for the error there, please make a ticket on our\\n\\nHelp Board\\n\\nIs it possible to obtain publication-quality figures by using these tools?\\n\\nSome of the Apps can produce publication-quality images that can be views by opening in a separate window. We hope to add more in the future.\\n\\nI have too much data to move.\\n\\nIs there a better way for handling large datasets? For large files or collections of files, we recommend using Globus. We also have a\\n\\nguide for importing using Globus\\n\\nCan I import more than one file into a Narrative at the same time?\\n\\nYes, bulk import allows you to check multiple files in the staging area and import them all by clicking \"Import Selected.\" More instructions on this feature are\\n\\ndescribed in the Narrative Guide\\n\\nHow many users can view a Narrative?\\n\\nThere is no limit to the number of users that can be invited to view or edit Narratives. However, it is possible for users who are accessing the Narrative simultaneously to overwrite each other\\'s changes. For that reason, we recommend only one person edit the Narrative at a time.\\n\\nI work with datasets with many samples, outputs from each step often mixed together and often hard for me to select the ones for the next step. How do I keep them organized?\\n\\nThe best way to organize many inputs is to create sets. There are Apps for building these sets, such as\\n\\nBuild AssemblySet\\n\\nthat allow you to group those outputs.\\n\\nWhat languages are supported in code cells?\\n\\nCurrently we only support code cells in Python.\\n\\nHow does KBase compare with the standalone tools available with command line interface? Does KBase always use the most recent version?\\n\\nProvided the same data, parameters, and tool version are consistent, the results are the same. We make every attempt to use the latest versions of Apps and databases, but they do not always correspond to the most recent version.\\n\\nHow do I access advanced parameters?\\n\\nYou can view some advanced parameters depending on the app by clicking the Advanced Parameters button. For some apps there are few or no advanced parameters. In creating these Apps we aim to balance configurability and usability. As a result, for some Apps we do not make certain rarely changed parameters visible.\\n\\nHow do I get a tool I like added?\\n\\nThere are a couple venues:\\n\\n1. Submit a Help Desk New Feature Request. We can\\'t promise if or when new feature requests are added, but user input helps us plan our future development.\\n\\n2. Add as a 3rd-party developer: If you are a developer of an app, you can find more information on\\n\\nthe developer\\'s page\\n\\nand you can read the documentation for our\\n\\nSoftware Development Kit\\n\\n. If you are not the developer, you can reach out to the developer and ask them about wrapping it for KBase.\\n\\nAlternatively, a beta version of the App might already be in KBase. In the Apps panel, click on the \"R\" button to switch from release to beta. When a \"B\" appears in place of the \"R\", you are viewing beta Apps.\\n\\nWhat pipelines does KBase provide?\\n\\nKBase does not provide pipelines per se; we try to be as flexible as possible. We do have\\n\\nexample Narratives\\n\\nthat replicate the paths many users follow, but we encourage everyone to plot their own course.\\n\\nWill my data be made public?\\n\\nAll user data in KBase is\\n\\nprivate by default until you share it\\n\\n. You can keep all your data private if you wish. We do encourage users to make data public in the interest of FAIR -\\n\\nFindable, Accessible, Interoperable, and Reusable\\n\\ndata practices.\\n\\nWhat organisms are available in KBase?\\n\\nIn KBase as a whole, anything can be uploaded, the only exception being that human data is not permitted. There are connections to some bacterial, plant, and fungal species in the Public Data tab, but any user data or data found from public sources other than those integrated into KBase can be uploaded. On an app-by-app basis, some tools may work sub-optimally or not at all with certain species.\\n\\nIs there an API to allow access to the methods from the command line?\\n\\nCommand line is not supported for the narratives. However, you may find more flexibility with the KBase Software Development Kit (SDK).\\n\\nCan I get an in-person KBase workshop at my institution?\\n\\nIf you would like an in-person workshop, contact us by email at\\n\\n[email\\xa0protected]\\n\\nCan I see the background code for KBase?\\n\\nYes, the relevant Github repo for each App can be found by following the links in the App Catalog. You can access the version and parameters that were used during an App run by clicking on the ellipsis button and clicking \"Show code.\"\\n\\nWhat is the difference between Public and Private organizations?\\n\\nThe existence and contents of Private organizations are only visible to members of the organization. The organization will not be findable by non-members.\\n\\nAny KBase user can see a list of all Public organizations plus a description of the organization. Non-members are given the option to request membership. Members of the organization can access more information than non-members.\\n\\nWhy can\\'t I open a Narrative in an Org?\\n\\nClick on ‘Click for Access’ and you are automatically granted View access to the Narrative. You can then click on the linked Narrative to open.\\n\\nWhat if I lose access to my account?\\n\\nYou can log into KBase through Google, Globus, and ORCiD. It is recommended to always have at least one personal account linked to your KBase account, so that in the event you lose access to an identity supplier i.e., Google, Globus, or ORCiD because of an email provider change, then you will still have another method to login.\\n\\nIf you do lose access completely, please submit a ticket through the\\n\\nHelp Board\\n\\nPreviousOrganizations\\n\\nNextWorking with Data\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_faq.html'}),\n",
       " Document(page_content='Working with Data\\n\\nGuides for working with data in KBase\\n\\nContents\\n\\n1.\\u200bUpload and Download Guide\\u200b\\n\\n2.\\u200bHow to search, add and upload data in KBase video tutorial\\u200b\\n\\n3.\\u200bHow to filter, manage, and view data in KBase video tutorial\\u200b\\n\\n4.\\u200bLinking data and metadata with Samples\\u200b\\n\\n5.\\u200bUsing public data within KBase\\u200b\\n\\n6.\\u200bHow to transfer data using Globus\\u200b\\n\\n7.\\u200bHow to transfer data from JGI\\u200b\\n\\nPreviousFAQs\\n\\nNextData Upload and Download Guide\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data.html'}),\n",
       " Document(page_content='Data Upload and Download Guide\\n\\nThis guide provides instructions for uploading different types of data to your KBase account and downloading output data from your analysis results.\\n\\nA Quick Video Tutorial on Uploading and Downloading Data\\n\\nContents\\n\\n1.\\u200bData Types\\u200b\\n\\n2.\\u200bImporting Data\\u200b\\n\\n3.\\u200bAssemblies\\u200b\\n\\n4.\\u200bGenomes\\u200b\\n\\n5.\\u200bFASTQ/SRA Reads\\u200b\\n\\n6.\\u200bFBA Models\\u200b\\n\\n7.\\u200bMedia\\u200b\\n\\n8.\\u200bExpression Matrix\\u200b\\n\\n9.\\u200bPhenotype Set\\u200b\\n\\n10.\\u200bAmplicon Matrix\\u200b\\n\\n11.\\u200bChemical Abundance Matrix\\u200b\\n\\n12.\\u200bSampleSet\\u200b\\n\\n13.\\u200bCompressed Files\\u200b\\n\\n14.\\u200bDownloading Data\\u200b\\n\\nPreviousWorking with Data\\n\\nNextData Types\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide.html'}),\n",
       " Document(page_content='Data Types\\n\\nData within KBase covers a wide range of data types relevant to systems biology research, including genomes and their annotations, metagenomes, expression, and protein-protein interactions, inferred models of organismal and community metabolism and gene regulation, even geographical information about populations.\\n\\nRemember to follow the KBase\\n\\nData Policy and Sources\\n\\nagreement.\\n\\nData Type Descriptions\\n\\nKBase handles data as objects versus files for interoperability with KBase Apps. The following are data types and their adjacent file types:\\n\\nAssembly ⏤ FASTA files; extensions .fasta, .fna, .fa, .fas\\n\\nFASTQ/SRA Reads ⏤ Interleaved, Non-interleaved, paired-end reads, single-end reads; extensions .fastq, .fq, .sra\\n\\nGenome ⏤ GenBank or GFF3 with a FASTA file; extensions .genbank, .gb, .gbk, or .gbff, .gff, and .fasta, .fna, .fa, .fas\\n\\nMetagenome ⏤ GFF Metagenome with a FASTA file; extensions .gbff, .gff, and .fasta, .fna, .fa, .fas\\n\\nFBA Model ⏤ SBML, Excel, or TSV; extensions .sbml, .xml, .tsv, .xls, .xlsx\\n\\nMedia ⏤ Excel or TSV; .xls, .xlsx, .tsv\\n\\nExpression Matrix ⏤ Excel or TSV; .xls, .xlsx, .tsv\\n\\nPhenotype Set ⏤  Excel or TSV; .xls, .xlsx, .tsv .tab\\n\\nAmplicon Matrix ⏤ Excel and FASTA; .xls, .xlsx, and .fasta, .fna, .fa, .fas\\n\\nChemical Abundance Matrix ⏤ Excel, CSV, or TSV; .xls, .xlsx, .csv, .tsv\\n\\nSampleSet ⏤ Excel or TSV; .xls, .xlsx, .tsv\\n\\nThe\\n\\nData Type Catalog\\n\\nlists and describes the key data types representing different classifications of biological data within KBase.\\n\\nPreviousData Upload and Download Guide\\n\\nNextImporting Data\\n\\nLast modified \\n\\n7mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_data-types.html'}),\n",
       " Document(page_content='Importing Data\\n\\nUsing the Data Browser in the KBase Narrative Interface.\\n\\nUsing the Data Browser\\n\\nThe Import tab lets you drag & drop data files from your computer into the Staging Area to upload data into your Narrative. Visit\\n\\nthis page\\n\\nof the Narrative User Guide for general instructions on how to use the Staging Area.\\n\\nData Panel and Import to Staging Area\\n\\nThe first step in uploading your data is to locate the\\n\\nalong the left side of the Narrative Interface window and click the “Add Data” button, the circular “+” icon, or the arrow at the upper right of the panel to access the slide-out\\n\\n. The Data Browser has several tabs that allow you to access data within KBase and the\\n\\ntab for uploading your own data.\\n\\nTo close the Import panel and return to the Narrative Interface, simply click the “Close” button on the bottom right of the import panel.\\n\\nIf you have a small screen, you might not be able to see that button. Another way to close the Data Browser is to click the arrow icon on the top right of your Data Panel (the same one that opens the Data Browser slide-out).\\n\\nOpening and Closing Data Panel\\n\\nImporting Data\\n\\nThe import type selection dropdown can detect the data type of the file based on its extension where possible. Some types will not automatically select, such as .fasta or .fastq. In these cases more than one importer can use the FASTA file extension. However, suggested importers will be listed at the top of the drop down under “Suggested Types.”\\n\\nLearn how to import multiple files under the\\n\\nBulk Import Guide\\n\\nFor example, files with the extension FASTQ can import as a single interleaved file, as a pair of forward and reverse non-interleaved files to create a PairedEndLibrary, or as a single non-interleaved file to create a SingleEndLibrary.\\n\\nSuggested Types for Importing Data into the Narrative\\n\\nTo import a file, make sure you have an import type selected and the check box is active, then click \"Import Selected.\"\\n\\nDrag & Drop Limitations\\n\\nThe drag & drop from your local computer works for many files, but there is a size limit (about 5GB) dependent on your computer and browser. For larger files around 20GB, use the\\n\\nGlobus Online transfer\\n\\nor a direct upload with a URL.\\n\\nUsing Globus to upload large data files to the Staging Area\\n\\nData Privacy\\n\\nAny data you upload to KBase is private unless you choose to share it. You can share any of your Narratives (including their associated data) with one or more specific users, or make it publicly available to all KBase users. Please see the\\n\\nShare Narratives\\n\\npage for more information and how to share data and Narratives. KBase policies on data are described on the\\n\\nTerms and Conditions\\n\\nand\\n\\nData Policy\\n\\nThe next sections of this guide describe the specific steps involved in uploading the currently supported data types, and show examples for each type.\\n\\nTransferring JGI data to KBase\\n\\nIf you are a JGI user, you can transfer public genome reads and assemblies (including your private data and annotated genomes) from JGI to your KBase account—see\\n\\nthis page\\n\\nfor instructions.\\n\\nBulk Import Guide\\n\\nTo import multiple files, ensure all the files you want to import have an import type selected and the check box is active, then click \"Import Selected.\"\\n\\nBulk import currently supports one parameter set per data type. If you need to upload files of a given data type with different parameters, perform a separate set of bulk imports for each parameter set.\\n\\nYou’ll see a new import cell created for bulk imports. This cell contains a tab for each of the data types you selected from the Staging Area, which you can view by clicking on the available types within the Data type column on the left. This cell is a bulk wrapper for existing import apps, so you can fill out the parameters the same as you would for a single import.\\n\\nBulk Import App in the Narrative\\n\\nOnce you have filled out all file paths and parameters, run the import. While the import is running you can see all the logs and status details in the Job Status tab, formatted for the bulk import.\\n\\nExpand each of the child jobs by clicking on the corresponding line to view details, then expand further to see individual logs. When the bulk import cell is set to run, each child job for the individual data object will have a status, including\\n\\nto\\n\\nor\\n\\nindividual jobs.\\n\\nRunning Bulk Import App\\n\\nAlternatively, queued and running jobs can be cancelled using the bulk action dropdown. Or the retry action can be used for multiple import jobs within the bulk import cell.\\n\\nBulk Action Dropdown\\n\\nWhen the jobs are completed successfully, you’ll be able to see all the information on the data object, repackaged for viewability in bulk. The first new feature is a table of all successful jobs. Clicking on object names adds a viewer widget to the Narrative. All the reports are available under the Reports tree. The full list is collapsed by default to conserve memory. The list can be expanded to show each report, and each report can be viewed in a separate window.\\n\\nResult tab of the Bulk Import Cell\\n\\nWhat data types does this apply to?\\n\\nWhile this will be applicable to all data types and file extensions in the future, bulk import currently supports the following:\\n\\nAssembly - FASTA\\n\\nSRA Reads\\n\\nFASTQ Reads Interleaved - FASTQ Interleaved reads\\n\\nFASTQ Reads Noninterleaved - FASTQ paired-end reads or FASTQ single-end reads library\\n\\nGenBank Genome - GenBank\\n\\nGFF Metagenome - FASTA and GFF3\\n\\nWhen selected files are not supported data types for Bulk Import within the Staging Area, separate, single import app cells will open in the Narrative for each type and file.\\n\\nWhat are the limitations?\\n\\nThis is a new feature in KBase, and the current release should be considered a beta version with future development still to come. Some of the most notable limitations are listed below. This is not a comprehensive list, but does contain the known bugs, issues, and limitations that are the highest priority for future releases. Please report any new bugs you find to the Help Board.\\n\\nView\\n\\ncommon bugs and limitations\\n\\nPreviousData Types\\n\\nNextBulk Import Limitations\\n\\nLast modified \\n\\n1mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_uploads.html'}),\n",
       " Document(page_content='Bulk Import Limitations\\n\\nThis page contains a list of the most commonly observed bugs and known limitations to the bulk import process.\\n\\nKnown Issues and Limitations\\n\\nIssue: Only a single set of parameters can be used for a given type in a given bulk import. Workaround: Subset your imports and run separate imports for each subset of data with a single set of parameters. Note that this only applies to parameters within an import type. For example, you have a set of Illumina reads, PacBio reads, and assemblies, you can import one of the sets of reads and the assemblies in one import, and the other type of reads in a separate import.\\n\\nIssue: Staging Area allows uploading incomplete files, leading to data corruption.Workaround: This is an existing issue in the staging area that exists for single-file imports as well. Ensure files have fully uploaded and the file size in the Staging Area matches the size of the file from your machine. You can use MD5 checksums to verify the file uploaded correctly. You can view the MD5 inside the Staging Area by viewing the info for the file (shown below). Then, verify that the MD5 in the Staging Area matches the MD5 on your local machine using the following command depending on your operating system (replacing \"assembly_1.fasta\" with your file name):MacOS: md5 assembly_1.fasta  Windows: certutil -hashfile assembly_1.fasta md5Linux: md5sum assembly_1.fasta\\n\\nExample for viewing file size information in the Staging Area to ensure the file is completely uploaded\\n\\nIssue: File paths are not autofilled. Workaround: During import pairs of files that go together, such as pairs of non-interleaved FASTQ files or FASTA and GFF Metagenome files, will need to be manually assigned to a forward or reverse read.\\n\\nSelecting import pairs\\n\\nScientific Name input disappears and becomes undefined within the input parameters when switching between tabs. Workaround: This is a visual issue and the selected Scientific Name will still be included when running the import - even though it is not searchable in the undefined state. To verify the selected name is correct when the parameter does switch to the undefined state, 1) search for and select a different scientific name, 2) click between Data type tabs, 3) return to Data type tab and search for and select the original, preferred scientific name, 4) then Run the import.\\n\\nIssue: Job Status/Logs/Results do not appear or do not update.Workaround: There have been infrequent bugs found in which logs failed to appear, jobs reported as successful but no report could be generated, and similar issues. Switching between tabs or reloading the page has been found to fix these issues.\\n\\nPreviousImporting Data\\n\\nNextAssembly\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_uploads_bulk-limitations.html'}),\n",
       " Document(page_content='Assembly\\n\\nHow to upload FASTA files into KBase.\\n\\nAn assembly file is a single file containing one or more contiguous DNA sequences in FASTA format. It can be uploaded to KBase from your local computer (with file extension .fasta, .fna, .fa, or .fas) or directly from a publicly accessible FTP or HTTP URL.\\n\\n“Assembly” is the KBase data type for assembled, unannotated DNA sequence contigs or  predicted coding sequences. If you want to upload annotated sequences in GenBank or GFF format, please see the\\n\\nGenome\\n\\npage.\\n\\nImporting an Assembly from your computer\\n\\nUsing a file on your computer, open the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the assembly file into the Staging Area box or select from your computer files.\\n\\nOpen the Import As pulldown menu to the right of the filename in your Staging Area and select “Assembly.”\\n\\nImport Fasta Assembly File\\n\\nMake sure the correct file type is selected and the checkbox is active, then click the “Import Selected” button. The data slide-out will close and an app called “Import FASTA File as Assembly from Staging Area” will be added to your Narrative.\\n\\nImport Assembly / FASTA file Import App Cell\\n\\nThe name of the Assembly file is filled in, as is a suggested name for the Assembly data object that will be created by the import (you can change the Assembly object name). Adjust the minimum contig length if needed and then click the green \"Run\" button to start the import. When the import is finished, your Data Panel will update to show the new Assembly object, and a report will appear in the Import App.\\n\\nUpload an Assembly from other sources\\n\\nYou can upload data into your KBase Staging Area using\\n\\nGlobus\\n\\n, or by supplying a URL for a publicly accessible FTP location, Google Drive, Dropbox, or a direct HTTP link to import into the Narrative. Options for adding data to your Staging Area are described\\n\\nhere\\n\\nDrag and Drop, Upload with Globus, Upload with URL options to upload and import data into KBase\\n\\nDrag & Drop Limitations\\n\\nThe drag & drop option from your local computer works for many files, but there is a size limit that depends on your computer and browser. For larger files (around 20GB), use\\n\\nGlobus Online transfer\\n\\nBulk Import\\n\\nAssemblies can be imported as one of the supported bulk import types. You can select multiple assemblies simultaneously from the staging area to import them at once. See the bulk import section of\\n\\nthe guide to importing data into the Narrative.\\n\\nTransfer assemblies from JGI\\n\\nIf you are a JGI user, you can transfer public genome reads and assemblies (as well as your private data and annotated genomes) from JGI to your KBase account—see\\n\\nthis page\\n\\nfor instructions.\\n\\nPreviousBulk Import Limitations\\n\\nNextGenome\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_assembly.html'}),\n",
       " Document(page_content='Genome\\n\\nFormatting and uploading annotated assemblies and GenBank or GFF and FASTA files.\\n\\nIn KBase, a Genome object is the annotated version of an\\n\\nAssembly\\n\\nor annotated predicted coding sequences and can encompass several types of feature calls. If you want to upload solely the DNA sequence from a FASTA file (without annotations), go to the\\n\\nAssembly\\n\\npage.\\n\\nThe Genome importer supports only GenBank and GFF formatted files.\\n\\nA GenBank-formatted input file should include sequence contig(s), feature calls (annotations), and taxonomy information for the organism. KBase parses the input file into two data objects: an assembly object with the sequence and a genome object containing the original feature calls and annotations.\\n\\nGenBank-formatted files with no features can be uploaded as Genomes.\\n\\nA GenBank-formatted file can be uploaded into the Staging Area from your local computer (files with.gb, .gbff, or .gbk extensions) or directly from an FTP or HTTP URL.\\n\\nA GFF-formatted file must be paired with a corresponding FASTA file of the DNA sequence. These  will be parsed into two data objects: an assembly object with the sequence and a genome object containing the original feature calls and annotations.\\n\\nFurther instructions for adding data to your Staging Area can be found\\n\\nhere\\n\\nImporting a GenBank-formatted genome\\n\\nUsing a file on your computer, open the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the genome file into your Staging Area.\\n\\nOpen the Import As... pulldown menu to the right of the filename in your Staging Area and select “GenBank Genome.”\\n\\nMake sure the correct file type is selected and the checkbox is active, then click the “Import Selected” button.\\n\\nImport GenBank Genome\\n\\nThe Data Browser will close and the “Import GenBank File as Genome from Staging Area” App will be added to your Narrative.\\n\\nNotice that the name of the Genome file is filled in, as is a suggested name for the Genome and Assembly data objects that will be created by the import, which can be changed. Adjust the Genome Type and source of the GenBank file or any advanced parameters if needed.\\n\\nImport GenBank Genome App\\n\\nClick the green \"Run\" button to start the import. When the import is finished, your Data Panel will update to show the new Genome and Assembly objects, and a report will appear in the Import App.\\n\\nImporting a GFF-formatted Genome\\n\\nOpen the\\n\\nImport tab in the Data Browser\\n\\nand drag and drop\\n\\nthe genome\\n\\ncorresponding FASTA file into your Staging Area. In your Staging Area, open the\\n\\npulldown menu to the right of the GFF filename and select “GFF Genome.\"\\n\\nNote the name of the corresponding FASTA file.\\n\\nMake sure the correct file type is selected and the checkbox is active, then click the \"Import Selected\" button. The data slide-out will close and the “Import GFF/FASTA File as Genome from Staging Area” App will be added to your Narrative. The GFF File Path name will be filled in.\\n\\nYou will need to fill in the name of the FASTA file. Using the dropdown for the “FASTA File Path”, select the FASTA file in the Staging Area. Ensure the file type is a FASTA file type.\\n\\nGFF/FASTA Genome Import\\n\\nThe name of the Scientific Name may be filled in, as is a suggested name for the Genome data object that will be created by the import. You can edit the name of the output Genome Object Name, Scientific Name, and any advanced options as needed. Click the green \"Run\" button. When the import is finished, your Data Panel will update to show the new Genome object, and a report will appear in the Import App.\\n\\nUploading a Genome from other sources\\n\\nYou can upload data into your KBase Staging Area using\\n\\nGlobus\\n\\n, or by supplying a URL for a publicly accessible FTP location, Google Drive, Dropbox, or a direct HTTP link to import into the Narrative. Options for adding data to your Staging Area are described\\n\\nhere\\n\\nDrag and Drop, Upload with Globus, Upload with URL options to upload and import data into KBase\\n\\nDrag & Drop Limitations\\n\\nThe drag & drop option from your local computer works for many files, but there is a size limit that depends on your computer and browser. For larger files (around 20 gigabases), use the\\n\\nGlobus Online transfer\\n\\nBulk Import\\n\\nBoth GenBank and GFF genomes can be imported as one of the supported bulk import types. You can select multiple assemblies simultaneously from the staging area to import them at once. See the bulk import section of\\n\\nthe guide to importing data into the Narrative.\\n\\nUploading a Genome using a URL\\n\\nOpen the\\n\\nImport tab in the Data Browser\\n\\nand click on the Upload with URL button (below the drag & drop area) to open an Upload App Cell.\\n\\nThe Data Browser will close and the “Upload a File to Staging from Web” App will appear in your Narrative. Alternatively, you can open the app directly from the Apps Panel. From the app, click on the dropdown for the URL Type and select the URL type.\\n\\nUpload File to Staging from Web App\\n\\nWhen uploading a GenBank Genome, you will only need to use one link. When uploading a GFF Genome, you will need to use two links for the GFF file and the FASTA file.\\n\\nIn the App, click the \"+\" button for the URLs and paste in the name of the Genome file (GenBank or GFF). Hit the \"+\" button again and paste in the name of the FASTA file.\\n\\nThen click the green \"Run\" button to start the upload. After the App completes the files will appear in your Staging Area, which you can access via the Import tab in the Data Browser.\\n\\nThe genome file(s) are now in your Staging Area. Now you need to import them to your Narrative to use them in analyses.\\n\\nPreviousAssembly\\n\\nNextFASTQ/SRA Reads\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_genome.html'}),\n",
       " Document(page_content='FASTQ/SRA Reads\\n\\nFormatting and uploading FASTQ and SRA reads files.\\n\\nIn KBase, reads from FASTQ and SRA files can be imported to create reads library data objects. The objects will either be a SingleEndLibrary or a PairedEndLibrary. The tools in KBase can then be used to assemble reads into an “Assembly” data object or to align reads to an “Assembly.\" After uploading and importing reads data, you may want to refer to the documentation about\\n\\nAssembly and Annotation\\n\\n. Reads can also be used in\\n\\nRNA-seq and expression analysis\\n\\nSingle-end and paired-end reads can be uploaded in FASTQ or SRA format. For FASTQ files, please ensure that your filename ends with the .fastq, .fnq, or .fq file extension. SRA files should have the .sra file extension. The uploader also accepts compressed files in .zip, .gz, .bz2, .tar.gz, .tar.bz2 formats.\\n\\nFiles can be uploaded into your KBase Staging Area from your local computer or directly from a publicly accessible FTP or HTTP URL.\\n\\nImporting reads from your computer\\n\\nSingle-end library in FASTQ or SRA format\\n\\nUsing a file on your computer, open the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the single-end library into your Staging Area. Open the pulldown menu to the right of the filename in the Staging Area and select “FASTQ Reads NonInterleaved\" or \"SRA Reads.\"\\n\\nFASTQ Single-end Library\\n\\nSRA Single End Library\\n\\nMake sure the correct file type is selected and the checkbox is active, then click \"Import Selected\". The slide-out Data Browser will close and an app called “Import FASTQ/SRA File as Reads from Staging Area” will be added to your Narrative.\\n\\nNotice that the name of the Reads Library file is already filled in, as is a suggested Reads Object Name that will be created by the import (you can change that if you like). Adjust the Sequencing Technology and any of the advanced options if needed. Note that this was a metagenomic sample, we would uncheck the box next to Single Genome. When ready, click the green \"Run\" button to start the import. When the import is finished, your Data Panel will update to show the new SingleEndLibrary object, and a report will appear in the import app cell.\\n\\nPaired-end library in FASTQ format\\n\\nThere are two ways that KBase and\\n\\nGenBank SRA\\n\\nrecognize a paired-end library. A paired-end library can be either two files which typically have the same name and are designated as forward and reverse or a single interleaved file. Interleaved files use an 8-line format where forward and reverse reads alternate.\\n\\nOpen the\\n\\nImport tab in the Data Browser\\n\\nand drag either one interleaved file or two paired files into the Staging Area.\\n\\nOpen the pulldown menu to the right of the filename under the Import As... column in your Staging Area and select “FASTQ Reads NonInterleaved” for the first file in the pair. Make sure the correct file type is selected and the checkbox is active, then click \"Import Selected\". The Data Browser slide-out will close and the “Import FASTQ/SRA File as Reads from Staging Area” App will open.\\n\\nNotice that the name of the FASTA/FASTQ file is filled in. A suggested Reads Object Name is also created by the import and can be changed.\\n\\nIf the files are two paired file, you will need to select a file for “Reverse/Right FASTA/FASTQ File Path” from the dropdown.\\n\\nPaired-end or non-interleaved FASTQ Import\\n\\nAs with the single-end library example, you can make adjustments to the available parameters. Adjust the Sequencing Technology and any of the advanced options if needed. If the file is an interleaved paired-end library, check the box to the right of Interleaved.\\n\\nInterleaved FASTQ Import\\n\\nWhen ready, click the green \"Run\" button to start the import. When the import is finished, your Data Panel will update to show the new PairedEndLibrary object, and a report will appear in the Import App.\\n\\nImporting a reads library from other sources\\n\\nIn the Staging Area, beneath the box for Drag and Drop, there are other options for adding data to your staging area. You can import reads into KBase using\\n\\nGlobus Online\\n\\n, or by supplying a URL for a publicly accessible FTP location, Google Drive, Dropbox, or a direct HTTP link.\\n\\nIf your reads are in a publicly accessible URL, you can bypass the Staging Area and directly import reads into your Narrative using one of these three apps (which you can find in the Apps panel or the\\n\\nApp Catalog\\n\\n):\\n\\n\\u200bImport SRA File as Reads From Web\\u200b\\n\\n\\u200bImport Single-End Reads From Web\\u200b\\n\\n\\u200bImport Paired-End Reads From Web\\u200b\\n\\nNote that directly importing these files into KBase from the web behaves the essentially the same as uploading to the Staging Area and then importing to a Narrative, except the transfer is carried out by the Importer App. Use the link available to directly download the file as if you were going to save it to your computer.\\n\\nFor SRA reads from NCBI, this generally means navigating to the Data access tab of the Run Browser in the NCBI Sequence Read Archive (SRA) for the reads to import, as seen here:\\n\\nFor this example, SRR18272216, navigate to the Data access tab and copy the SRA-download link under Name and paste the full link into the Import from Web App.\\n\\nCopy the SRA-download link located under the Name heading and paste the link into the URL input within the Importer App.\\n\\nFor how to search and download or locate download links for SRA sequences,\\n\\nsee the NCBI Search and Download\\n\\ndocumentation.\\n\\nDrag & Drop Limitations\\n\\nThe drag & drop from your local computer works for many files, but there is a size limit that depends on your computer and browser. Some users have reported problems around 20GB. For larger files, use\\n\\nGlobus Online transfer\\n\\nBulk Import\\n\\nFASTQ and SRA reads can be imported as one of the supported bulk import types. You can select multiple assemblies simultaneously from the staging area to import them at once. See the bulk import section of\\n\\nthe guide to importing data into the Narrative.\\n\\nTransfer reads from JGI\\n\\nIf you are a JGI user, you can transfer public genome reads and assemblies (as well as your private data and annotated genomes) from JGI to your KBase account—see\\n\\nthis page\\n\\nfor instructions.\\n\\nPreviousGenome\\n\\nNextFlux Balance Analysis (FBA) Model\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_reads.html'}),\n",
       " Document(page_content='Flux Balance Analysis (FBA) Model\\n\\nData handling to create a model to predict biomass yield given a certain amount of input nutrient.\\n\\nFlux Balance Analysis (FBA) models can be uploaded into KBase as a Systems Biology Markup Language (SBML) file using the .sbml or .xml file extension, an Excel file using the .xls extension, or a tab-separated values (TSV) file using the .tsv extension.\\n\\nWhen uploading an FBA model in the TSV format, there will be two files corresponding to chemical compounds and reactions.\\n\\nFormatting Flux Balance Analysis Models\\n\\nSBML\\n\\nMore SBML FBA Models for various organisms can be found here:\\n\\nhttp://systemsbiology.ucsd.edu/Downloads\\n\\n.\\nFor a description of how to create a COBRA-compliant SBML file that can be imported as an FBA model into KBase, please see\\n\\nSBML Level Three Specifications\\n\\nExcel\\n\\nIn the Excel format, the tables containing the information about the chemical compounds and reactions in an FBA model are stored in two separate tabs respectively. The most important requirement for the Excel file is naming the tabs “ModelCompounds” and “ModelReactions.”\\n\\nTSV\\n\\nFor the Tab-Separated Values file format, the chemical compounds and reactions tables are saved as two separate files named “FBAModelCompounds.tsv” and “FBAModelReactions.tsv” that will be transformed\\n\\nFBAModelCompounds\\n\\nid name formula charge aliases  cpd00113  Isopentenyldiphosphate  C5H10O7P2  -2 \\u200b  cpd02590  all-trans-Heptaprenyl diphosphate  C35H58O7P2  -2 \\u200b\\n\\nid: Compound identifier; see the KBase Biochemistry reference for a list of compounds [link]\\n\\nname: Name of chemical compound\\n\\nformula: Chemical formula using the Hill system\\n\\ncharge: Formal charge of the molecule\\n\\naliases: Alternative names for chemical compound\\n\\nFBAModelReactions\\n\\nImporting an FBA model into KBase requires specified reaction IDs for the biomass producing equations in the model. An individual organism FBA model needs a single biomass equation, while a community model may have multiple biomass producing equations. In the FBAModelReactions table, denote the biomass producing reaction of an individual organism, or the first in a community model, with an ID such as “bio1” so the reaction is easily identifiable. For subsequent biomass producing reactions in a community model, it is recommended that you use “bio2” for the second biomass producing reaction, “bio3” for the third, and so on.\\n\\nScroll table below from left to right to see complete format\\n\\nid direction compartment gpr name enzyme pathway reference equation rxn00001_c0 -> c0 fig|211586.9.peg.3751 Pyrophosphate phosphohydrolase_c0 \\u200b \\u200b \\u200b (1) cpd00001[c0] + (1) cpd00012[c0] -> (2) cpd00067[c0] + (2) cpd00009[c0] rxn00056_c0 <-> c0 fig|211586.9.peg.1038 Fe(II):oxygen oxidoreductase_c0 \\u200b \\u200b \\u200b (4) cpd00067[c0] + (4) cpd10515[c0] + (1) cpd00007[c0] <-> (4) cpd10516[c0] + (2) cpd00001[c0]\\n\\nid: Reaction identifier; see the KBase Biochemistry reference for a list of reactions\\u200b\\n\\ndirection: Directionality of chemical reaction specified as forward (->), reversed (<-), or equilibrium (<->)\\n\\ncompartment: Cellular compartment that the reaction takes place in; see the Metabolic Modeling FAQ for more information\\u200b\\n\\ngpr: Gene-Protein Reaction in PATRIC identifier format\\n\\nname: Protein associated with the reaction\\n\\nenzyme: Optional, can be left empty\\n\\npathway: Optional, can be left empty\\n\\nreference: Optional, can be left empty\\n\\nequation: Chemical reaction equation specified with compound coefficients in parentheses, compound identifiers, and directional arrows\\n\\nImporting FBA Models\\n\\nIn order to successfully upload an FBA model into your KBase workspace, you first need to\\n\\nadd the Genome\\n\\nthat corresponds to referenced in the FBA Model you wish to upload. Add the Genome to the Narrative before importing the FBA model file.\\n\\nUsing a file on your computer, open the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the FBA model file (plus compound file if using TSV format) into your Staging Area. Open the pulldown menu to the right of the filename in the Staging Area and select “FBA Model.\"\\n\\nFBA Model selection from the Staging Area\\n\\nNow click the import icon (up arrow) to the right of “FBA Model.\" The Data Browser slide-out will close and an app called “Import TSV/XLS/SBML File as an FBAModel from Staging Area” will be added to your Narrative.\\n\\nNotice that the Model file path is filled in, as is a suggested name for the FBA Model data object that will be created by the import (you can edit this).\\n\\nImport TSV/XLS/SBML File as an FBAModel from Staging Area App\\n\\nAt this point, the corresponding Genome needs to be linked to the FBA Model. Add the name of the Genome using the dropdown to select the Genome that you added to the Data Panel. Although specifying the Genome is optional, adding a Genome with gene IDs that match genes IDs in the model is required to import gene-protein-reactions (GPRs). GPRs are necessary to perform gene KO in the Run Flux Balance Analysis App.\\n\\nEnsure the correct Model file type is selected (SBML, Excel, Tab-separated values).\\n\\nClick the \"+\" button to the right of Biomass and type in the name of the first biomass. If there are additional biomass values to enter, click the \"+\" button to open another parameter field for another biomass. Note: If the biomass reaction name starts with “R_”, do NOT include the “R_” when entering the name in this field.\\n\\nIf importing a TSV-formatted model, select the corresponding Compounds file path from the dropdown.\\n\\nEdit the FBA Model object name for the output if necessary. Click the green \"Run\" button to start the import.\\n\\nWhen the import is finished, your Data Panel will update to show the new FBA Model object, and a report will appear in the Import App.\\n\\nDrag & Drop Limitations\\n\\nDrag & drop upload from your local computer works for many files, but there is a size limit  dependent on your computer and browser. For larger files around 20GB or more, use\\n\\nGlobus Online transfer\\n\\nAssociating an FBA Model with Genome\\n\\nFor KBase Metabolic Modeling Apps, associating a genome with user uploaded models is necessary for ensuring fidelity between the genome annotations and gene-protein-reactions in the metabolic model. This means that the gene IDs between the model and genome need to match. Occasionally, researchers will use locus tags or gene IDs for constructing metabolic models that are different than those found commonly in reference genomes. If the gene IDs do not match, you will need to change either the model or the genome to ensure their gene IDs match.\\n\\nOne way to accomplish this is to locate the genome used as the reference for building the model – based on shared gene IDs – and to upload this genome into KBase before uploading the model and associating it with the genome. If you cannot locate the genome with the gene IDs used to build the model, the best option for reconciling the gene IDs is to edit the model file (as SBML or TSV) to change the gene IDs contained therein to match the gene IDs of the genome.\\n\\nThis\\n\\npublic Help Board ticket\\n\\nhas a detailed example of how to accomplish this.\\n\\nPreviousFASTQ/SRA Reads\\n\\nNextMedia\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_fba-model.html'}),\n",
       " Document(page_content='Media\\n\\nHow to format reaction media files for KBase to use in FBA and modeling analysis.\\n\\nKBase offers several options for defining specific growth or culture media for metabolic modeling. Within the Narrative Interface, you can use the\\n\\nEdit Media\\n\\ntool to build media based on the hundreds of conditions available in KBase or import your own custom media as a TSV or Excel file using the instructions below.\\n\\nFormatting Media TSV or Excel files\\n\\nIf you choose to upload a custom media or use media from an external source, please ensure that it is formatted properly for use with KBase.\\n\\nMedia can be uploaded as a TSV (tab-separated values) or Excel file using four columns:\\n\\nMaxflux (maximum allowed uptake/excretion of compound)\\n\\nMinflux (minimum allowed uptake/excretion of compound)\\n\\nConcentration (concentration of compound in mol/L)\\n\\nCompound identifier (e.g., ModelSEED ID, KEGG ID, PubChem ID, or compound name such as “glucose”)\\n\\nA list of all the biochemical compounds is available in the\\n\\nKBase ModelSEED Biochemistry Database\\n\\nfor a reference list of reactions and compounds or use the\\n\\nBiochemistry Search\\n\\n. From the list you can find the IDs for compounds to include in custom media.\\n\\nThe concentration field is not used by any of the the metabolic modeling tools in KBase. It is best to account for these values accurately so that your media files will retain their utility pending updates.\\n\\nThe minflux and maxflux columns connect media to a flux balance analysis (FBA) metabolic model. The values in these columns are absolute units representing the range of possible fluxes for reactions that transport the media compounds into and out of the cell. A negative flux value corresponds to excretion (transport out of the cell), and a positive flux value corresponds to uptake (transport into the cell).\\n\\nThe combination of maximum and minimum flux will dictate whether uptake and/or excretion of a reagent is allowed. Assigning a positive number to minflux for a particular compound essentially forces uptake of that compound. Assigning a negative number to maxflux will force excretion of the compound.\\n\\nMinimum and maximum flux are measured in mmol per gram cell dry weight per hour.\\n\\nNote: When creating a media Excel file, the name of the worksheet that contains your media conditions must be named “MediaCompounds” or it will not upload. By default, the first sheet of any Excel file is called “Sheet1.” To rename it, find the worksheet tabs at the bottom left of the Excel file. Double-click on the default name and type in “MediaCompounds.”\\n\\nBelow is an example of a media file in TSV format. In this case, the compound IDs in the first column are from the\\n\\nKBase ModelSEED Biochemistry Database\\n\\n(scroll to the right to see full table):\\n\\ncompounds name formula minflux maxflux concentration cpd00149 Co2+ Co -100 100 0.001 cpd00099 Cl- Cl -100 100 0.001 cpd00063 Ca2+ Ca -100 100 0.001 cpd00007 O2 O2 -100 10 0.001 cpd00027 D-Glucose C6H12O6 -100 5 0.001\\n\\nImport media from a TSV or Excel file\\n\\nFor this example, you can click on either link to download a small media file onto your computer:\\n\\nGlucose_minimal.xls\\n\\n6KB\\n\\nBinary\\n\\nExample Excel file\\n\\nGlucose_minimal.tsv\\n\\n706B\\n\\nText\\n\\nExample TSV file\\n\\nUsing a file on your computer, open the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the media file to upload it to your Staging Area. Choose \"Media\" from the Import As... data type dropdown menu in your Staging Area. Click the upload icon next to the field that now says “Media.”\\n\\nSelecting Media for import from the Staging Area\\n\\nThe “Import Media file (TSV/Excel) from Staging Area” App will be added to your Narrative.\\n\\nImport Media file App\\n\\nYou can change the name that will be assigned to the Media Object. Then click the green \"Run\" button. After the import process has completed, the Media data object will appear in your Data Panel.\\n\\nPreviousFlux Balance Analysis (FBA) Model\\n\\nNextExpression Matrix\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_media.html'}),\n",
       " Document(page_content='Expression Matrix\\n\\nThe Expression Matrix data type contains gene expression values taken under given sampling conditions.\\n\\nExpression Matrices can be uploaded and imported using the Tab-separated values (TSV) file extension .tsv or the Excel file extension .xls.\\n\\nFormatting Expression Matrix files\\n\\nIf you are importing expression data from an external source or want to populate a file with your own data, please ensure that it is formatted properly for use with KBase.\\n\\nTSV\\n\\nA tab-separated values (TSV) file is a tab delimited text file that has genes across the rows and sample observations across the columns. Make sure the first label in the first column is “feature_ids” followed by tab-delimited labels for samples.\\n\\nExcel\\n\\nIn Excel, the first column matches Feature IDs from the genome. This column could have also been a gene alias. Gene aliases supported by KBase include NCBI, EMBL, UniProt, BioCyc, and ASAP. The column headings are sample conditions.\\n\\nThe remaining cells in the table contain expression values for the appropriate gene and sample. Be sure to exclude gene features that are missing all expressions or are composed of non-changing expressions across the samples.\\n\\nBelow is an example of a properly formatted expression data file in TSV format. In this case, the gene-ids in the first correspond to gene identifiers for E. coli K-12 MG1655 genes and the sample conditions are derived from the\\n\\nMany Microbe Microarrays Database (M3D)\\n\\nfeature_ids dinI_U_N0025_r1 dinI_U_N0025_r2 dinI_U_N0025_r3 b4634 9.05367 9.07827 9.10114 b3241 7.20924 7.08695 7.07071 b3240 7.21535 7.14312 7.19478\\n\\nTo see an example Expression Matrix format,\\n\\ndownload\\n\\nthe \"Shewanella_MR-1_M3D_ExpData\" ExpressionMatrix data set from the Example tab in the Data Panel. Unzip the downloaded file and examine the \"matrix.tsv\" file.\\n\\nPlant Expression Data\\n\\nFor KBase plant genomes, the gene IDs retain the data structure from the external source databases (Ensembl or Phytozome) and do not have aliases as mentioned above. When constructing an expression dataset, append gene IDs with the transcript IDs followed by “.CDS.\" You can check that you have the correct gene IDs\\n\\nUpload and Import an Expression Matrix\\n\\nExpression Matrices can be uploaded into KBase without a Genome, but they will have limited use. In order to successfully upload an Expression Matrix into KBase, you first need to add the\\n\\nGenome\\n\\nthat corresponds to references in the Expression Matrix.\\n\\nUsing a file on your computer, open the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the expression matrix file into the Staging Area.\\n\\nNow that the expression matrix is in your Staging Area and you can import it from there into your Narrative. Open the pulldown menu to the right of the filename under the Import As... column in your Staging Area and select “Expression Matrix.\"\\n\\nClick the import icon (up arrow) to the right of “Expression Matrix.” The Data Browser will close and the “Import TSV File as Expression Matrix From Staging Area” App will be added to your Narrative.\\n\\nThe name of the Tab-delimited (TSV) file is already filled in, as is a suggested name for the Expression Matrix data object that will be created by the import, which can be edited or changed.\\n\\nImport Expression Matrix App\\n\\nAt this point, the corresponding genome is optional and hasn’t been linked to the expression matrix. To add the name of the Genome, click on the ‘show advanced’ link to the right of ‘Input Objects’ in the Import App.\\n\\nImport Expression Matrix App showing advanced parameters\\n\\nUse the Genome dropdown to select the corresponding genome. Adjust any of the other advanced options if needed, then click the green \"Run\" button to start the import. When the import is finished, your Data Panel will update to show the new Expression Matrix object, and a report will appear in the Import App.\\n\\nExpression Matrix Object\\n\\nEach gene measured in the expression dataset should have an identifier listed in the first column of the TSV file. To ensure that the gene identifiers listed in your dataset correspond to the aliases in KBase, click the name of the Genome object to open up the genome viewer.\\n\\nClick on the tab labeled Browse Features and locate a gene of interest by searching for the name of the function or protein associated with the gene.\\n\\nClick the Feature ID of the gene of interest to open a new tab with additional information about the gene.\\n\\nLocate the section titled Aliases and crosscheck the gene labels contained within your expression dataset with either the Feature ID or one of these aliases to ensure that these labels will correspond to features in KBase.\\n\\nDrag & Drop Limitations\\n\\nThe drag & drop from your local computer works for many files, but there is a size limit that depends on your computer and browser. For larger files around 20GB, use the\\n\\nGlobus Online transfer\\n\\nPreviousMedia\\n\\nNextPhenotype Set\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_expression-matrix.html'}),\n",
       " Document(page_content='Phenotype Set\\n\\nThe Phenotype Set data type represents experimental data about an organism’s ability to grow in specific media conditions, recorded as growth or no growth.\\n\\nFormatting Phenotype files\\n\\nA Phenotype Set can be uploaded from a TSV (tab-separated values) file with a .tsv or .tab file extension, or from Excel spreadsheet with a .xls extension. The spreadsheet needs to have exactly these five columns:\\n\\nGene knockout (geneko) – List of genes knocked out in the phenotype; use \\'none\\' for wild-type phenotypes. Gene IDs should be in the same format that appears in your metabolic model (e.g., kb|g.220339.CDS.2927)\\n\\nWorkspace information (mediaws) – Workspace where the media for the phenotype data was loaded into KBase. The workspace information can be found in the Narratives tab by clicking Show Advanced Controls at the end of the list of Narratives\\n\\nMedia – ID of the media condition loaded in KBase where the phenotype was observed.\\n\\nAdditional Compounds (addtlCpd) – Additional media compound IDs to be added alongside the primary media formulation. See the KBase ModelSEED Biochemistry Database for a reference list of reactions and compounds or use Biochemistry Search.\\n\\nGrowth – Indication of whether or not the organism grew in the specified media with the specified knockouts. 1 meaning growth; 0 meaning no growth.\\n\\ngeneko mediaws media addtlCpd growth SO_0009 KBaseMedia C-D-Glucose \\u200b 0 SO_0009 KBaseMedia C-D-Lactate \\u200b 1 SO_0009 KBaseMedia C-acetate \\u200b 1\\n\\nTo download an example phenotype set data table, go to the \"Example\" tab in the data Data Browser slide-out, add a Phenotype Data Object to the Narrative and download the TSV file.\\n\\nExample Data Objects\\n\\nImporting a Phenotype Set\\n\\nFirst\\n\\nadd the Genome\\n\\nthat corresponds to referenced in the Phenotype Set. Add the Genome object to the Narrative before importing the Phenotype Set file.\\n\\nUsing a file on your computer, open the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the phenotype set file into your Staging Area. Open the pulldown menu to the right of the filename in the Staging Area and select “Phenotype Set.\"\\n\\nImporting a Phenotype Set from the Staging Area\\n\\nClick the import icon (up arrow) to the right of “Phenotype Set”. The data slide-out will close and the “Import TSV File as Phenotype Set From Staging Area” App will be added to your Narrative.\\n\\nNotice that the name of the Phenotype TSV file is already filled in, as is a suggested name for the Phenotype Set data object that will be created by the import, which you can edit.\\n\\nImport Phenotype Set App\\n\\nAdd the name of the Genome, click on the dropdown to select the corresponding genome that you added to your Narrative earlier.\\n\\nClick the green \"Run\" button to start the import. When the import is finished, your Data Panel will update to show the new Phenotype Set object, and a report will appear in the Import App.\\n\\nDrag & Drop Limitations\\n\\nThe drag & drop from your local computer works for many files, but there is a size limit that depends on your computer and browser. For larger files around 20GB, use the\\n\\nGlobus Online transfer\\n\\nPreviousExpression Matrix\\n\\nNextAmplicon Matrix\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_phenotype-set.html'}),\n",
       " Document(page_content='Amplicon Matrix\\n\\nGenerating and handling Amplicon Matrices to link to Samples\\n\\nFormatting Amplicon Matrix files\\n\\nAny approach to creating an amplicon matrix should be usable in KBase, but may require additional curation. Amplicon matrices or taxonomic abundance matrices include OTU (operational taxonomic units) and ASV (amplicon sequence variants).\\n\\nWe recommend to upload raw counts as analysis (\\n\\nrarefaction, standardization, and normalization\\n\\n) will be tracked and provenance maintained on system. There are a number of\\n\\ntools in KBase to analyze Amplicon Matrices and metadata\\n\\nAmplicon matrices can be uploaded from a TSV (tab-separated values) file with a .tsv file extension.\\n\\nWhen uploading, ensure that\\n\\nand\\n\\n. A\\n\\nConsensus sequences must also be included as a FASTA file with extension .fa, .fasta. Sequence names must match exactly with the taxa names in the amplicon matrix.\\n\\nTwo files are needed for uploading amplicon data:\\n\\nFASTA file with consensus sequences\\n\\nTaxonomic abundance matrix (OTU or ASV) row = taxoncolumn = sampleoptional column = taxonomy\\n\\nAdditional information to have on hand for parameters include:\\n\\nSample processing metadata (SampleSet), i.e. sequencing technology or platform, target gene or region\\n\\nBioinformatic processing metadata, i.e. clustering method, quality filtering steps\\n\\nUploading and Importing\\n\\nAmplicon Matrices can be uploaded into KBase without linking metadata. For full functionality of analysis tools using metadata, first\\n\\nupload and import the SampleSet\\n\\nUsing a file on your computer, open the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the amplicon matrix file and the FASTA file with consensus sequences into the Staging Area.\\n\\nOnce the files are in your Staging Area, you can import the data into your Narrative.\\n\\nGo to the APPS panel and open the Upload category to locate the \"Import Amplicon Matrix from TSV/FASTA File in Staging Area\" App. Click on the Import Amplicon App to add it to your Narrative.\\n\\nImport Amplicon Matrix from TSV/FASTA File in Staging Area. Double click to view.\\n\\nIn the first section for Input Objects, select the previously imported SamplesSet file from the dropdown menu for linking metadata.\\n\\nUnder Parameters, select the Amplicon Matrix or Taxonomic Abundance TSV file using the dropdown and then the corresponding sequences FASTA file.\\n\\nParameters that require inputs are the target gene, target subfragment, taxon calling method, i.e. denoising or clustering, and sequence error cutoff.\\n\\nAdditional processing metadata, i.e. primer sequences, library kit, sequencing center, the denoise or clustering methods can be input by clicking \"show advanced\" parameters. While these parameters are not required, they are recommended to enhance workflow documentation and reproducibility.\\n\\nFill in the Amplicon Matrix Object Name and click the \"Run\" button.\\n\\nOnce your amplicon matrix is imported into the Narrative, you can use the \"View Matrix as Table\" App to view your new AmpliconMatrix object.\\n\\namplicon_example.zip\\n\\n4KB\\n\\nBinary\\n\\nExample of test amplicon matrix with consensus FASTA\\n\\nThe amplicon matrix may be applied to an existing\\n\\nSampleSet\\n\\nand\\n\\nAttribute Mapping\\n\\non-system to facilitate the interoperability of objects in KBase.\\n\\nTutorial Webinar\\n\\nImporting Samples and Amplicon\\n\\nPreviousPhenotype Set\\n\\nNextChemical Abundance Matrix\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_amplicon-matrix.html'}),\n",
       " Document(page_content='Chemical Abundance Matrix\\n\\nMetabolomics, exometabolite, and chemical abundance data can be integrated with metabolic modeling and flux balance analysis tools in KBase.\\n\\nWhat is \"chemical abundance\" in KBase?\\n\\nThe name of this data type “chemical abundance” is a broad term that we use to represent a wide array of measurements associated with chemicals. This data type can be used to upload and store diverse types of chemical data in the system such as metabolomics (intracellular and/or extracellular) that is derived based on microbiomes/isolate organism growth experiments etc., computationally predicted compounds, or data collected on the concentration or from elemental analysis. These data could be collected on environmental samples, such as soil, sediment, or water. Currently, the metabolomics data derived from the samples are the most popular data that is uploaded and stored in the system.\\n\\nOnce chemical abundance data matrices are\\n\\nuploaded\\n\\n, they can be analyzed using\\n\\nKBase Apps for metabolomics\\n\\n, such as Escher mapping. Additional statistical analysis of the chemical abundance attribute maps, such as PCA and clustering, can also be performed.\\n\\nA Chemical Abundance Matrix can be uploaded from a TSV (tab-separated values) file with a .tsv or .tab file extension, or from Excel spreadsheet with a .xls extension.\\n\\nEach Chemical Type can be either a specific compound or element, aggregate (totals), exometabolites (measurements of compounds or elements that are consumed or excreted into the medium).\\n\\nChemical Abundance Upload Webinar\\n\\nFormatting chemical abundance matrices\\n\\nThe\\n\\nCreate Chemical Abundance Matrix Template App\\n\\ncreates an Excel spreadsheet for direct download that can be populated with chemical abundance data. While Chemical abundance data works best and more meaningful when linked with an existing SampleSet in the system, linking a SampleSet is not required (See section Linking SampleSet). While Chemical abundance data works best and more meaningful when linked with an\\n\\nexisting SampleSet\\n\\nin the system, linking a SampleSet is not required.\\n\\nThe minimal set of metadata in a chemical abundance matrix includes an ID (unique value) field, a chemical type (aggregate, exometabolite, specific), and one or more of the following:  Compound ID (e.g; ModelSEED, KEGG, ChEBI), mass, formula, inchikey, inchi, smiles, or compound name. Additional metadata such as units are strongly encouraged to provide with proper information that fits your scientific use cases or be kept as ‘unknown’.  (see section \"Template Fields Descriptions\" for an explanation of each field)  Providing additional metadata may enhance the downstream analysis of use cases for you and other readers.\\n\\nIf a SampleSet exists, it can be applied to the chemical abundance data. Chemical abundance data needs to be formatted to ensure Samples are correctly linked.\\n\\nNote that linking to Samples is not required, but highly recommended. When linking to using this app, the template will be automatically populated with Sample IDs to ensure the chemical abundance data is properly linked to corresponding Samples in the system.\\n\\nCreate Chemical Abundance Matrix Template\\n\\nThis App generates a spreadsheet onto which you can copy your data to ensure it links to the SampleSet when uploaded.\\n\\nWhen creating the chemical abundance template, there are 10 columns in the default sheet shown in the sheet below. Column headings in italics come pre-filled with validated values to choose form a dropdown.\\n\\nID (unique value) The identifier for the chemical/element/compound or peak value. This can be user-defined but must be unique for each compound. Chemical Type Specific - Specific compounds in the sample (e.g; TiO2, Tyrosine)  Aggregate - Measurement of total concentration of individual elements (e.g; Total C, N)  Exometabolite - Measurements of compounds that are consumed or excreted into the medium by an organism or a microbial community Measurement Type unknown, FTICR, Orbitrap, Quadrapole Units mg/Kg, g/Kg, mg/L, mg/g DW, mM (millimolar), M (molar), % (percentage), Total Weight  %, unknown Unit Medium soil, solvent, water Chemical Ontology Class Used for grouping chemicals by their functional groups (e.g; aromatics) or could use according to ontologies defined in public databases (e.g; ChEBI). Chromatography Type unknown, HPLC, MS/MS, LCMS, GS Chemical Class Used for defining major classes like lanthanides Protocol Any term or sentence defining the protocol used in your lab Identifier An optional identifier/abbreviation that you use for the compound and or element (e.g; Menaquinone => MK-8, Fatty acid => (Iso-C16:1) )\\n\\nSelect chemical data to include, such as aggregate M/Z, compound name, predicted formula, and more, depending on what data you have for upload.\\n\\nAggregate M/Z M/Z represents the ratio of mass (M) divided by net charge (Z) of the chemical. Compound Name Optional name for the compound Predicted Formula Formula of the compound (if available). The formula could be a predicted formula based on techniques such as Fourier-transform ion cyclotron resonance (FTICR) mass spectrometry, computationally predicted formula based on molecular weight, derived from instruments based on known protocols, or based on the chemical structure determination from NMR. Predicted Structure (smiles) Predicted structure using Simplified Molecular Input Line Entry System (SMILES) notation Predicted Structure (inchi-key) Predicted structure using International Chemical Identifier (InChi) key notation. Theoretical Mass Theoretical molecular mass of the compound Retention Time The retention time of the chemical; not validated, so users can include units or not depending on needs Polarity Polarity of the chemical.\\n\\nFinally, select at least one form of standard Chemical IDs to include in the template. These can be selected from the\\n\\nKEGG\\n\\nChEBI\\n\\n, and\\n\\nModelSEED\\n\\ndatabases. You can use their respective websites to get the identifier for your compound. If you are creating a chemical abundance sheet from scratch, you don\\'t have to include one of these chemical IDs, but it is recommended that you do so in order to compare similar chemical once uploaded.\\n\\nIf you plan to use your data with metabolic modeling analysis pipelines (see Use cases section) we highly encourage to have at least one type of identifier to be listed (if available), as the compound identifiers will be used to map compounds in metabolic models. (Alternatively, you can provide inchikeys which we able to map to compounds in metabolic models).\\n\\nUploading and Importing\\n\\nFor full functionality of analysis tools using metadata, first\\n\\nupload and import the SampleSet\\n\\nOnce you have added and\\n\\nformatted\\n\\nyour data to the Chemical Abundance Matrix template, you can upload it using the\\n\\nImport Chemical Abundance Matrix from CSV/Excel/TSV File in Staging Area App\\n\\nUsing a file on your computer, open the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the chemical abundance matrix file into the Staging Area.\\n\\nOnce the matrix is in your Staging Area, you can import the data into your Narrative.\\n\\nImport Chemical Abundance Matrix\\n\\nIn the first section for Input Objects, select the previously imported SamplesSet file from the dropdown menu for linking metadata.\\n\\nUnder Parameters, select the Chemical Abundance Matrix using the FilePath dropdown.\\n\\nFill in the Matrix Object Name and click the \"Run\" button to add the metabolomics data to the Narrative.\\n\\nUsing the Uploaded Data\\n\\nOnce you\\'ve uploaded your chemical abundance data, you can explore apps using that type of data or use one of our use case demonstration Narratives. See\\n\\nthe chemical abundance section of in Using Apps for more info.\\n\\nPreviousAmplicon Matrix\\n\\nNextSampleSet\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_chemical-abundance-matrix.html'}),\n",
       " Document(page_content='SampleSet\\n\\nA guide to importing and formatting sampling data and metadata to comply with standard templates.\\n\\nSampleSets are still a work in progress, and much of the documentation here is not currently applicable in production. Watch your inbox for notifications from KBase Outreach when these workflows go live.\\n\\nIf you are interested in beta testing these Apps, please contact us at [email\\xa0protected]\\n\\nA Sample in KBase represents physical material from an experiment with attributes or metadata (environment, location, temperature, etc). It has a unique ID internal to KBase, aliases, related Samples, and structured metadata. Samples occur within the Narrative as a SampleSet, and multiple SampleSets may share the same Sample.\\n\\nFormatting SampleSets\\n\\nIGSN SESAR\\n\\nThe\\n\\nSESAR\\n\\n(System for Earth SAmple Registration) format is one format that can be used in KBase for uploading Samples. Their samples use a\\n\\ncontrolled vocabulary for fields of data and metadata\\n\\nthat can be used in KBase.\\n\\nDownload a template of the SESAR format below and fill in your own data, or generate a new dataset file\\n\\nbased on their documentation\\n\\n. The .xlsx file has pre-filled options for the columns with\\n\\ncontrolled vocabulary\\n\\n, whereas the CSV file is a much smaller file that only contains a heading with the possible fields.\\n\\nSESAR_sample_template.xlsx\\n\\n156KB\\n\\nBinary\\n\\nDownloadable SESAR Template File\\n\\nSESAR_sample_template.csv\\n\\n61KB\\n\\nText\\n\\nDownloadable SESAR CSV Template\\n\\nSamples can be directly imported from IGSN SESAR using the \"Import Samples from IGSN\" App, which takes SESAR number(s) separated by commas as input and imports the sample information directly from SESAR.\\n\\nNCBI BioSamples\\n\\nNCBI BioSamples can be directly imported using the \"Import Samples from NCBI\" App, which takes the NCBI BioSample accession number(s), again separated by commas when importing multiple samples into the same SampleSet.\\n\\nENIGMA\\n\\nThe Ecosystems & Networks Integrated with Genes & Molecular Assemblies (ENIGMA) Science Focus Area (SFA) partners with KBase for new functionality development. The ENIGMA metadata template is available as a template for the \"Import Samples\" App.\\n\\nenigma_format.xlsx\\n\\n6KB\\n\\nBinary\\n\\nDownloadable ENIGMA XLSX Template\\n\\nenigma_format.csv\\n\\n2KB\\n\\nText\\n\\nDownloadable ENIGMA CSV Template\\n\\nUploading and Importing\\n\\nWhen uploading and importing Samples, new Samples can either be added to an existing SampleSet or used to create a new SampleSet. To add Samples to an existing SampleSet, choose the appropriate SampleSet from the drop-down in the importer. To create a new set, leave it blank and a new SampleSet will be created from the newly imported Samples.\\n\\nfrom IGSN or NCBI\\n\\nSamples can be directly imported from IGSN or NCBI using the \"Import Samples from IGSN\" or \"Import from NCBI\" Apps, respectively, searchable in the Upload section of the Apps panel. Samples imported using this method can be added to an existing SampleSet through selecting it from the SampleSet drop-down or can be used to create a new SampleSet. Additional parameters are the same as those for importing Samples manually (from local computer).\\n\\nfrom local\\n\\nSamples data is uploaded to the Staging Area and then imported into a Narrative. When importing new Samples, the default action is to create a new SampleSet. Individual samples must belong to a SampleSet, even if that SampleSet contains only one Sample.\\n\\nSamples must be in\\n\\nSESAR or ENIGMA formats\\n\\nto not limit analysis as these standards are required to validate most of the data. Non-validated data can be uploaded as user terms, but this will not allow comparisons across Samples. We recommend to change the column headers to correspond with the relevant SESAR or ENIGMA headers, and ensure the data matches.\\n\\n\\u200bData Validation\\u200b\\n\\nSampleSets that are uploaded to KBase are validated to ensure the data format is correct. If data does not match format requirements, the upload may result in an\\n\\nerror\\n\\nCritical errors, such as unrecognized units for a measurement, will prevent upload entirely. Minor errors, such as an unrecognized controlled vocabulary term, can be ignored by selecting the \"Ignore Warnings\" checkbox in the importer.\\n\\nViewing\\n\\nin the Narrative\\n\\nOnce imported, the SampleSet viewer widget allows you to see the Samples data that has been imported. This includes all metadata, such as collection information, and can be selectively viewed using the search bar.\\n\\nthe Landing Page\\n\\nMore details for the SampleSets can be found on the landing page. Click on the SampleSet name in the viewer widget or the binoculars button in the data pane to view the landing page.\\n\\nUploading and Importing SampleSets\\n\\nPermissions\\n\\nPermissions are managed separately from the Narrative workspace with some exceptions. The permissions for a Sample can be viewed in the Sample landing page under the Access tab.  When a SampleSet is shared with a user, that user automatically has read access to  all Samples included in the SampleSet. Read access will automatically be given with the first access to the SampleSet by visiting its landing page or viewing it in the Narrative.\\n\\nViewing Samples Access Permissions\\n\\nTo grant a user increased permissions (e.g. to modify Samples), you should use the Update SampleSet Access Controls App in the Narrative.\\n\\nUpdate SampleSet Access Controls App to change user permissions for specific SampleSets.\\n\\nPermissions will be granted for all Samples in the SampleSet. There is not currently a mechanism to change the permissions on a single app and there is not a method to easily revoke permissions.\\n\\nUpdating Attributes\\n\\nYou will need to export a CSV file of the Samples from a SampleSet, modify the columns and/or their values, then re-import the Samples CSV using the Samples Importer App to edit Sample attributes (for example adding, modifying or removing columns or updating values).\\n\\nThis will not modify existing data or analysis that has already been run. To update the analysis results, you will need to re-run the analysis apps using the new SampleSet with your recently updated Samples.\\n\\nSubsets and Supersets\\n\\nCreating subsets and supersets in the Narrative is under development. The most efficient way manage subsets of Samples is to upload separate SampleSets for each subset, and a complete SampleSet to represent the parent set. In the same way, a superset can be created by downloading SampleSets, combining them, and then uploading the combined superset.\\n\\nErrors\\n\\nThe Samples Importer Apps are designed to encourage the use of validated data, but we are aware that not all information is accounted for in our validator, and sometimes wrangling the data to fit validation formats is not worth the effort. While we do encourage validation as much as possible, our importers use a two-tiered system of error reporting.\\n\\nThere are two, color-coded error types: red and yellow.\\n\\n1.Red errors indicate the validation failed to recognize a value as a validated field, such as the type \"playdoh\" for material. This prevents the importer from creating a SampleSet for any cell with values that do not comply with controlled vocabulary.\\n\\n2.Yellow errors indicate a column (in this case, \"color\") could not be validated as a controlled field. These result in warnings, but the SampleSet can still be created.\\n\\nTwo kinds of errors - red errors are critical, while yellow are warnings\\n\\nHandling Errors\\n\\nSample imports can return a \"successful\" job that does not create an object in the Data panel. In this case, you can interpret a successful job run to indicate that the sample file was read, but the contents may or may not be valid.\\n\\nA cell in a validated column that does not comply with the controlled vocabulary will fail to upload. Consult the format guide or source of the template to determine if the value is accepted and if the vocabulary contains a valid term for your value. If your value is not part of the controlled vocabulary and there is not a suitable alternative, you can change the column header to a user defined field.\\n\\nIf you would like your term added to the controlled vocabulary,\\n\\nplease contact us\\n\\nIf there is a warning due to an unvalidated field, you can choose to accept the warning or decline it and fix the spreadsheet. If you accept an unvalidated field, it will be stored as a user-defined field. This allows you to include the information, but does not allow comparisons across Samples. For additional descriptive information, accepting the warning may be beneficial. But if the information is critical to Samples which you plan to use in analysis, we recommend manually fixing the values to comply with controlled vocabulary terms.\\n\\nNo mechanism exists to accept warnings after the import has run. You can run the Importer App with the \"Ignore Warnings\" option unchecked. This allows you to see the errors and warnings and run does not create a data object. Then proceed to edit your spreadsheet and re-run the App with the warnings fixed, or run the App with the option checked and delete the resulting object if you find warnings you want to fix.\\n\\nPreviousChemical Abundance Matrix\\n\\nNextCompressed/Zipped Files\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_sampleset.html'}),\n",
       " Document(page_content='Compressed/Zipped Files\\n\\nHow to handle compressed files when uploading and importing data.\\n\\nCompressed Format Types\\n\\nData importers accept compressed files in these formats: .zip, .gz, .bz, .bzip2, .bz2, .tar, .tgz.\\n\\nUploading and Importing Compressed Files\\n\\nTo add compressed data files from your computer to the KBase Narrative, follow the same three-step process for all data files.\\n\\n1.Drag & drop the data file(s) from your computer to the new Import tab to upload them to your Staging Area.\\n\\n2.Choose a format for importing data from your Staging Area into your Narrative.\\n\\n3.Run the Import App that is created.\\n\\nWhen files have been uploaded to the Staging Area, compressed files have an outward-facing double arrow icon to the left of the file name.\\n\\nClick the uncompress icon to the left of the file name to uncompress the compressed/zip uploaded files before clicking the import icon. To ensure this works, the file extension must match the type of packing/compression.\\n\\nNote that importers will accept compressed files, but not archives, as imports. For instance, you can import a single reads file as reads.fastq.gz or decompress it in the staging area and import reads.fastq; however, if you have a series of files in an archive (e.g. reads1.fastq, reads2.fastq,  and reads3.fastq inside all-reads.zip), the archive will first have to be unpacked and then the files can be imported. This applies to both single and bulk imports.\\n\\nPreviousSampleSet\\n\\nNextBulk Import Specification\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_compressed-files.html'}),\n",
       " Document(page_content='Bulk Import Specification\\n\\nImport multiple files at once through bulk upload using an import specification file with the extension .csv, .tsv, or .xls following the instructions below.\\n\\nImport Specification applies to\\n\\nselect data types\\n\\nA TSV (tab-separated values) file with a .tsv or .tab file extension, a CSV (comma-separated values) file with a .csv extension, or an Excel spreadsheet with a .xls extension can be used to upload multiple files at once.\\n\\nThis approach treats the Import Specification file as a directory or manifest containing the upload information for multiple files to fill out an import cell. The TSV, CSV, or Excel file must be formatted exactly to import correctly.\\n\\nImport Specification files (TSV, CSV, Excel) require several columns for uploading sequencing data when selecting files. These will vary for each data type, but make sure you have the following information:\\n\\nFile Path(s) – These will be relative to your home directory of the Staging Area, so you only need to include the file name if the files are at the top level. If they are within a folder called “folder1” then the file input fields should be \"folder1/file1\", \"folder1/file2\", etc.\\n\\nObject name – Designate an object name for the imported file, which will be the object name to use as the input for analysis Apps.\\n\\nParameters – Any required input parameters.\\n\\nUse the bulk Importer App with a subset of files to generate an Import Specification template. Download and modify the template locally to fill out file paths and parameters. Then upload the Import Specification file to import all files at once.\\n\\nCreating an import specification template from the Narrative\\n\\nTo create an import specification template CSV, TSV or Excel file for many files from the Narrative, begin with the Bulk Import directions using a subset of the files or file types to import.\\n\\nDrag and drop all files or a folder containing the files to upload to Staging.\\n\\nOnce the files appear in the Staging Area, select at least one that represents each file type you wish to import, select the Import As file type and make sure the check box is active, then click \"Import Selected.\"\\n\\nThis will create a bulk Importer App in the Narrative. Fill out the parameters for the import. Then click the \"Create Import Template\" button.\\n\\nImport from Staging Area Bulk Import App\\n\\nThis will create a pop-up file to prompt the template creation. Select the import type(s), if multiple use key controls (such as command-hold and select with Mac or control-shift and select on Microsoft). Choose the output type as either Comma-separated (CSV), Tab-separated (TSV), or Excel (XLS). Then select or rename the output destination within the Staging Area. Click \"Generate template!\" to create the templates selected.\\n\\nIf you have an Import Specification Template within the destination that you designate in the pop-up, it will be overwritten with the newly created template.\\n\\nTo import multiple data types at once, either create a CSV/TSV for each type, or a tab for each type in a single Excel file.\\n\\nA pop-up will confirm the Import Specification template has been generated and its location in the Staging Area.\\n\\nNavigate back to the Staging Area  and click \"Refresh\" if you do not see the folder or template in the destination designated in the previous step. Download the Import Specification template using the download button on the left of the trash icon.\\n\\nOpen the Import Specification template. In the template files 3 rows are already filled; these are required for the staging service to parse the file. In the Excel file, the top two rows are hidden to simplify the view. The third row displays the column headers for inputs, outputs, and parameters included in the Importer App. The following rows are for each of the files to import, the files selected before will be included here.\\n\\nContinue filling out the spreadsheet as if you would the Importer App with the remaining files to import. Each row corresponds to one data object to import. Some data objects will include more than one file path. For instance, a GFF Genome requires both GFF and FASTA files. All text fields should match exactly to the display text in the Importer. Use the pre-filled parameters to fill out the rest of the rows with designated file paths.\\n\\nTo manually fill in the Scientific name into a template, navigate to an Import cell with a GenBank Genome Data type. Under Parameters, go to the \\'Scientific name\\', search for the genus/species name you want and select it. Click the copy button on the right. You can now paste the name into the template.\\n\\nSearch for the Scientific name where the orange arrow is located, then copy using the icon to the right with the orange circle.\\n\\nImporting a bulk Import Specification file\\n\\nOpen the\\n\\nImport tab within the Data Browser\\n\\n. Then drag & drop the Import Specification file and the files to be imported if they are not already in the Staging Area.\\n\\nDrag and Drop a CSV file\\n\\nOnce the file appears in the Staging Area, select \"\\n\\n\" from the dropdown in the\\n\\ncolumn. Make sure\\n\\nwith a filled checkbox (do not select the independent files). Click\\n\\nat the bottom right corner.\\n\\nOnce the Import Specification file has been read, it will create a bulk Importer App in your Narrative and you can proceed as usual following\\n\\nthe bulk import\\n\\ndirections. Verify the file names and import settings within the App. Click \"Run\" on the upper lefthand corner of the App to import the files from the Import Specification into the Narrative.\\n\\nOnce the Importer App has run, the App will signal that the files have been imported or any errors that occurred across individual jobs.\\n\\nLimitations\\n\\nCurrently, 500 rows or data objects per import is supported with good performance. The current limit is 10000 rows.\\n\\nIf you upload an invalid file, you will receive an error that lists the problem encountered. For errors that can be fixed through the interface, the Importer cell will be created but you must correct the error before running the import.\\n\\nAt this stage of development only a single set of parameters can be used per import job. While we plan to add multi-parameter support in the future and have made the Import Specification templates forward-compatible, the import will apply parameters from the first row to all files.\\n\\nTroubleshooting tips:\\n\\nDo not change the first 3 rows of the TSV or CSV file or the first row or hidden rows of Excel files. This will cause errors or incorrect data.\\n\\nBe careful of stray cells in Excel - ensure that all the cells outside your data block are empty. Consider deleting all the rows and columns outside your data block.\\n\\nFor CSV and TSV files, be careful to always include the correct number of separator characters for every row, even if some values are blank. The importer is picky about this to help the user avoid off-by-one (or more) errors in their data.\\n\\nOnly one file or Excel tab is allowed per data type. Do not include rows for the same data type in more than one Excel tab, more than one file, a tab and an Excel file, etc.\\n\\nIn CSV and TSV files, white space around non-numerical data is ignored, and whitespace prior to numerical data is ignored.\\n\\nIn CSV and TSV files, you can use quotes (\") to surround text that contains the separator character (respectively a comma or a tab).\\n\\nIf individual files in an import specification are selected along with the import specification, a bug will occur where extra rows that cannot be filled in are added to the bulk import cell for the individual files. To workaround this issue, either do not select files that are within the selected import specification or delete the uncompletable row(s) from the import cell.\\n\\nIf the scientific name lookup ignores spaces, the workaround is to create (and optionally delete) a new bulk import cell for the scientific name input.\\n\\nPreviousCompressed/Zipped Files\\n\\nNextDownloading Data\\n\\nLast modified \\n\\n1mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_bulk.html'}),\n",
       " Document(page_content='Downloading Data\\n\\nAfter analyzing data in KBase, you may wish to download results to your local computer. The Narrative Interface offers the ability to download many of the data types stored in KBase in common form.\\n\\nFrom the Narrative\\n\\nLocate the data you wish to download in the Data Panel under the Analyze tab\\n\\nHover your mouse over the data object until a “…” icon appears to the right of its name\\n\\nClick the “…” icon to reveal advanced options\\n\\nClick the seventh icon from the left \"Export / Download Data\"\\n\\nSelect a format for exporting and downloading the data\\n\\nThe data will download to the default downloads folder on your computer\\n\\nWhen downloading data from KBase, the data will be compressed into a Zip file (.zip) containing files (or a directory containing files) in the format you selected and a metadata file (in JSON format). If you choose to re-upload data you have downloaded from the system, take note that you cannot directly import the Zip file – you must first extract the file and then re-upload using the specified uploader for the data type.\\n\\nFrom App Cells\\n\\nA few KBase Apps have file links as part of the output at the bottom of the App Cell.\\n\\nRight click on the file name to download and choose “Save Link As\" to download and save to your local computer.\\n\\nLarge Data Objects\\n\\nLarge data objects (over 10 gigabases) include many Reads, Assemblies, Genomes, and Alignment (BAM) objects. They encounter many of the same obstacles as uploading large data sets. Just like uploading, this is a 2-stage process; copy the file to the Staging Area and download from Staging to your local computer with\\n\\nGlobus\\n\\nCopy the file to the Staging Area (version 1)\\n\\nLocate the data you wish to download in the Data Panel under the Analyze tab\\n\\nHover your mouse over the data object until a “…” icon appears to the right of its name\\n\\nClick the “…” icon to reveal advanced options\\n\\nClick the seventh icon from the left \"Export/Download Data\"\\n\\nSelect \"Staging\"\\n\\nThis will open “Export Data Object to Staging Area” App.\\n\\nThe name of the Input Object is filled in. A default Destination Directory is filled in, but can be edited and changed. A directory will be created in the Staging Area, and all created files will be copied to the directory. The \"show advanced\" Parameters may have more options relevant to your data object.\\n\\nOnce in the Staging Area, large files (>10GB) should be transferred to your local machine using\\n\\nGlobus\\n\\n. Instead of “To” the “KBase Bulk Share” endpoint, this will be a transfer “From” the “KBase Bulk Share” endpoint.\\n\\nCopy the file to the Staging Area (version 2)\\n\\nIn the App Panel, search for the app “Export Data Object to Staging Area,” put in the name of your Input Object for the data object to download, and follow the prior instructions.\\n\\nPreviousBulk Import Specification\\n\\nNextSearching, Adding, and Uploading Data\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_downloads.html'}),\n",
       " Document(page_content='Searching, Adding, and Uploading Data\\n\\nVideo tutorial on searching for, adding to Narratives, and how to upload data in KBase.\\n\\nYou can add data to a Narrative through a variety of methods. The Data Browser allows you to search data in KBase or import data from your computer or a database.\\n\\nMy Data shows your data objects available within Narratives.\\n\\nShared With Me includes data associated to Narratives shared with you.\\n\\nPublic displays Public datasets.\\n\\nExample shows datasets that have been pre-loaded by the KBase team.\\n\\nImport allows you to upload your own datasets or external datasets for analysis.\\n\\nAdd data by navigating to the Data Browser tab and hover over the data object to reveal the\\n\\nbutton. Click on the \"\\n\\n\" button to import the data object into the Narrative. Now you can view the data object and run analysis.\\n\\nAdd Data already in KBase.\\n\\nPreviousDownloading Data\\n\\nNextFiltering, Managing, and Viewing Data\\n\\nLast modified \\n\\n7mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_search-add-upload.html'}),\n",
       " Document(page_content='Filtering, Managing, and Viewing Data\\n\\nVideo tutorial on filtering, managing, and viewing data in KBase.\\n\\nPreviousSearching, Adding, and Uploading Data\\n\\nNextLinking Metadata\\n\\nLast modified \\n\\n3yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_filter-manage-view.html'}),\n",
       " Document(page_content='Linking Metadata\\n\\nHow to link environmental sampling metadata with experimental data.\\n\\nSampleSets are still a work in progress. Watch your inbox for notifications from KBase Outreach when these workflows go live.\\n\\nIf you are interested in beta testing these Apps, please contact us at [email\\xa0protected]\\n\\nA Sample in KBase represents measured material from an experiment. It has a unique ID internal to KBase, aliases, related Samples, and structured metadata. Samples occur within a Narrative as a SampleSet, but the same Sample may be shared by multiple SampleSets.\\n\\nOne key feature is linking the same data across many users and analyses. To do this, Sample aliases allow the same data uploaded by different users to be linked and connect different analyses of the same data. These aliases can also be used to link Samples in KBase with other methods of sample registration, such as ESS-DIVE or SESAR.\\n\\nBecause the metadata about a sample is critically important for, KBase provides the option to\\n\\nupload metadata spreadsheet\\n\\ns to link Samples for more detailed analysis.\\n\\nManaging Linked Data\\n\\nDuring upload and import\\n\\nOnce a\\n\\nSampleSet\\n\\nis in system, it can be used when uploading\\n\\namplicon\\n\\nor\\n\\nchemical abundance\\n\\nmatrices into KBase. This allows the amplicon and chemical abundance analysis to remain linked to the sample metadata and enrich analysis.\\n\\nWhen uploading and importing either an amplicon matrix or chemical abundance matrix, select the corresponding SampleSet from the dropdown menu for the input before running the Import App.\\n\\nAdditional Data Links\\n\\nAdditional existing data can be linked to samples. The app \"Link Workspace Objects to Samples\" provides a graphical interface for creating the links. In this app, you can select the SampleSet to create links from, then choose the Sample from the set and object from the Narrative. You can add multiple Samples and objects by clicking the + button.\\n\\nIf you have a large number of links to create, it would be advisable to use the \"Batch Link Workspace Objects to Samples\" app.\\n\\nPreviousFiltering, Managing, and Viewing Data\\n\\nNextOntologies and Validated Terms\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_samples.html'}),\n",
       " Document(page_content=\"Ontologies and Validated Terms\\n\\nMake the most out of your metadata by using controlled vocabulary\\n\\nTL;DR: Use vocabulary that can be identified in KBase so your metadata is functional.\\n\\nWhen uploading Samples to KBase, there are many metadata terms that are built in and can be validated. Doing so allows your data to be more easily compared with other data and allows us to integrate data when different sources may use different conventions.\\n\\nMany standard terms, such as NCBI BioSamples and IGSN SESAR, aim to describe sample metadata. These standards often overlap and may differ significantly depending on the focus of the organization or institution that manages the samples. At the same time, many labs and smaller institutions have their own templates and organization of their data.\\n\\nThus, there must be a way to integrate similar metadata from different sources and account for the differences between the sources.\\n\\nWhen Samples are uploaded into KBase via a spreadsheet, they are transformed into an internal representation. Columns within the spreadsheet are mapped and transformed based on the template used. Columns that correspond to recognized terms are validated to ensure they properly formatted, such as checking that the value is the proper type (e.g. string versus a number), in the correct range, match an enumerated list, or appear in an ontology.\\n\\nControlled terms are useful both because they undergo this validation and they provide a more precise meaning for the value and comparisons accounting for units.\\n\\nWhen the uploader encounters terms it doesn’t recognize, those terms and values will be stored in a user section of the samples. These values still serve a purpose and can be used in analysis within that data set and they can provide contextual information that the original uploader understands.  However, unrecognized terms can not reliably be compared across SampleSets and other samples in the system. For example, two projects may use the same term to represent different concepts (e.g. depth below sea-level or depth below surface).\\n\\nUsing templates and controlled terms clarifies the exact meaning of a term and enforces additional validation. For a full list of terms KBase recognizes, see the\\n\\nvalidated metadata table\\n\\nbelow. If there are terms you would like added, please contact us at [email\\xa0protected] to suggest additions.\\n\\nOntology Landing Pages\\n\\nThe Ontology API supports multiple ontology systems. Currently supported systems are Gene Ontology (GO) and Environmental Ontology (ENVO). You can see more information on both ontologies from their own home pages, or view the KBase landing page for a given ontology page with the links and URL format below (respectively).\\n\\n\\u200bGene Ontology (GO)\\u200bKBase: https://narrative.kbase.us/#ontology/term/go_ontology/GO:#######\\n\\n\\u200bThe Environmental Ontology (ENVO)\\u200bKBase: https://narrative.kbase.us/#ontology/term/envo_ontology/ENVO:########\\n\\nValidated Metadata Terms\\n\\nThe file TSV file below contains a list of the metadata terms currently supported for validation along with a description that provides general formatting direction. You can view the\\n\\nmost up-to-date version of this list\\n\\nin the KBase Samples GitHub.\\n\\nmetadata_validation.tsv\\n\\n19KB\\n\\nText\\n\\nValidated Metadata TSV\\n\\nIf you have Sample Attribute columns that are represented as validated terms in SESAR, ENIGMA, or KBase formats, you may request the terms to be added to KBase's vocabulary.\\n\\nKBase will review newly proposed attributes to assess: overlap with current supported attributes; existing or relevant ontological frameworks; and presence of and interoperability of units for measured data.\\n\\nUnit Conversions\\n\\nYou can see a full list of the units and their defined relations to each other in\\n\\nGitHub\\n\\nTerms\\n\\nTo determine if your term is valid, the easiest way is to search the\\n\\nENVO Database\\n\\nCommonly used units:\\n\\nLength\\n\\nBase: meter, m\\n\\nMicron/micrometer: µm or um, 1*10^-6 m\\n\\nMillimeter: mm, 1*10^-3 m\\n\\nCentimeter: cm, 1*10^-2 m\\n\\nKilometer: km, 1*10^3 m\\n\\nTime\\n\\nBase: second, s\\n\\nMinute: min, 60 s\\n\\nHour: hr, 3600 s or 60 min\\n\\nDay: d, 24 hr\\n\\nYear: yr or a, 365.25 d\\n\\nMass\\n\\nBase: gram, g\\n\\nMicrogram: µg, 1*10^-6 g\\n\\nMilligram: mg, 1*10^-3 g\\n\\nKilogram: kg, 1*10^3 g\\n\\nPreviousLinking Metadata\\n\\nNextPublic Data in KBase\\n\\nLast modified \\n\\n1yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_samples_ontology.html'}),\n",
       " Document(page_content=\"Public Data in KBase\\n\\nKBase provides users with a unified resource for analyzing a range of public data together with the data generated from their own experiments. The data in KBase, which includes public data from various public sources as well as internally-generated analysis results, consists of an extensive set of prokaryotic annotated genomes; a selected set of eukaryotic genomes; and thousands of biochemical compounds and reactions.\\n\\nPlease review\\n\\nKBase's Data Policy\\n\\nand the sources of our public reference data.\\n\\nTo search public data in KBase, use the\\n\\nPublic tab in the Data Browser\\n\\nwithin the Narrative Interface.\\n\\nIf you are a JGI user, you can transfer public genome reads and assemblies (as well as your private data and annotated genomes) from JGI to your KBase account using\\n\\nthese instructions\\n\\nPreviousOntologies and Validated Terms\\n\\nNextTransferring Data with Globus\\n\\nLast modified \\n\\n7mo ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_public.html'}),\n",
       " Document(page_content='Transferring Data with Globus\\n\\nA how-to on large file transfer into KBase.\\n\\nTransfer large files (~20GB and more) to KBase using\\n\\nGlobus\\n\\n. Drag & drop from your local computer works for many files, but there is a size limit dependent on your computer and browser.\\n\\nGlobus is a data management and file transfer system that can facilitate bulk transfer of data (large data files or a large number of files) between two endpoints. The endpoints that apply here are KBase, JGI, and your local computer. The KBase endpoint is called “KBase Bulk Share,” and JGI has their own way to link to Globus. To do any transfer using Globus, you will need a\\n\\nGlobus account\\n\\nYour local computer can become a Globus endpoint with a\\n\\nGlobus Connect Personal Endpoint\\n\\n. This is necessary for transfers to or from your local computer.\\n\\nOnce you have a Globus account, link to your KBase account to facilitate file transfers. See\\n\\nLinking Accounts\\n\\nfor further guidance.\\n\\nVideo Tutorial on Using Globus\\n\\nA quick 4-min video on using Globus with KBase\\n\\nStarting a Globus Data Transfer\\n\\nTo initiate a transfer using Globus, click \\'Or upload to this staging area by using Globus Online\\' in the\\n\\ntab of the\\n\\nbelow the Staging Area\\n\\nThis brings you to a page that looks like this:\\n\\nThe panel on the right should be set to “KBase Bulk Share,” which directs to the Staging Area. In the example image above, the share point is at the root directory, and the user needs to click on their individual directory. If needed, Globus can be used to create new directories in your Staging Area for your data transfer and to delete staging files when you are done importing into KBase.\\n\\nSet the left panel to the location of the data you want to transfer (i.e. you could choose the JGI endpoint they send via email). On the KBase endpoint, if you don’t see a directory with your name as in the example above, add /username to the Path for the KBase endpoint (/user/ is used in the example above).\\n\\nYou are then be able to drag data files or folders from the left endpoint to the right one, inside the Globus interface, and the files will appear in your KBase Staging Area.\\n\\nYour transfer request will be submitted to Globus. If there are network or other problems, Globus will retry several times before giving up. Globus will send an email when the transfer is complete. In the KBase Staging Area, you may need to click on the refresh icon (two encircling arrows to the left of the user name) to update the Import tab.\\n\\nFirewalls\\n\\nSome government institutions have a firewall between users and the outside world. In an effort to increase security, many web links can be blocked, including Globus traffic. If you see a message that ends with “Details: 500 Command failed….”, a firewall may be the cause.\\n\\nNo ACL Rules\\n\\nAnother common error seen when using Globus shows and \"No ACL rules\" error on the KBase Bulk Share endpoint in the Globus UI. The most common cause for this error is multiple Globus accounts where the wrong account is linked to KBase. Be sure to log out of all Globus accounts, then link the correct account at\\n\\nhttps://narrative.kbase.us/#/auth2/account\\n\\n. If you still get this error, contact the help desk.\\n\\nUploading data from JGI\\n\\nIf you are a JGI user, you can transfer public genome reads and assemblies (and your private data and annotated genomes) from JGI to your KBase account—see\\n\\nthis page\\n\\nfor instructions.\\n\\nPreviousPublic Data in KBase\\n\\nNextTransferring Data from JGI\\n\\nLast modified \\n\\n7mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_globus.html'}),\n",
       " Document(page_content='Transferring Data from JGI\\n\\nDOE’s Joint Genome Institute (JGI) users can transfer genome reads, assemblies, and annotated genomes to KBase’s Narrative Interface for assembly, annotation, metabolic modeling, and more.\\n\\nBefore transferring data from JGI to KBase, you must:\\n\\n\\u200bGet a JGI account\\u200b\\n\\n\\u200bGet a Globus account\\u200b\\n\\n\\u200bGet a KBase account and link it to Globus (see the section “Create or modify user profile” here for help)\\n\\n\\u200bUse either Chrome, Firefox, or Safari browser\\u200b\\n\\nThe Search tool on the KBase menu is a quick and easy way to copy public genome reads and assemblies from JGI to your\\n\\nKBase Staging Area\\n\\n(this page describes the process for transferring those and also private data and annotated genomes).\\n\\nJGI’s Globus service lets users select data from the JGI Genome Portal’s sequencing projects and make them transferable to KBase, all in a few simple steps. There are some similarities to the KBase-initiated\\n\\nGlobus service\\n\\n, but also some differences.\\n\\nTransferring Data from JGI to KBase\\n\\nStarting at the\\n\\nJGI Genome Portal\\n\\n, search for the genome or project of interest and click on the download link. Click on the green \"Agree\" button on the next page to accept the data usage policy.\\n\\nNear the top of the download page is a link to JGI’s Globus v.2 service. This version works by 1) transferring the entire project to a JGI/Globus staging area and 2) sending you an email when it is ready.\\n\\nClick on the blue \"Download\" button and one of two windows will pop up:\\n1. A window to enter a Globus ID or register for a Globus ID\\n\\nOR\\n\\n2. Or a window saying “There is not enough disk space in the staging area for this data.”\\n\\nIf you do get the message about disk space, this is an issue between JGI and Globus (it has nothing to do with KBase), and it is dependent on other users staging data for transfer. The issue should eventually resolve itself if you try again later.\\n\\nOnce you successfully enter your Globus account name and hit the blue \"Submit\" button, your data will be transferred to the JGI staging area. A successful transfer request will look like this:\\n\\nYou will receive an email from [email\\xa0protected] with a link for transferring data with Globus. The email also states that JGI only keeps this link active for 14 days. After that time, they will remove the data from their staging area to make room for other users.\\n\\nClicking on the link in the email will take you to a Globus file transfer window:\\n\\nDouble click on the directories until you get o the file you want. Click on the file name to highlight it and then click on the \"Transfer or Sync to...\" link on the right.\\n\\nThe newly created JGI endpoint is filled in on the left.\\n\\nSet the “To” endpoint (right) to “KBase Bulk Share”, which is the KBase Staging Area. If you do not see a directory with your name, add /username (i.e. /landml for the user below) to the Path for the KBase endpoint.\\n\\nYou should then be able to drag data files or folders from the left endpoint to the right one, inside the Globus interface, and the files will be copied to your KBase Staging Area.\\n\\nIf the connection to the server is broken during transfer, Globus will start up again where it left off once the connection is reestablished. The link in the JGI email may be used to reestablish the connection.\\n\\nWhile it may be possible to drag ALL the project data from the JGI endpoint to the KBase endpoint, not all files can be imported into KBase. Reads files (.fastq or .fastq.gz), assemblies (.fna or .fasta), and annotated genomes (.gbk or .gff) have a corresponding KBase object.\\n\\nYou can import data from your Data Browser Staging Area to any of your KBase Narratives (see\\n\\nthis page\\n\\nfor more information). The imported objects are permanent in your KBase account, but the files in the Staging Area will be removed after 90 days.\\n\\nImportant differences between JGI and KBase data staging:\\n1. The JGI staging area keeps file links for 14 days. The KBase staging area keeps files for 90 days.\\n2. The JGI staging area has a limited amount of space that is shared with other users. The KBase staging area has 2.8 terabytes available to the user.\\n3. The JGI notifies users by email when files have been added to the staging area. The KBase Import tab shows files in your Staging Area and has an update button to refresh the list.\\n4. JGI stages an entire project at one time. KBase stages specific files selected by the user.\\n\\nPreviousTransferring Data with Globus\\n\\nNextUsing Apps\\n\\nLast modified \\n\\n7mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_jgi-transfer.html'}),\n",
       " Document(page_content='Using Apps\\n\\nAll about analysis tools available in KBase\\n\\n1.How to access and run analysis Apps in KBase\\n\\n2.How to access Beta Apps in KBase\\n\\n3.Visit the KBase App Catalog\\u200b\\n\\nPreviousTransferring Data from JGI\\n\\nNextAnalysis Apps in KBase\\n\\nLast modified \\n\\n2yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps.html'}),\n",
       " Document(page_content='Analysis Apps in KBase\\n\\nKBase Apps are analysis tools that you can use in KBase. Apps interoperate seamlessly to enable a range of scientific workflows (see figure below). Some of the apps were written by KBase scientists and developers; others are third-party tools that were integrated into KBase with our\\n\\nSoftware Developer Kit (SDK)\\n\\n. The number of apps available in KBase increases as members of the community use our SDK to integrate their analysis tools into the KBase platform.\\n\\nThe App Catalog lists currently available Apps.\\n\\nExternal App Catalog (no user account or sign-in required for browsing).\\n\\nWithin KBase by clicking on the Catalog icon in the Menu\\n\\nIn a Narrative by clicking the right arrow at the top of the Apps Panel.\\n\\nThe majority of KBase Apps fall into the following categories:\\n\\n\\u200bAssembly\\u200b\\n\\n\\u200bAnnotation\\u200b\\n\\n\\u200bSequence Analysis\\u200b\\n\\n\\u200bComparative Genomics\\u200b\\n\\n\\u200bMetabolic Modeling\\u200b\\n\\n\\u200bExpression\\u200b\\n\\n\\u200bMicrobial Communities\\u200b\\n\\nShow me!\\nGo straight to the\\n\\nApp Catalog\\n\\n\\u200b\\u200b\\n\\nNote: you will need a\\n\\nKBase user account\\n\\nto use our tools.\\n\\nTips on using the App Catalog\\u200b\\n\\nEach app links to a reference page (which includes technical details about the inputs and outputs) called an App Details Page.\\n\\nTo run apps, you will need to sign in to the Narrative Interface.\\n\\nYou can access the App Catalog from inside the Narrative Interface by clicking the small arrow in the upper right corner of the Apps Panel.\\n\\nYou can click the star at the lower left of any app to add it to your “favorites.” The gray star will turn yellow to indicate that you have favorited the app. The number to the right of the star shows how many people have favorited that app.\\n\\nBy default, the apps are sorted by category. Try the options in the “Organize by” menu to sort the apps by My Favorites, Run Count and other options.\\n\\nRunning Analysis Apps\\n\\nOnce you have added an app to the Narrative, you will need to select a data object as the Input Object and set Parameters. When working within an application, a red bar or banner will highlight empty fields and errors.\\n\\nMore info\\n\\nFor more information about using the App Catalog, see the\\n\\nNarrative Interface User Guide\\n\\nPreviousUsing Apps\\n\\nNextAssembly & Annotation\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis.html'}),\n",
       " Document(page_content='Assembly & Annotation\\n\\nSome of the tools in KBase available for Assembly and Annotation\\n\\nKBase provides multiple Apps for de novo\\n\\nassembly\\n\\nof prokaryotic Next-Generation Sequencing (NGS) reads from various sequencing platforms. These assemblies can then be\\n\\nannotated\\n\\nto explore structural and functional features of a Genome or use it in other analyses. The\\n\\ninteractive tutorials\\n\\nare a good way to learn about these workflows\\n\\nRead Processing\\n\\n\\u200bTrim Reads with Trimmomatic – Read trimming and adaptor removal\\n\\n\\u200bFilter Out Low-Complexity Reads with PRINSEQ – Filter low complexity reads\\n\\n\\u200bAssess Read Quality with FastQC – Quality assessment and reporting\\n\\n\\u200bCutadapt – Custom adapter removal\\n\\nAssembly\\n\\nDe novo\\n\\nassembly\\n\\nof Illumina and Ion Torrent next-generation sequencing reads. Supports single-end and paired-end read libraries.\\n\\n\\u200bAssemble with HipMer – HipMer is a highly-parallelized port of JGI’s Meraculous assembler. Meraculous is a de Bruijn graph-based which increases speed by not performing error correction. Instead, it bases contigs on already high-quality scores and fills the gaps based on localized assemblies from the reads. HipMer enhances the speed of Meraculous.\\n\\n\\u200bAssemble with IDBA-UD – IDBA-UD is an iterative graph-based assembler for single-cell and standard short read data and is good for data of highly uneven sequencing depth. This assembler uses an iterative approach for selecting k-mer size that compensates for the information loss associated with single k-mer based de Bruijn graphs, making IDBA-UD one of the more accurate microbial assemblers.\\n\\n\\u200bAssemble with MaSuRCA – MaSuRCA is a short read assembler that combines the benefits of de Bruijn graph and overlap layout consensus assembly approaches. The main concept is the creation of super-reads that contain sequence information present in the original reads, which super-reads are then extended in both directions using an efficient k-mer lookup table. MaSuRCA is one of a smaller set of assemblers biologists use for eukaryotic assembly.\\n\\n\\u200bAssemble with MEGAHIT – MEGAHIT is a single node assembler for large and complex metagenomics NGS reads. It makes use of succinct de Bruijn graph (SdBG) to achieve low memory assembly, making it fast and especially suitable for assembly of small metagenomes, metatranscriptomes or low-coverage data in general.\\n\\n\\u200bAssemble with SPAdes – SPAdes is a single-cell and standard assembler based on paired de Bruijn graphs, considered to be one of the most accurate microbial assemblers. SPAdes employs a multisized de Bruijn graph which detects and removes bubble and chimeric reads, estimates insert distance from paired kmers, and computes contigs based on paired assembly graph.\\n\\n\\u200bAssemble with Velvet – Velvet is a classic de Bruijn graph based assembler that works by efficiently manipulating de Bruijn graphs through simplification and compression. It eliminates errors and resolves repeats by first using an error correction algorithm that merges sequences together. Repeats are then removed from the sequence via the repeat solver that separates paths which share local overlaps.\\n\\n\\u200bCompare assemblies with QUAST – Assess the output assemblies from different configurations of the same assembler, or compare assemblies from multiple assemblers to determine which one is optimal for downstream analysis.\\n\\nAnnotation\\n\\nGenomes can be\\n\\nannotated\\n\\nwith Prokka or RAST.\\n\\n\\u200bAnnotate Domains in a Genome – identifies protein domains from widely used domain libraries (COGs, TIGRfams, Pfam).\\n\\n\\u200bAnnotate Assembly with Prokka – combines multiple open-source annotation tools in a quick and thorough annotation pipeline for prokaryotic sequences for genomes, plasmids, and metagenomes.\\n\\n\\u200bAnnotate Microbial Assembly – uses components from the RAST (Rapid Annotations using Subsystems Technology) toolkit to annotate an assembled bacterial or archaeal genome.\\n\\n\\u200bAnnotate Microbial Genome – uses RAST to annotate a prokaryotic genome, to update the annotations of a genome, or to perform computations on a set of genomes so that they are consistent.\\n\\n\\u200bAnnotate Plant Coding Sequences with Metabolic Functions – performs functional annotation of plant cDNA or protein sequences.\\n\\n\\u200bBulk Annotate Genomes/Assemblies – uses components from the RAST (Rapid Annotations using Subsystems Technology) toolkit to annotate a set of genomes or assemblies.\\n\\nThe output of the annotation apps is a Genome, which is displayed in a tabular genome viewer (see below) that shows information about the Genome as well as a list of contigs and the genes that were called on each contig.\\n\\nViewContig\\n\\nPreviousAnalysis Apps in KBase\\n\\nNextComparative Genomics\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_assembly-and-annotation.html'}),\n",
       " Document(page_content='Comparative Genomics\\n\\nSome of the tools in KBase available for comparative genomics and phylogenetics\\n\\nKBase provides multiple\\n\\ncomparative genomics and phylogenetic analysis Apps\\n\\nto enable researchers to understand evolutionary relationships between organisms and explore structural and functional variance across genomes.\\n\\nPhylogenetic and Taxonomic Analysis\\n\\n\\u200bInsert Genome Into Species Tree – Build a phylogenetic tree with your genomes and RefSeq microbial genomes based on conserved orthologous groups\\n\\n\\u200bBuild Microbial Species Tree (in beta) – Build a phylogenetic tree with your genome and phylum exemplars from RefSeq\\n\\n\\u200bBuild Gene Tree (in beta) – Build an evolutionary reconstruction tree for a collection of sequence homology related genes\\n\\n\\u200bBuild Phylogenetic Tree with FastTree2  – Build a tree with FastTree2 directly on a user-created Multiple Sequence Alignment (MSA)\\n\\nSequence Homology and Functional Analysis\\n\\n\\u200bBLASTx, tBLASTn, and tBLASTx – Search genes in genomes and Annotated Metagenome Assemblies (AMAs) and compare translated protein sequences\\n\\n\\u200bBLASTn – Search genes and AMAs and compare nucleotide sequences\\n\\n\\u200bBLASTp and psiBLAST – Search genes and AMAs and compare protein sequences\\n\\nMUSCLE for nucleotide or protein sequences – Build and analyze multiple sequence alignments (MSA)\\n\\n\\u200bHMMER MSA Collection – Search and profile genes in genomes and Annotated Metagenome Assemblies (AMAs) with a collection of user-defined MSAs\\u200bHMMER Search from MSA – Search genes in genomes and AMAs with a single user-defined MSA  \\u200bHMMER Search with dbCAN2 of CAZy families – Search and profile genes in genomes and AMAs with the suite of CAZy dbCAN2 hidden markov models (HMMs)\\u200bHMMER Search for Phylo Markers – Search and profile genes in genomes and AMAs with the suite of GTDB protein phylogenetic marker HMMs\\u200bHMMER Search for Env BioElement families – Search and profile genes in genomes and AMAs with the suite of Environmental BioElement metabolic families\\n\\n\\u200bAnnotate Domains in a GenomeSet – Compute canonical gene family membership using COG, Pfam, and TIGRFAMs\\n\\n\\u200bView Function Profile for Genomes and View Function Profile for a Phylogenetic Tree – Examine functional content from canonical domain family annotation\\n\\nPangenome Exploration\\n\\n\\u200bCompare Two Proteomes – Compare proteomes and produce a synteny map dot plot matrix and table of gene differences\\n\\n\\u200bBuild Pangenome with OrthoMCL – Group orthologous protein sequences and identify core and non-core genes across multiple species\\n\\n\\u200bCompute Pangenome – Build a pangenome and evaluate protein family conservation\\n\\n\\u200bPangenome Circle Plot – View a microbial pangenome as a circle plot using\\n\\n\\u200bCompare Genomes from Pangenome – Compare isofunctional and homologous gene families within a pangenome\\n\\n\\u200bPhylogenetic Pangenome Accumulation – View the pangenome in a phylogenetic context\\n\\nPreviousAssembly & Annotation\\n\\nNextMetabolic Modeling\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_comparative-genomics.html'}),\n",
       " Document(page_content='Metabolic Modeling\\n\\nSome of the tools in KBase available for metabolic modeling\\n\\nKBase has a suite of\\n\\nApps supporting the reconstruction, prediction, and design of metabolic models\\n\\nin microbes and plants. Genome-scale metabolic models can be used to explore an organism’s growth in specific media conditions, determine which biochemical pathways are present, optimize production of an important metabolite, identify high flux pathways, and more.\\n\\nFlux Balance Analysis\\n\\n\\u200bBuild Metabolic Model – Construct genome-scale metabolic models based on a bacterial genome and media condition Red pathways is increased and blue pathways\\n\\n\\u200bGapfill Metabolic Model – Fill in missing reactions based on stoichiometry\\n\\n\\u200bRun Flux Balance Analysis – Predict metabolic fluxes\\n\\n\\u200bCompare FBA solutions – Determine optimal conditions of flux\\n\\n\\u200bCheck Model Mass Balance – Ensure accuracy\\n\\n\\u200bCompare Models – View multiple models side by side\\n\\nEditing Models\\n\\n\\u200bEdit Metabolic Model – Build unique models suited to specific experiments.\\n\\n\\u200bCreate or Edit Media – Create specialized growth conditions\\n\\n\\u200bBulk Download Modeling Objects – Save modeling data for future analysis\\n\\nComparative Genomics\\n\\n\\u200bPropagate Model to New Genome – Translate metabolic models from one organism to another\\n\\nExpression\\n\\n\\u200bCompare Flux with Expression – Compare reaction fluxes with gene expression values to identify metabolic pathways where expression and flux data agree or conflict and compare with chemical abundance data\\n\\n\\u200bSimulate Growth on Phenotype Data – Reconcile models with empirical data\\n\\nMetabolomics\\n\\n\\u200bEscher Pathway Viewer App (in beta) – Display Escher metabolic pathways and combine pathways with flux balance analysis and metabolomics data to view flux or expression. When viewing flux is in green and expression is in brown. The oval size and color intensity reflects metabolite abundance.\\n\\n\\u200bPickAxe App (in beta) – Generate novel compounds from an FBA Model and a Chemical Abundance Matrix Pickaxe is part of the MINE Databases, documentation for which can be found here: https://github.com/tyo-nu/MINE-Database/blob/master/minedatabase/pickaxe.py.\\n\\n\\u200bFit Model to Exometabolite Data (in beta) – Identify biochemical reactions to add to a draft metabolic model for production and consumption of exometabolites\\n\\nMicrobial Communities\\n\\n\\u200bMerge Metabolic Models into Community Model – Investigate community metabolism\\n\\n\\u200bBuild Metagenome Metabolic Model – Build a metagenome metabolic model from an annotated assembly or bins\\n\\nPreviousComparative Genomics\\n\\nNextMetagenomics & Community Exploration\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_metabolic-modeling.html'}),\n",
       " Document(page_content='Metagenomics & Community Exploration\\n\\nSome of the tools in KBase available for analyzing microbial communities and metagenomics\\n\\nKBase has a multitude of\\n\\nApps designed to analyze microbial communities and metagenomic reads data\\n\\n. They provide a way to understand the functional interactions between species in microbial communities. Annotated genomes extracted from metagenomic sequence libraries can be used for metabolic modeling, comparative phylogenomics, functional profiling, and more.\\n\\nMetagenomic Read Analysis & Assembly\\n\\n\\u200bFastQC  – Raw read quality control and analysis\\n\\n\\u200bTrim Reads with Trimmomatic – Read trimming and adaptor removal\\n\\n\\u200bKaiju – Taxonomy assignment of raw metagenomic reads\\n\\n\\u200bMEGAHIT – Assemble large and complex metagenomic reads\\n\\n\\u200bMetaSPAdes – Assemble shotgun metagenomic reads\\n\\n\\u200bIDBA-UD – Assemble short metagenomic reads\\n\\nContig Binning, Optimization, and Quality Analysis\\n\\n\\u200bMaxBin2 – Group metagenomic contigs using depth-of-coverage, nucleotide composition, and marker genes\\n\\n\\u200bMetaBAT2 – Group metagenomic contigs using strain abundance and nucleotide composition\\n\\n\\u200bCONCOCT – Group metagenomic contigs using depth-of-coverage and nucleotide composition\\n\\n\\u200bDAS-Tool – Optimizes bacterial or archaeal genome bins using a de-replication, aggregation and scoring strategy\\n\\n\\u200bCheckM – Assess genome and MAG quality\\n\\n\\u200bJorg (in beta)– Improve and circularize single-genome assemblies and MAGs\\n\\n\\u200bCircos (in beta) – Visualize assembly coverage\\n\\nCommunity Exploration\\n\\nRAST for Multiple Microbial Genomes and Multiple Microbial Assemblies – Annotate multiple bacterial or archaeal genomes and assemblies or sets using RASTtk\\n\\n\\u200bAnnotate and Distill Genomes with DRAM – Annotates Metagenome Assembled Genomes (MAGs) and provides an interactive functional summary per genome\\n\\n\\u200bdRep (in beta) – Dereplicate binned contigs, assemblies, and genomes using nucleotide similarity\\n\\n\\u200bVirSorter – Identify viral sequences from viral and microbial metagenomes\\n\\n\\u200bClassify Microbes with GTDB-Tk  – Assign taxonomy to isolate and classify Metagenome Assembled Genomes (MAGs) using Genome Taxonomy Database (GTDB) genome taxonomy\\n\\n\\u200bModelSEED – Generate and merge metabolic models from annotated genomes\\n\\n\\u200bPhylogenetic Pangenome Accumulation – Produce phylogenetic trees from annotated genomes\\n\\n\\u200bMetabolic modeling tools for microbial communities – analyze community metabolism\\n\\nPreviousMetabolic Modeling\\n\\nNextData Matrices - Amplicon, Stats\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_metagenomics.html'}),\n",
       " Document(page_content='Data Matrices - Amplicon, Stats\\n\\nSome of the tools in KBase available for rarefaction, standardization, and analyzing amplicon matrices and OTU tables.\\n\\nData Cleaning\\n\\nRarefaction and standardization can be performed in KBase to track data provenance on how the data was transformed. These Apps can be used in any order and/or repeatedly. For instance to remove singletons, rarefy samples, and then calculate relative abundance of rarefied samples.\\n\\n\\u200bTransform Matrix App – uses an AmpliconMatrix object as inputprovides abundance filtering, which can remove rows or columns if all values fall below a user-specified threshold and/or if the sum of all values falls below a user-specified thresholdperform standardization, with options for centering the data before scaling, scaling the data to unit variance, and whether to perform the standardization on columns or rowsperform ratio transformation, by either Centre or Isometric Log Ratio Transformation methods, and be applied to columns or rows\\n\\n\\u200bRarefy Matrix App – allows manual selection of a seed value to aid reproducibility to columns or rows\\n\\nTaxonomic Classification\\n\\nClassify rRNA with taxonomy using naïve Bayes with RDP Classifier App (in dev/beta) – assigns taxonomies to amplicon matrices based on several gene databases including RDP Classifier 2.13 - 16srna, RDP Classifier - fungallsu, and SILVA 138 - SSU, Full Length. Silva databases are currently under development.\\n\\n\\u200bTaxonomy Abundance Barplot App (in beta) – visualize the taxonomic abundance\\n\\nStatistical Analysis\\n\\nStatistical comparisons can be calculated, such as clustering (hierarchical or k-means) and PCA analysis.\\n\\n\\u200bPerform NMDS Analysis App (in beta) – performs Non-metric Multidimensional Scaling Analysis on amplicon matrix data to visualize it in two dimensions\\n\\n\\u200bPerform Categorical Variable Statistics Analysis App – perform Categorical Variable Statistics Analysis on an amplicon matrix object\\n\\n\\u200bPerform Similarity Percentage (SIMPER) Statistics Analysis App – performs similarity percentage or SIMPER (Clarke 1993) analysis based on the decomposition of Bray-Curtis dissimilarity index\\n\\nFunctional Prediction for Amplicon Matrices\\n\\n\\u200bMap Tax to Functions using FAPROTAX App – identify functions and assign them to a new AmpliconMatrix using a manually-curated database that focuses on marine and lake biochemistry, such as sulfur, nitrogen, hydrogen, and carbon cycling.\\n\\nLouca Lab FAPROTAX site\\n\\nhttp://www.loucalab.com/archive/FAPROTAX/\\n\\nLouca, S., Parfrey, L.W., Doebeli, M. (2016) Decoupling function and taxonomy in the global ocean microbiome. Science 353: 1272-1277.\\n\\nhttps://www.science.org/doi/10.1126/science.aaf4507\\n\\n\\u200bPredict prokaryote EC, KO, & MetaCyc function abundances with PICRUSt2 App – identify functions from many different ontologies, including COG, EC, and KO, or any combination of which can be chosen in the parameters. Create new FunctionalProfile objects for the SampleSet, AmpliconMatrix, or both, and outputs a new AmpliconMatrix Object.\\n\\nDouglas, G.M., Maffei, V.J., Zaneveld, J.R.\\n\\n(2020). PICRUSt2 for prediction of metagenome functions.\\n\\n685–688 (2020).\\n\\nhttps://doi.org/10.1038/s41587-020-0548-6\\n\\nPICRUSt2: An improved and extensible approach for metagenome inference\\n\\n(preprint)\\n\\nMost of the relevant KBase Apps can be found in the Utilities section of the Apps panel and filtered through searching with \"GenericsAPI\" or the\\n\\nGenericsAPI Module Catalog\\n\\nPreviousMetagenomics & Community Exploration\\n\\nNextChemical Abundance\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_matrix.html'}),\n",
       " Document(page_content='Chemical Abundance\\n\\nUse abundance matrices with element, compound, or exometabolite data for analysis.\\n\\nData Cleaning and Statistical Analysis\\n\\nAs both chemical abundance and amplicon matrices are similar, similar apps for cleaning, normalizing, and performing statistical comparisons can be done on both types. See the previous page on\\n\\nData Matrices\\n\\nfor information on using these apps.\\n\\nPopular Use Cases of Chemical Abundance Data\\n\\nYou can view a few public Narratives that demonstrate popular use cases for chemical abundance tables in KBase.\\n\\nMetabolomics (and other omics) data representation in Escher pathway maps\\u200bhttps://narrative.kbase.us/narrative/55494\\u200b\\u200bhttps://narrative.kbase.us/narrative/86637\\n\\nCheminformatics expansion of known metabolites (based on known compounds in a metabolic model) and mapping onto unknown peaks from metabolomics data (PNNL Summer School workflows - WHONDRS)\\u200bhttps://narrative.kbase.us/narrative/61753\\u200b\\n\\nReconcile metabolic models to exo-metabolite data - Web of Microbes\\u200bhttps://narrative.kbase.us/narrative/39597\\u200b\\n\\nPreviousData Matrices - Amplicon, Stats\\n\\nNextExpression & Transcriptomics\\n\\nLast modified \\n\\n9mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_chemical-abundance.html'}),\n",
       " Document(page_content='Expression & Transcriptomics\\n\\nSome of the tools in KBase available for expression analysis\\n\\nKBase offers a powerful suite of\\n\\nexpression analysis Apps\\n\\n. Starting with short reads, you can use the tool suite to analyze transcriptomic data in a . You can also compare the expression data with the flux when studying metabolic models in KBase and identify pathways where expression and flux agree or conflict.\\n\\nKBase offers a powerful suite of\\n\\nexpression analysis tools\\n\\n. Starting with short reads, you can use the tool suite to assemble, quantify long transcripts, and identify differentially expressed genes. You can also compare the expression data with the flux when studying metabolic models in KBase and identify pathways where expression and flux agree or conflict\\n\\nReads Management\\n\\n\\u200bAssess Read Quality with FastQC – Read quality analysis\\n\\n\\u200bCutadapt – Remove adapter sequences from reads\\n\\n\\u200bTrim Reads with Trimmomatic – Read trimming and removing Illumina adapters\\n\\nAlignment\\n\\n\\u200bAlign Reads using Bowtie2 – aligns the sequencing reads for a set of two or more samples to long reference sequences of a prokaryotic genome using Bowtie2 and outputs a set of alignments for the given sample set in BAM format.\\n\\n\\u200bAlign Reads using TopHat2 – aligns the sequencing reads for a set of two or more samples to an eukaryotic genome using TopHat2 in order to identify splice junctions between exons with the help of Bowtie2 mapping program.\\n\\n\\u200bAlign Reads using HISAT2 – aligns the sequencing reads for a set of two or more samples to long reference sequences of a genome using HISAT2 and outputs a set of alignments for the given sample set or reads set in BAM format.\\n\\n\\u200bAlign Reads using STAR – aligns the sequencing reads of a single or a set of two (paired end) reads to long reference sequences of a prokaryotic genome using the STAR alignment program.\\n\\nAssembly\\n\\n\\u200bAssemble Transcripts using Cufflinks – assembles transcripts for a given sample or a sample set using Cufflinks so that you can view the relative abundances of the assembled transcripts in a histogram, obtained as output from Cufflinks.\\n\\n\\u200bAssemble Transcripts using StringTie – assembles transcripts for a given sample or a sample set using StringTie so that you can view the relative abundances of the assembled transcripts in a histogram.\\n\\nDifferential Expression\\n\\n\\u200bIdentify Differential Expression using Cuffdiff – uses the Cufflinks transcripts for two or more samples to calculate gene and transcript levels in more than one condition and finds significant changes in the expression levels.\\n\\n\\u200bCreate Differential Expression Matrix using Ballgown – uses the transcripts for two or more samples obtained from either Cufflinks or StringTie to calculate gene and transcript levels in more than one condition and finds significant changes in the expression levels.\\n\\n\\u200bCreate Differential Expression Matrix using DESeq2 – uses the transcripts for two or more samples obtained from either Cufflinks or StringTie to calculate gene and transcript levels in more than one condition and finds significant changes in the expression levels.\\n\\nDownstream Analyses\\n\\n\\u200bFilter Expression Matrix – Filter an expression matrix using either Log Odds Ratio (LOR) or ANalysis of VAriance (ANOVA) algorithms.\\n\\n\\u200bCluster Expression Data – Hierarchical – Perform hierarchical clustering to group gene expression data into a dendrogram.\\n\\n\\u200bEstimate K for K-means Clustering – Generate reasonable numbers of clusters (K) for use in the Cluster Expression Data - K-Means App.\\n\\n\\u200bCluster Expression Data – K-means – Perform K-means clustering to group expression data for observing and analyzing patterns of gene expression.\\n\\n\\u200bCluster Expression Data – WGCNA – Perform weighted gene co-expression network analysis (WGCNA) to detect gene clusters and expression patterns.\\n\\n\\u200bView Interactive Heatmap – display heatmap of expressed genes.\\n\\n\\u200bView Multi-cluster Heatmap – display multi-cluster heatmap of expressed genes.\\n\\nPreviousChemical Abundance\\n\\nNextApps in Beta\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_expression.html'}),\n",
       " Document(page_content='Apps in Beta\\n\\nThese KBase Apps are being tested and debugged.\\n\\nApp Panel Menu\\n\\nTo use apps still in development, you can click the “R” in the Apps Panel, that changes it to a “B” which shows apps that are in Beta. Click the \"OK\" button to start using Beta Apps.\\n\\nChoosing Beta will only show the apps that are still being tested and debugged. Click the \"B\" icon to go back to Released apps.\\n\\nClick the \"R\" icon in the Apps menu to toggle to \"B\" version of available apps.\\n\\nPreviousExpression & Transcriptomics\\n\\nNextRunning Common Workflows\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_beta.html'}),\n",
       " Document(page_content='Running Common Workflows\\n\\nUser guides on how to run different workflows in KBase.\\n\\nHere is an example outline of the major workflows and datatypes in KBase. The unboxed labels represent datatypes, while each colored box represents a single KBase App. The box colors signify the category of functionality, and the numbers in parentheses indicate the number of alternative apps that implement each function. Apps that require a genome as input are marked with a green “G” icon.\\n\\nFind example workflows here:\\n\\n1.\\u200bAssembling & Annotating Microbial Genomes\\u200b\\n\\n2.\\u200bComparative Genomics & Phylogenetic Analysis\\u200b\\n\\n3.\\u200bConstructing Metabolic Models\\u200b\\n\\n4.\\u200bMetagenomic & Community Analysis\\u200b\\n\\n5.\\u200bRNAseq & Expression Analysis\\u200b\\n\\n6.\\u200bSamples & Metadata\\u200b\\n\\nPreviousApps in Beta\\n\\nNextAssembling & Annotating Microbial Genomes\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows.html'}),\n",
       " Document(page_content='Assembling & Annotating Microbial Genomes\\n\\nRunning assembly and annotation in KBase\\n\\nKBase enables de novo assembly of prokaryotic NGS reads from various sequencing platforms. Assemblies can then be annotated with RAST or Prokka, enabling you to explore structural and functional features of a Genome or use it in other analyses.\\n\\nSteps for\\n\\nassembly and annotation\\n\\ninclude:\\n\\n1.Upload reads to assemble.\\n\\n2.\\u200bAdd your Reads data object to the Narrative.\\n\\n3.Search the App Catalog and insert an Assembly App Use one of these tools to create an Assembly object.\\n\\n4.Search the App Catalog and insert an Annotation App\\u200bUse one of these tools to annotate the Assembly object.\\n\\n5.Examine assembly job statistics and genome annotation.\\n\\nNarrative Tutorials\\n\\n\\u200bMicrobial Genome Assembly and Annotation Tutorial – guides through assembling a set of NGS short reads into contigs and annotate the assembled contigs, assigning biological functions derived from RAST\\n\\n\\u200bGenome Analysis Tools and Features – a how-to guide for various tools to analyze genomic sequencing data in KBase from quality control and assessment of NGS reads, de novo assembly, and genome annotation\\n\\n\\u200bGenome Analysis 1: Drafting Isolate Genomes – demonstrates how to draft isolate genomes from quality assessment to assembly and annotation\\n\\n\\u200bGenome Analysis 2: Identifying Features of Interest in Genomes – shows how to build a workflow for searching and identifying features within your genomes\\n\\nGenome Analysis 3: Domain Annotation and Feature Profiling – guides users through annotating protein families and analyzing features across a set of genomes for comparative analysis\\n\\nVideo Tutorial\\n\\nIntroduction to Genome Analysis\\n\\nPreviousRunning Common Workflows\\n\\nNextFAQ: Assembly and Annotation\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_assembly-annotation.html'}),\n",
       " Document(page_content='FAQ: Assembly and Annotation\\n\\nBefore submitting them for annotation. Is there a direct way to import to JGI or do we have to save FASTA files and separately submit to JGI/IMG?\\n\\nThis page on transferring data from JGI to KBase takes you through the process. You can also start from the KBase search and use the JGI tab.\\n\\nWhat is the typical threshold to determine whether our assembled genome is contaminated?\\n\\nIn general, a genome that is >90% complete and <5% contaminated is high quality. A rough guide to the quality of MAGs and SAGs can be found here:\\n\\nhttps://www.nature.com/articles/nbt.3893\\n\\nIs there a limit on the size of the FASTQ files I can upload to KBase?\\n\\nAssemblers currently have an upper limit of between 180,263,840 paired reads and 240,351,788 reads depending on complexity. If the job has been run twice, exceeded the 7 day limit, and your data is in this size range, it may be too big for KBase at this time.\\n\\nWhat is the best assembler?\\n\\nThe “best” assembler often depends on the user. Sometimes the user may want the most contiguous metagenome, or an assembly with minimized assembly artifacts. Users can use multiple assemblers and choose whichever results in the best assembly for their purpose.\\n\\nHow do I compare the results of various assemblers?\\n\\nYou can use the\\n\\nQUAST App\\n\\nto compare the contig distributions of Assembly objects.\\n\\nDoes KBase support co-assembly?\\n\\nYes, but the results are unstable currently above 200 M reads (Illumina 150bp x 2). Use the\\n\\nMerge Reads Libraries App\\n\\nto get a combined Reads Library object.\\n\\nJGI/IMG recommends KBase for genome assembly before submission to them for annotation. Is there a direct way to import to JGI or do we have to save FASTA files and separately submit to JGI/IMG?\\n\\nThese\\n\\ninstructions on transferring data from JGI to KBase\\n\\ntakes you through the process. You can also start from the KBase search and use the JGI tab.\\n\\nWhat is the difference between RAST and Prokka annotations?\\n\\nIn addition to having different options in the app, their method for assigning the annotation is different. The determination of better or worse is in the eye of the beholder. The primary advantage of RAST is its linking to our metabolic modeling. The RAST functional roles are considered a controlled vocabulary where we map specific RAST annotations to biochemical reactions in the model, so if you plan to build metabolic models, you should annotate with RAST. Because RAST tends to assign more hypothetical proteins, some people will run Prokka first, and then reannotate with RAST using the Retain old annotation for hypotheticals option.\\n\\nWhat is the difference between Annotate Microbial Genome and Annotate Microbial Assembly?\\n\\nAnnotate Microbial Assembly\\n\\ntakes an Assembly object, follows it by gene calling using algorithms from Prodigal and Glimmer, and then functional annotation.\\n\\nAnnotate Microbial Genome\\n\\ntakes a Genome object, does not call genes, and instead preserves the original gene calls. It then re-annotates the genes, overwriting previous annotations.\\n\\nIs it possible or necessary to manually curate annotations, or are the RAST and Prokka annotations sufficient?\\n\\nManual curation of annotations is not supported on-system. RAST and Prokka are likely sufficient for many applications, but as you mention, for difficult-to-annotate or highly divergent metabolic genes you may need to use additional tools. In addition to RAST and Prokka, there are on-system tools available for feature annotation using pre-generated hidden Markov models, which could be useful for higher-resolution annotation.\\n\\nDoes RAST annotate archaea and protists?\\n\\nRAST\\n\\nprimarily annotates bacteria and archaea, and may be limited with protists.\\n\\nCan I annotate plants or fungi?\\n\\nTools available for plant annotations include\\n\\nAnnotate Plant Transcripts with Metabolic Functions\\n\\nand\\n\\nAnnotate Plant Enzymes with OrthoFinder\\n\\n. These tools use the PlantSEED curated Database.\\n\\nFor fungi, it is recommended to use external tools and then import the annotated genomes into KBase.\\n\\nAre there tools specialized for fungal data?\\n\\nCurrently there is a tool to construct draft metabolic models of fungal species (\\n\\nBuild Fungal Model\\n\\n) that uses highly curated published fungal models as the underlying biochemistry data.\\n\\nPreviousAssembling & Annotating Microbial Genomes\\n\\nNextComparative Genomics & Phylogenetic Analysis\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_assembly-annotation_faq-assembly-annotation.html'}),\n",
       " Document(page_content='Comparative Genomics & Phylogenetic Analysis\\n\\nRunning comparative and phylogenetic Workflows in KBase\\n\\nKBase’s comparative genomics and phylogenetic analysis\\n\\ntools\\n\\nenable researchers to understand evolutionary relationships between organisms and explore structural and functional variance across genomes. By integrating public datasets with user data and tools to search for sequence homology, explore gene orthology, and construct phylogenetic linkages between organisms, KBase provides an advanced and useful platform for comparative genomics research.\\n\\nNarrative Tutorials\\n\\n\\u200bBuild a Gene Tree – guides the user through the process of building a gene tree for the \"Sulfate Adenylyltransferase\" (Sat) gene from Candidatus Desulforudis audaxviator MP104C\\n\\n\\u200bPhylogenomics of Sulfate Reducing Clostridia - Tutorial - Part 1: Comparative Functional Assignment – guides the user through the process of whole-genome phylogeny, homology, and domain family functional profiling.\\n\\n\\u200bGenome Analysis Tools and Features – demonstrates some workflows for comparative analysis of genomes, including searching for features, annotating protein families, and performing phylogenetic analysis.\\n\\nVideo Tutorial\\n\\nPreviousFAQ: Assembly and Annotation\\n\\nNextFAQ: Comparative Genomics\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_comparative-genomics.html'}),\n",
       " Document(page_content='FAQ: Comparative Genomics\\n\\nWhat does the tail <1% each taxon mean in the Kaiju output?\\n\\nThis option allows users to exclude rare lineages which would congest the plot. However, underlying analysis is not affected and the data can be replotted.\\n\\nCan I use the GTDB database within Kaiju?\\n\\nYes, you can run the our GTDB-tk Classify app on the same data as you run through Kaiju and compare. The GTDB-tk app is in the Microbial Communities section of the Apps Panel in the Narrative interface. One distinction: Kaiju runs on raw sequencing data (reads) and GTDB-tk works on assembled contigs.\\n\\nDoes the Kaiju analysis give a text output that can be used in R, for example?\\n\\nYes, at the bottom of the App cell you can see an option to download kaiju_classification.zip and kaiju_summaries.zip which both contain text files.\\n\\nPreviousComparative Genomics & Phylogenetic Analysis\\n\\nNextMetagenomic & Community Analysis\\n\\nLast modified \\n\\n3yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_comparative-genomics_faq-comparative-genomics.html'}),\n",
       " Document(page_content='Metagenomic & Community Analysis\\n\\nRunning Metagenomic and community analysis in KBase\\n\\nKBase has a multitude of tools designed to analyze\\n\\nmicrobial communities\\n\\nand\\n\\nmetagenomic reads\\n\\ndata. They provide a way to understand the functional interactions between species in microbial communities. Annotated genomes extracted from metagenomic samples can be used for metabolic modeling, comparative phylogenomics, functional profiling, and more.\\n\\nNarrative Tutorials\\n\\n\\u200bGenome Extraction from Shotgun Metagenome Sequence Data – this is a good example for understanding a metagenomic pipeline.\\n\\n\\u200bUsing Metagenomes to Discover Novel Microbial Linages in KBase – demonstrates the application of bioinformatic tools for metagenome assembly and genome binning. This Narrative highlights the effective use of employing multiple genome binners and a bin optimization approach.\\n\\n\\u200bMetagenome-Assembled Genome Extraction from a Compost Microbiome Enrichment – demonstrates an in-depth example of metagenome-assembled genome extraction and analysis to obtain high-quality genomes.Chivian et al. (2023) Metagenome-assembled genome extraction and analysis from microbiomes using KBase. Nature Protocols 18, 208-238. https://doi.org/10.1038/s41596-022-00747-x\\u200bCompare and contrast the enriched environment sample analysis to a Moab Desert Crust environmental sample.\\n\\n\\u200bThe PlantSEED Resource in KBase – shows the streamlined process of annotating plant genome sequences, automatically constructing metabolic models based on genome annotations, and using models to test annotations.\\n\\nMicrobiome Fractionation by Genome Extraction\\n\\nFunctional and Taxonomic Profiling of Metagenome-Assembled Genomes\\n\\nMetagenome-assembled Genome Extraction and Analysis Tutorial Webinar\\n\\nPreviousFAQ: Comparative Genomics\\n\\nNextFAQ: Metagenomics & Community Analysis\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metagenomic-analysis.html'}),\n",
       " Document(page_content=\"FAQ: Metagenomics & Community Analysis\\n\\nI am very new to shotgun metagenomics based assembly and annotations, and there are many apps are listed in KBase. Does KBase have a pre-developed workflow for shotgun metagenomics, starting from assembly, annotations and metabolic pathways mining?\\n\\nKBase tries to be as flexible as possible, so there are many options. One App that you could consider is the JGI Metagenome Assembly App. It is a beta app with a complete workflow, optimized by the JGI, that goes from raw reads to an assembly using BFC, BBTools for read QC and metaSPAdes. The assembled sequence can be binned using the available binning tools. And the individual bins annotated using the standard prokaryotic annotation Apps (Prokka, RAST).\\n\\nFor annotating the complete metagenome assembly, the Prokka App is being updated to allow this, but it is still in beta. For metabolic modeling of an entire community, there is currently one app, Build Community Metabolic Model that requires a set of genomes or bins as input. It doesn't take the entire metagenome annotation as input since it attempts to model the individual members and the transfer between them. It is possible to make a mixed bag model using the entire metagenome annotation, that can be useful to see entire pathways are present in the metagenome.\\n\\nWhy extract genome sequences from metagenomes rather than working with unassembled genome sequences?\\n\\nWhen you have a contiguous fragment of a genome, there will be 1) full-length genes and their protein products, 2) genomic context of the genes [to have a better chance of understanding of which genes are being used as part of the same system/pathway, especially if they are polycistronic (operons)], 3) more accurate phylogenetic placement with consensus placement from multiple genes, and possibly even a clade-specific phylogenetic marker.\\n\\nWhat is the best way to assemble 100-150bp sequencing data in order to recover MAGs?\\n\\nEffective MAG recovery is highly dependent on your sample. If there is a lot of diversity in the sample and/or low read coverage, MAGs are more challenging to recover, regardless of the tools used. This is partly why we recommend evaluating your data prior to assembly, so you can get some idea of what your data look like.\\n\\nWhat causes the contamination in the bins? And what is considered a high quality bin for filtering them out?\\n\\nContamination in the bins can occur for a variety of reasons, including but not limited to contig mis-assembly, limited diversity in kmer space, horizontal gene transfer. In general, a genome that is 90% complete and <5% contaminated is high-quality. A rough guide to the quality of MAGs and SAGs can be found\\n\\nhere\\n\\nHow would I exclude eukaryotes and viruses from a marine metagenome?\\n\\nNote that this is hard to do, not just in KBase, but in general. To do this 1) annotate the entire metagenome assembly, 2) identify contigs that fall into the different categories of interest (bacteria/archaea/viruses/eukaryotes), 3) filter out contigs belonging to eukaryotes/viruses, 4) summarize remaining results. A major challenge here will be the unambiguous identification of the different domains of life, which is sometimes tricky (e.g. prophages). Another note: file manipulation outside of KBase would be required to perform this task - as currently there are no KBase Apps to complete this task.\\n\\nPreviousMetagenomic & Community Analysis\\n\\nNextTranscriptomic Analysis\\n\\nLast modified \\n\\n1yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metagenomic-analysis_faq-metagenomics.html'}),\n",
       " Document(page_content='Transcriptomic Analysis\\n\\nRunning RNA-seq analyses pipelines in KBase\\n\\nKBase offers a powerful suite of\\n\\nexpression analysis tools\\n\\n. Starting with short reads, you can use the tool suite to assemble, quantify long transcripts, and identify differentially expressed genes. You can also compare the expression data with the flux when studying metabolic models in KBase and identify pathways where expression and flux agree or conflict.\\n\\nPrerequisites\\n\\nKBase requires a reference genome to guide the analysis of short reads.\\n\\n1.\\u200bImport Genome\\u200b\\n\\n2.\\u200bImport Short Reads: The reads must be a set of single-end, paired-end, or interleaved paired-end reads in FASTA, FASTQ, or SRA format.\\n\\n3.\\u200bCreate a SampleSet: Run the Create RNA-seq Sample Set App to group together your reads into an RNA-seq sample set with associated experimental metadata to run RNA-seq Apps in batch mode wherever appropriate.\\n\\n4.\\u200bQC SampleSet: Run FastQC to assess the read quality of the reads set from the previous step and if needed, run Trimmomatic, Cutadapt, or PRINSEQ to pre-process or filter the reads before starting RNA-seq analysis.\\n\\nRNA-seq Pipeline\\n\\nThe RNA-seq pipeline in KBase is modular and consists of three steps. You can pick any of the multiple Apps available for a given step depending on your preference or individual characteristics of the App.\\n\\n1.Read Alignment: Align reads to map short reads to the reference genome. The output is a set of BAM alignments and Qualimap report. You can download the alignment output object generated by aligner Apps for further analysis.\\n\\n2.Transcriptome Assembly and Quantification: Assemble aligned reads to generate full-length transcripts and quantify transcripts and genes as appropriate. You can view downloadable normalized full expression matrices in FPKM (fragments per kilobase of exon model per million mapped reads) and TPM (transcripts per million).\\n\\n3.Differential Gene Expression: Generate gene- or transcript-level differential expression based on the quantification. Run Create Feature Set/Filtered Expression Matrix From Differential Expression after selecting appropriate q-value and fold change cutoffs as input parameters for the filtering of the differential gene expression.\\n\\nDownstream Expression Analysis\\n\\n1.Filtering: You can create a filtered expression matrix and associated feature set based on fold-change or adjusted p-value. You can also filter an expression matrix based on LOR or ANOVA.\\n\\n2.Clustering: Depending on preference, run the Hierarchical, K-Means or WGCNA clustering App to group features into clusters based on gene expression. You can also visualize the clusters as an interactive heatmap.\\n\\n3.Functional Enrichment: Assess the functional enrichment in plant genomes for a set of features using associated GO terms.\\n\\n4.Integration into Metabolic Models: Assimilate the expression data from RNA-seq into the metabolic models to compare reaction fluxes with gene expression and thus identify pathways where expression and flux agree or conflict.\\n\\nNarrative Tutorials\\n\\n\\u200bE. coli RNA-seq Analysis Tutorial – bacterium-based example of an RNAseq workflow using a HISAT2/StringTie/DESeq2 pipeline\\n\\n\\u200bArabidopsis RNA-seq Analysis Tutorial – plant-based example of an RNAseq workflow using a HISAT2/StringTie/DESeq2 pipeline\\n\\n\\u200bCase Study: Genome-wide Transcriptomics and Plant Primary Metabolism in response to Drought Stress in Sorghum – plant-based example of using KBase to integrate full genome RNA-seq analysis and a metabolic model to generate a reaction matrix Kumari et al. (2021) A KBase case study on genome-wide transcriptomics and plant primary metabolism in response to drought stress in Sorghum. Current Plant Biology 28. https://doi.org/10.1016/j.cpb.2021.100229\\u200b\\n\\nVideo Tutorial\\n\\nFunctional Genomics\\n\\nPreviousFAQ: Metagenomics & Community Analysis\\n\\nNextFAQ: RNA-seq Analysis\\n\\nLast modified \\n\\n9mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_rnaseq.html'}),\n",
       " Document(page_content='FAQ: RNA-seq Analysis\\n\\nCan I assemble a transciptome for a species that does not have a reference genome using KBase?\\n\\nCurrently KBase only offers RNA-seq using the Tuxedo suite, which usese reference genome-guided transcriptome assembly. We do not have stand-alone tools like Trinity for de novo transcriptome assembly.\\n\\nWhere can I find reference genomes?\\n\\nPublic genomes are available through the public data tab. You can directly import genomes from NCBI ReqSeq. You can also import genomes if they are not found in RefSeq. You can use the Upload From Web App if the files are available publicly online, or upload via Globus or the drag and drop upload if they are on your local machine.\\n\\nAre there some references to what tools (aligner, assembler, etc) we should apply to certain situations (e.g. transciptomics, metatranscriptomics, extreme-conditions communitites)? Or should we apply every combination of tools to every analysis we perform, and see which combination performs better?\\n\\nCertain tools are recommended only for prokaryotes, such as Bowtie2. We recommend HISAT2 for aligner, StringTie for assembler, and DEseq2 for differential gene expression for prokaryotes. We do offer an additional tool, Cuffdiff, for differential gene expression. For eukaryotes, we offer Tophat2 or HISAT2 for assembly and Ballgown, Cuffdiff, and DESeq2 for differential gene expression. Our suggested pipelines are below: For prokaryotes - Bowtie2/HISAT2 ? StringTie ? DESeq2 (DESeq2 is based on gene counts). If you want to compare it with gene abundance, you can run Cuffdiff instead of DESeq2 at the last step. For Eukaryotes - HISAT2 ? StringTie ? DEseq2/Ballgown (you can run with Ballgown only if you want to compare these two different algorithms for differential gene expression). Though Cuffdiff can also work with eukaryotes but it is quite slow and too compute intensive for eukaryotes. So, Cuffdiff is only recommended if you have very few samples.\\n\\nWhen you have replicates, is the analysis done for each of the replicates, or are the replicates pooled to do one analysis?\\n\\nThe RNA-seq apps such as aligners (HiSat, TopHat2, Bowtie) and transcriptome assemblers (StringTie, Cufflinks) work on each of the replicates and generate corresponding analysis output objects. However, these apps also wrap the individual results into a set to pass conveniently for the downstream app. However, the Apps such as DESeq2 and ?Create Feature Set/Filtered Expression Matrix From Differential Expression? by definition need to operate on two or more conditions and operate on pools of replicates as opposed to individual replicates.\\n\\nCan I compare RNA sequencing data with genome-scale metabolic modes?\\n\\nYes, the Compare Flux with Expresson allows you to compare the two. You can also use the ?Run Flux Balance Analysis? app to predict fluxes using your expression data. In this approach, you pick a threshold expression level (as a percentile), and the app tries to turn off reactions associated with genes expressed below the threshold and turn ?on? reactions associated with genes expressed above the threshold. We are actively working on this method now, but you are welcome to try it out. It?s fully deployed. Just load you expression data into KBase using the upload tool or by running the RNA-seq pipeline (which will make a matrix for you). Then select this matrix in the advanced parameters of the Run Flux Balance Analysis app, and specify which column of the matrix you want to use to constrain your fluxes.\\n\\nDoes KBase support metatransciptomic assembly?\\n\\nCufflinks and Stringtie are available for metatranscriptomics.\\n\\nPreviousTranscriptomic Analysis\\n\\nNextConstructing Metabolic Models\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_rnaseq_faq-rnaseq.html'}),\n",
       " Document(page_content='Constructing Metabolic Models\\n\\nRunning metabolic modeling workflows in KBase\\n\\nKBase has a suite of\\n\\nApps\\n\\nand data that support the reconstruction, prediction, and design of metabolic models in bacteria, fungi, and plants using functionality from\\n\\nModelSEED\\n\\nand\\n\\nPlantSEED\\n\\ndatabases. Genome-scale metabolic models are primarily reconstructed from protein functional annotations. These genome-scale metabolic models can be used to explore an organism’s growth in specific media conditions, determine which biochemical pathways are present, optimize production of an important metabolite, identify high flux pathways, and more.\\n\\nMetabolic models can be used to evaluate an organism’s metabolic capability by simulating growth under different conditions to answer important biological questions such as:\\n\\nWhat biochemical pathways are present?\\n\\nWhat are the high flux pathways under a certain growth condition?\\n\\nCould the organism grow anaerobically?\\n\\nWould it grow under certain minimal media conditions?\\n\\nCould the organism be optimized to produce a particular drug molecule or industrially important biofuel?\\n\\nWorkflow\\n\\nFlowchart of\\n\\nApps\\n\\nused in\\n\\nMetabolic Modeling\\n\\nLearn More\\n\\nThe flowchart above shows KBase’s metabolic modeling tools (green) as well as some other analysis tools. Check the\\n\\nApp Catalog\\n\\nfor the latest set of metabolic modeling analysis tools in KBase.\\n\\nThe\\n\\nvideo tutorial\\n\\nbelow presents an introduction to building metabolic models in KBase. The\\n\\n“Microbial Metabolic Model Reconstruction and Analysis” Narrative tutorial\\n\\nlets you see and try out for yourself some examples of KBase’s metabolic modeling functionality in action. Common questions and answers about KBase’s metabolic modeling tools can be found in the\\n\\nMetabolic Modeling FAQ\\n\\nNarrative Tutorials\\n\\n\\u200bMicrobial Metabolic Model Reconstruction and Analysis Tutorial – Guides through a workflow on how to draft a metabolic model and use KBase Apps to gapfill, run flux balance analysis and simulate growth.\\n\\n\\u200bConstructing and Analyzing Metabolic Flux Models of Microbial Communities – Provides a case study on current research in modeling the growth and behavior of microbial communities.\\n\\nModeling Central Metabolism and Energy Biosynthesis across Microbial Life: Publication; Narrative\\u200b\\n\\nMicrobial Community Metabolic Modeling: A Community Data-Driven Network Reconstruction: Publication; Narrative\\u200b\\n\\nCase Study: Genome-wide Transcriptomics and Plant Primary Metabolism in response to Drought Stress in Sorghum – plant-based example of using KBase to integrate full genome RNA-seq analysis and a metabolic model to generate a reaction matrix. Publication; Narrative.\\n\\nMetabolomics\\n\\n\\u200bModeling and Integration of Omics Data in KBase – Demonstrates model construction, integration of transcriptomics and metabolomics data, and an example of data representation on metabolic maps based on a Yeast example.\\n\\n\\u200bPredicting Novel Compounds and Reactions using PickAxe – Demonstrates cheminformatics expansion of known metabolites, based on compounds within a metabolic model, and using the PickAxe App to map metabolomics data.\\n\\nVideo Tutorials\\n\\nBuild Metabolic Model Tutorial\\n\\nIntroduction to Metabolic Modeling in KBase\\n\\nPreviousFAQ: RNA-seq Analysis\\n\\nNextConstructing and Analyzing Metabolic Flux Models of Microbial Communities\\n\\nLast modified \\n\\n3mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metabolic-models.html'}),\n",
       " Document(page_content='Constructing and Analyzing Metabolic Flux Models of Microbial Communities\\n\\nKBase has a suite of tools and data that support the reconstruction, prediction, and design of metabolic networks in individual microbes and microbial communities. A recent review article in Hydrocarbon and Lipid Microbiology Protocols discusses and compares approaches to modeling the growth and behavior of microbial communities, and shows how to perform the analyses in KBase. The analysis workflows described in the article are available as Narratives (linked below) that you can copy to your own KBase account and rerun, perhaps changing some of the parameters or adding your own datasets.\\n\\nAbstract\\n\\nHere we provide a broad overview of current research in modeling the growth and behavior of microbial communities, while focusing primarily on metabolic flux modeling techniques, including the reconstruction of individual species models, reconstruction of mixed-bag models, and reconstruction of multi-species models. We describe how flux balance analysis may be applied with these various model types to explore the interactions of a microbial community with its environment, as well as the interactions of individual species with each other. We demonstrate all discussed model reconstruction and analysis approaches using the Department of Energy’s Systems Biology Knowledgebase (KBase), constructing and importing genome-scale metabolic models of Bacteroides thetaiotaomicron and Faecalibacterium prausnitzii and subsequently combining them into a community model of the gut microbiome. We also use KBase to explore how these species interact with each other and with the gut environment, exploring the trade-offs in information provided by applying each metabolic flux modeling approach. Overall, we conclude that no single community modeling approach is better than the others, and often there is much to be gained by applying multiple approaches synergistically when exploring the ecology of a microbial community.\\n\\nNarratives\\n\\nSingle-Species Model Reconstruction and Analysis: https://narrative.kbase.us/narrative/ws.10779.obj.1\\u200b\\n\\nMixed-Bag Model Reconstruction and Analysis: https://narrative.kbase.us/narrative/ws.10786.obj.1\\u200b\\n\\nMixed-Bag Model Reconstruction and Analysis (VERSION 2): https://narrative.kbase.us/narrative/notebooks/ws.15122.obj.2\\u200b\\n\\nMulti-Species Model Reconstruction and Analysis: https://narrative.kbase.us/narrative/ws.10824.obj.1\\u200b\\n\\nSpecies Interaction Analysis: https://narrative.kbase.us/narrative/ws.10778.obj.1\\u200b\\n\\nReference: Faria JP, Khazaei T, Edirisinghe J, Weisenhorn PB, Seaver S, Conrad N, Harris N, DeJongh M, Henry CS (2016)\\n\\nConstructing and Analyzing Metabolic Flux Models of Microbial Communities\\n\\n. Hydrocarbon and Lipid Microbiology Protocols, 10.1007/8623_2016_215.\\n\\nVisit our\\n\\nPublications page\\n\\nfor more articles that describe research done using KBase.\\n\\nPreviousConstructing Metabolic Models\\n\\nNextFAQ: Metabolic Modeling\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metabolic-models_metabolic-flux-models.html'}),\n",
       " Document(page_content='FAQ: Metabolic Modeling\\n\\nKBase has a suite of analysis apps and data that support the reconstruction, prediction, and design of metabolic networks in microbes and plants. These tools can help advance efforts to optimize microbial production of a certain biofuel, find the minimal media conditions under which that fuel is generated, or predict soil amendments that improve the productivity of plant bioenergy feedstocks.\\n\\nThe\\n\\n“Microbial Metabolic Model Reconstruction and Analysis” Narrative tutorial\\n\\nlets you see some of this functionality in action.\\n\\nBelow are answers to common questions about KBase’s metabolic modeling tools.\\n\\nFor more information — including descriptions of the underlying algorithms, inputs, and outputs — see the\\n\\nMetabolic Modeling Apps\\n\\nWhat is Flux Balance Analysis?\\n\\nOne of the best papers on the fundamentals of FBA is the aptly titled\\n\\n\"What is Flux Balance Analysis\" by Jeffrey Orth, Ines Thiele, and Bernhard Palsson.\\n\\nWe highly recommend starting there to understand how FBA works in general.\\n\\nFor a detailed description of Model SEED, which KBase uses for modeling, see\\n\\nHigh-throughput generation, optimization and analysis of genome-scale metabolic models by Chris Henry, et al.\\n\\nCan Prokka annotations be used for metabolic modeling?\\n\\nYou can use Prokka to annotate, but we recommend using RAST for metabolic models because we use the RAST functional roles as a controlled vocabulary to derive metabolic reactions.\\n\\nWhat does the KBase Gapfill Metabolic Models App do?\\n\\nDraft metabolic models lack a number of essential reactions due to missing or inconsistent annotations. Some of the most common problems arise from having missing transporters (reactions that move metabolites across cell membranes) because they are difficult to annotate well. Consequently, draft models often are unable to generate biomass on media where the organism typically is capable of growing.\\n\\nThe KBase gapfilling process compares the set of reactions in your metabolic model to a database of all known reactions and attempts to find a minimal set of reactions that, when added to your model, will allow it to grow. The set of reactions (known as the gapfilling solution) can then be integrated into your model, creating a new model capable of growth on the media used in the gapfilling process.\\n\\nThe gapfilling app uses a cost function associated with each internal reaction and transporter to find a solution that uses the fewest reactions to fill all gaps, and it does this without extra knowledge about the organism’s biochemistry.\\n\\nFor more information on gapfilling, see the Henry et al. 2010\\n\\nModelSEED paper\\n\\nin Nature Biotechnology.\\n\\nHow does the underlying gapfilling algorithm work (i.e., what kind of programming formulation does it use)?\\n\\nAt one time, KBase used a mixed-integer linear programming (MILP) formulation in gapfilling but replaced it with simple linear programming (LP), which minimizes the sum of flux through gapfilled reactions. From extensive experience with both formulations, KBase modeling experts have found that LP solutions are just as minimal as MILP solutions but require far less time to compute. In rare cases where an LP gapfilling would not be minimal, the lower wait time to obtain and adjust a new solution is typically worth a slightly larger solution.\\n\\nThe LP insists on using minimal flux, which almost always corresponds with minimal reactions in a stoichiometrically consistent database. In fact, inefficient solutions were more likely to result from the former MILP formulation because KBase sometimes would cut off the solver before an optimal solution was found in order to limit run-times to 24 hours.\\n\\nIn gapfilling, it is important to recognize that not all reactions are created equal. For example, transporters and non-KEGG reactions are penalized, along with reactions that have missing structures or unknown deltaG. To understand all the penalties, see the Henry et al. 2010\\n\\nModelSEED\\n\\npaper in Nature Biotechnology.\\n\\nThe gapfilling app always has reasons for adding the reactions it does, but those reasons admittedly are not always obvious or biologically relevant. KBase is working on an algorithm that would provide detailed explanations of why each gapfilled reaction is added; we hope to make this feature available soon.\\n\\nWhat is the solver used in your Gapfill optimization?\\n\\nThe Gapfill optimization uses the\\n\\nSCIP solver\\n\\nWe actually use two solvers in KBase. We use GLPK for most pure-linear optimizations. We use SCIP for larger more complex problems, particularly when integer variables are involved. Gapfilling is an example of a problem where we use SCIP.\\n\\nWhat is “complete” media?\\n\\nThe “Complete” media is an abstraction of what’s available in our biochemistry database. Every compound that can be transported into the extracellular compartment–or, in other words, for which a transport reaction is available–is used in the complete media. This list is built in real-time, meaning that whenever you run FBA with complete media, the available transporters are parsed from the media database–and is therefore not stored permanently in any media object in the workspace.\\n\\nTo find out which transport reactions are available when running FBA using Complete media, use the\\n\\nKBase ModelSEED Biochemistry Database\\n\\nfor a reference list of reactions and compounds or\\n\\nBiochemistry Search\\n\\nwithin KBase.\\n\\nTo see the specific compounds that got consumed/excreted when a model was gapfilled in complete media, you can run FBA on complete media. In the ‘FBA output table’ there is a tab called “Exchange Fluxes” which gives a list of excrete/uptake compounds. You can also view the gapfilled reactions in the ‘model output table’ by choosing the “Gapfilling” tab.\\n\\nHow do I choose a media condition for gapfilling, and what is the default media used?\\n\\n“Gapfilling” refers to choosing the minimal set of reactions to add to a draft metabolic model that will enable it to produce biomass in a specified media. KBase’s gapfilling app lets you set the media condition (i.e., the metabolites available in the environment in which you want to analyze the growth of your organism). For more information, see the\\n\\nGapfilling App details page\\n\\nand the previous question.\\n\\nIf you leave the Media field blank, “complete” media will be used by default. Please see the next question for more details about complete media. When used on complete media, the gapfilling app is not limited by the transport reactions already available in the model and therefore invariably will add more transport reactions to the model.\\n\\nIn addition to the default complete media, KBase provides more than 500 media conditions you can use for gapfilling. The\\n\\nNarrative Interface User Guide\\n\\ndescribes how to add these media and other data to your Narrative. You can also\\n\\nupload your own custom media condition\\n\\nChoosing minimal media for the initial gapfilling is often a good idea because it ensures that the gapfilling algorithm will add the maximal set of reactions to the genome-inferred model. This will allow the model to biosynthesize many common substrates necessary for growth—substrates that otherwise would be present in the media. The gapfilling algorithm looks for incomplete metabolic pathways and tries to make an informed guess as to which protein-encoding genes may be responsible for the missing step in the pathway. This process works best when informed by the a priori knowledge that a particular organism can grow on a defined media type (e.g., a model of Helicobacter pylori, an endosymbiont, might require a media containing substrates it can’t biosynthesize in vivo.) Thus, specifying a growth media is beneficial.\\n\\nNote that gapfilling runs can be stacked, meaning that multiple gapfilling solutions are incorporated into the same model, one after the other. In other words, if you gapfill on Complete media first (i.e., don’t specify a media, so it defaults to Complete), gapfilling will add all reactions needed to get the model to grow assuming it can transport all compounds for which a transporter is available in the KBase biochemistry database. If you then gapfill the same (already gapfilled once) model on minimal media, the model will only add any additional reactions needed to grow on the minimal media.\\n\\nIf you want to gapfill on the two media independently, be careful not to overwrite your original model during your gapfilling simulations and then use the same original model in both gapfilling runs.\\n\\nThe\\n\\ntutorial for the Microbial Metabolic Modeling and Analysis of Shewanella oneidensis\\n\\nexplores the differences between two gapfilled solutions: one generated on complete media, and one generated on minimal media.\\n\\nHow can I see the compounds comprising a media used in gapfilling?\\n\\nTo see the list of media compounds for a particular model, open a viewer for the model by either clicking on its name in the Data Panel or dragging it from the panel and dropping it into the main Narrative window. Select the Compounds tab and filter by “e0” (extracellular compartment). This gives you a complete list of compounds transportable by your model.\\n\\nWhich reactions were added in the gapfilling process and why were they added?\\n\\nAfter performing gapfilling on a metabolic model, you can sort the “Reactions” tab of the output table by the “Gapfilling” column to see which reactions were added. In order to compare the differences between the draft model and the gapfilled model, look at the directionality of the reaction in the “Equation” column of the “Reactions” tab of the output table. If the directionality is reversible in the gapfilled model, i.e. “<=>”, this means that the reaction was already in the draft model, albeit previously irreversible, and gap-filling made it reversible. However, if the reaction is irreversible e.g., “=>” or “<=”, then this means it is a new reaction that wasn’t previously present in the draft model, and has been added to the model by the gapfilling algorithm.\\n\\nThe primary function of gapfilling is to ensure that the model in question can produce the biomass of the organism. As gapfilling itself is a heuristic, the result of the algorithm is essentially a prediction and requires manual curation. If the addition or reversibility of a reaction is not desired, you can use the “Custom flux bounds” field to force the reaction to be zero, and re-run gap-filling to find another possible solution. For the gapfilling process, it does not necessarily matter whether a reaction was previously set to be irreversible as long as it produces biomass. However, even if it was determined that a reaction was irreversible a priori, the thermodynamics algorithm used to determine this is also a heuristic.\\n\\nWhat conventions are used for metabolic IDs, reaction IDs, etc?\\n\\nIf you generate a metabolic model based on the\\n\\nBuild Metabolic Model App\\n\\n, the biochemistry represented is based on modelSEED. However, if you generated a model based on a published model and propagate the biochemistry into a new model based on proteome comparison data you could have a mix of ModelSEED and the native reaction and compound IDs from the published model. We recommend using the ModelSEED ontology; you can convert a mixed model using the\\n\\nIntegrate Imported Model into KBase Namespace App\\n\\n(in beta).\\n\\nWhen I import a metabolic model, how do I associate it with a genome?\\n\\nLoad the genome you wish to associate with a metabolic model into your Narrative. Next, when importing your metabolic model, click on the “show advanced options” link shown towards the bottom of the import metabolic model menu. When you click on this link, a new input will appear that will permit you to select the genome for your model. Refer to the\\n\\nFBA Model section of the Data Upload and Download Guide\\n\\nto learn more about correctly importing and associating your model and genome.\\n\\nWhen viewing a model, what do the compartment IDs mean?\\n\\nCompartments are the subcellular localization of compounds, enzymes, and reactions. The table below lists compartment abbreviations for microbes and plants.\\n\\nReactions and compounds belonging to each compartment are identified using compartment notation in square brackets after the five-digit reaction or compound ID (e.g., rxn00001[c0], cpd00001[c0]).\\n\\nThe integer associated with the compartment (e.g., the 0 in c0) represents the index number of the model. For a single-species model, this number will always be zero, but if individual models are merged into a community model, each sub-model will then be assigned a distinct index.\\n\\nCompartment Microbes   Plants   Cytosol (c) ✓ ✓ Plastid (d) \\u200b ✓ Extracellular (e) ✓ ✓ Golgi apparatus (g) \\u200b ✓ Mitochondria (m) \\u200b ✓ Nucleus (n ) \\u200b ✓ Periplasm (p) ✓ \\u200b Endoplasmic reticulum (r) \\u200b ✓ Vacuole (v) \\u200b ✓ Cell wall (w) \\u200b ✓ Peroxisome (x) \\u200b ✓\\n\\nAfter merging models, why do I sometimes end up with one or two more compartments?\\n\\nWhen you combine models into a community model, you have two options:\\n\\n1.Make a compartmentalized multi-species model where each species is contained in a separate compartment leading to the replication of compartments that you describe\\n\\n2.Make a “mixed bag” model where the component models are merged non-redundantly into a single model with no additional compartments.\\n\\nThe first model type is better for predicting potential interactions between species. The second model type is a simpler model and at times, easier to work with. These papers and the narratives associated with these papers explain this in greater detail:\\n\\nHenry CS, Bernstein H, Weisenhorn P, Taylor RC, Lee JY, Zucker J, Song HS. Microbial Community Metabolic Modeling: A Community Data-Driven Network Reconstruction. Journal of Cellular Physiology (2016) 10.1002/jcp.25428.\\n\\nFaria JP, Khazaei T, Edirisinghe J, Weisenhorn PB, Seaver S, Conrad N, Harris N, DeJongh M, Henry CS. Constructing and Analyzing Metabolic Flux Models of Microbial Communities. Hydrocarbon and Lipid Microbiology Protocols (2016)\\n\\nWhat should I know about performing metabolic modeling on plants?\\n\\nThe Plant Metabolic Modeling in KBase is based on the metabolic subsystems described in the\\n\\nPlantSEED\\n\\n. Because PlantSEED was focused on the primary metabolism of higher plants, performing metabolic modeling on lower plants such as bryophytes and algae is not recommended. In principal, the majority of the metabolic networks for plants will be similar. What will be different in the plant primary metabolic network between organisms the gene – reaction associations. From metabolic modeling, you can derive a prediction of the enzymes that would catalyze each of the reactions in the primary metabolic network.\\n\\nHow can I find and download all the biochemical compounds and reactions in KBase, as well as linked information such as Enzyme Commission (E.C.) Numbers?\\n\\nThe biochemistry compounds and reactions can found in the\\n\\nKBase ModelSEED Biochemistry Database\\n\\nfor a reference list of reactions and compounds or using\\n\\nBiochemistry Search\\n\\n. This spreadsheet includes the EC numbers. You can also consult the\\n\\nEnzyme database\\n\\nto see a description of each type of characterized enzyme that has an EC number.\\n\\nDoes media in KBase include sources other than carbon (e.g., nitrogen, phosphorus, sulfur)?\\n\\nEvery KBase media has all the compounds necessary for minimal growth. In most cases the significant difference among media is the carbon source, which generally dictates the media’s name (e.g., “Carbon-D-Glucose”).\\n\\nHow can I specify my own media?\\n\\nYou can generate and import your own media file by specifying the biochemical compounds found in a particular growth environment. Go to\\n\\nKBase ModelSEED Biochemistry Database\\n\\nto see a list of all the biochemical compounds in KBase and find the IDs for those you wish to include in your custom media.\\n\\nMedia can be uploaded as a TSV (tab-separated values) or Excel file with four columns:\\n\\nCompound identifier (e.g., ModelSEED ID, KEGG ID, PubChem ID, or compound name such as “glucose”)\\n\\nConcentration (concentration of compound in mol/L)\\n\\nMin flux (minimum allowed uptake/excretion of compound)\\n\\nMax flux (maximum allowed uptake/excretion of compound)\\n\\nBelow is an example of a media file in TSV format. In this case, the compound IDs in the first column are ModelSEED IDs.\\n\\nNote that when creating a media Excel file, the name of the worksheet that contains your media conditions must be named “MediaCompounds” or it will not upload. To rename a worksheet, find the worksheet tabs at the bottom left of the Excel file. Double-click on the default name of the spreadsheet containing your custom media and name it “MediaCompounds.\"\\n\\nThe values in the “minflux” and “maxflux” columns are absolute units that represent the range of possible fluxes for reactions that transport the media compounds into and out of the cell. A negative flux value corresponds to excretion (transport out of the cell), and a positive flux value corresponds to uptake (transport into the cell).\\n\\nThe combination of maximum and minimum flux will dictate whether uptake and/or excretion of a reagent is allowed. Assigning a positive number to “minflux” for a particular compound essentially forces uptake of that compound. Conversely, assigning a negative number to “maxflux” will force excretion of the compound.\\n\\nIf you specify a range between “minflux” and “maxflux,” then you will need to look at the list of “Exchange Fluxes” in the output after running flux balance analysis to observe what actually happened during the growth simulation.\\n\\nPlease see the\\n\\nMedia section of the Data Guide\\n\\nfor more information on how to configure your media file.\\n\\nHow are units of flux measured within the media?\\n\\nMinimum and maximum flux are measured in mmol per gram cell dry weight per hour.\\n\\nWhat is the unit of measurement for the estimated biomass from a metabolic model?\\n\\nThe biomass components are normalized so that a result of 1 (objective value/growth/flux through biomass reaction) is estimated to be 1 gram of dry weight of biomass.\\n\\nHow is media data structured in KBase?\\n\\nAs with all KBase data types, the technical “spec” files of media data can be viewed at\\n\\nnarrative.kbase.us/#spec/type/KBaseBiochem.Media\\n\\nFrom here, you can find details on the structure of media data in KBase by navigating the following path: “KBaseBiochem” → “Types”→ “Media” → “Spec file.” Scroll to the bottom of the media spec file to see the overall data structure, which should look like this:\\n\\ntypedef structure {\\n\\nmedia_id id;\\n\\nstring name;\\n\\nsource_id source_id;\\n\\nstring source;\\n\\nstring protocol_link;\\n\\nbool isDefined;\\n\\nbool isMinimal;\\n\\nbool isAerobic;\\n\\nstring type;\\n\\nstring pH_data;\\n\\nfloat temperature;\\n\\nstring atmosphere;\\n\\nstring atmosphere_addition;\\n\\nlist<MediaReagent> reagents;\\n\\nlist<MediaCompound> mediacompounds;\\n\\n} Media;\\n\\nWhat is the function of the .gffba file? What is the difference between the .gffba output and Flux Balance Analysis output?\\n\\nThe .gffba object is generated automatically by gap filling and it holds the fluxes corresponding to the solution selected by gap filling to make your model grow in the specified environment. It is useful as a means of exploring why various reactions were added, as well as diving into the gap filling result in detail. We are working to improve the functionality of this data object.\\n\\nThe .gffba output and Flux Balance Analysis output are the same. It is important to understand that a gap filing simulation is a form of FBA, and as such, it generates a flux solution, which is exposing to you via the .gffba object.\\n\\nHow can I determine if my metabolic model will grow in a certain media?\\n\\nOnce you have built a metabolic model, you can perform flux balance analysis ( FBA) to calculate the flow of metabolites through your model. FBA results can be used to predict the growth rate of an organism under certain conditions or the production rates for particular metabolites of interest.\\n\\nTo determine whether your organism grew in the media that you specified for FBA, find the Overview tab in the FBA results table. Among the summary information in this tab is the objective value. This value is is significant because it represents the maximum achievable flux through the biomass reaction of the metabolic model. An objective value of 0 (or something very close to 0) means that the model did not grow on the specified media.\\n\\nFor more information on the\\n\\nRun Flux Balance Analysis\\n\\nApp.\\n\\nAfter running flux balance analysis, what does the sign of a reaction flux mean?\\n\\nA positive flux value indicates that a reaction is proceeding from left to right as written in the equation, while a negative value indicates the reaction is running from right to left.\\n\\nFor example, the following table shows a portion of the “Reaction fluxes” tab from the output of the\\n\\nRun Flux Balance Analysis\\n\\nApp. Here, the positive flux in the “Flux” column indicates that rxn00543 is proceeding from left to right—that is, ethanol is being oxidized to acetaldehyde.\\n\\nReaction Name Flux Min flux (Lower bound) Max flux (Upper bound) Class Equation rxn00543[c0] Ethanol-NAD\\noxidoreductase 100 -100\\n(-1000)  100\\n(1000)  Variable NAD[c0] + Ethanol[c0] <=>\\nAcetaldehyde[c0] + NADH[c0] + H+[c0]\\n\\nThe “Min flux,” “Max flux,” and “Class” columns show that other valid FBA solutions for the model have flux values for rxn00543 that range from -100 to 100. This reaction therefore is classified as “Variable” because it can proceed in either direction.\\n\\nWhat do the various reaction classifiers mean?\\n\\nReaction classifiers are assigned when using Flux Variability Analysis, FVA. In FVA, the global objective (biomass) is fixed at its optimal value, then each reaction, iteratively, is optimized independently to find both the maximal and minimal value that is possible given that the global objective must still be reached.\\n\\nVariable – the reaction has positive maximal and negative minimal values, meaning that it can go in either direction.\\n\\nPositive variable – the reaction has a positive maximal, and a zero minimal, meaning that it can either be zero, or it can go from left to right.\\n\\nNegative variable –  the reaction has a zero maximal, and a negative minimal, meaning it can either be zero, or it can go from right to left.\\n\\nIA (InActive) – the reaction is blocked and cannot have a non-zero value.\\n\\nCan I customize the flux boundaries for reactions in my model?\\n\\nTo constrain the flux of a reaction, you can export your model, edit it, and re-import it.\\n\\nAlternatively, in the case of a transport reaction, you can change the uptake of a media compound by editing the media object itself. For example, setting the maximum flux of any media compound to zero (or a small number) will prevent its uptake. This is contingent, of course, on the media itself being defined, so if you performed gapfilling on the default complete media, you will have to re-gapfill on an actual defined media object of your choice.\\n\\nAlso, in the\\n\\nRun Flux Balance Analysis\\n\\nApp, you can change the “bounds” on a specific reaction to whatever you want using the “Custom flux bounds” field under advanced options.\\n\\nNote that the help text next to the custom flux field provides an example of the proper format for setting bounds:\\n\\nSpecify a number for the lower bound (e.g. “0”)\\n\\nInsert a semicolon (no spaces!)\\n\\nType in the reaction ID or compound ID\\n\\nInsert another semicolon\\n\\nSpecify a number for the upper bound (e.g. “5”)\\n\\nSo to set “rxn00006” at a flux of “5,” you would set this custom bound: 5;rxn00006;5.\\n\\nHow can I edit my media or metabolic model (add or delete reactions, compounds, or biomass)?\\n\\nAfter you have drafted your metabolic model with the Build Metabolic Model app, you can use the\\n\\nGapfill Metabolic Model App\\n\\nto automatically fill in the gaps of necessary reactions for completing the pathway for a set of specified reactions (this can be either the biomass producing reaction or other specified pathways). You can alternatively use our\\n\\nEdit Media\\n\\nor\\n\\nEdit Metabolic Model\\n\\nApps. This works best if the reactions you\\'re adding at ModelSEED reactions (ID looks like rxn#####). In those cases, you can manually list several reactions to add by ID and associate genes and specify directionality. You can add completely custom reactions as well, but this is harder.\\n\\nWhy did gapfilling fail for a model containing a custom media or custom biomass reaction? What can I do about it?\\n\\nDraft metabolic models tend to have gaps by default. If you simulate FBA on a given media, you would see the objective value to be zero; means the model is not able to grow on the simulated media. In the model output viewer gapfill reactions can be seen and easy to recognize as there are no features mapped to the gapfilled reactions. There is a tool in KBase to try and find gene candidates for gapfilled reactions. If a candidate is found, this can give you more confidence that the gapfilled reaction is real. It\\'s in beta and it\\'s called\\n\\nFind Candidate Genes for a Reaction\\n\\n. Just plug in the list of gapfilled reactions and see the candidates that come out.\\n\\nIf the\\n\\nGapfill Metabolic Model App\\n\\nfails with custom media, first check whether the compound that you are gapfilling for is already available in the reconstructed model. (See instructions for viewing model compounds above.)\\n\\nIf the compound is not in the model, it may be rare or uncommon. Check\\n\\nKBase ModelSEED Biochemistry Database\\n\\nor\\n\\nBiochemistry Search\\n\\nto ensure that:\\n\\nThe compound is actually available in KBase’s biochemistry collection.\\n\\nThere are reactions that involve the compound in question.\\n\\nThere are exchange reactions (transporters) that would enable the compound to be consumed from the media.\\n\\nTo help mitigate this issue, the “Gapfill Metabolic Model” app includes a field for “Source Gapfill Model” under the advanced options. You can specify a source model to supply reactions missing from the KBase database that might prevent the gapfilling from working successfully. If your biomass reaction is from an existing published model, you should import this model and use it as a source database for your gapfilling (because that model should contain all reactions needed to grow on its own biomass reaction). The source model simply supplements the KBase database, which still will be used in gapfilling.Successfully running the gapfilling app with a custom biomass reaction depends on whether the custom biomass includes compounds that can be produced given all the balanced, approved reactions in the KBase database. For now, non-standard compartments for biomass components may also be a problem.\\n\\nKBase uses a\\n\\nKBase ModelSEED Biochemistry Database\\n\\nthat includes all ModelSEED reactions, and some additional reactions. However, the gapfilling is restricted to a subset of the reactions in the database. First and foremost, KBase filters out imbalanced reactions from gapfilling because they cause mass leakage and a bad gapfilling solution. We also filter out several reactions that cause problems in gapfilling due to strange formulations (e.g., reactions containing lumped or undefined compounds).\\n\\nWhat if my source and translated metabolic models are the same after running the “Propagate Genome-scale Model to Close Genome” App (in beta)?\\n\\nThe\\n\\nPropagate Model to New Genome App\\n\\n(in beta) builds a new metabolic model by translating the gene associations in an existing model to a new genome. This translation is conducted based on gene correspondence tables contained in a proteome comparison object generated in Step 1 of the app. A gapfilling is performed to permit the translated model to grow on a specific media. More information on the model translation step can be found on the\\n\\ndetails page\\n\\nIf the translated model resulting from this app is the same as your source model, run through the following checks:\\n\\nMake sure the reactions in the source model have genes associated with them. If they don’t, then the source and translated models will be the same because reactions with no genes are automatically retained in the model propagation process. For this reason, it is very important to associate a genome with your model when you import it. It is equally important that the gene IDs in this associated genome match the gene IDs mapped to the reactions in your imported model exactly. Otherwise, the genes in the genome will not be properly mapped to the reactions in the model. This proper mapping is absolutely essential for apps such “Propagate Genome-scale Model to Close Genome” to work.\\n\\nLook at the proteome comparison object. How different are the genomes that are being compared? How many hits are there? If the genomes are very close, the source and translated model could end up being the same. Such a result is unlikely but not impossible.\\n\\nCheck the translated model before gapfilling. Did you lose some reactions only to have them added back in the gapfilling process?\\n\\nWhat do the “Prediction class” values mean in the output table generated by the “Simulate Growth on Phenotype Data” App?\\n\\nThe\\n\\nSimulate Growth on Phenotype Data\\n\\nApp performs multiple flux balance analyses (FBA) using each media condition in a phenotype dataset and compares the results, which are listed as “Observed normalized growth” (experiments done in a laboratory) vs. “Simulated growth” (simulation of FBA on the model). You can see these results in the Phenotypes tab of the output cell.\\n\\nIn the “Prediction class” column, simulated growth results against observed growth are represented in four classes:\\n\\n— Correct positive (model was predicted to grow, and did)\\n\\n— Correct negative (model was predicted not to grow, and did not)\\n\\n— False positive (model was predicted to grow, but it did not)\\n\\n— False negative (model was predicted not to grow, but it did)\\n\\nWhat do the different colors mean in the visualization of metabolic pathways?\\n\\nBlue means that the reaction is present, white means the reaction is absent. Red means negative flux, green means positive flux. Shade represents intensity of flux. The arrows on the map do not always agree with the direction of reactions as defined in our biochemistry database, so these colors can appear confusing at times. For this reason, we are actually currently working on replacing our KEGG-based metabolic map viz with a new system using dynamically drawn maps in Escher.\\n\\nWhy did I get false negatives after simulating growth, and what can I do about it?\\n\\nFalse negative (FN) results from the “\\n\\nSimulate Growth on Phenotype Data\\n\\n” app mean that the model is not able to grow on laboratory-tested media conditions. The most likely explanation for a FN prediction is that the model is missing reactions required for viable growth in the specified media conditions (e.g., no asparagine metabolism pathway in a media where asparagine is the only carbon source). This error can be corrected by gapfilling the model on that media to add a minimal set of reactions to the model such that growth is permitted.\\n\\nIs it possible to use KBase to integrate RNA sequencing data into existing genome-scale models?\\n\\nYes, there is an app called\\n\\nCompare Flux with Expression\\n\\nthat allows you to compare your GSM with RNA-seq data to identify metabolic pathways where expression and flux data agree or conflict. You can also use the\\n\\nRun Flux Balance Analysis App\\n\\nto predict fluxes using your expression data. In this approach, you pick a threshold expression level (as a percentile), and the app tries to turn off reactions associated with genes expressed below the threshold and turn on reactions associated with genes expressed above the threshold. Just load your expression data into KBase using the upload tool or by running the RNA-seq pipeline (which will make a matrix for you). Then select this matrix in the advanced parameters of the\\n\\nRun Flux Balance Analysis App\\n\\n, and specify which column of the matrix you want to use to constrain your fluxes.\\n\\nWhat exactly is the criteria for gapfilling/ How do we know which reaction should be maximized, added, etc? Should we consider the reactions which are blocked and do gapfilling for those?\\n\\nGap filling is an optimization problem; by default, we gap fill for growth, meaning that the goal of the algorithm is to ensure that given the media provided by the user, all compounds in the biomass reaction (necessary for growth) can be produced by the model. The algorithm searches our biochemistry database and finds reactions that fill gaps that enable the production of the biomass precursors that the draft model without gap-filling can’t produce.\\n\\nThe algorithm tries to minimize the number of reactions that are added to the model to achieve growth since, at the moment, we don’t have a way to map the reactions being added by gap filling to genes with missing or wrong annotations in your genome. This is why gap-filling reactions don’t have gene-protein-reaction associations. If you look at the algorithm in detail in the paper mentioned above, you will see the penalty system we use for adding reactions to a model with gap-filling; for example, thermodynamic feasibility weighs heavily in deciding if a reaction is added to the model or not. Another factor can be the penalty scoring of adding a transporter vs. adding a full biosynthesis pathway if a compound is present in the media.\\n\\nReactions can be gap-filled for many reasons:\\n\\n1.The genome used to build the model has missing or incorrect annotations for a given gene(s), meaning the reaction associated with the gene won’t be placed in the model.\\n\\n2.Our template is missing or has incorrect mappings between gene functional roles and reactions. While we try our best to have accurate mappings in our templates, knowledge for some pathways keep evolving, EC number get re-assigned, and more specific enzyme complex roles are mapped to individual genes, etc. If you find that a reaction missing in your model should have been included, given that the gene is properly annotated in the source genome, please let us know so we can fix it in our templates.\\n\\n3.There is truly unknown knowledge of how some organisms perform certain processes.\\n\\n4.As mentioned above, gap filling will try to make sure all biomass precursors needed for growth are generated by the model. This biomass reaction is mostly generic/universal with some differences in cell wall biomass precursors between gram-negative and gram-positive organisms. Your organism may not have the same biomass requirements as the universal biomass formulation we use, meaning the gap-filling algorithm may gap-fill reactions to produce biomass precursors that are irrelevant to your organism.\\n\\nIf in the gap-filling problem, you decide to maximize the production of a different compound, for example, maximize the production of ethanol by selecting the ethanol transporter instead of the biomass reaction, the gap-filling will act in the same way. Still, this time around will only care to add reactions that fulfill this objective and address gaps in the ethanol production pathway. And important to note that gap filling may add reactions even if your organism cannot produce ethanol. The algorithm does not know the specifics of your organism, but the gap-filling results can help elucidate if this is something or organism can do. It is important to analyze the gap-filling results. Continuing with the previous example. If to produce ethanol in your organism, gap-filling fill only one gap in the pathway, that probably means there was a missing or wrong annotation in your genome, and gap-filling correctly filled the necessary gap. If, on the other hand, gap-filling fills in most of the reactions present in that pathway, it might mean that this is a metabolic capability that your organism doesn\\'t have. Or your source genome has bad quality annotations that miss our reactions mapping completely for this pathway.\\n\\nTLDR: it is very important to analyze the gap-filling results critically to understand if some reactions are truly missing or are being added to satisfy a biomass precursor that is not relevant to your organism. I also recommend you do further literature research on the subject since multiple gap-filling algorithms exist that employ different methodologies.\\n\\nIs there a tutorial, public narrative, publication, or documentation of using a high-quality, curated metabolic model to develop a new metabolic model for a genome of a different strain of the same species?\\n\\nWe have an app Propagate Model to New Genome which can be used to do that. The\\n\\napp catalog page\\n\\nhas good documentation about using the app.\\n\\nThere’s a public Narrative by Chris Henry that you can view as an example here:\\n\\n\\u200bKBase\\n\\n. For context, it is part of this publication:\\n\\nhttps://onlinelibrary.wiley.com/doi/full/10.1002/jcp.25428\\n\\nIn addition, check out this protocol paper:\\n\\nhttps://link.springer.com/protocol/10.1007/978-1-0716-1585-0_13\\n\\nWe welcome your feedback on the information provided in this FAQ. If you have a question about the metabolic modeling tools and data in KBase that is not answered here, or the answer provided didn’t work for you, please contact us.\\n\\nPreviousConstructing and Analyzing Metabolic Flux Models of Microbial Communities\\n\\nNextTroubleshooting\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metabolic-models_faq-metabolic-modeling.html'}),\n",
       " Document(page_content=\"Troubleshooting\\n\\nThis section provides guidance on troubleshooting and issue reporting. We greatly appreciate your feedback as we strive to make KBase an increasingly useful resource for the community. Please don't hesitate to contact us if you encounter a bug or want to suggest a new feature.\\n\\nContents\\n\\n1.\\u200bExperiencing problems with the Narrative User Interface\\u200b\\n\\n2.\\u200bThe Help Board\\u200b\\n\\n3.\\u200bReporting issues to the Help Board\\u200b\\n\\n4.\\u200bFinding an error in the Job Log\\u200b\\n\\n5.\\u200bCommon error messages and their meanings\\u200b\\n\\nPreviousFAQ: Metabolic Modeling\\n\\nNextProblems with the User Interface\\n\\nLast modified \\n\\n1yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting.html'}),\n",
       " Document(page_content='Problems with the User Interface\\n\\nIf you experience problems with KBase’s Narrative user interface (e.g., data, apps, or viewers won’t load or the interface appears to be frozen), here are a few things to try that may help:\\n\\nSee our list of recommended web browsers to make sure your browser is supported by KBase.\\n\\nTry a hard reload of your Narrative (shift-reload in your browser window).\\n\\nLog out of KBase and log in again.\\n\\nClear your browser cache or open the Narrative Interface in an “incognito” or “private” browser window.\\n\\nSelect the “Shutdown and Restart” option in the menu at the top left of the Narrative Interface. A popup will ask you to confirm that you want to restart the Narrative Interface and remind you to save your latest changes.\\n\\nWhat should I do if my Narrative freezes or won’t load properly?\\n\\nSee our list of recommended browsers to ensure yours is supported by KBase.\\n\\nTry a hard reload (shift-reload in your browser window).\\n\\nIf you’ve used earlier versions of KBase, they may be lurking in your cache and messing things up. Try either clearing your cache or launching KBase from an “incognito” or “anonymous” window (which doesn’t use your cache).\\n\\nShut down and restart your Narrative by selecting this option from the top left menu of the Narrative Interface.\\n\\nPreviousTroubleshooting\\n\\nNextHelp Board\\n\\nLast modified \\n\\n2yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_narrative.html'}),\n",
       " Document(page_content='Help Board\\n\\nContact Us\\n\\nThe\\n\\nKBase Help Board\\n\\nis an interactive way for users to report bugs, ask questions, or suggest new features, using an issue tracking system called Jira.\\n\\nUsing the\\n\\nHelp Board\\n\\n, you’ll be able to:\\n\\n\\u200bSubmit bug reports, questions and suggestions without revealing your email address\\n\\nEngage in a two-way dialog with KBase staff as they track down and resolve your issue\\n\\nSearch the issues submitted by other users. If someone else has already reported your problem or suggested the same new feature, you can vote for that issue or add a “Me too!” comment.\\n\\nIncrease your productivity in KBase by seeing other users’ suggestions and workarounds\\n\\nAccess the Help Board\\n\\n1.To get into Jira to use the KBase Help Board, go to https://kbase-jira.atlassian.net/.\\n\\n2.Log in with your Google account or click “Sign up for an account.”\\n\\nSee more on how to report issues to the Help Board to find out how to\\n\\nsubmit your bug reports, questions and suggestions\\n\\nto us.\\n\\nPreviousProblems with the User Interface\\n\\nNextHow to Report Issues\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_support.html'}),\n",
       " Document(page_content='How to Report Issues\\n\\nTo report issues, suggest new features, or get help using KBase tools, please go to our\\n\\nHelp Board\\n\\nand create a ticket.\\n\\nCreate a Ticket\\n\\nOn the\\n\\nKBase Help Board\\n\\n, click the blue Create button in the menu. A window pops up to gather information about your issue. Required fields are marked with a red *; the other fields are optional.\\n\\n1.Choose the issue type (Bug, Question, or New Feature)\\n\\n2.Enter a short summary\\n\\n3.Enter a longer description with more details\\n\\n4.Enter the URL of the affected Narrative (remember to share the Narrative with \"kbasehelp\")\\n\\n5.Include the specific Job ID found within the Job Log\\u200b\\n\\n6.Select the best fit Category\\n\\n7.Include you KBase Username\\n\\n8.The Environment field can be used to let us know which operating system you’re using (e.g., Mac, Windows XP, etc.) and which web browser (e.g., IE 10, Chrome, etc.)\\n\\n9.Attach files, such as screenshots or small data files that demonstrate the problem\\n\\n10.Include the Incident Start and End Times if applicable\\n\\n11.Click the blue Create button to submit your ticket\\n\\nRemember, the more information provided about what went wrong, the more efficiently our team can reproduce and resolve the problem.\\n\\nPreviousHelp Board\\n\\nNextJob Errors and Their Meanings\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_report.html'}),\n",
       " Document(page_content=\"Job Errors and Their Meanings\\n\\nSome job errors give an explanation of the problem and others can be much more cryptic. Here are more common job error messages found within the\\n\\nJob Log\\n\\nalong with the context of the App or job-associated, the meaning of the error, and steps to take toward a solution.\\n\\nIf you have an error message and are looking to find a solution, start with the Common Job Errors or select the specific app category page to use the browser's search function to look for the error message and solutions.\\n\\n1.\\u200bCommon Job Errors – occur across all apps\\n\\n2.\\u200bImport Job Errors – specific to importing data\\n\\n3.\\u200bAssembly App Errors – messages specific to Assembly apps\\n\\n4.\\u200bAnnotation App Errors – messages specific to Annotation apps\\n\\n5.\\u200bFunctional Genomics App Errors – messages specific to apps relevant to Functional Genomics analysis\\n\\n6.\\u200bModeling App Errors – messages specific to Modeling apps\\n\\nPreviousHow to Report Issues\\n\\nNextCommon Job Errors\\n\\nLast modified \\n\\n1yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors.html'}),\n",
       " Document(page_content=\"Common Job Errors\\n\\nThis is a list of common job error messages found within the Job Log for all KBase Apps, what they mean and how to go about fixing them or if a job ticket needs to be submitted.\\n\\nYou can always submit a ticket for help, questions, or follow-up to the\\n\\nKBase Help Board\\n\\nTypes of Errors\\n\\nUE=User fixable or possible user error\\n\\nKE=KBase error\\n\\nUK=Unknown or multiple possibles causes\\n\\nTE=Temporary system problem\\n\\nUse your Browser's search tool to paste in your error message to locate your error message and next steps.\\n\\nCommon Job Error Messages\\n\\n(2, No such file or directory)\\n\\nUK: The app did not produce any output or it produced output but none passed the filters. There may also be app-specific reasons described below.\\n\\nCheck the logs for more information and adjust filters as needed.\\n\\nUnable to build output viewer\\n\\nUK: The job ran for more than 7 days and crashed. There is no output. Dependent on the reason for the crash, rerunning the job may work.\\n\\nResubmit job.\\n\\nOutput file is not found, exit code 123\\n\\nUK: No space left on device. The job may be too big for KBase in the current configuration.\\n\\nResubmit job to make certain the issue isn’t a conflict with other jobs. If this doesn’t work, submit a ticket to the\\n\\nHelp Board\\n\\nOutput file is not found, exit code is 137\\n\\nUK: Cause unknown. Job probably cancelled by another process but not the user.\\n\\nResubmit job.\\n\\nJob was cancelled as it ran overMax allotted time (604800000) milliseconds (10080) minutes\\n\\nUE: Job ran for more than 7 days and finished cleanly. Likely the job is too big.\\n\\nResubmit job, dependent on the reason it took so long.\\n\\n'listener timeout after waiting for [600000] ms'\\n\\nTE: The utility ElasticSearch went down. The fix is a manual process and may not get fixed during nights, weekends, and holidays.\\n\\nSubmit a ticket the the\\n\\nHelp Board\\n\\n. Resubmitting the job every couple of hours until it runs may also work.\\n\\nkafka\\n\\nTE: Any messages with the word kafka in the error. Something went wrong with the system.\\n\\nResubmit job. Report the issue to the\\n\\nHelp Board\\n\\nif resubmitting doesn’t fix the problem.\\n\\nProtocolError, Connection aborted., BadStatusLine\\n\\nUK: Something went wrong with the reporting and cleanup at the end of the job. Intermittent error. The data is fine, but there will be no report at the end. If an object was created, clicking on it in the data panel will create a viewer for the object, which is likely missing.\\n\\nResubmit job if you need the end report.\\n\\nUser XXXX may not read workspace nnnnn\\n\\nUE: The app requires data owned by another user and you do not have access. In another scenario, either you or the user have deleted the original file.\\n\\nAsk the user for access to the needed file or re-import file if deleted.\\n\\nGET http://elasticsearch.kbase.us:9200/kbaseprod.taxon_*,-*_sub/data/_search: HTTP/1.1 503 Service Unavailable\\n\\nTE: A component in KBase needed to be rebooted and will recover in a few minutes.\\n\\nTry resubmitting the job in 10-15 minutes.\\n\\nNo such container:...\\n\\nTE: A known temporary error.\\n\\nResubmit job.\\n\\nToken validation failed: Too many open files'\\n\\nTE: A known temporary error .\\n\\nTry resubmitting the job in 10-15 minutes.\\n\\n502 Server Error: Bad Gateway….\\n\\nTE: A known temporary error.\\n\\nResubmit job.\\n\\nGee whiz, I sure am sorry, but an error occurred. Gosh!...... Object 19 cannot be accessed\\n\\nUE: User’s browser cache is retaining old information. In this case, it is 'Object 19'. One symptom is that the user gets the error but the narrative looks fine to others.\\n\\nTry the following:\\n\\nReload the page\\n\\nClose and reopen your narrative\\n\\nLog out and log back in again\\n\\nClear your cache\\n\\nIf you find something that works, no need to try the others.\\n\\nDetails: 500 globus_xio: ICE negotiation failed.\\n\\nUE: The error generally indicate a firewall issue.\\n\\nPlease refer to this question on the Globus forum for a similar question with its solution. If you are working from home, you can view the Globus documentation for configuring the firewall. If you are on an institution network, you'll probably have to request an exception from IT.\\n\\nPreviousJob Errors and Their Meanings\\n\\nNextThe Job Log\\n\\nLast modified \\n\\n2yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_common.html'}),\n",
       " Document(page_content=\"The Job Log\\n\\nThe Job Log contains run information from apps and other jobs, including important error information.\\n\\nThe Job Log can be found in two places: an app's Job Status tab and the Job Browser (in the sidebar Menu).\\n\\nJob Status tab\\n\\nThe Job Log starts at the bottom with the error highlighted in pink. There may be useful information located just above the pink section.\\n\\nJob Browser\\n\\nFind the line with the job of interest and click on the icon for the piece of paper. This will open up the log, starting at the top.\\n\\nIn the Job Browser, jobs not associated to an app i.e., export, are included within the job list. The Job ID and worker node are easily located and the log can be scrolled through within the pop-out. Job Logs can be downloaded in CSV, TSV, JSON, and TEXT formats using the download (downward-facing arrow into tray) icon.\\n\\nEven if you find an explanation for your error in the\\n\\nlist of common errors\\n\\n, it is okay to\\n\\nsubmit a Help Board ticket\\n\\nwith follow-up questions.\\n\\nHelpful Points\\n\\nIf the output object is created but the job ends in an error, assume that the error didn’t affect the data. Errors sometimes occur in the cleanup and reporting at the end while not affecting the output data.\\n\\nJobs do not run more than 7 days. If the narrative and/or the Job Browser say that a job is still running after 7 days, assume that it is dead. Resubmitting may solve the problem, but it is possible the job might be too big for KBase.\\n\\nAssemblers currently have an upper limit of between 180,263,840 paired reads and 240,351,788 reads depending on complexity. If the job has been run twice, exceeded the 7 day limit, and your data is in this size range, it may be too big for KBase at this time.\\n\\nBowtie appears to have an upper limit between 3.5E+10 and 4.0E+10 bases due to space limitations.\\n\\nIf asked for a job ID, it is a 24-digit hexadecimal number that looks similar to this: 5e1e3234e4b0fb2e6517240a. It is in the first line of the job log.\\n\\nIf asked for the node where the job ran, it is on the first line of the job log and looks similar to Running on chicagoawe163 or Running on uploadworker or Running on uploadworker-prod-dtn2\\n\\nAny file uploaded from a Windows machine might have  DOS-style carriage-return line files along with new-lines. This has been known to cause problems with importing FASTQ files  and may affect other files as well.\\n\\nPreviousCommon Job Errors\\n\\nNextImport Job Errors\\n\\nLast modified \\n\\n2yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_common_job-log.html'}),\n",
       " Document(page_content='Import Job Errors\\n\\nThis is a list of error messages found within the Job Log for Import Jobs, what they mean and how to go about fixing them or if a job ticket needs to be submitted.\\n\\nTypes of Import Job Errors\\n\\nUE=User fixable or possible user error\\n\\nKE=KBase error\\n\\nUse your Browser\\'s search tool to paste in the error to locate possible fixes or next steps.\\n\\nYou can always submit a ticket for help, questions, or follow-up to the\\n\\nKBase Help Board\\n\\nCommon Import Job Error Messages\\n\\nCannot connect to URL: ftp://ftp.imicrobe.us. ...\\n\\nUE: The provided URL cannot be accessed from within KBase.\\n\\nRecheck the URL and permissions. Then try resubmitting the job.\\n\\nInvalid FTP Link: ...\\n\\nUE: The provided URL cannot be accessed from within KBase. Perhaps the option for ‘Direct’ download should be specified instead of ‘FTP’ (e.g., when downloading from the SRA)\\n\\nRecheck the URL and permissions. Then try resubmitting the job.\\n\\nInvalid Google Drive Link ...\\n\\nUE: The provided URL cannot be accessed from within KBase.\\n\\nRecheck the URL and its permission. Then try resubmitting the job.\\n\\nBulk or Batch Imports\\n\\n(2, No such file or directory)\\n\\nUE: The file is not in the Staging Area.\\n\\nVerify that the name is correct and upload is complete. Then try resubmitting the job.\\n\\n\\u200bImport FASTQ/SRA File as Reads from Staging Area\\u200b\\n\\n(2, No such file or directory)\\n\\nKE: The fastqdump ran but the file names are not the expected names.\\n\\nUse the long workaround\\n\\nhere\\n\\nSRA input file type selected. But missing SRA file\\n\\nUE: The format of the file is not recognized.\\n\\nRecheck the file and try resubmitting the job.\\n\\nInvalid FASTQ file ...\\n\\nKE/UE: Sometimes the user has specified the file name wrong. It can also happen because the importer has problems with file names that end in \".1\"\\n\\nUse the long workaround\\n\\nhere\\n\\nError running command: /kb/deployment/bin/fastq-dump ...\\n\\nUE: The file does not appear to be in the expected SRA format.\\n\\nRecheck the file and try resubmitting the job.\\n\\nError running command:pigz ...\\n\\nUE: The file could not be unzipped by KBase and most likely couldn’t be unzipped by the user either.\\n\\nVerify the file is can be unzipped locally.\\n\\nBoth SRA and FASTQ/FASTA file given.\\n\\nUE: The inputs should be either all fastq/a or all SRA.\\n\\nModify the inputs, then try resubmitting the job.\\n\\nSame file [XXX.XXXX.gz] is used for forward and reverse. Please select different files and try again.\\n\\nUE: There are names for both a forward and reverse strand and they are identical.\\n\\nA Single-end read library only needs one name. A Paired-end read library needs two files with different names.\\n\\nFile /kb/XXX.fasta is not a FASTQ file\\n\\nUE: Either the file is not in fastq format or the file extension is not recognized.\\n\\nRecheck that the file is in the right format. Change the extension to .fastq if needed, then try resubmitting the job.\\n\\nInvalid FASTQ file\\n\\nUE: Possible issues\\n\\nThe fastq file includes one or more sequences that are less than 10 bases. Short reads are a problem for some tools.\\n\\nThe fastq file doesn\\'t have the right number of lines. For example, the lines in a single-end file needs to be a multiple of four and interleaved paired-end library should be a multiple of eight.\\n\\nThe options haven\\'t been selected correctly. For example, using an interleaved fastq file but failing to check the Interleaved box. The documentation on FASTQ/SRA Reads may be helpful.\\n\\nThe file might not have the right filename to be recognized.The file is an SRA file and not FASTQ.\\n\\nDOS-style carriage-return line files along with new-lines. Our fasta validation doesn\\'t handle this properly. To remove the carriage return characters use this unix command: tr -d \\'\\\\015\\' < 1.fastq >cleaned_1.fastq\\n\\nReading FASTQ record failed - non-blank lines are not a multiple of four.\\n\\nUE: The number of lines in the FASTQ file are not a multiple of four.\\n\\nRecheck the file and try resubmitting the job.\\n\\nInterleave failed - reads files do not have an equal number of records….\\n\\nUE: Something went wrong trying to interleave the Paired-end files.\\n\\nRecheck the line count of the files. Hidden carriage returns or linefeeds in the file could contribute to the problem.\\n\\nDeinterleave failed - line count is not divisible by 8\\n\\nUE: The interleaved file does not appear to be the correct format.\\n\\nRecheck the file and try resubmitting the job.\\n\\nObject 1: Illegal character in object name\\n\\nUE: The name of the output reads object can’t have spaces or special characters.\\n\\nRename the output file and then try resubmitting the job.\\n\\n\\u200bImport FASTA File as Assembly from Staging Area\\u200b\\n\\nThere are no contigs to save, thus there is no valid assembly.\\n\\nUE: There are no contigs that passed the minimum contig size.\\n\\nAdjust the minimum contig size or other optional parameters. Then try resubmitting the job.\\n\\nThe FASTA header key XXX appears more than once in the file\\n\\nUE: The FASTA header lines may not be unique.\\n\\nRecheck the format of the header lines and try resubmitting the job.\\n\\nThis FASTA file has non nucleic acid characters\\n\\nUE: The file appears to be proteins or special characters instead of DNA.\\n\\nRecheck the file contents, and then try resubmitting the job.\\n\\nThis FASTA file may have amino acids in it instead of the required nucleotides.\\n\\nUE: The file appears to be proteins instead of DNA.\\n\\nRecheck the file contents, and then try resubmitting the job.\\n\\nFASTQ/FASTA input file type selected. But missing FASTQ/FASTA file\\n\\nUE: Selected file does not match the import selected.\\n\\nSelect a valid combination and try resubmitting the job.\\n\\n(\\\\utf-8\\\\, b\\\\PK\\\\x03\\\\x04\\\\x14\\\\x00\\\\x08…….\\n\\nUE: Attempt to import a zip file with multiple files as a single data object.\\n\\nRun the App \\'\\n\\nUnpack a Compressed File in Staging Area\\n\\n\\' on the file and retry resubmitting the job.\\n\\n\\u200bImport GenBank File as Genome from Staging Area\\u200b\\n\\nDuplicate gene ID: XXXX_xxxx\\n\\nUE: Gene IDs within the input file are not unique.\\n\\nEdit gene IDs and try resubmitting the job.\\n\\nThe input directory does not have any files with one of the following extensions .gbff,.gbk,.gb,.genbank,.dat,.gbf\\n\\nUE: The app only recognizes the listed file extensions as valid GenBank files.\\n\\nChange the file extension and try resubmitting the job.\\n\\nXXX is not a valid KBase taxon ID\\n\\nUE: The Taxonomy ID in the advanced parameters is optional and needs to be an integer when specified. The user provided the text string ‘XXX’.\\n\\nUse an integer taxon ID or leave it blank. The information will be picked up from the GenBank file or from the scientific name.\\n\\n\\u200bImport GFF3/FASTA file as Genome from Staging Area\\u200b\\n\\nEvery feature sequence id must match a fasta sequence id\\n\\nUE: The IDs in the ‘sequence source’ lines must match the header lines in the FASTA file in the GFF format.\\n\\nCorrect the GFF format and try resubmitting the job.\\n\\nunable to parse >.....\\n\\nUE: The file may not be in GFF format.\\n\\nRecheck the file format and try resubmitting the job.\\n\\nFeatures must be completely contained within the Contig in the Fasta file.\\n\\nUE: The coordinates for the feature are outside the bounds of the contig.\\n\\nRecheck the file where indicated and try resubmitting the job. In rare instances, the GFF file contains a feature that wraps around the 0 position and the coordinates look like the feature goes off the end of the sequence. The options are to 1) remove the feature from the GFF file, 2) edit the feature so that it is in two parts, or 3) find a GenBank formatted version of the file and resubmit.\\n\\n\\u200bImport Media file (TSV/Excel) from Staging Area\\u200b\\n\\nStarch.csv\" is not a valid EXCEL nor TSV file\\n\\nUE: File format is not recognized.\\n\\nRecheck the file format and try resubmitting the job.\\n\\n\\u200bImport TSV/XLS/SBML File as an FBAModel from Staging Area\\u200b\\n\\ndata file 4.xml either does not use commas or tabs as a separator\\n\\nUE: File format is not recognized.\\n\\nRecheck the file format and try resubmitting the job.\\n\\nNo object with name _Nostoc_azollae__0708\\n\\nUE: The genome does not exist in the narrative\\n\\nFix the genome name and try resubmitting the job.\\n\\n\\u200bUnpack a Compressed File in Staging Area\\n\\nError running command:pigz ...\\n\\nUE: The file could not be unzipped by KBase and most likely couldn’t be unzipped by the user either.\\n\\nVerify the file can be unzipped locally.\\n\\nPreviousThe Job Log\\n\\nNextAssembly App Errors\\n\\nLast modified \\n\\n2yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_import.html'}),\n",
       " Document(page_content=\"Assembly App Errors\\n\\nThis is a list of error messages found within the Job Log for Assembly App Jobs, what they mean and how to go about fixing them or if a job ticket needs to be submitted.\\n\\nTypes of Errors\\n\\nUE=User fixable or possible user error\\n\\nKE=KBase error\\n\\nUK=Unknown or multiple possibles causes\\n\\nTE=Temporary system problem\\n\\nUse your Browser's search tool to paste in your error message to locate your error message, information about the error and next steps.\\n\\nYou can always submit a ticket for help, questions, or follow-up to the\\n\\nKBase Help Board\\n\\nCommon errors across SPAdes Assembly Apps\\n\\n\\u200bAssemble Reads with SPAdes; Assemble Reads with HybridSPAdes; Assemble Reads with metaSPAdes; Filter Assembled Contigs by Length\\u200b\\n\\nThere are no contigs to save, thus there is no valid assembly.\\n\\nUE: There are no contigs that passed the minimum contig size.\\n\\nAdjust the minimum contig size or other optional parameters. Then try resubmitting the job.\\n\\nError running SPAdes, return code: 1\\n\\nUE: Potential explanations\\n\\nThe coverage of your input is so uneven that everything is disconnected.\\n\\nThe reads contain too many k-mers to fit into available memory.\\n\\nIncomplete write! Reason: No space left on device.\\n\\nCoverage not uniform hybridSPAdes ended abnormally\\n\\nFailed to align paired reads left paired reads is not equal to right paired reads cannot specify any data types except a single paired-end library (optionally accompanied by a single library of TSLR-contigs, or PacBio reads, or Nanopore reads) in metaSPAdes mode\\n\\nThe program was terminated by segmentation fault\\n\\nToo many erroneous kmers, the estimates might be unreliable\\n\\nLook for an explanation from SPAdes in the log. If possible, correct the error and try resubmitting the job.\\n\\nDeinterleave failed - line count is not divisible by 8\\n\\nUE: The interleaved file does not appear to be the correct format.\\n\\nRecheck that the files are labeled properly and try resubmitting the job.\\n\\nReads object XXX is marked as containing metagenomic data but the assembly method was not specified as metagenomic\\n\\nUE: The input object and the selected parameters disagree\\n\\nChange the data input or the app parameters and try resubmitting the job.\\n\\nPlasmid assembly requires that one and only one library as input.\\n\\nUE: There is more than one input library.\\n\\nChange the input to be just one library, change the library source, or merge the libraries and try resubmitting the job.\\n\\nInvalid type for object 27459/32/1…..\\n\\nUE: The user is allowed to enter an Assembly for the long reads (like in the MaSuRCA app) but this results in an error.\\n\\nEnsure objects correspond with the object types listed and try resubmitting the job.\\n\\nQUAST reported an error, return code was 4\\n\\nUE: None of the assembly files contains correct contigs (none greater than the minimum contig filter).\\n\\nProvide different files or decrease --min-contig threshold. Then try resubmitting the job.\\n\\n\\u200bVelvet Assembler\\u200b\\n\\nError on Object #1: Illegal character in object name\\n\\nUE: Velvet did not assemble any contigs longer than the minimum length.\\n\\nChange the configuration and try resubmitting the job.\\n\\n\\u200bAssemble with HipMer\\u200b\\n\\n'list index out of range'\\n\\nUE: There is no input data\\n\\nAdd\\n\\ndata\\n\\nto the app and try resubmitting the job.\\n\\nError in HipMER execution\\n\\nUE: Possible causes -\\n\\nThe data was a metagenome, but the metagenome flag was not enabled\\n\\nThe  metagenome flag was enabled but the data was not a metagenome\\n\\nExceeded time limit at NERSC\\n\\nOther data/parameter combinations that result in no returned assembly\\n\\nChange the configuration and try resubmitting the job.\\n\\nUsing input parameters, you have filtered all contigs from the HipMer assembly. Decrease the minimum contig size and try again\\n\\nUE: The minimum contig length is too large.\\n\\nLower the minimum contig length in the parameters and try resubmitting the job.\\n\\n(2, No such file or directory)\\n\\nTE: Hipmer was in the queue too long and did not run.\\n\\nTry resubmitting the job.\\n\\n\\u200bMaSuRCA Assembler\\u200b\\n\\nError running command: XXX/config.txt Exit Code: 1\\n\\nUE: invalid file for PACBIO invalid file for NANOPORE.\\n\\nRecheck the file and try resubmitting the job.\\n\\n\\u200bAssemble Transcripts using StringTie\\u200b\\n\\nGenome at 48203/40/2 does not have reference to the assembly object\\n\\nUE: A previous step (e.g., HISAT) used an assembly instead of a genome as input.\\n\\nGo back to the previous step and use a genome as input.\\n\\ncoercing to unicode: need string or buffer\\n\\nKE: genome related issue with some of the NCBI RefSeq genomes.\\n\\nWorkaround: import the GFF-format genome with correct feature annotations and try resubmitting the job. See\\n\\nhttps://kbase-jira.atlassian.net/browse/PUBLIC-969\\n\\n\\u200bMerge Reads Libraries\\u200b\\n\\nincompatible read library types in ReadsSet\\n\\nUE: Every library in the input to merge must be of the same type, either all Paired-End or Single-End.\\n\\nChange the inputs and try resubmitting the job.\\n\\n[Errno 28] No space left on device\\n\\nUE/KE: Most likely case: The combined size of the libraries exceeds the current KBase hardware. Efforts are being made to increase capacity and to provide a cleaner more informative message. Workarounds should be considered.\\n\\nThe  workarounds include:\\n\\n1.Reduce the number of libraries in the merge\\n\\n2.Run Kaiju to determine which samples have similar profiles. Then try the co-assemblies pairwise using two-at-a-time based on which profiles are most similar.\\n\\nPreviousImport Job Errors\\n\\nNextAnnotation App Errors\\n\\nLast modified \\n\\n2yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_assembly.html'}),\n",
       " Document(page_content=\"Annotation App Errors\\n\\nThis is a list of error messages found within the Job Log for Annotation App Jobs, what they mean and how to go about fixing them or if a job ticket needs to be submitted.\\n\\nTypes of Errors\\n\\nUE=User fixable or possible user error\\n\\nKE=KBase error\\n\\nUK=Unknown or multiple possibles causes\\n\\nTE=Temporary system problem\\n\\nUse your Browser's search tool to paste in your error message to locate your error message, information about the error and next steps.\\n\\nYou can always submit a ticket for help, questions, or follow-up to the\\n\\nKBase Help Board\\n\\nCommon Errors across Annotation Apps\\n\\nRAST: Annotate Microbial Genome / Annotate Multiple Microbial Genomes / Annotate Microbial Assembly / Annotate Multiple Microbial Assemblies\\n\\nProkka:  Annotate Assembly and Re-annotate Genomes with Prokka\\u200b\\n\\n(2, No such file or directory)\\n\\nUE: One or more contigs had header lines longer than 37 characters.\\n\\nEdit the fasta file, upload again, and try resubmitting the job.\\n\\nError invoking method call_features_rRNA_SEED\\n\\nUE: One or more of the FASTA header lines are extremely long.\\n\\nShorten them, re-upload, and try resubmitting the job. .\\n\\nObject #1, 0ae7adb1-6351-4d26-a526-cc09c15e46ee.report has invalid reference: ….\\n\\nUE: There are no input datasets listed.\\n\\nSupply the dataset(s) and try resubmitting the job.\\n\\ntoo many contigs\\n\\nUE: RAST has a limit of 10,000 contigs. Prokka has a limit of 30,000 contigs.\\n\\nDivide the file or bin the contigs before running RAST again by resubmitting the job.\\n\\nError on ObjectSpecification #1: Unable to parse version portion of object reference nnnn/2/1, nnnn to an integer\\n\\nUE: The wrong delimiter was used in the free text of the Assembly list or Genome list.\\n\\nChange the delimiter to a semicolon (;) and try resubmitting the job.\\n\\nError on ObjectSpecification #1: Illegal number of separators /\\n\\nUE: The wrong delimiter was used in the free text of the Assembly list or Genome list.\\n\\nChange the delimiter to a semicolon (;) and try resubmitting the job.\\n\\nname assembly_info is not defined\\n\\nUE: You are running the beta version of the app.\\n\\nChange to the released version.\\n\\n\\u200bAnnotate Plant Transcripts with Metabolic Functions\\u200b\\n\\nThe genome does not contain any CDSs or features!\\n\\nUE: The input genome does not have the needed features.\\n\\nDouble check the file and try resubmitting the job.\\n\\n\\u200bAssess Genome Quality with CheckM\\u200b\\n\\nObject NNN cannot be accessed: User user_name may not read workspace XXXXX\\n\\nUE: The app requires data which is owned by another user and you do not have access. In another scenario, either you or the user have deleted the original assembly.\\n\\nAsk the user for access to the needed file.\\n\\nPreviousAssembly App Errors\\n\\nNextFunctional Genomics App Errors\\n\\nLast modified \\n\\n2yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_annotation.html'}),\n",
       " Document(page_content='Functional Genomics App Errors\\n\\nThis is a list of error messages found within the Job Log for Functional Genomics App Jobs, what they mean and how to go about fixing them or if a job ticket needs to be submitted.\\n\\nTypes of Errors\\n\\nUE=User fixable or possible user error\\n\\nKE=KBase error\\n\\nUK=Unknown or multiple possibles causes\\n\\nTE=Temporary system problem\\n\\nUse your Browser\\'s search tool to paste in your error message to locate your error message, information about the error and next steps.\\n\\nYou can always submit a ticket for help, questions, or follow-up to the\\n\\nKBase Help Board\\n\\nCluster Expression Data: Hierarchical, Estimate K, K-Means\\u200b\\n\\nError running service CLI for method \\'ClusterServiceR.cluster_hierarchical\\' with exit code 1 …...\\n\\nKE: The app is failing and needs maintenance.\\n\\nNothing can be done at this time. KBase is working on a solution to this problem.\\n\\n\\u200bBLASTp prot-prot Search\\u200b\\n\\ncannot have both input_one_sequence and input_one_ref parameter\\n\\nUE: The app requires either a single Query Sequence or a Query Object with single sequence.\\n\\nChange the inputs and try resubmitting the job.\\n\\noutput_one_name parameter required if input_one_sequence parameter is provided\\n\\nUE: If a Query Sequence is used as input, an Output Query Sequence must be named.\\n\\nAdd an output name or change the input. Then try resubmitting the job.\\n\\n\\u200bBuild a Feature Set\\u200b\\n\\nFeature ID AT2G30640 does not exist in the supplied genome\\n\\nUE The selected feature does not exist in the genome.\\n\\nRecheck the selection and try resubmitting the job.\\n\\nInsert Genome or Set of Genomes into SpeciesTree\\n\\nError processing genome [xxx/xxx/x] (Not one protein family member found)\\n\\nUE: The species tree is built using the predicted proteins in the genomes.\\n\\nRun an app that does gene calling/genome annotation on the listed genome(s). Either RAST or Prokka annotation will work.\\n\\n\\u200bPhylogenetic Pangenome Accumulation\\u200b\\n\\nfloat division by zero\\n\\nThe app will fail with this message if there are only 2 genomes in the pangenome.\\n\\nTry running the app with 3 or more genomes.\\n\\nView Function Profile for FeatureSet, Genomes, or a Phylogenetic Tree\\u200b\\n\\nABORT: You must run the RAST SEED Annotation App or use SKIP option….\\n\\nUE: The app is expecting RAST annotation for the input genome(s).\\n\\nThe easiest option is to check the \\'Skip missing genomes\\' box in the advanced parameters. For more complete output 1) rune Annotate Microbial Genome(s) with RAST on the listed genomes and 2) use the genome(s) newly created by RAST as the input to the View Function Profile. Because the names are different, the app does not know how to find the newer version of the genome.\\n\\nABORT: You must run the \\'Domain Annotation\\' App or use SKIP option ….\\n\\nUE: The app is expecting Domain annotation for the input genome(s).\\n\\nEither run the Domain Annotation on the listed genome(s) or check the ‘Skip missing genomes’ box in the advanced parameters.\\n\\n\\u200bClassify Taxonomy of Metagenomic Reads with Kaiju\\u200b\\n\\nmissing or empty krona input file\\n\\nUE: The filters were too restrictive and no output was generated.\\n\\nChange the input parameters and try resubmitting the job.\\n\\n\\u200bGTDB-Tk classify\\u200b\\n\\n(1, \\'/bin/bash -c \"source activate py2 && GTDBTK_DATA_PATH\\n\\nKE: Unknown error.\\n\\nNothing can be done at this time. KBase is working on a solution to this problem.\\n\\n(2, \\'/bin/bash -c \"source activate py2 && GTDBTK_DATA_PATH\\n\\nUE: The ‘Minimum Alignment Percent’ was left blank\\n\\nAssign a minimum percent in the advanced parameters and try resubmitting the job.\\n\\n\\u200bCreate Differential Expression Matrix using DeSEQ2\\n\\nInvalid input:nselect Run All Paired Condition Combinations or provide partial condition pairs. Dont do both\\n\\nUE: The user failed to select ALL vs Partial conditions. It cannot be blank and you cannot select both.\\n\\nDo one of the following:\\n\\n1.Check the box next to \\'Run All Paired Condition Combinations.\\'\\n\\n2.Add a \\'Run Partial Condition Combinations.\\'\\n\\n\\u200bClassify Taxonomy of Metagenomic Reads with GOTTCHA2\\u200b\\n\\n(2, No such file or directory)\\n\\n1.UE: The wrong reference database was used. Change the reference database and try resubmitting the job.\\n\\n2.KE: More than two reads libraries were included in the input. The developers have a task to fix this. The workaround is to only submit a maximum of two libraries at a time. Merging reads libraries can help with this process.\\n\\n\\u200bKraken2 Taxonomic Sequence Classifier\\u200b\\n\\nYou must enter either an input genome or input reads\\n\\nUE: There are two input options and one of them must be selected.\\n\\nAdd an input and try resubmitting the job.\\n\\n\\u200bVirSorter\\u200b\\n\\nCannot write data to fasta\\n\\nKE: The user interface allows users to enter a genome as input but the app will reject it. The error has been reported.\\n\\nChange to an input that is not a genome and try resubmitting the job.\\n\\nPreviousAnnotation App Errors\\n\\nNextModeling App Errors\\n\\nLast modified \\n\\n2yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_functional-genomics.html'}),\n",
       " Document(page_content=\"Modeling App Errors\\n\\nThis is a list of error messages found within the Job Log for Modeling App Jobs, what they mean and how to go about fixing them or if a job ticket needs to be submitted.\\n\\nTypes of Errors\\n\\nUE=User fixable or possible user error\\n\\nKE=KBase error\\n\\nUse your Browser's search tool to paste in your error message to locate your error message, information about the error and next steps.\\n\\nYou can always submit a ticket for help, questions, or follow-up to the\\n\\nKBase Help Board\\n\\n\\u200bBulk Download Modeling Objects\\u200b\\n\\nMust provide one and only one of workspace name\\n\\nUE: No modeling objects were found the the user’s narrative.\\n\\nAdd some objects and try resubmitting the job.\\n\\nGLPK does not support ID's longer than 256 characters\\n\\nKE: This option fails when this condition exists and the app needs maintenance.\\n\\nNothing can be done at this time. KBase is working on a solution to this problem.\\n\\n\\u200bView Flux Network\\u200b\\n\\nCan't use an undefined value as an ARRAY reference\\n\\nKE: The app is failing and needs maintenance.\\n\\nNothing can be done at this time. KBase is working on a solution to this problem.\\n\\nAuthentication required for AbstractHandle but no authentication header was passed\\n\\nKE: The app is failing and needs maintenance.\\n\\nNothing can be done at this time. KBase is working on a solution to this problem.\\n\\n\\u200bCompare Models\\u200b\\n\\nMust select at least two models to compare\\n\\nUE: The app requires two or more models for the comparison.\\n\\nAdd at least one more model and and try resubmitting the job.\\n\\nPreviousFunctional Genomics App Errors\\n\\nNextDeveloping Apps\\n\\nLast modified \\n\\n2yr ago\", metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_modeling.html'}),\n",
       " Document(page_content='Developing Apps\\n\\nInterested in developing an App for KBase? Here you can find information for developers.\\n\\nThe Principles\\n\\nReproducibility\\n\\nKBase is a scientific platform. A cornerstone of science is that scientific experiments and analyses are reproducible - that is a scientist must be able to take a description of an experiment and/or analysis, perform said experiment / analysis independently, and get the same results. Science that is not reproducible is not science, and knowledge that cannot be verified independently is not knowledge.\\n\\nProvenance\\n\\nProvenance in the context of KBase explains how data comes to exist - the sequence of operations that transformed a set of units of data into a different set of units of data along with who caused those transformations to occur and when. Information on job performance and the error log should be considered part of the data produced. Accurate provenance enables many of the other KBase principles, including reproducibility and Credit where credit is due. Data without provenance is not useful, as there is no way to determine how the data was created and therefore assess the data’s reliability.\\n\\nSharing and data privacy\\n\\nUsers loading data into and creating data within KBase are guaranteed that their data and activities are private unless they explicitly share their data or make it public. KBase will not mine, collate, or otherwise use their private data or activities/jobs (other than for internal tasks required to administer the platform), and their data/activities/jobs will not be visible in any data view to users that are not granted access.\\n\\nWhen data or analyses are shared, they are expected to be viewable and runnable by the users they are shared with. If a user makes a copy of data or analyses, those data and analyses are expected to be viewable and runnable just as the sources are viewable and runnable. Information about the original generators of the data are expected to propagated along with the data to ensure proper credit.\\n\\nCredit where credit is due\\n\\nUsers that add data, apps, or analyses to the system and make them available for other users to rerun, reuse, or copy and modify must receive credit for their work.\\n\\nSupport of cross-references\\n\\nData in KBase should reference appropriate and relevant related data e.g., a Genome object referencing the Taxon object to which it belongs. Data should be richly connected to other data types when feasible. Additionally, data imported into the system should reference the data source.\\n\\nFAIR data management\\n\\nFindable, Accessible, Interoperable, Reusable\\n\\ndata meets the criteria set out in the linked paper. FAIR overlaps with many of the other principles set out here but is an emerging standard for data management. Further, we wish to ensure data of the same type are comparable i.e., data are in comparable units derived from similar “workflows” so for example, numbers from two different data sets can be fairly combined.\\n\\nOpen source software and data\\n\\nKBase software is publicly available through\\n\\nGitHub\\n\\n. You may reuse KBase code under the terms of our\\n\\nopen source license\\n\\n. All software and data used in KBase must be open source and restriction free. This, along with reproducibility and provenance allows users to examine and understand each step in analyses workflows.\\n\\nExtensible by 3rd parties\\n\\n3rd parties can extend the functionality of KBase by contributing application modules that run in the KBase execution environment. These modules are expected to also follow the principles outlined here.\\n\\n\\u200bDevelop and release your own App\\u200b\\n\\nIf you want to develop Apps using the SDK, please apply for a KBase developer account by going to the Requesting a\\n\\nNew KBase Developer Account\\n\\n. If you are a US citizen, your account can be created within a few days. For foreign nationals, it may take several weeks (and, in a few cases, you may not be able to get a developer account). Non-US citizens will be asked for additional information in order to process their application.Once your account is approved,\\n\\ncontact us\\n\\nwith your username and ask to be added to the developer list.\\n\\nKBase System Architecture\\n\\nKBase is based loosely on a service-oriented architecture that bundles related functionality into a set of independently scalable services that are managed to provide responsive interaction via the Narrative Interface.\\n\\nFor more details, please see the KBase\\n\\narchitecture overview\\n\\n, which outlines the components and relationships between KBase’s user interfaces, services and databases.\\n\\nPreviousModeling App Errors\\n\\nNextThe KBase SDK\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/development.html'}),\n",
       " Document(page_content='The KBase SDK\\n\\nKBase Software Development Kit for collaborators\\n\\nInformation for Developers\\n\\nThe\\n\\nKBase Software Development Kit (SDK)\\n\\noffers members of the KBase community a mechanism to add open-source, open-license (as\\n\\ndefined by OSI\\n\\n) analysis tools to KBase so that they run on\\n\\nKBase’s computational architecture\\n\\nand are available through the\\n\\nKBase Narrative Interface\\n\\n. Community developers can write new tools that take advantage of the\\n\\nKBase Structured Data Types\\n\\nand large\\n\\nKBase Reference Data Resources\\n\\n. Check which tools have already been made available in our\\n\\nApp Catalog\\n\\nand add your own!\\n\\nGet Started\\n\\n1. Sign up for a user account\\u200b\\n\\nYou need a user account to be able to access KBase’s user interfaces, APIs, workspaces, and more.\\n\\n2. Create a KBase developer account\\u200b\\n\\n3. Read the SDK Documentation\\u200b\\n\\nLearn to use\\n\\nKBase SDK\\n\\nto begin crafting your tools to add to the KBase\\n\\nApp Catalog\\n\\nPreviousDeveloping Apps\\n\\nNextCreate a KBase Developer Account\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/development_kbase-sdk.html'}),\n",
       " Document(page_content='Create a KBase Developer Account\\n\\nThe following instructions outline how to create a\\n\\nKBase Developer Account\\n\\n. This has two phases 1) apply for an Argonne collaborator account and 2) join the KBase-Developers project.\\n\\nApply for an Argonne Collaborator Account.\\n\\n1.Go to the Computing, Environment and Life Sciences Account Management Create a Collaborator Account Page: https://apps.anl.gov/registration/collaborators.\\n\\n2.Read Terms and then click \\'Accept Terms and Conditions\\' button.\\n\\u200b\\n\\n3.Fill out sponsor form with the following information: \\nSponsor E-mail: [email\\xa0protected]\\nReason for Account: KBase developer account\\nThen click \\'Verify Sponsor E-mail\\'.\\n\\u200b\\n\\n4.Fill out the Argonne Collaborator Registration form. \\nPlease note: Make sure to set “Are you a US citizen” correctly. If you are not, you will be required to fill out additional information and provide full documentation.\\n\\u200b\\n\\n5.Click the Create Collaborator button when complete.\\n\\n6.Email [email\\xa0protected] to let us know that you have requested a collaborator account!\\n\\n7.Once your account has been approved, you will receive an email to set your password. Follow the instructions to change your initial password. \\n\\u200b\\n\\n8.You can now login to https://accounts.cels.anl.gov. \\n\\u200b\\n\\nJoin the KBase-Developers Project\\n\\n1.Login to https://accounts.cels.anl.gov.\\n\\n2.To join the KBase-Developers Group, click on \\'Join Project\\' in the left menu bar. \\n\\u200b\\n\\n3.Search for \"KBase-Developers\" and click on \\'KBase-Developers\\'.\\n\\u200b\\n\\n4.Click on the \\'Request Membership\\' button.\\n\\u200b\\n\\n5.Using the same thread as started above, email [email\\xa0protected] to let us know that you have requested to join the KBase-Developers Group.\\n\\nPreviousThe KBase SDK\\n\\nNextKBase GitHub Repository\\n\\nLast modified \\n\\n4mo ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/development_create-a-kbase-developer-account.html'}),\n",
       " Document(page_content='KBase GitHub Repository\\n\\nAccess the KBase GitHub Repository here.\\n\\nFor more details, see the KBase\\n\\narchitecture overview\\n\\n, which outlines the components and relationships between KBase’s user interfaces, services and databases.\\n\\nPreviousCreate a KBase Developer Account\\n\\nNextkbase.us\\n\\nLast modified \\n\\n2yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/development_github-repository.html'}),\n",
       " Document(page_content='kbase.us\\n\\nSpecific how to edit and use constraints in WordPress.\\n\\nResponsive Image Sizing\\n\\nFor webpages viewed on a larger screen i.e., computer, the maximum width is 800px.\\n\\nIf image width is greater than 800px\\n\\nwidth=\"100%” height=\"auto”\\n\\nIf width is less than 800px\\n\\nmax-width=\"100%” height=\"auto”\\n\\nIf width is not 800px or greater, but you want to set a specific maximum size.\\n\\nwidth=\"100%” max-width=”nnnpx” height=\"auto”\\n\\nCreating boxes around text\\n\\n<p style=\"border: Npx; border-style:solid; border-color: #HTMLtag; padding: 1em;\"> text here </p>\\n\\nTeal boxes for developer highlights are 5px and #009688.\\n\\nAnchors\\n\\nStop text from wrapping around an align left image\\n\\n<a id=\"anchor\"></a>\\n\\n<br clear=\"left\">\\n\\nAdd anchor to end of slug for web address\\n\\nhttps://siteaddress/slug/#anchor\\n\\nEmbedding and Linking from Drive\\n\\n1.Open the PDF or image file in a new preview window\\n\\n2.Open the expanded menu and click on \"Embed item...\"1.To embed an image, select the link or frame.2.To link a PDF, copy the Drive link.\\n\\n3.Insert the link in desired location.\\n\\nPop up messages\\n\\nBest practices for using Hustle for messaging on kbase.us\\n\\n1.Add or edit descriptive Title including dates.\\n\\n2.Include informative Main Content.\\n\\n3.Add enable the \"Never see this message again\" link\\n\\n4.Optional, schedule specific dates, but revert to \"Draft\" after maintenance period.\\n\\n5.Publish!\\n\\nPreviousKBase GitHub Repository\\n\\nLast modified \\n\\n1yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/kbase.us.html'}),\n",
       " Document(page_content='Please enable cookies.\\n\\nEmail Protection\\n\\nYou are unable to access this email address docs.kbase.us\\n\\nThe website from which you got to this page is protected by Cloudflare. Email addresses on that page have been hidden in order to keep them from being accessed by malicious bots. You must enable Javascript in your browser in order to decode the e-mail address.\\n\\nIf you have a website and are interested in protecting it in a similar way, you can sign up for Cloudflare.\\n\\nHow does Cloudflare protect email addresses on website from spammers?\\n\\nCan I sign up for Cloudflare?\\n\\nCloudflare Ray ID: 7d768743196c22e5\\n    •\\n    \\n      Your IP:\\n      Click to reveal\\n      140.221.43.116\\n      •\\n    \\n    Performance & security by Cloudflare', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/cdn-cgi_l_email-protection.html'})]"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a1698632",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm = ChatOpenAI(temperature=0, model=\"gpt-4\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "27f3a2f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "408"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_splitter = RecursiveCharacterTextSplitter(\n",
    "    chunk_size = 1000,\n",
    "    chunk_overlap  = 20,\n",
    "    length_function = len,\n",
    ")\n",
    "documents_html = text_splitter.split_documents(data_html)\n",
    "len(documents_html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "85f2a37f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Document(page_content='KBase Documentation\\n\\nHere you will find guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.\\n\\nHow to cite KBase:\\n\\nArkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163\\n\\nFunding:\\n\\nThis work is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase (KBase) funded by the\\n\\nU.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\\n\\nunder Award Numbers DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.\\n\\nNextKBase Terms & Conditions\\n\\nLast modified \\n\\n2yr ago', metadata={'source': '/scratch/ac.pgupta/convLLM/kbasedocs/index.html'})"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "documents_html[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "21958fad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Supported Browsers\n",
      "\n",
      "The KBase website and Narrative Interface should work on most modern web browsers, but we primarily focus on supporting Chrome.\n",
      "\n",
      "Supported Browsers\n",
      "\n",
      "The following desktop web browsers have been tested against the KBase web site and Narrative user interface. If you experience any issue using these browsers with our site please report it through the\n",
      "\n",
      "Help Board\n",
      "\n",
      "OS Supported Browsers Mac Chrome Firefox Safari 10 (macOS Sierra) and up Windows Firefox Microsoft Edge - not recommended Linux Chrome Firefox\n",
      "\n",
      "Ad blockers may prevent KBase from working in your web browser\n",
      "\n",
      "Some ad-blockers and browser extensions may cause errors within the Narrative Interface. We are working to identify issues with commonly used browser extensions, but you may want to disable browser extensions if you encounter issues within KBase.\n",
      "\n",
      "Mobile\n"
     ]
    }
   ],
   "source": [
    "# UnstructuredHTMLLoader DirectoryLoader will inline tables this way\n",
    "# original page: https://docs.kbase.us/getting-started/browsers\n",
    "print(documents_html[17].page_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3f3654d",
   "metadata": {},
   "source": [
    "### Few shot example prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "9197810f",
   "metadata": {},
   "outputs": [],
   "source": [
    "few_shot_examples = [\n",
    "  {\n",
    "    \"Instruction\": \"\"\"I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
    "                I split the docs into small text pieces, based on each text provided, generate 3-5 questions and answers pairs from the text. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
    "                The questions can be, if a statement is correct or not? what is xxx? how to do xxx in KBase?\n",
    "                text is:\n",
    "                KBase Documentation\\n\\nHere you will find guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.\\n\\nHow to cite KBase:\\n\\nArkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163\\n\\nFunding:\\n\\nThis work is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase (KBase) funded by the\\n\\nU.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\\n\\nunder Award Numbers DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.\\n\\nNextKBase Terms & Conditions\\n\\nLast modified \\n\\n2yr ago\n",
    "                \"\"\",\n",
    "\n",
    "    \"Generated content\":\"\"\"\n",
    "    ### question: How to cite KBase? ## answer: Arkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al.\n",
    "    ### question: What is KBase? ## answer: The United States Department of Energy Systems Biology Knowledgebase.\n",
    "    ### question: Who supports and funds KBase? ## answer: The United States Department of Energy (DOE)\n",
    "    \"\"\"\n",
    "  },\n",
    "  {\n",
    "    \"Instruction\": \"\"\"I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
    "                I split the docs into small text pieces, based on each text provided, generate 3-5 questions and answers pairs from the text. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
    "                The questions can be, if a statement is correct or not? what is xxx? how to do xxx in KBase?\n",
    "                text is:\n",
    "                Supported Browsers\\n\\nThe KBase website and Narrative Interface should work on most modern web browsers, but we primarily focus on supporting Chrome.\\n\\nSupported Browsers\\n\\nThe following desktop web browsers have been tested against the KBase web site and Narrative user interface. If you experience any issue using these browsers with our site please report it through the\\n\\nHelp Board\\n\\nOS Supported Browsers Mac Chrome Firefox Safari 10 (macOS Sierra) and up Windows Firefox Microsoft Edge - not recommended Linux Chrome Firefox\\n\\nAd blockers may prevent KBase from working in your web browser\\n\\nSome ad-blockers and browser extensions may cause errors within the Narrative Interface. We are working to identify issues with commonly used browser extensions, but you may want to disable browser extensions if you encounter issues within KBase.\\n\\nMobile\\n\\nKBase is currently not supported on mobile browsers. The KBase website and services should work with tablet browsers on iPad and Android devices in landscape orientation, but we do not currently test on these platforms so there may be compatibility issues. Watch for future updates as KBase goes responsive and mobile.\\n\\nInstallation of Supported Browsers\\n\\nChrome\\n\\nChrome is the best browser to use for KBase. The Chrome browser is produced by Google, Inc., and is freely available. Chrome operates on a rolling release, meaning that it is updated frequently and it is not easy to track specific versions. Our best advice is to make sure you are using the most recent version available. On Mac and Windows this is built into the Chrome application itself.\\n\\n\\u200bDownload Chrome\\u200b\\n\\nFirefox\\n\\nFirefox, produced by the non-profit Mozilla Corporation, is a freely available, open source browser for Mac, Windows, Linux, and other systems. Like Chrome, Firefox is updated frequently, so it is not feasible for us to recommend a specific (or even minimum) version. Our best advice is to make sure you are using the most recent version available. Recent versions of Firefox will automatically update themselves on Mac and Windows.\\n\\n\\u200bDownload Firefox\\u200b\\n\\nSafari\\n\\nSafari is distributed with Mac OS X. (We do not support the Windows version of Safari.) Specific versions of Mac OS X support specific versions of Safari – see this\\n\\nWikipedia entry\\n\\nfor details.\\n\\nMicrosoft Edge\\n\\nMicrosoft Edge\\n\\nis the new web browser for Windows 10; it replaces Internet Explorer. You should be able to use KBase on Edge, but this has not been extensively tested.\\n\\nUnsupported Browsers\\n\\nKBase is not supported on Internet Explorer. It should work on Microsoft Edge, but it has not been extensively tested. For the best chance of success on Edge, be sure you are using the most recent version.\\n\\nOlder versions of supported browsers are also not supported. Please ensure that your browser and system are up-to-date before reporting a problem with a supported browser.\\n\\nOther modern browsers, such as Opera, may work with KBase, but as we do not test with them we cannot make any assurances about their operation.\\n\\nPreviousAuthentication Update\\n\\nNextNarrative Quick Start\\n\\nLast modified \\n\\n1yr ago\n",
    "                \"\"\",\n",
    "\n",
    "    \"Generated content\":\n",
    "    \"\"\"\n",
    "    ### question: What browswers are supported by KBase? ## answer: There are four browswers that are supported by KBase, they are\n",
    "    1. Chrome:Chrome is the best browser to use for KBase. \n",
    "    2. Firefox: Firefox, produced by the non-profit Mozilla Corporation, is a freely available, open source browser for Mac, Windows, Linux, and other systems. \n",
    "    3. Safari: Safari is distributed with Mac OS X. (We do not support the Windows version of Safari.)\n",
    "    4. Microsoft Edge: Microsoft Edgeis the new web browser for Windows 10; it replaces Internet Explorer. You should be able to use KBase on Edge, but this has not been extensively tested.\n",
    "    ### question: What browswers are supported for Mac? ## answer: Chrome, Firefox and Safari 10 (macOS Sierra) and up\n",
    "    ### question: Why Kbase is not working in my browswer? ## answer: Ad blockers may prevent KBase from working in your web browserSome ad-blockers and browser extensions may cause errors within the Narrative Interface. We are working to identify issues with commonly used browser extensions, but you may want to disable browser extensions if you encounter issues within KBase.\n",
    "    \"\"\"\n",
    "  }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "0261550c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Instruction: I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
      "                I split the docs into small text pieces, based on each text provided, generate 3-5 questions and answers pairs from the text. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
      "                The questions can be, if a statement is correct or not? what is xxx? how to do xxx in KBase?\n",
      "                text is:\n",
      "                KBase Documentation\n",
      "\n",
      "Here you will find guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.\n",
      "\n",
      "How to cite KBase:\n",
      "\n",
      "Arkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163\n",
      "\n",
      "Funding:\n",
      "\n",
      "This work is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase (KBase) funded by the\n",
      "\n",
      "U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\n",
      "\n",
      "under Award Numbers DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.\n",
      "\n",
      "NextKBase Terms & Conditions\n",
      "\n",
      "Last modified \n",
      "\n",
      "2yr ago\n",
      "                \n",
      "Generated content: \n",
      "    ### question: How to cite KBase? ## answer: Arkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al.\n",
      "    ### question: What is KBase? ## answer: The United States Department of Energy Systems Biology Knowledgebase.\n",
      "    ### question: Who supports and funds KBase? ## answer: The United States Department of Energy (DOE)\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "example_prompt = PromptTemplate(input_variables=[\"Instruction\", \"Generated content\"], template=\"Instruction: {Instruction}\\nGenerated content: {Generated content}\")\n",
    "\n",
    "print(example_prompt.format(**few_shot_examples[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "7c6438d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Instruction: I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
      "                I split the docs into small text pieces, based on each text provided, generate 3-5 questions and answers pairs from the text. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
      "                The questions can be, if a statement is correct or not? what is xxx? how to do xxx in KBase?\n",
      "                text is:\n",
      "                KBase Documentation\n",
      "\n",
      "Here you will find guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.\n",
      "\n",
      "How to cite KBase:\n",
      "\n",
      "Arkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163\n",
      "\n",
      "Funding:\n",
      "\n",
      "This work is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase (KBase) funded by the\n",
      "\n",
      "U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\n",
      "\n",
      "under Award Numbers DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.\n",
      "\n",
      "NextKBase Terms & Conditions\n",
      "\n",
      "Last modified \n",
      "\n",
      "2yr ago\n",
      "                \n",
      "Generated content: \n",
      "    ### question: How to cite KBase? ## answer: Arkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al.\n",
      "    ### question: What is KBase? ## answer: The United States Department of Energy Systems Biology Knowledgebase.\n",
      "    ### question: Who supports and funds KBase? ## answer: The United States Department of Energy (DOE)\n",
      "    \n",
      "\n",
      "Instruction: I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
      "                I split the docs into small text pieces, based on each text provided, generate 3-5 questions and answers pairs from the text. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
      "                The questions can be, if a statement is correct or not? what is xxx? how to do xxx in KBase?\n",
      "                text is:\n",
      "                Supported Browsers\n",
      "\n",
      "The KBase website and Narrative Interface should work on most modern web browsers, but we primarily focus on supporting Chrome.\n",
      "\n",
      "Supported Browsers\n",
      "\n",
      "The following desktop web browsers have been tested against the KBase web site and Narrative user interface. If you experience any issue using these browsers with our site please report it through the\n",
      "\n",
      "Help Board\n",
      "\n",
      "OS Supported Browsers Mac Chrome Firefox Safari 10 (macOS Sierra) and up Windows Firefox Microsoft Edge - not recommended Linux Chrome Firefox\n",
      "\n",
      "Ad blockers may prevent KBase from working in your web browser\n",
      "\n",
      "Some ad-blockers and browser extensions may cause errors within the Narrative Interface. We are working to identify issues with commonly used browser extensions, but you may want to disable browser extensions if you encounter issues within KBase.\n",
      "\n",
      "Mobile\n",
      "\n",
      "KBase is currently not supported on mobile browsers. The KBase website and services should work with tablet browsers on iPad and Android devices in landscape orientation, but we do not currently test on these platforms so there may be compatibility issues. Watch for future updates as KBase goes responsive and mobile.\n",
      "\n",
      "Installation of Supported Browsers\n",
      "\n",
      "Chrome\n",
      "\n",
      "Chrome is the best browser to use for KBase. The Chrome browser is produced by Google, Inc., and is freely available. Chrome operates on a rolling release, meaning that it is updated frequently and it is not easy to track specific versions. Our best advice is to make sure you are using the most recent version available. On Mac and Windows this is built into the Chrome application itself.\n",
      "\n",
      "​Download Chrome​\n",
      "\n",
      "Firefox\n",
      "\n",
      "Firefox, produced by the non-profit Mozilla Corporation, is a freely available, open source browser for Mac, Windows, Linux, and other systems. Like Chrome, Firefox is updated frequently, so it is not feasible for us to recommend a specific (or even minimum) version. Our best advice is to make sure you are using the most recent version available. Recent versions of Firefox will automatically update themselves on Mac and Windows.\n",
      "\n",
      "​Download Firefox​\n",
      "\n",
      "Safari\n",
      "\n",
      "Safari is distributed with Mac OS X. (We do not support the Windows version of Safari.) Specific versions of Mac OS X support specific versions of Safari – see this\n",
      "\n",
      "Wikipedia entry\n",
      "\n",
      "for details.\n",
      "\n",
      "Microsoft Edge\n",
      "\n",
      "Microsoft Edge\n",
      "\n",
      "is the new web browser for Windows 10; it replaces Internet Explorer. You should be able to use KBase on Edge, but this has not been extensively tested.\n",
      "\n",
      "Unsupported Browsers\n",
      "\n",
      "KBase is not supported on Internet Explorer. It should work on Microsoft Edge, but it has not been extensively tested. For the best chance of success on Edge, be sure you are using the most recent version.\n",
      "\n",
      "Older versions of supported browsers are also not supported. Please ensure that your browser and system are up-to-date before reporting a problem with a supported browser.\n",
      "\n",
      "Other modern browsers, such as Opera, may work with KBase, but as we do not test with them we cannot make any assurances about their operation.\n",
      "\n",
      "PreviousAuthentication Update\n",
      "\n",
      "NextNarrative Quick Start\n",
      "\n",
      "Last modified \n",
      "\n",
      "1yr ago\n",
      "                \n",
      "Generated content: \n",
      "    ### question: What browswers are supported by KBase? ## answer: There are four browswers that are supported by KBase, they are\n",
      "    1. Chrome:Chrome is the best browser to use for KBase. \n",
      "    2. Firefox: Firefox, produced by the non-profit Mozilla Corporation, is a freely available, open source browser for Mac, Windows, Linux, and other systems. \n",
      "    3. Safari: Safari is distributed with Mac OS X. (We do not support the Windows version of Safari.)\n",
      "    4. Microsoft Edge: Microsoft Edgeis the new web browser for Windows 10; it replaces Internet Explorer. You should be able to use KBase on Edge, but this has not been extensively tested.\n",
      "    ### question: What browswers are supported for Mac? ## answer: Chrome, Firefox and Safari 10 (macOS Sierra) and up\n",
      "    ### question: Why Kbase is not working in my browswer? ## answer: Ad blockers may prevent KBase from working in your web browserSome ad-blockers and browser extensions may cause errors within the Narrative Interface. We are working to identify issues with commonly used browser extensions, but you may want to disable browser extensions if you encounter issues within KBase.\n",
      "    \n",
      "\n",
      "Instruction:  I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
      "                I split the docs into small text pieces, based on each text provided, generate 3-5 questions and answers pairs from the text. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
      "                The questions can be, if a statement is correct or not? what is xxx? how to do xxx in KBase?\n",
      "                text is:\n"
     ]
    }
   ],
   "source": [
    "instruction_prompt=\"\"\"I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
    "                I split the docs into small text pieces, based on each text provided, generate 3-5 questions and answers pairs from the text. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
    "                The questions can be, if a statement is correct or not? what is xxx? how to do xxx in KBase?\n",
    "                text is:\"\"\"\n",
    "prompt = FewShotPromptTemplate(\n",
    "    examples=few_shot_examples,\n",
    "    example_prompt=example_prompt,\n",
    "    suffix=\"Instruction:  {input}\",\n",
    "    input_variables=[\"input\"]\n",
    ")\n",
    "prompt\n",
    "print(prompt.format(input=instruction_prompt))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "fd314096",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generateQA(instruction_prompt, text):\n",
    "    system_message = SystemMessage(content=prompt.format(input=instruction_prompt))\n",
    "    human_message = HumanMessage(content=text)\n",
    "    res = llm([system_message, human_message])\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "069bab0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "answers = generateQA(instruction_prompt, documents_html[0].page_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "ac13055f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AIMessage(content='Generated content: \\n    ### question: What is the purpose of the KBase Documentation? ## answer: The KBase Documentation provides guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.\\n    ### question: How should KBase be cited in academic work? ## answer: KBase should be cited as follows: Arkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163.\\n    ### question: Who funds the work of KBase? ## answer: KBase is funded by the U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research.\\n    ### question: What are the award numbers associated with the funding of KBase? ## answer: The award numbers associated with the funding of KBase are DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.', additional_kwargs={}, example=False)"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "answers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "eebdc025",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extractQA(answers):\n",
    "    answers = answers.content.split(\"###\")\n",
    "    res = []\n",
    "    for i in range(1, len(answers)):\n",
    "        q = answers[i].split(\"##\")[0][11:].strip()\n",
    "        a = answers[i].split(\"##\")[1][9:].strip()\n",
    "        res.append({'question': q, \"answer\": a})\n",
    "    return res\n",
    "# extractQA(answers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "dcb9cf3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8275, Requested 2150. Please try again in 2.55s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7945, Requested 2137. Please try again in 492ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8099, Requested 2158. Please try again in 1.542s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7916, Requested 2090. Please try again in 36ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7889, Requested 2165. Please try again in 324ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7950, Requested 2185. Please try again in 810ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7852, Requested 2184. Please try again in 216ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8368, Requested 2105. Please try again in 2.838s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8150, Requested 2176. Please try again in 1.955s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7831, Requested 2186. Please try again in 102ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7830, Requested 2186. Please try again in 96ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8618, Requested 2077. Please try again in 4.17s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7938, Requested 2077. Please try again in 90ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7834, Requested 2176. Please try again in 60ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8049, Requested 2187. Please try again in 1.416s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7996, Requested 2181. Please try again in 1.062s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8365, Requested 2182. Please try again in 3.282s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8363, Requested 2185. Please try again in 3.288s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7933, Requested 2182. Please try again in 690ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7910, Requested 2155. Please try again in 390ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8404, Requested 2105. Please try again in 3.054s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7881, Requested 2181. Please try again in 372ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7932, Requested 2174. Please try again in 636ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7987, Requested 2119. Please try again in 636ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7933, Requested 2183. Please try again in 696ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7996, Requested 2173. Please try again in 1.013s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7973, Requested 2129. Please try again in 612ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8240, Requested 2173. Please try again in 2.478s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8098, Requested 2177. Please try again in 1.65s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7989, Requested 2147. Please try again in 816ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8468, Requested 2181. Please try again in 3.894s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8334, Requested 2177. Please try again in 3.066s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8003, Requested 2191. Please try again in 1.164s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8883, Requested 2129. Please try again in 6.072s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8203, Requested 2129. Please try again in 1.992s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8453, Requested 2113. Please try again in 3.396s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8298, Requested 2185. Please try again in 2.898s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8117, Requested 2140. Please try again in 1.542s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8316, Requested 2154. Please try again in 2.82s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8249, Requested 2169. Please try again in 2.508s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7898, Requested 2121. Please try again in 114ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7938, Requested 2167. Please try again in 630ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8103, Requested 2188. Please try again in 1.746s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8271, Requested 2161. Please try again in 2.592s. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8252, Requested 2187. Please try again in 2.634s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8114, Requested 1994. Please try again in 648ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8304, Requested 2150. Please try again in 2.724s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8146, Requested 2179. Please try again in 1.95s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7999, Requested 2167. Please try again in 996ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8124, Requested 2186. Please try again in 1.86s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7883, Requested 2158. Please try again in 246ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7987, Requested 2046. Please try again in 198ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8161, Requested 2154. Please try again in 1.89s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8212, Requested 2181. Please try again in 2.358s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7857, Requested 2192. Please try again in 294ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8507, Requested 2188. Please try again in 4.17s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7826, Requested 2188. Please try again in 84ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8251, Requested 2187. Please try again in 2.628s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8167, Requested 2160. Please try again in 1.962s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8014, Requested 2136. Please try again in 900ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8310, Requested 2059. Please try again in 2.214s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8265, Requested 2099. Please try again in 2.184s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8062, Requested 2179. Please try again in 1.446s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7910, Requested 2156. Please try again in 396ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8575, Requested 2164. Please try again in 4.434s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7897, Requested 2164. Please try again in 366ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8140, Requested 2148. Please try again in 1.728s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8212, Requested 2098. Please try again in 1.86s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8300, Requested 2156. Please try again in 2.736s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8789, Requested 2158. Please try again in 5.682s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8105, Requested 2158. Please try again in 1.578s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8144, Requested 2141. Please try again in 1.71s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7992, Requested 2188. Please try again in 1.08s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8721, Requested 2181. Please try again in 5.412s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8041, Requested 2181. Please try again in 1.332s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8248, Requested 2139. Please try again in 2.322s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8378, Requested 2189. Please try again in 3.402s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8311, Requested 2188. Please try again in 2.994s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8313, Requested 2174. Please try again in 2.922s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8268, Requested 2136. Please try again in 2.424s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7945, Requested 2144. Please try again in 534ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7914, Requested 2189. Please try again in 618ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8571, Requested 2146. Please try again in 4.302s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7889, Requested 2146. Please try again in 210ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8106, Requested 2119. Please try again in 1.35s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8278, Requested 2175. Please try again in 2.718s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8398, Requested 2175. Please try again in 3.438s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8167, Requested 2118. Please try again in 1.71s. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8280, Requested 2174. Please try again in 2.724s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8062, Requested 2075. Please try again in 822ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8151, Requested 2092. Please try again in 1.458s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8037, Requested 2174. Please try again in 1.266s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8477, Requested 2170. Please try again in 3.881s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8637, Requested 2010. Please try again in 3.881s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8816, Requested 2181. Please try again in 5.982s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8135, Requested 2181. Please try again in 1.896s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8431, Requested 2128. Please try again in 3.354s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8309, Requested 2179. Please try again in 2.928s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8022, Requested 2156. Please try again in 1.068s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8267, Requested 2178. Please try again in 2.67s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8115, Requested 2181. Please try again in 1.776s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8306, Requested 2183. Please try again in 2.934s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8070, Requested 2188. Please try again in 1.548s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8927, Requested 2186. Please try again in 6.678s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8244, Requested 2186. Please try again in 2.58s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7920, Requested 2104. Please try again in 144ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7996, Requested 2048. Please try again in 264ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8262, Requested 2136. Please try again in 2.388s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8065, Requested 2096. Please try again in 966ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8171, Requested 2186. Please try again in 2.142s. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8099, Requested 1974. Please try again in 438ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8212, Requested 2176. Please try again in 2.328s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 9250, Requested 2187. Please try again in 8.622s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8569, Requested 2187. Please try again in 4.536s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7891, Requested 2187. Please try again in 468ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7908, Requested 2168. Please try again in 456ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8083, Requested 2132. Please try again in 1.29s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8139, Requested 2132. Please try again in 1.626s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7962, Requested 2189. Please try again in 906ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7869, Requested 2178. Please try again in 282ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8294, Requested 2130. Please try again in 2.544s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8560, Requested 2125. Please try again in 4.11s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7879, Requested 2125. Please try again in 24ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8106, Requested 2060. Please try again in 996ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8501, Requested 2183. Please try again in 4.104s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7832, Requested 2183. Please try again in 90ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8516, Requested 2183. Please try again in 4.194s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7819, Requested 2183. Please try again in 12ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8442, Requested 2164. Please try again in 3.636s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8509, Requested 2180. Please try again in 4.134s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7824, Requested 2180. Please try again in 24ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8601, Requested 2157. Please try again in 4.548s. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7913, Requested 2157. Please try again in 420ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8316, Requested 2153. Please try again in 2.814s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8076, Requested 2153. Please try again in 1.374s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8191, Requested 2187. Please try again in 2.268s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 9232, Requested 2102. Please try again in 8.003s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8552, Requested 2102. Please try again in 3.924s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8401, Requested 2186. Please try again in 3.522s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8065, Requested 2191. Please try again in 1.536s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8120, Requested 2177. Please try again in 1.782s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8103, Requested 1988. Please try again in 546ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8472, Requested 2188. Please try again in 3.96s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8462, Requested 2169. Please try again in 3.786s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8449, Requested 1997. Please try again in 2.676s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8660, Requested 2169. Please try again in 4.974s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7971, Requested 2169. Please try again in 840ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8015, Requested 2090. Please try again in 630ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8208, Requested 2154. Please try again in 2.172s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8359, Requested 2157. Please try again in 3.096s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8392, Requested 2184. Please try again in 3.456s. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "content='Generated content: \\n    ### question: What ontology systems does the Ontology API support? ## answer: The Ontology API supports Gene Ontology (GO) and Environmental Ontology (ENVO).\\n    ### question: Where can I find more information on the supported ontologies? ## answer: You can find more information on the supported ontologies on their own home pages or the KBase landing page. For Gene Ontology (GO), the link is https://narrative.kbase.us/#ontology/term/go_ontology/GO:####### and for The Environmental Ontology (ENVO), the link is https://narrative.kbase.us/#ontology/term/envo_ontology/ENVO:########.\\n    ### question: What is the metadata_validation.tsv file? ## answer: The metadata_validation.tsv file contains a list of the metadata terms currently supported for validation along with a description that provides general formatting direction. You can view the most up-to-date version of this list in the KBase Samples GitHub.' additional_kwargs={} example=False\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 9174, Requested 2168. Please try again in 8.052s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8493, Requested 2168. Please try again in 3.966s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8684, Requested 2185. Please try again in 5.214s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8002, Requested 2185. Please try again in 1.122s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7983, Requested 2184. Please try again in 1.002s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8120, Requested 2119. Please try again in 1.434s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8115, Requested 2187. Please try again in 1.812s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8137, Requested 2170. Please try again in 1.842s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8004, Requested 2169. Please try again in 1.038s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8593, Requested 2004. Please try again in 3.582s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8676, Requested 2192. Please try again in 5.208s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7993, Requested 2192. Please try again in 1.11s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8052, Requested 2157. Please try again in 1.254s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8027, Requested 2036. Please try again in 378ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8293, Requested 2152. Please try again in 2.67s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7957, Requested 2128. Please try again in 510ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8064, Requested 2158. Please try again in 1.332s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8123, Requested 2174. Please try again in 1.782s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8106, Requested 2182. Please try again in 1.728s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8103, Requested 2179. Please try again in 1.692s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7932, Requested 2181. Please try again in 678ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8296, Requested 2171. Please try again in 2.802s. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8095, Requested 2088. Please try again in 1.098s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8139, Requested 2189. Please try again in 1.968s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8022, Requested 2176. Please try again in 1.188s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8247, Requested 2153. Please try again in 2.4s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8253, Requested 1992. Please try again in 1.47s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7940, Requested 2187. Please try again in 762ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7817, Requested 2186. Please try again in 18ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8918, Requested 2081. Please try again in 5.994s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8238, Requested 2081. Please try again in 1.913s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8491, Requested 2152. Please try again in 3.857s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8311, Requested 2178. Please try again in 2.934s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7954, Requested 2148. Please try again in 612ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8084, Requested 2175. Please try again in 1.554s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8128, Requested 2184. Please try again in 1.872s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8438, Requested 2079. Please try again in 3.102s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8382, Requested 2184. Please try again in 3.396s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8526, Requested 2018. Please try again in 3.264s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8951, Requested 2172. Please try again in 6.738s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8263, Requested 2172. Please try again in 2.61s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7922, Requested 2157. Please try again in 474ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8527, Requested 2125. Please try again in 3.911s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8079, Requested 2106. Please try again in 1.11s. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8145, Requested 2179. Please try again in 1.944s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8564, Requested 2163. Please try again in 4.362s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7883, Requested 2163. Please try again in 276ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7989, Requested 2189. Please try again in 1.068s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8566, Requested 2186. Please try again in 4.512s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7886, Requested 2186. Please try again in 432ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7946, Requested 2187. Please try again in 798ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8896, Requested 2177. Please try again in 6.438s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8215, Requested 2177. Please try again in 2.352s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8135, Requested 2151. Please try again in 1.716s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7881, Requested 2162. Please try again in 258ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8060, Requested 2110. Please try again in 1.02s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8064, Requested 2108. Please try again in 1.032s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8099, Requested 2187. Please try again in 1.716s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8425, Requested 2184. Please try again in 3.654s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8433, Requested 2179. Please try again in 3.672s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8313, Requested 2178. Please try again in 2.946s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8099, Requested 2153. Please try again in 1.512s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8008, Requested 2164. Please try again in 1.032s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8620, Requested 2127. Please try again in 4.482s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7936, Requested 2127. Please try again in 378ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7925, Requested 2093. Please try again in 108ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8100, Requested 2162. Please try again in 1.572s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8250, Requested 2098. Please try again in 2.088s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8202, Requested 2139. Please try again in 2.046s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7953, Requested 2158. Please try again in 666ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8108, Requested 2170. Please try again in 1.668s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8371, Requested 2132. Please try again in 3.018s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8421, Requested 2191. Please try again in 3.672s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8033, Requested 2174. Please try again in 1.242s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7918, Requested 2186. Please try again in 624ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7847, Requested 2175. Please try again in 132ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8253, Requested 2145. Please try again in 2.388s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7973, Requested 2158. Please try again in 786ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7890, Requested 2183. Please try again in 438ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7838, Requested 2170. Please try again in 48ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8341, Requested 2143. Please try again in 2.904s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8203, Requested 2084. Please try again in 1.722s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7941, Requested 2098. Please try again in 234ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7999, Requested 2073. Please try again in 432ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8299, Requested 2176. Please try again in 2.85s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8198, Requested 2143. Please try again in 2.046s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8185, Requested 2134. Please try again in 1.913s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8795, Requested 2180. Please try again in 5.85s. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8113, Requested 2180. Please try again in 1.758s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8092, Requested 2085. Please try again in 1.062s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8066, Requested 2065. Please try again in 786ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8078, Requested 2187. Please try again in 1.59s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8077, Requested 2177. Please try again in 1.524s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8454, Requested 2094. Please try again in 3.288s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8358, Requested 2182. Please try again in 3.24s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8232, Requested 2188. Please try again in 2.52s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 9082, Requested 2179. Please try again in 7.565s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8395, Requested 2179. Please try again in 3.444s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7816, Requested 2192. Please try again in 48ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8431, Requested 2176. Please try again in 3.642s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8115, Requested 2174. Please try again in 1.734s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8465, Requested 2188. Please try again in 3.918s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8302, Requested 2182. Please try again in 2.904s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8006, Requested 2058. Please try again in 384ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8068, Requested 2179. Please try again in 1.482s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7954, Requested 2181. Please try again in 810ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8856, Requested 2188. Please try again in 6.264s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8162, Requested 2188. Please try again in 2.1s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8457, Requested 2177. Please try again in 3.804s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7933, Requested 2191. Please try again in 744ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7872, Requested 2183. Please try again in 330ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8012, Requested 2191. Please try again in 1.218s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 7830, Requested 2185. Please try again in 90ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8739, Requested 2077. Please try again in 4.896s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8060, Requested 2077. Please try again in 822ms. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8372, Requested 2153. Please try again in 3.15s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8173, Requested 2160. Please try again in 1.998s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8448, Requested 1944. Please try again in 2.352s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 9142, Requested 2178. Please try again in 7.92s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8463, Requested 2178. Please try again in 3.846s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8233, Requested 2140. Please try again in 2.238s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8977, Requested 2162. Please try again in 6.834s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8290, Requested 2162. Please try again in 2.712s. Visit https://platform.openai.com/account/rate-limits to learn more..\n",
      "Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised RateLimitError: Rate limit reached for gpt-4 in organization org-zw6y0pzavFVsBHYYRW7WQpeU on tokens_usage_based per min: Limit 10000, Used 8423, Requested 2122. Please try again in 3.27s. Visit https://platform.openai.com/account/rate-limits to learn more..\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'question': 'What can you find in the KBase Documentation?',\n",
       "  'answer': 'You can find guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.'},\n",
       " {'question': 'How to cite KBase in a research paper?',\n",
       "  'answer': 'You can cite KBase as \"Arkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163\".'},\n",
       " {'question': 'Who funds the work of KBase?',\n",
       "  'answer': 'The work of KBase is funded by the U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research.'},\n",
       " {'question': 'What are the award numbers for the funding received by KBase?',\n",
       "  'answer': 'The award numbers are DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.'},\n",
       " {'question': 'What can I find in the \"Getting Started\" section of KBase?',\n",
       "  'answer': 'In the \"Getting Started\" section of KBase, you can find information on Signing Up and Signing In, KBase Supported Browsers, A Quick Start Guide to KBase, KBase Narrative Interface User Guide, and FAQs.'},\n",
       " {'question': 'Where can I find a guide on how to get started with KBase?',\n",
       "  'answer': 'You can find a guide on how to get started with KBase in the \"Getting Started\" section of the KBase website.'},\n",
       " {'question': 'Where can I find information on KBase supported browsers?',\n",
       "  'answer': 'You can find information on KBase supported browsers in the \"Getting Started\" section of the KBase website.'},\n",
       " {'question': 'How can I sign up for a KBase account?',\n",
       "  'answer': 'To create a KBase account, you will need to use one of the supported identity providers. You can sign up for a free KBase account using your existing Google, ORCiD, or Globus identity.'},\n",
       " {'question': 'What do I need to do after linking my Google, ORCiD, or Globus account to KBase?',\n",
       "  'answer': 'After you link your Google, ORCiD, or Globus account, you will be able to sign in and use the KBase Narrative Interface to perform your computational systems biology analyses.'},\n",
       " {'question': 'What do I need to accept when linking an identity to KBase?',\n",
       "  'answer': 'When you link an identity to KBase, you will need to accept the Terms & Conditions, including the User Agreement. You may also want to review the Privacy Policy and Data Policy. Users shall follow the KBase Code of Conduct.'},\n",
       " {'question': 'Is KBase free to use?',\n",
       "  'answer': 'Yes, KBase is sponsored by the U.S. Department of Energy and is free for all to use, even if you are not in the United States.'},\n",
       " {'question': 'How can I sign in to KBase?',\n",
       "  'answer': \"To sign in, click the Sign In button on the KBase website or KBase Narrative Interface. Select the identity provider you chose when you signed up for your KBase account, and use the credentials you would normally use when signing into the that provider's web site.\"},\n",
       " {'question': 'How can I reset my KBase password?',\n",
       "  'answer': 'KBase itself does not manage passwords of its own, so all password management must be done through your identity provider.'},\n",
       " {'question': 'Can I link multiple accounts to sign into KBase?',\n",
       "  'answer': 'Yes, you can link multiple accounts to sign into KBase.'},\n",
       " {'question': 'How can I sign up for KBase?',\n",
       "  'answer': 'You can sign up for KBase using your existing Google, ORCiD, or Globus account. From the KBase Narrative login, click the “Sign Up” button and choose the account you want to use for signing up.'},\n",
       " {'question': \"What if I don't have a Google, ORCiD, or Globus account?\",\n",
       "  'answer': 'You can easily and freely get a new account with one of these identity providers.'},\n",
       " {'question': 'How can I sign up for KBase using my Google account?',\n",
       "  'answer': 'From the KBase Narrative login, click the “Sign Up” button, then click on the “Sign in with Google” button. You will be prompted to choose a Google account (if you have more than one) and sign in to it (if you aren’t already signed in).'},\n",
       " {'question': 'What is required to create a KBase account?',\n",
       "  'answer': 'The KBase account creation form will prompt you to enter some required information, which includes choosing a username and identifying your organization and department. You must also agree to the KBase User Agreement and KBase Data Policy.'},\n",
       " {'question': 'What happens after I agree to the KBase User Agreement and KBase Data Policy?',\n",
       "  'answer': 'Once you’ve checked both boxes for User Agreement and Data Policy, you will be able to click the blue button at the bottom that says “Create KBase Account”.'},\n",
       " {'question': 'What should I do if I encounter difficulties or have questions when creating a KBase account?',\n",
       "  'answer': 'If you encounter any difficulties or have any questions, please feel free to contact us.'},\n",
       " {'question': 'What can I do after creating a KBase account?',\n",
       "  'answer': 'After creating a KBase account, you can click “Continue” and you’ll see your KBase landing page. You may want to take a look at the Narrative Quick Start or other documentation before you plunge headfirst into KBase.'},\n",
       " {'question': 'What is the previous step before contacting KBase?',\n",
       "  'answer': 'The previous step is Signing Up and Signing In.'},\n",
       " {'question': 'What is the next step after contacting KBase?',\n",
       "  'answer': 'The next step is Linking KBase to ORCiD.'},\n",
       " {'question': 'When was the last modification made to the contact us section?',\n",
       "  'answer': 'The last modification was made 7 months ago.'},\n",
       " {'question': 'How can I link my ORCiD account to KBase?',\n",
       "  'answer': 'First, sign into KBase through your Globus or Google accounts. Then, on the left sidebar menu, click the “Account” icon, and then choose the tab labeled “Linked Sign-In Accounts”. At the bottom of the tab, choose \"ORCiD” from the pulldown list of identity providers, and then click the “Link” button. The next screen will prompt you to Sign in using your ORCiD account to link.'},\n",
       " {'question': 'What is ORCiD in KBase?',\n",
       "  'answer': 'ORCiD is an account that you can sign into KBase using it and link it to your KBase account.'},\n",
       " {'question': 'Can I link my ORCiD account to KBase if I am already a KBase User?',\n",
       "  'answer': 'Yes, if you are already a KBase User, first sign into KBase through your Globus or Google accounts, then you can link your ORCiD account to KBase.'},\n",
       " {'question': 'How can I link my ORCiD account to KBase?',\n",
       "  'answer': 'After clicking the Link button, you will be returned to the Linked Sign-in Accounts screen in your Account Manager, where you will see your newly linked ORCiD account.'},\n",
       " {'question': 'Can I sign-in to KBase using my ORCiD ID?',\n",
       "  'answer': 'Yes, you can sign-in to KBase using your ORCiD ID.'},\n",
       " {'question': 'What should I do if I encounter difficulties while signing in to KBase with my ORCiD credentials?',\n",
       "  'answer': 'If you encounter any difficulties or have any questions, you can contact KBase for assistance.'},\n",
       " {'question': 'What is the benefit of linking accounts in KBase?',\n",
       "  'answer': 'Linking Accounts can be helpful for data transfers, ease of signing in, and sharing data.'},\n",
       " {'question': 'Which accounts can be linked to KBase?',\n",
       "  'answer': 'You can link your KBase Account to Google, Globus, and ORCiD.'},\n",
       " {'question': 'How to link accounts in KBase?',\n",
       "  'answer': 'First sign into KBase and go through the normal sign-in process. On the left sidebar menu, click the “Account” icon, and then choose the tab labeled “Linked Sign-In Accounts”. At the bottom of the tab, choose an account type from the pulldown list of identity providers, and then click the “Link” button. The next screen will prompt you to Sign in using your to the chosen account type to link. After you click that Link button, you will be returned to the Linked Sign-in Accounts screen, where you will see your newly linked account.'},\n",
       " {'question': 'Can I link my Google account to KBase?',\n",
       "  'answer': 'Yes, if you have a Google (gmail) account, you can link it to your KBase account.'},\n",
       " {'question': 'How can I link my Google account to my KBase account?',\n",
       "  'answer': 'You can link your Google account to your KBase account by following these steps:\\n1. Sign in to KBase.\\n2. On the left sidebar of your menu, click the “Account” icon, and then choose the tab labeled “Linked Sign-In Accounts”.\\n3. At the bottom of the tab, choose “Google” from the pulldown list of identity providers.\\n4. Click the “Link” button. The next screen will prompt you to choose a Google account to link.\\n5. After you click that Link button, you will be returned to the Linked Sign-in Accounts screen, where you will see your newly linked Google account.'},\n",
       " {'question': 'How can I sign in to KBase using my Google account?',\n",
       "  'answer': 'Once you have linked your Google account to your KBase account, you can sign in to KBase using the “Sign in with Google” button. This will bring you straight to the “Choose an account” Google page.'},\n",
       " {'question': 'What is the \"Linked Sign-In Accounts\" tab for in KBase?',\n",
       "  'answer': 'The \"Linked Sign-In Accounts\" tab in KBase is where you can link your KBase account with other identity providers, such as Google.'},\n",
       " {'question': 'How can I sign in to KBase?',\n",
       "  'answer': 'You can sign in to KBase using your linked Google account.'},\n",
       " {'question': 'What happens if I had signed out of my linked Google account?',\n",
       "  'answer': 'If you had signed out of your linked Google account, you will be directed to sign into it again before you login to KBase.'},\n",
       " {'question': 'What should I do if I encounter any difficulties or have any questions while signing in to KBase?',\n",
       "  'answer': 'If you encounter any difficulties or have any questions, you can contact KBase for assistance.'},\n",
       " {'question': 'What is the new sign in process for KBase?',\n",
       "  'answer': 'You will be able to sign into KBase with a Google, Globus, or ORCiD once accounts have been linked.'},\n",
       " {'question': 'How can new users create a KBase account?',\n",
       "  'answer': 'New users can create a KBase account by following the step-by-step signup guide provided on the KBase website.'},\n",
       " {'question': 'What should existing KBase users do after the authentication update?',\n",
       "  'answer': 'Existing KBase users will need to go through a few steps to convert their existing KBase (through Globus) account to the new system. After going through the conversion, they’ll be able to sign in to KBase with their existing account or their Google account, if they choose to link their KBase account.'},\n",
       " {'question': 'How can I sign in to KBase?',\n",
       "  'answer': 'You can sign in to KBase using Globus. Click the Sign In button below “Use KBase” and sign-in options will appear. Then click “Sign in with Globus”, a page to Log in with your Globus credentials will open. Enter your usual KBase username and password and click the “Log In” button. You will be taken to a page where KBase requests access to your email address and Globus identities. Click the blue “Allow” button.'},\n",
       " {'question': 'What is my KBase username?',\n",
       "  'answer': 'Your KBase username is not your email address – it should not contain an “@” sign.'},\n",
       " {'question': 'What should I do if I forgot my KBase password?',\n",
       "  'answer': 'If you forgot your password, use the “Forgot password?” link at the lower right.'},\n",
       " {'question': 'What do I need to do to continue using KBase?',\n",
       "  'answer': 'You need to read and agree to the KBase User Agreement and KBase Data Policy. After checking the checkboxes that say “I have read and agree to this policy”, you will be able to click the blue button at the bottom that says “Continue to the KBase account [username]”.'},\n",
       " {'question': 'What happens after I click the \"Continue\" button?',\n",
       "  'answer': 'After clicking the \"Continue\" button, you should see the KBase User Interface landing page.'},\n",
       " {'question': 'What should I do if my profile looks bare-bones after the system changes?',\n",
       "  'answer': 'You should fill in the information in your profile when you get a chance, and choose or upload a new avatar if you like.'},\n",
       " {'question': 'What is the process to sign into KBase using Google credentials?',\n",
       "  'answer': 'The process to sign into KBase using Google credentials is as follows:\\n    1. Go to narrative.kbase.us\\n    2. Click the Sign In button below “Use KBase” and more options will appear\\n    3. Click “Sign in with Globus”\\n    4. Click the blue button at the bottom that says “Sign in to KBase account [username]” and you will be at your KBase landing page!'},\n",
       " {'question': 'What happens if I had signed out completely from Globus after the last time I used KBase?',\n",
       "  'answer': 'If you had signed out completely from Globus after the last time you used KBase, you might have to go through two extra steps to sign-in to Globus. Click \"Continue\" and then proceed to enter Globus Log in information.'},\n",
       " {'question': 'What should I do if I encounter any difficulties or have any questions while signing into KBase?',\n",
       "  'answer': 'If you encounter any difficulties or have any questions, you should contact KBase support.'},\n",
       " {'question': 'When was the last modification made to the Supported Browsers section?',\n",
       "  'answer': 'The last modification was made 7 months ago.'},\n",
       " {'question': 'What is the primary browser supported by KBase?',\n",
       "  'answer': 'Chrome is the primary browser supported by KBase.'},\n",
       " {'question': 'What browsers are supported by KBase on a Mac operating system?',\n",
       "  'answer': 'Chrome, Firefox, and Safari 10 (macOS Sierra) and up are supported by KBase on a Mac operating system.'},\n",
       " {'question': 'What could prevent KBase from working in my web browser?',\n",
       "  'answer': 'Ad blockers and some browser extensions may prevent KBase from working in your web browser.'},\n",
       " {'question': 'What should I do if I encounter issues within KBase?',\n",
       "  'answer': 'You may want to disable browser extensions if you encounter issues within KBase. If the problem persists, report it through the Help Board.'},\n",
       " {'question': 'Is KBase supported on mobile browsers?',\n",
       "  'answer': 'KBase is currently not supported on mobile browsers.'},\n",
       " {'question': 'Is KBase supported on mobile browsers?',\n",
       "  'answer': 'No, KBase is currently not supported on mobile browsers.'},\n",
       " {'question': 'Can I use KBase on my tablet?',\n",
       "  'answer': 'Yes, the KBase website and services should work with tablet browsers on iPad and Android devices in landscape orientation, but there may be compatibility issues as they do not currently test on these platforms.'},\n",
       " {'question': 'Which browser is recommended for using KBase?',\n",
       "  'answer': 'Chrome is the best browser to use for KBase.'},\n",
       " {'question': 'How often is Chrome updated?',\n",
       "  'answer': 'Chrome operates on a rolling release, meaning that it is updated frequently.'},\n",
       " {'question': 'What is Firefox?',\n",
       "  'answer': 'Firefox, produced by the non-profit Mozilla Corporation, is a freely available, open source browser for Mac, Windows, Linux, and other systems.'},\n",
       " {'question': 'What is Safari?',\n",
       "  'answer': 'Safari is a web browser distributed with Mac OS X. KBase does not support the Windows version of Safari.'},\n",
       " {'question': 'What is Microsoft Edge?',\n",
       "  'answer': 'Microsoft Edge is the new web browser for Windows 10; it replaces Internet Explorer. You should be able to use KBase on Edge, but this has not been extensively tested.'},\n",
       " {'question': 'Is it recommended to use a specific version of Firefox for KBase?',\n",
       "  'answer': 'No, it is not feasible to recommend a specific (or even minimum) version. The best advice is to make sure you are using the most recent version available.'},\n",
       " {'question': 'Is Safari supported on Windows for KBase?',\n",
       "  'answer': 'No, KBase does not support the Windows version of Safari.'},\n",
       " {'question': 'Is KBase supported on Internet Explorer?',\n",
       "  'answer': 'No, KBase is not supported on Internet Explorer.'},\n",
       " {'question': 'Can I use KBase on Microsoft Edge?',\n",
       "  'answer': 'Yes, KBase should work on Microsoft Edge, but it has not been extensively tested. For the best chance of success on Edge, be sure you are using the most recent version.'},\n",
       " {'question': 'Are older versions of supported browsers compatible with KBase?',\n",
       "  'answer': 'No, older versions of supported browsers are not supported. Please ensure that your browser and system are up-to-date before reporting a problem with a supported browser.'},\n",
       " {'question': 'Can I use KBase on Opera browser?',\n",
       "  'answer': 'Other modern browsers, such as Opera, may work with KBase, but as KBase does not test with them, no assurances about their operation can be made.'},\n",
       " {'question': 'What is a Narrative in KBase?',\n",
       "  'answer': 'In KBase, you can create shareable, reproducible workflows called Narratives that include data, analysis steps, results, visualizations, and commentary.'},\n",
       " {'question': 'How to start using KBase?',\n",
       "  'answer': 'To begin using KBase, you will first need to sign up for a KBase Account. You can now use your existing Google or Globus username and password to get a free KBase account.'},\n",
       " {'question': 'How to access Narratives in KBase?',\n",
       "  'answer': 'After you sign in, you will be taken to your Narratives. From Narratives, you can open your existing Narratives, access others that have been shared with you, and create new Narratives. These Narratives can be accessed at narrative.kbase.us.'},\n",
       " {'question': 'How to create a new Narrative in KBase?',\n",
       "  'answer': 'Click the “+ New Narrative” button in Narratives to open a new untitled Narrative. A “Welcome to the Narrative Interface” box offers information and links to documentation for adding content to your Narrative. You can collapse or delete this box using the “…” dropdown menu in the top right corner.'},\n",
       " {'question': 'What is the Narrative tour in KBase?',\n",
       "  'answer': 'The \"Narrative Tour\" is a new feature in KBase that walks you through the user interface, pointing out various useful aspects of it. It can be selected from the Help menu.'},\n",
       " {'question': 'How to load data to your Narrative in KBase?',\n",
       "  'answer': 'To load data to your Narrative, click the “Add Data” button in the Data Panel, which can be found under the Analyze tab near the top left of the Narrative.'},\n",
       " {'question': 'How can I add data to my Narrative in KBase?',\n",
       "  'answer': 'When you hover over a data object in the Data Browser, a blue “< Add” button will appear to its left. Click the button to add the data to your Narrative.'},\n",
       " {'question': 'How can I upload my own data to KBase?',\n",
       "  'answer': 'The Import tab allows you to upload your own data. KBase currently supports upload of a variety of data types.'},\n",
       " {'question': 'Is my data private in KBase?',\n",
       "  'answer': 'Any data you upload to KBase is private unless you choose to share it.'},\n",
       " {'question': 'What does the Example tab in KBase contain?',\n",
       "  'answer': 'The Example tab contains sample data from KBase’s reference data collection that can be used to try out the apps.'},\n",
       " {'question': 'How can I exit the Data Browser in KBase?',\n",
       "  'answer': 'You can exit the Data Browser by clicking either the “Close” button at the bottom left of the browser window or the arrow at the top right of the Data Panel.'},\n",
       " {'question': 'How can you save your work in KBase?',\n",
       "  'answer': 'You can save your work in KBase by clicking the “Save” button at the top right of the interface.'},\n",
       " {'question': 'How can you add an app to your Narrative in KBase?',\n",
       "  'answer': 'To add an app to your Narrative from the App Panel, click on its name or icon.'},\n",
       " {'question': 'How can you run an app in KBase?',\n",
       "  'answer': 'To run an app in KBase, fill in all the required fields in your App. Some app fields are “smart” and can recognize which data in your Narrative are valid input for those fields. These smart fields have a pulldown list of all the data objects from your Data Panel that you can choose from as input.'},\n",
       " {'question': 'How can you get more information about an app in KBase?',\n",
       "  'answer': 'You can get more information about an app by hovering over it until “…” appears and clicking on it.'},\n",
       " {'question': 'How can you filter apps in KBase?',\n",
       "  'answer': 'In KBase, there are options for filtering apps by category, name, input type, and more. You can also designate apps you like as favorites using the star icon.'},\n",
       " {'question': 'How to start the analysis in KBase?',\n",
       "  'answer': 'Fill out all required fields and click the green “Run” button near the top left of the app cell.'},\n",
       " {'question': 'Where can I see the progress of the analysis?',\n",
       "  'answer': 'You can see the progress in the Job Status tab of the app box.'},\n",
       " {'question': 'How to share my Narrative in KBase?',\n",
       "  'answer': 'You can make a Narrative public or share it with specific collaborators by clicking the “share” button near the top right of your Narrative.'},\n",
       " {'question': 'Are the Narratives private or public by default?',\n",
       "  'answer': 'All Narratives that you create are private by default.'},\n",
       " {'question': 'Where can I find more details on running apps in KBase?',\n",
       "  'answer': 'You can find more details in the KBase Apps section of the Narrative Interface User Guide.'},\n",
       " {'question': 'What is the Narrative Interface User Guide?',\n",
       "  'answer': \"The Narrative Interface User Guide is a detailed introduction to KBase's Narrative interface. It includes topics such as accessing the narrative, exploring data in KBase, adding data to a narrative, analyzing data using Apps, and sharing a narrative.\"},\n",
       " {'question': 'How can I create a new Narrative in KBase?',\n",
       "  'answer': \"You can refer to the 'Create a new Narrative' section in the Narrative Interface User Guide.\"},\n",
       " {'question': 'How can I share a Narrative in KBase?',\n",
       "  'answer': \"You can refer to the 'Share a Narrative' section in the Narrative Interface User Guide.\"},\n",
       " {'question': 'How can I analyze data using Apps in KBase?',\n",
       "  'answer': \"You can refer to the 'Analyze data using Apps' section in the Narrative Interface User Guide.\"},\n",
       " {'question': 'How can I add data to a Narrative in KBase?',\n",
       "  'answer': \"You can refer to the 'Add data to a Narrative' section in the Narrative Interface User Guide.\"},\n",
       " {'question': 'What is the Narrative Interface?',\n",
       "  'answer': 'The Narrative Interface enables researchers to design and carry out computational experiments while creating interactive, reproducible records of the data, computational steps, and thought processes underpinning their results. These records can be shared with collaborators and published as “active papers,” called Narratives, which allow others to repeat the computational experiment and alter parameters or input data to produce different or improved results. The Narrative Interface is built on top of the Jupyter Notebook framework.'},\n",
       " {'question': 'What is the purpose of this guide?',\n",
       "  'answer': 'This guide is for the Narrative Interface, the graphical user interface for accessing KBase’s functionality and data.'},\n",
       " {'question': 'What are Narratives in KBase?',\n",
       "  'answer': 'Narratives are interactive, reproducible records of the data, computational steps, and thought processes underpinning the results of computational experiments. They can be shared with collaborators and published as “active papers,” allowing others to repeat the computational experiment and alter parameters or input data to produce different or improved results.'},\n",
       " {'question': 'What is the Narrative Interface built on top of?',\n",
       "  'answer': 'The Narrative Interface is built on top of the Jupyter Notebook framework.'},\n",
       " {'question': 'How can I access the Narrative Interface?',\n",
       "  'answer': 'The Narrative Interface is available at narrative.kbase.us. Once you sign in, you will see the Narratives Navigator.'},\n",
       " {'question': 'What are the ways to enter the Narrative Interface from KBase?',\n",
       "  'answer': 'There are several ways to enter the Narrative Interface from KBase: 1. Use the menu at the top left of the landing page and select “Narrative Interface”. 2. Open a public Narrative or one that has been shared with you. 3. Start a new Narrative by clicking the “+ New Narrative” button.'},\n",
       " {'question': 'How can I update to the latest version of the Narrative Interface?',\n",
       "  'answer': 'New versions of the Narrative Interface are released periodically. In most cases, KBase will automatically update to the latest version. However, if you had an older version already running, you will see a green button near the top right of your Narrative that says “New Version Available.” Click this button to update and reload the interface. You also may need to do a hard refresh (shift-reload) in your browser window.'},\n",
       " {'question': 'What is the \"Welcome\" cell in KBase?',\n",
       "  'answer': 'When you open a new Narrative in KBase, a “Welcome” cell will appear at the top, offering quick tips for getting started and links for accessing help.'},\n",
       " {'question': 'How can I collapse the \"Welcome\" cell in KBase?',\n",
       "  'answer': 'You can collapse the \"Welcome\" cell by clicking the \"-\" button in the top right corner of the cell.'},\n",
       " {'question': 'How can I delete the \"Welcome\" cell in KBase?',\n",
       "  'answer': 'You can delete the \"Welcome\" cell by selecting \"Delete cell\" from the \"...\" dropdown menu.'},\n",
       " {'question': 'What is the purpose of the \"Welcome\" cell in KBase?',\n",
       "  'answer': 'The \"Welcome\" cell is a quick link to the KBase documentation site and shares information on any service status for kbase.us.'},\n",
       " {'question': 'What is the Narratives Navigator in KBase?',\n",
       "  'answer': 'The Narratives Navigator lets you access other Narratives from a Narrative.'},\n",
       " {'question': 'What does the \"About the Narrative\" option do?',\n",
       "  'answer': 'It opens a window describing Narrative Interface properties (including version number) and provides a link to notes about the latest KBase release.'},\n",
       " {'question': 'What is the function of \"Shutdown and Restart\"?',\n",
       "  'answer': 'It closes and relaunches the Narrative Interface, which can be useful if your content fails to load or the Narrative appears frozen.'},\n",
       " {'question': 'What can be done through the top-level menu?',\n",
       "  'answer': 'The top-level menu offers additional possibilities such as going to the main Narrative Interface, opening a new Narrative, going to the Data Search tab, going to the Compound and Reaction search tabs, opening a page that describes the KBase User Interface and connecting to the KBase GitHub, and going to a page that describes the current version of the Narrative Interface.'},\n",
       " {'question': 'What does the \"New Narrative\" option do?',\n",
       "  'answer': 'It opens a new Narrative in a new window.'},\n",
       " {'question': 'What does the \"KBase Services Status\" option do?',\n",
       "  'answer': 'It goes to a page that describes the current version of the Narrative Interface.'},\n",
       " {'question': 'What does the \"Contact KBase\" feature do?',\n",
       "  'answer': 'It allows you to contact the KBase Help Desk with your question, bug report, or feature suggestion.'},\n",
       " {'question': 'What does the \"Support\" feature do?',\n",
       "  'answer': 'It brings you to the Narrative User Guide.'},\n",
       " {'question': 'What does the \"KBase Services Status\" feature do?',\n",
       "  'answer': 'It directs to information on service versions and performance.'},\n",
       " {'question': 'What does the \"KBase UI Plugins\" feature do?',\n",
       "  'answer': 'It lists user interface plugins and their version info and links to GitHub.'},\n",
       " {'question': 'What does the \"Developer\" feature do?',\n",
       "  'answer': 'It links to developer tools and the configuration editor.'},\n",
       " {'question': 'What does the \"Public Search\" feature do?',\n",
       "  'answer': 'It can be used to locate data and filter by data types.'},\n",
       " {'question': 'What does the \"About\" feature do?',\n",
       "  'answer': 'It directs to information about the KBase User Interface and current version.'},\n",
       " {'question': 'What does the \"KBase (home)\" feature do?',\n",
       "  'answer': 'It will open a new window to kbase.us.'},\n",
       " {'question': 'What does the \"Docs\" feature do?',\n",
       "  'answer': 'It will open a new window to the KBase Documentation site at docs.kbase.us.'},\n",
       " {'question': 'What does the \"Support\" feature do?',\n",
       "  'answer': 'It will open www.kbase.us/support to access the Help Board or contact the KBase team.'},\n",
       " {'question': 'What is the function of view-only mode in KBase?',\n",
       "  'answer': 'In view-only mode, you can look at the Narrative but not run or change it.'},\n",
       " {'question': 'What is the function of the Help menu in KBase?',\n",
       "  'answer': 'The Help menu links to documentation as well as to the Narrative Tour.'},\n",
       " {'question': 'What is the function of the Share button in KBase?',\n",
       "  'answer': 'The Share button allows you to make your Narrative public or give specific KBase users the ability to view, edit, or share it further.'},\n",
       " {'question': 'What is the function of the Save icon in KBase?',\n",
       "  'answer': 'The Save icon allows you to save your work often.'},\n",
       " {'question': 'What is the function of the User icon in KBase?',\n",
       "  'answer': 'The User icon allows you to sign in/out or access your user profile.'},\n",
       " {'question': 'What is the function of the Analyze tab in KBase?',\n",
       "  'answer': 'The Analyze tab consists of the Data Panel and the Apps Panel below it.'},\n",
       " {'question': 'What is the function of the Narratives tab in KBase?',\n",
       "  'answer': 'The Narratives tab allows you to create a new Narrative, copy an existing Narrative, and see a list of other Narratives created by or shared with you.'},\n",
       " {'question': 'How can you add data to your Narrative in KBase?',\n",
       "  'answer': 'You can add data to your Narrative by searching KBase’s reference collection, importing datasets from external sources, or generating new data from KBase analyses.'},\n",
       " {'question': 'What is the function of the Data Panel controls in KBase?',\n",
       "  'answer': 'The Data Panel controls are used to search, filter, sort, and refresh the data in your panel. The right arrow will open the Data Browser.'},\n",
       " {'question': 'What information can be seen when a data object in the panel is expanded?',\n",
       "  'answer': 'When a data object in the panel is expanded, you can see more information about the data and several options for downloading, viewing data provenance, and more.'},\n",
       " {'question': 'What does the \"Add Data\" function do in KBase?',\n",
       "  'answer': 'The \"Add Data\" function opens the Data browser, which provides options for working with your own data or data from KBase’s reference collection.'},\n",
       " {'question': 'What is the function of the Apps Panel in KBase?',\n",
       "  'answer': 'The Apps Panel lists KBase analysis functions. Clicking the app name or icon beside it will add an analysis cell to your Narrative.'},\n",
       " {'question': 'What are the App Controls in KBase?',\n",
       "  'answer': 'The App Controls are icons used to search and refresh the list of apps, as well as expose apps that are available but still in beta. The right arrow will open the App Catalog, which provides advanced options for browsing apps; designating them as favorites; and filtering them based on analysis type, popularity, and more.'},\n",
       " {'question': 'What is a Markdown Cell in KBase?',\n",
       "  'answer': 'A Markdown Cell is used to add rich text commentary and graphics to your Narrative.'},\n",
       " {'question': 'What is an App Cell in KBase?',\n",
       "  'answer': 'An App Cell is one of several types of cells, including those displaying app inputs and outputs, text and commentary, and code cells for writing custom scripts. The cell controls are at the top right of each cell.'},\n",
       " {'question': 'What are the Markdown and Code Cell buttons in KBase?',\n",
       "  'answer': 'The Markdown and Code Cell buttons are used to add and control Markdown and Code cells in your Narrative.'},\n",
       " {'question': 'What can be added to your Narrative in KBase?',\n",
       "  'answer': 'You can add different types of cells to your Narrative in KBase.'},\n",
       " {'question': 'Is there documentation on how to use code cells in KBase?',\n",
       "  'answer': 'There will be more documentation on how to use code cells in KBase in the future.'},\n",
       " {'question': 'When was the last modification made to this information?',\n",
       "  'answer': 'The last modification was made 1 year ago.'},\n",
       " {'question': 'What are the five tabs for exploring and locating existing Narratives in KBase?',\n",
       "  'answer': 'The five tabs are My Narratives, Shared with Me, Tutorials, and Public.'},\n",
       " {'question': 'How are the Narratives sorted in KBase?',\n",
       "  'answer': 'The most recently updated Narratives appear first, but can be searched or sorted using the dropdown menu - Recently updated, Least recently updated, Recently created, Oldest, Lexographic (A-Z), Reverse Lexicographic (Z-A).'},\n",
       " {'question': 'How can I find a specific Narrative in KBase?',\n",
       "  'answer': 'You can use the search function or sort the Narratives using the dropdown menu options.'},\n",
       " {'question': 'How can I see an overview of the Data within the Narrative and a Preview?',\n",
       "  'answer': 'A list of available Narratives will be on the left panel and information on the selected Narrative on the right panel. Here you can see an overview of the Data within the Narrative and a Preview.'},\n",
       " {'question': 'What do the icons under a Narrative name indicate?',\n",
       "  'answer': 'The icons under a Narrative name indicate which apps the Narrative includes, how many markdown and code cells it contains, the number of times the Narrative has been shared, whether there are any running jobs, total running time for all completed jobs, and more.'},\n",
       " {'question': 'How can I open a selected Narrative?',\n",
       "  'answer': 'Click on the selected Narrative name to open it.'},\n",
       " {'question': 'How can I manage a Narrative?',\n",
       "  'answer': 'Click on the Gear icon next to the Narrative title for management options to: Manage Sharing, Copy this Narrative, Rename, Link to Organization, Delete.'},\n",
       " {'question': 'How can I create a new Narrative?',\n",
       "  'answer': 'Click the green + New Narrative button at the top to create a new Narrative.'},\n",
       " {'question': 'What are Public Narratives in KBase?',\n",
       "  'answer': 'Public Narratives are Narratives that have been made public by their owners in KBase. They have not necessarily been reviewed or tested by the KBase team.'},\n",
       " {'question': 'What does the Narratives Navigator offer?',\n",
       "  'answer': 'The Narratives Navigator offers many launch points, including access to your Narratives and a list of the Organizations within KBase.'},\n",
       " {'question': 'How can I join an Organization in KBase?',\n",
       "  'answer': 'Under \"All Orgs\" in the Narratives Navigator, you can search and select different Organizations and request to join by clicking on the \"Join this Organization\" button.'},\n",
       " {'question': 'What information can I find about Organizations in KBase?',\n",
       "  'answer': 'You can find descriptions and statistics for the Organizations, including Narratives and Apps.'},\n",
       " {'question': 'What is the App Catalog in KBase?',\n",
       "  'answer': 'The App Catalog in KBase allows you to browse, search for, and see the specifications of each app KBase has to offer.'},\n",
       " {'question': 'What does the Data Search tab in KBase allow you to do?',\n",
       "  'answer': 'The Data Search tab gives you the ability to search JGI data (Beta) as well as KBase user and reference data. It allows you to search for data present in your own Narratives, those shared with you, and those made public by their owners. It also enables you to search for reads and assemblies contained in the Joint Genome Institute (JGI) Genomes Online Database (GOLD) and add them directly to your Narratives.'},\n",
       " {'question': 'What is the Job Browser in KBase?',\n",
       "  'answer': 'The Job Browser allows you to view apps that you have run recently, see their progress, search for specific runs, and more. You can filter the runs by Finished, Queued, Running, Success, and Error. Query results also display basic information surround the runs such as the Narrative they can be found in, App ID, Submission Time, Queue Time, Run Time, and Status.'},\n",
       " {'question': 'What can you do in the Account Manager in KBase?',\n",
       "  'answer': 'In the Account Manager you have access to edit your profile information.'},\n",
       " {'question': 'How can I access my Profile Page in KBase?',\n",
       "  'answer': 'You can access your Profile Page from the tab which gives an overview of your Narratives and your Collaborators.'},\n",
       " {'question': 'How can I access the user profile of my Collaborators in KBase?',\n",
       "  'answer': 'You can access the user profile of any of your Collaborators by simply clicking on the name under the Your Collaborators panel while in your Profile Page.'},\n",
       " {'question': 'What is the function of the Feeds tab in KBase?',\n",
       "  'answer': 'The Feeds tab takes you to the Notification Feeds page to view notifications from KBase.'},\n",
       " {'question': 'How can I create a new Narrative in KBase?',\n",
       "  'answer': 'You can create a new Narrative from your Narratives by clicking the “+ New Narrative” button, which will take you to a new Narrative Interface. If you are already in the Narrative Interface, you can start a new Narrative by selecting the Narratives tab (next to Analyze) and clicking the button labeled “+ New Narrative.”'},\n",
       " {'question': 'What happens when I create a new Narrative?',\n",
       "  'answer': 'Your Narrative will open in a new browser tab and contain a default “Welcome to the Narrative” cell that provides brief instructions for using the Narrative Interface. You can retain this cell or delete it with the “Delete cell” option in the “…” dropdown menu in the upper right corner of the cell.'},\n",
       " {'question': 'How can I name my new Narrative?',\n",
       "  'answer': 'You can name the Narrative now by clicking on “Untitled” and entering a new name. The new Narrative will not appear in your Narratives list until it is named something other than ‘Untitled’ and you have clicked the Save button to save it.'},\n",
       " {'question': 'Can I share my Narrative with other users on KBase?',\n",
       "  'answer': 'Yes, any Narrative you create is private until you choose to share it.'},\n",
       " {'question': \"What can I do if I don't feel ready to create my own Narrative?\",\n",
       "  'answer': 'You can copy and run a Narrative that someone else has shared with you.'},\n",
       " {'question': 'What will be explained in the following sections?',\n",
       "  'answer': 'The following sections will explain how to explore data, add data to your Narrative, and use apps to analyze it.'},\n",
       " {'question': 'Is my Narrative automatically shared with other users when I create it?',\n",
       "  'answer': 'No, any Narrative you create is private until you choose to share it.'},\n",
       " {'question': 'How can I find data in KBase?',\n",
       "  'answer': 'You can select data available within KBase, upload your own files, or import datasets from external resources such as the DOE Joint Genome Institute (JGI) or NCBI.'},\n",
       " {'question': 'What is the Data Browser in KBase?',\n",
       "  'answer': 'The Data Browser is a tool in KBase that allows you to explore the wide range of data available. It can be accessed by clicking the right-facing arrow in the menu or the red rectangular \"Add Data\" button in the Data Panel.'},\n",
       " {'question': 'What are the tabs in the Data Browser?',\n",
       "  'answer': 'The Data Browser has two tabs: \"My Data\", which shows data you have already loaded or analyzed in another Narrative, and \"Shared With Me\", which shows data included in Narratives that have been shared with you.'},\n",
       " {'question': 'What is the Public tab in KBase?',\n",
       "  'answer': 'The Public tab in KBase displays a list of data objects available in the KBase reference collection that is accessible to everyone.'},\n",
       " {'question': 'How can you change the type of data displayed under the Public tab in KBase?',\n",
       "  'answer': 'You can change the type of data displayed under the Public tab by using the data-type selector at the top left of the browser window.'},\n",
       " {'question': 'How can you search for specific data objects in KBase?',\n",
       "  'answer': 'You can use the “Search/Filter data...\" field in the Data Browser to find data objects whose names include the text you’ve typed in the search box. Searches are not case-sensitive.'},\n",
       " {'question': 'What does the Import tab in KBase do?',\n",
       "  'answer': 'The Import tab in KBase allows you to import your own data of supported types to your Narrative.'},\n",
       " {'question': 'What is the Example tab in KBase?',\n",
       "  'answer': 'The Example tab in KBase provides example datasets that can be used as inputs to apps and methods.'},\n",
       " {'question': 'How can you add the Vibrio brasiliensis LMG 20546 genome to your Narrative in KBase?',\n",
       "  'answer': 'By clicking the \"< Add\" button.'},\n",
       " {'question': 'What does the Data Panel in KBase show?',\n",
       "  'answer': 'The Data Panel shows all the data that you’ve added to your Narrative.'},\n",
       " {'question': 'How can you access more details about a particular data object in KBase?',\n",
       "  'answer': 'You can access more details about a particular data object by hovering over the object and clicking the ”...” that appears in the right-hand side of the data object.'},\n",
       " {'question': 'What can you do with the icons in the Data Panel header in KBase?',\n",
       "  'answer': 'You can search, sort, or filter the list using the icons in the Data Panel header.'},\n",
       " {'question': 'What happens when you add or generate more data during the course of your analyses in KBase?',\n",
       "  'answer': 'The number of objects in the Data Panel will increase.'},\n",
       " {'question': 'What does the arrow in icon represent in KBase?',\n",
       "  'answer': 'The arrow in icon for \"Show Apps with this as input\" shows the apps (in the App Panel) that can use this type of data for input.'},\n",
       " {'question': 'What does the binoculars icon do in KBase?',\n",
       "  'answer': 'The binoculars icon is the \"Explore data\" button that open a Data Landing page for the data object.'},\n",
       " {'question': 'What does the page with three lines and folded upper right corner icon do in KBase?',\n",
       "  'answer': 'The page with three lines and folded upper right corner icon is the \"View associated report\" button and will open the Data View report page for the data object.'},\n",
       " {'question': 'What does the counter clockwise circle with clock hands icon do in KBase?',\n",
       "  'answer': 'The counter clockwise circle with clock hands is the \"View history to revert changes\" icon allows you to see and revert to previous versions of the data.'},\n",
       " {'question': 'What does the graph-like icon do in KBase?',\n",
       "  'answer': 'The graph-like icon is the \"View data provenance and relationships\" button that opens the Provenance page.'},\n",
       " {'question': 'What does the download icon do in KBase?',\n",
       "  'answer': 'The download icon is the \"Export/Download data\" button that lets you download a data object to your local computer. Currently, most data objects can be downloaded only in the JSON format.'},\n",
       " {'question': 'What does the “A” icon do in KBase?',\n",
       "  'answer': 'The “A” icon is the \"Rename data\" button that lets you rename your data object. Object names can contain only letters, digits, and underscores; no spaces or other special characters are allowed.'},\n",
       " {'question': 'What does the trash can icon do in KBase?',\n",
       "  'answer': 'The trash can icon is the \"Delete data\" button that lets you delete a data object from your Narrative. If the data came from a public or shared data source, only the copy within your Narrative will be deleted.'},\n",
       " {'question': 'What should I be cautious about when renaming a data object in KBase?',\n",
       "  'answer': 'If a Narrative was already using the old object name in analyses, they might stop working after the renaming.'},\n",
       " {'question': 'How can you access data viewers in KBase?',\n",
       "  'answer': 'There are two ways to access data viewers in KBase: 1. Click the name of a data object, and its viewer will be added to your Narrative. 2. Drag the object from the Data Panel and drop it into the main part of your Narrative.'},\n",
       " {'question': 'What is the function of data viewers in KBase?',\n",
       "  'answer': 'Data viewers in KBase allow you to learn more about the data.'},\n",
       " {'question': 'What information can be found in the genome viewer for a genome in the Data Panel?',\n",
       "  'answer': 'The genome viewer has tabs for an overview (including GC content, taxonomy information, size, and more) and a list of contigs and genes. Each contig and gene entry in these lists is clickable, opening either a contig browser or a tab with expanded information about the gene.'},\n",
       " {'question': 'How can you sort table entries in KBase?',\n",
       "  'answer': 'You can sort the table entries by clicking on a column header to sort by that field. Clicking the same column header again will reverse the sort order.'},\n",
       " {'question': 'How can you sort by more than one column at a time in KBase?',\n",
       "  'answer': 'You can sort by more than one column at a time by clicking one column header and then Shift-clicking other column headers.'},\n",
       " {'question': 'Are there different types of viewers in KBase?',\n",
       "  'answer': 'Yes, there are many types of viewers in KBase in addition to the Genome viewer. Different viewers may have different options to explore.'},\n",
       " {'question': 'How can I remove a viewer from my Narrative in KBase?',\n",
       "  'answer': 'To remove a viewer from your Narrative, click the trashcan icon in the top right of the viewer cell.'},\n",
       " {'question': 'What happens when I remove a viewer from my Narrative in KBase?',\n",
       "  'answer': 'Removing a viewer from your Narrative doesn’t delete the data object itself. You can always re-add the viewer.'},\n",
       " {'question': 'What are Data Landing and Provenance pages in KBase?',\n",
       "  'answer': 'Data Landing (also known as Data Summary) and Provenance pages are ways to find out even more information about a data object.'},\n",
       " {'question': 'How can I access Data Landing and Provenance pages in KBase?',\n",
       "  'answer': 'You can access these pages using the icons that appear when you hover over a data object in the Data Browser or click on one of the objects in your Data Panel. The binoculars icon opens a new Data Landing page about that particular data object, while the graph-like icon opens a Provenance page.'},\n",
       " {'question': 'What is the purpose of Data Landing pages in KBase?',\n",
       "  'answer': 'Data Landing pages provide both known and contextual information about a data object, allowing users to examine various particulars about the data and, eventually, compare it to other data objects.'},\n",
       " {'question': 'What kind of information might be presented in the Data Landing pages?',\n",
       "  'answer': 'Depending on the type of data and where it came from, different sorts of information might be presented. For example, for a genome already in KBase, you will see a data object summary, a data provenance and reference network, a genome overview panel, a description of the species, publications that mention this genome, the taxonomic lineage, a species tree, a contig browser, functional categories, and more.'},\n",
       " {'question': 'What is a data object summary in KBase?',\n",
       "  'answer': 'A data object summary provides a brief overview of the data object, including its key characteristics and features.'},\n",
       " {'question': 'What is a data provenance and reference network in KBase?',\n",
       "  'answer': 'A data provenance and reference network provides information about the origin and history of the data object, as well as its relationships with other data objects. It is collapsed by default and can be opened by clicking the > on the left.'},\n",
       " {'question': 'What is a genome overview panel in KBase?',\n",
       "  'answer': 'A genome overview panel lists information about the genome, including its biological domain, DNA length, etc.'},\n",
       " {'question': 'What is the purpose of Provenance pages in KBase?',\n",
       "  'answer': 'Provenance pages are one way to facilitate reproducibility and transparency of scientific results, two key principles of KBase design.'},\n",
       " {'question': 'What information can be found on the Provenance page?',\n",
       "  'answer': 'The Provenance page records and illustrates how data is derived and modified in KBase, including how it entered the system, whether it was produced through analysis of other objects, who generated the data, and when. You also can identify the original “owners” of the data, allowing you to contact them or see their shared analyses.'},\n",
       " {'question': 'Can older versions of a data object be accessed in KBase?',\n",
       "  'answer': 'Yes, all data in KBase is versioned, and older versions of a data object can be accessed from the Provenance page.'},\n",
       " {'question': 'How can I add data to my Narrative in KBase?',\n",
       "  'answer': 'From the Data Panel in a Narrative, you can access the data slide-out, which allows you to search for data of interest and add it to your Narrative. In the Data Panel, click the \"Add Data\" button, the “+” button, or the arrow at the upper right of the panel to access the Data Browser slide-out.'},\n",
       " {'question': 'Is my data private in KBase?',\n",
       "  'answer': 'Any data that you upload to KBase is kept private unless you explicitly choose to share it. You can share any of your Narratives (including their associated data) with one or more specific users, or make it publicly available to all KBase users.'},\n",
       " {'question': 'How can I share my data in KBase?',\n",
       "  'answer': 'You can share any of your Narratives (including their associated data) with one or more specific users, or make it publicly available to all KBase users. Please see the Sharing page for more information about how to do that.'},\n",
       " {'question': \"What does the 'My Data' tab show in KBase?\",\n",
       "  'answer': \"The 'My Data' tab shows data objects that you have added to your project.\"},\n",
       " {'question': \"What can be found in the 'Shared With Me' and 'Public' tabs in KBase?\",\n",
       "  'answer': \"The 'Shared With Me' and 'Public' tabs display datasets that others have loaded and made accessible to you (or to everyone).\"},\n",
       " {'question': \"What is the purpose of the 'Example' tab in KBase?\",\n",
       "  'answer': \"The 'Example' tab shows datasets that have been pre-loaded for use with particular apps. These can be handy for trying out the Narrative Interface.\"},\n",
       " {'question': \"What does the 'Import' tab allow you to do in KBase?\",\n",
       "  'answer': \"The 'Import' tab allows you to upload your own datasets for analysis.\"},\n",
       " {'question': 'How can you add a data object to the Narrative in KBase?',\n",
       "  'answer': 'You can add a data object to the Narrative by hovering your cursor over the data object under the first four tabs, then options will appear allowing you to add that object to the Narrative.'},\n",
       " {'question': 'What does the icon to the left of each data object represent?',\n",
       "  'answer': 'The icon to the left of each data object represents its data type.'},\n",
       " {'question': 'How can you upload data from your computer to KBase?',\n",
       "  'answer': 'You can upload data from your computer to KBase by clicking on the Import tab and then the Upload button. This will open a window where you can select files to upload.'},\n",
       " {'question': 'What is the function of the Import tab in KBase?',\n",
       "  'answer': 'The Import tab in KBase lets you drag & drop data from your computer into your Staging Area to import into your Narrative, where you can then analyze it using KBase’s analysis apps.'},\n",
       " {'question': 'How can I get data from my computer to my KBase Narrative?',\n",
       "  'answer': 'It is a three-step process:\\n    1. Drag and drop the data file(s) from your computer to the new Import tab to upload them to your Staging Area.\\n    2. Choose a format for importing data from your Staging Area into your Narrative.\\n    3. Run the Import app that is created.'},\n",
       " {'question': 'What is the first step to get data from my computer to my KBase Narrative?',\n",
       "  'answer': 'The first step is to drag and drop the data file(s) from your computer to the new Import tab to upload them to your Staging Area.'},\n",
       " {'question': 'What is the second step to get data from my computer to my KBase Narrative?',\n",
       "  'answer': 'The second step is to choose a format for importing data from your Staging Area into your Narrative.'},\n",
       " {'question': 'What is the third step to get data from my computer to my KBase Narrative?',\n",
       "  'answer': 'The third step is to run the Import app that is created.'},\n",
       " {'question': 'What is a Staging Area in KBase?',\n",
       "  'answer': 'A Staging Area is a temporary holding place for your uploaded data files in KBase. It is private to you–no one else can see the data in your Staging Area. The files in your Staging Area stay there and can be seen from any of your Narratives.'},\n",
       " {'question': 'How can I add data to a Narrative from the Staging Area?',\n",
       "  'answer': 'You can choose a data type from the pulldown menu next to a data file in your staging area in order to add it to your Narrative as an object of that type.'},\n",
       " {'question': 'Why does KBase have a Staging Area?',\n",
       "  'answer': 'The Staging Area is more robust and extensible. It can handle large data files without timing out. The user interface is more intuitive: you can drag and drop one or more data files into your Staging Area, or even whole directories or zip files. Finally, the new importer is easier for developers to extend to new data types.'},\n",
       " {'question': 'What does \"Upload\" refer to in KBase?',\n",
       "  'answer': '\"Upload\" refers to getting data from your computer into your Staging Area in KBase.'},\n",
       " {'question': 'What does \"Import\" refer to in KBase?',\n",
       "  'answer': '\"Import\" refers to the process of converting a data file from your Staging Area into a data object in your Narrative that you can analyze with any of KBase\\'s analysis apps.'},\n",
       " {'question': 'What is the size limit for drag & drop data files into your Staging Area in KBase?',\n",
       "  'answer': 'The size limit for drag & drop data files into your Staging Area in KBase depends on your computer and browser. Some users have reported problems around 20GB. For larger files, use the Globus Online transfer.'},\n",
       " {'question': 'How can I import files into my KBase account?',\n",
       "  'answer': 'You can drag the files you want to import into the drop zone in the Staging Area. You can select multiple files from your computer and drag them all at once. You can also select a folder of data files and drag the folder into the Staging Area drop zone. Alternatively, you can click in the upload area to open a file chooser and select a file from your computer to upload.'},\n",
       " {'question': 'How will I know if my files are being uploaded to the Staging Area?',\n",
       "  'answer': 'While files are being uploaded to your Staging Area, you’ll see a green progress bar.'},\n",
       " {'question': \"What should I do if I don't see my file in the Staging Area after uploading?\",\n",
       "  'answer': 'If you don’t see your file, try clicking the refresh button.'},\n",
       " {'question': 'How are the files in the Staging Area sorted by default?',\n",
       "  'answer': 'By default, the files in the Staging Area are sorted by age, with the most recently uploaded file at the top. To sort the list by other fields, such as name or size, click a column header.'},\n",
       " {'question': 'What is the lifetime for files in the KBase Staging Area?',\n",
       "  'answer': 'Files in the KBase Staging Area are automatically removed after 90 days.'},\n",
       " {'question': 'What is the purpose of the Staging Area in KBase?',\n",
       "  'answer': 'The Staging Area is meant to be a temporary holding area for data you want to import into your KBase account.'},\n",
       " {'question': 'What happens to data objects imported into a Narrative in KBase?',\n",
       "  'answer': 'Data objects imported into a Narrative will exist indefinitely.'},\n",
       " {'question': 'What is Globus?',\n",
       "  'answer': 'Globus is a data management and file transfer system that can facilitate bulk transfer of data between two endpoints.'},\n",
       " {'question': 'What are the endpoints that apply for data transfer using Globus?',\n",
       "  'answer': 'The endpoints that apply here are KBase, JGI, and your local computer. The KBase endpoint is called “KBase Bulk Share,” and JGI has their own way to link to Globus.'},\n",
       " {'question': 'How can a JGI user transfer data to KBase?',\n",
       "  'answer': 'JGI users can transfer public genome reads and assemblies, as well as their private data and annotated genomes, from JGI to their KBase account. They can refer to the JGI data transfer page for instructions.'},\n",
       " {'question': 'How can data be uploaded to KBase from a URL?',\n",
       "  'answer': 'There is an option to use an App to upload from a public URL, such as a GenBank ftp URL, or a Dropbox or Google Drive URL that is publicly accessible. This can be done by clicking the link that says “Click here to use an App to upload from a public URL”.'},\n",
       " {'question': 'What types of files can be imported from a URL directly to your Narrative?',\n",
       "  'answer': 'There are several apps that import specific file types, such as single- or paired-end reads or SRA files, from a URL directly to your Narrative, bypassing your Staging Area. These apps are available from the Apps Panel and the App Catalog.'},\n",
       " {'question': 'How can I import a file from my Staging Area into my Narrative in KBase?',\n",
       "  'answer': 'To import a file from your Staging Area, choose a format (data type) from the pulldown menu to the right of the file’s age. Then click the import icon to the right of the format menu.'},\n",
       " {'question': 'What happens when I click the import icon in KBase?',\n",
       "  'answer': 'When you click the import icon, the Data Browser slides shut and an Import App cell for each data type or a Bulk Import App cell will be created in your Narrative, with the appropriate parameters filled in.'},\n",
       " {'question': 'Where can I find detailed instructions for importing different types of data in KBase?',\n",
       "  'answer': 'Detailed instructions for importing different types of data can be found in the Upload/Download Guide.'},\n",
       " {'question': 'How to import a GenBank file from a different source in KBase?',\n",
       "  'answer': 'Use the pulldown menu to select the source, change the output object name if desired, and then click the \"Run\" button to start the import.'},\n",
       " {'question': 'How to know if the import is successful in KBase?',\n",
       "  'answer': 'When the import is done, you should see the message “Finished with success” near the top of the app cell, and some information about the app run.'},\n",
       " {'question': 'How to find the new data object created by the import in KBase?',\n",
       "  'answer': 'Look at your Data Panel, you should see the new data object created by the import.'},\n",
       " {'question': 'How to find which apps accept a particular data type as input in KBase?',\n",
       "  'answer': 'Click the “…” menu in the data object cell that appears when you hover over it, and then use the “Show Apps with this as input” icon to filter the apps in the Apps Panel.'},\n",
       " {'question': 'What is a common cause of import failure in KBase?',\n",
       "  'answer': 'One of the most common causes of failure is attempting to import a file that’s the wrong data type, or not the expected format for that data type.'},\n",
       " {'question': 'What should I do if my import fails in KBase?',\n",
       "  'answer': 'If the importer objected to something in your file, check the Data Upload/Download guide for details about the relevant format. If you can’t figure out why your import isn’t working, please contact us (via the Help Board) for help.'},\n",
       " {'question': 'Can KBase support team access my Staging Area?',\n",
       "  'answer': 'No, no one besides you has access to your Staging Area, so the KBase support team will not be able to see the files you uploaded to your Staging Area. You may need to attach your input file to your Help Board ticket in order for them to diagnose the problem.'},\n",
       " {'question': 'How can multiple files be imported together in KBase?',\n",
       "  'answer': 'Multiple files can be imported together in a single Import App as part of the bulk import feature. When you select the file type from \"Import As...,\" you will be able to check multiple files in the staging area and import them all by clicking \"Import Selected.\"'},\n",
       " {'question': 'What happens when you import selected files in KBase?',\n",
       "  'answer': 'When you import selected, there will be a new cell type which contains a tab for each different file type you selected from the staging area. You will fill out the same parameters for each file type by clicking the file type in the \"Data Types\" column on the left; these are the same parameters as single imports.'},\n",
       " {'question': 'What file types are supported for import in KBase?',\n",
       "  'answer': 'Supported file types include Assembly - FASTA, SRA Reads, and FASTQ Reads Interleaved - FASTQ Interleaved reads.'},\n",
       " {'question': 'What does the list of files in your Staging Area include?',\n",
       "  'answer': 'The list of files in your Staging Area includes their name, size, and age (from when they were uploaded).'},\n",
       " {'question': 'How can you locate specific files in your Staging Area?',\n",
       "  'answer': 'If you have many files in your Staging Area, use the Search box to locate specific files.'},\n",
       " {'question': 'How can you unpack compressed or zipped files in your Staging Area?',\n",
       "  'answer': 'Compressed or zipped files have a little double-arrow icon next to the filename. You can click this icon to unpack these files.'},\n",
       " {'question': 'How can you remove a file from your Staging Area?',\n",
       "  'answer': 'Opening the information about a file in your Staging Area also reveals a trash can icon that allows you to remove the file from your staging area.'},\n",
       " {'question': 'How can you get more information about a file in your Staging Area?',\n",
       "  'answer': 'For more information about a file in your staging area, click the arrow to the left of the filename to open a tab.'},\n",
       " {'question': 'What happens when you delete a file in your Staging Area in KBase?',\n",
       "  'answer': 'The file is permanently deleted, but if you had already imported the file to your Narrative as a data object, that object won’t go away when you delete the file in your Staging Area.'},\n",
       " {'question': 'How can you delete a data object in KBase?',\n",
       "  'answer': 'You can delete a data object in your Data Panel.'},\n",
       " {'question': 'What can you do once you have added data to your Narrative in KBase?',\n",
       "  'answer': 'Once you have added data to your Narrative in KBase, you can analyze it by choosing and running an app.'},\n",
       " {'question': \"Is KBase's import functionality and user interface still under development?\",\n",
       "  'answer': 'Yes, KBase’s import functionality and user interface are under active development.'},\n",
       " {'question': 'What should you do if you encounter a bug or have a suggestion for KBase?',\n",
       "  'answer': 'You should report the bug or make the suggestion to KBase.'},\n",
       " {'question': 'How can I add analysis steps in a Narrative?',\n",
       "  'answer': 'You can add analysis steps in a Narrative by selecting one or more apps from the Apps Panel directly below your data.'},\n",
       " {'question': 'How are the apps in the Apps Panel organized?',\n",
       "  'answer': 'Apps are grouped by category in the Apps Panel.'},\n",
       " {'question': 'How can I browse KBase Apps?',\n",
       "  'answer': 'You can browse KBase Apps by (1) scrolling through the list in the Apps Panel; (2) searching for an app by name; or (3) using the App Catalog to further explore, filter, and designate apps as favorites.'},\n",
       " {'question': 'How can I open the App Catalog?',\n",
       "  'answer': 'You can open the App Catalog by clicking the right arrow in the top right corner of the Apps Panel.'},\n",
       " {'question': 'How can I search for apps in KBase?',\n",
       "  'answer': 'You can search for apps by clicking the search icon (magnifying glass) at the top of the panel and entering a term. The list will then show only the categories that have apps whose names include the text you typed.'},\n",
       " {'question': 'How can I see more details about an app in KBase?',\n",
       "  'answer': 'You can see more details about an app by hovering over the app in the list which will show the more “…” icon. Clicking the \"...\" icon will open an app details page featuring an expanded description of the app, along with technical information about its inputs, outputs, parameters, and more.'},\n",
       " {'question': 'How can I access the App Catalog in KBase?',\n",
       "  'answer': 'The App Catalog can be accessed in three ways: through the External App Catalog (no user account or sign-in required for browsing), from KBase by clicking on the Catalog icon in the Menu.'},\n",
       " {'question': 'How can I access the App Catalog in KBase?',\n",
       "  'answer': 'From the Narrative Interface, by clicking the right arrow at the top of the Apps Panel. The App Catalog will slide out to show available apps.'},\n",
       " {'question': 'Can I add my own app to the App Catalog in KBase?',\n",
       "  'answer': 'Yes, the App Catalog is growing to contain not only the apps generated by KBase staff, but also those created and contributed by outside developers. KBase’s Software Development Toolkit (SDK), provides a mechanism for users to add their own open-source, open-license tools to the system as new KBase apps.'},\n",
       " {'question': 'How can I sort and filter apps in the App Catalog?',\n",
       "  'answer': 'From the App Catalog, you can search for apps by name and use several options to organize and filter how they are displayed. Click the “Organize by” drop-down menu to see these options.'},\n",
       " {'question': 'What does \"My Favorites\" mean in the App Catalog?',\n",
       "  'answer': '\"My Favorites\" lists your favorited apps (those you’ve clicked the star icon to “favorite”) at the top of the catalog.'},\n",
       " {'question': 'What does the \"Favorites Count and Run Count\" filter do in KBase?',\n",
       "  'answer': 'It suggests some of the most popular and frequently used apps in KBase.'},\n",
       " {'question': 'How does the \"Category\" filter work in KBase?',\n",
       "  'answer': 'It displays apps in groups by analysis type including annotation, assembly, communities, comparative genomics, expression, metabolic modeling, and more.'},\n",
       " {'question': 'What is the purpose of the \"Input Types and Output Types\" filter in KBase?',\n",
       "  'answer': 'It groups apps based on the type of data that they take as input or generate as output. Each data type is linked to a specification page that provides some technical details, including version, description, structure, and more. This sort can be useful for determining the data you must add to your Narrative before using a specific app.'},\n",
       " {'question': 'What does the \"Developer\" filter do in KBase?',\n",
       "  'answer': 'It displays apps alphabetically by developers’ KBase usernames, which are linked to their profile pages.'},\n",
       " {'question': 'What is the \"Module\" filter in KBase?',\n",
       "  'answer': 'It sorts apps based on which SDK module they belong to.'},\n",
       " {'question': 'How can you see a short description of each app in the App Catalog?',\n",
       "  'answer': 'You can see a short description of each app by hovering over the “i” in the App Catalog.'},\n",
       " {'question': 'How can you add an app to your favorites in the App Catalog?',\n",
       "  'answer': 'You can add an app to your favorites by clicking the star in the App Catalog.'},\n",
       " {'question': 'What does the number beside the star icon indicate in the App Catalog?',\n",
       "  'answer': 'The number beside the star icon indicates how many other users have selected the app as a favorite.'},\n",
       " {'question': 'How can you access the app module in the App Catalog?',\n",
       "  'answer': 'You can access the app module by clicking its hyperlink in the App Catalog.'},\n",
       " {'question': 'How can you try out apps still in development in the App Catalog?',\n",
       "  'answer': 'To try out apps still in development, click the little “R” in the Apps Panel, changing it to a “B”. This will show apps that are in Beta.'},\n",
       " {'question': 'What will the next section discuss?',\n",
       "  'answer': 'The next section will discuss how to add data to your Narrative for analysis in KBase.'},\n",
       " {'question': 'What was the previous section about?',\n",
       "  'answer': 'The previous section was about adding data to your Narrative in KBase.'},\n",
       " {'question': 'What is the following section about?',\n",
       "  'answer': 'The following section is about analyzing data using KBase Apps.'},\n",
       " {'question': 'How can you add an App to your Narrative in KBase?',\n",
       "  'answer': 'To add a KBase analysis app, find the app of interest and click its name or the icon to the left of the app name. A box (called a cell) containing the chosen app will appear in the main Narrative panel.'},\n",
       " {'question': 'What are the three tabs in the App Cells?',\n",
       "  'answer': 'The three tabs in the App Cells are Configure, Info, and Job Status.'},\n",
       " {'question': 'What is the function of the up and down arrows in the App Cells?',\n",
       "  'answer': 'The up and down arrows let you move the cell up or down in your Narrative.'},\n",
       " {'question': 'What does the App “…” dropdown menu offer?',\n",
       "  'answer': 'The App “…” dropdown menu offers several options for showing the code that will execute the app, getting more info about the app, and deleting the cell.'},\n",
       " {'question': 'What does the \"-/+\" in the App Cell menu represent?',\n",
       "  'answer': 'The \"-/+\" in the App Cell menu is the collapse/expand cell option. When the App Cell is expanded, the farthest right option will be a \"-\" to collapse the cell. When the App Cell is collapsed, the farthest right option will be a \"+\" to expand the cell.'},\n",
       " {'question': 'How can you expand or collapse an App Cell?',\n",
       "  'answer': 'You can expand or collapse an App Cell by using the \"-/+\" option in the App Cell menu or by double clicking on the App Cell.'},\n",
       " {'question': 'What must be done before you can run an app in KBase?',\n",
       "  'answer': 'Every app has some parameters (fields) that must be filled in before you can run the app.'},\n",
       " {'question': 'What needs to be done after adding an app to your Narrative in KBase?',\n",
       "  'answer': 'The required parameters must be filled in before you can run it. For most apps, you will need to select the input data object(s). Other parameters may also need to be set.'},\n",
       " {'question': 'What does a red bar to the right of a field indicate in KBase?',\n",
       "  'answer': 'A red bar to the right of a field indicates that it is a required field and you have not yet entered a valid value in it.'},\n",
       " {'question': 'What are \"smart\" fields in KBase?',\n",
       "  'answer': 'Some app fields are “smart” and know which data in your Narrative is valid for that field. These “smart” fields have a pulldown list of data objects that you can choose from.'},\n",
       " {'question': 'How can you access your data from other Narratives in KBase?',\n",
       "  'answer': 'You can access your data from other Narratives via the My Data and Shared with me tabs in the Data Panel.'},\n",
       " {'question': 'What happens when all required fields are filled in an app?',\n",
       "  'answer': 'The green \"Run\" button is enabled when all required fields are filled in.'},\n",
       " {'question': 'How can you save your work in KBase?',\n",
       "  'answer': 'You can save your work frequently, using the Save button at the top right of the screen.'},\n",
       " {'question': 'How do you start running an app in KBase?',\n",
       "  'answer': 'When you have filled in all the required fields, the button on the left side of the app will be enabled; click it to start running the app.'},\n",
       " {'question': 'How long does an analysis take to run in KBase?',\n",
       "  'answer': 'Depending on the analysis, it might finish within a few seconds or could take hours to run.'},\n",
       " {'question': 'What happens when an app finishes running?',\n",
       "  'answer': 'When an app finishes running, you will see a summary in the Results tab, as well as an output cell below your app cell. Also, if the analysis generates a new data object, that object will be added to your Data Panel.'},\n",
       " {'question': 'What is the function of the \"Browse Contigs\" tab in the output cell?',\n",
       "  'answer': 'The \"Browse Contigs\" tab lists the large assembled pieces of a genome.'},\n",
       " {'question': 'What is the function of the \"Browse Features\" tab in the output cell?',\n",
       "  'answer': 'The \"Browse Features\" tab allows browsing of all the annotated genes.'},\n",
       " {'question': 'What do different apps create when they run?',\n",
       "  'answer': 'Different apps create different types of output cells when they run, depending on which type of data object is output by the app.'},\n",
       " {'question': 'What is a Data Viewer?',\n",
       "  'answer': 'A Data Viewer is a type of output cell created by an app in KBase. It is used to display the data object output by the app.'},\n",
       " {'question': 'What happens when an app runs in KBase?',\n",
       "  'answer': 'In addition to creating a new output cell, the app creates a new data object and adds it to the Narrative.'},\n",
       " {'question': 'How can you see more information and options for an object in the Data Panel?',\n",
       "  'answer': 'You can hover over an object in the Data Panel and click the “...\" that will appear to see more information and options.'},\n",
       " {'question': 'What options are available for downloading the object?',\n",
       "  'answer': 'You can download the object in GenBank or JSON format. In the future, you will be able to choose from a wider variety of common formats.'},\n",
       " {'question': 'What can you do with the newly generated data in KBase?',\n",
       "  'answer': 'You can use your newly generated data in additional analysis steps.'},\n",
       " {'question': 'How can you see which KBase Apps work with a certain data type?',\n",
       "  'answer': 'You can click on the data object in the Data Panel.'},\n",
       " {'question': 'What can you do with an annotated genome in KBase?',\n",
       "  'answer': 'An annotated genome can be used in a number of analyses because it now includes the standard KBase annotations. For example, you can use the Insert Genomes into Species Tree App to construct a phylogenetic tree, or use the Build Metabolic Model App to draft a metabolic model from the annotated genome.'},\n",
       " {'question': 'Can you re-run Apps with the same or different parameters in KBase?',\n",
       "  'answer': 'Yes, you can re-run Apps with the same or different parameters in KBase.'},\n",
       " {'question': 'What does the \"Reset\" button do in the App input cells?',\n",
       "  'answer': 'The \"Reset\" button allows you to redo the analysis, with the option to change any of the parameter settings (including, perhaps, the input data) before rerunning.'},\n",
       " {'question': 'What happens when you rerun an App in KBase?',\n",
       "  'answer': 'Rerunning an App will overwrite the information in the App cells, but will not overwrite any data objects in your Narrative unless you use the same output object name again.'},\n",
       " {'question': 'What should I do if a job fails in KBase?',\n",
       "  'answer': 'If a job fails, you will see an error message in the output cell and/or in the Result tab. You should contact KBase with details about what you were doing when the error occurred and what the error message said. Including the contents of the Job Status tab and the URL of your Narrative will help them debug the problem.'},\n",
       " {'question': 'What is the Job Browser in KBase?',\n",
       "  'answer': 'The Job Browser is one of the buttons on the left sidebar menu in KBase. It allows you to monitor and manage your jobs.'},\n",
       " {'question': 'What is the default timeframe for jobs shown in the Job Browser?',\n",
       "  'answer': 'By default, it shows all jobs submitted within the last 48 hours.'},\n",
       " {'question': 'How can I change the timeframe in the Job Browser?',\n",
       "  'answer': 'You can change the timeframe to an hour, a week, a month, or a custom range of dates.'},\n",
       " {'question': 'How can I refresh the results in the Job Browser?',\n",
       "  'answer': 'The two arrows in a circle to the right and below the timeframe will refresh the results in the table.'},\n",
       " {'question': 'Can I filter the jobs in the Job Browser?',\n",
       "  'answer': 'Yes, the check boxes allow you to filter the jobs. You can check as many filters you want.'},\n",
       " {'question': 'What information does the results display for the jobs in KBase?',\n",
       "  'answer': 'The results display basic information for the jobs such as the Narrative they are located within, App ID, Submission Time, Queue Time, Run Time, and Status.'},\n",
       " {'question': 'What happens when you click on the Narrative name in the results?',\n",
       "  'answer': 'Clicking on the Narrative name will open a new tab with the Narrative.'},\n",
       " {'question': 'What happens when you click on the App ID in the results?',\n",
       "  'answer': 'Clicking on the App ID will open a new tab with the catalog page for the app.'},\n",
       " {'question': 'How can you view the contents of the log in the Job Status Cell?',\n",
       "  'answer': 'You can view the contents of the log by clicking the log icon to expand the table.'},\n",
       " {'question': 'What happens when you cancel a job in KBase?',\n",
       "  'answer': 'Canceling the job will prompt you to make sure you didn’t click it accidentally. Jobs that have been canceled may continue to show up as “Queued” or “Running” until they clear the system.'},\n",
       " {'question': 'How can you download the Job Logs in KBase?',\n",
       "  'answer': 'Job Logs can be downloaded in CSV, TSV, JSON, and TEXT formats using the download (downward-facing arrow into tray) icon.'},\n",
       " {'question': 'What is the next step after analyzing data using KBase Apps?',\n",
       "  'answer': 'The next step is to revise your narrative.'},\n",
       " {'question': 'When was the last modification made to this section?',\n",
       "  'answer': 'The last modification was made 2 years ago.'},\n",
       " {'question': 'What is the previous step before revising your narrative in KBase?',\n",
       "  'answer': 'The previous step is to analyze data using KBase Apps.'},\n",
       " {'question': 'Where can I find a list of my Narratives in KBase?',\n",
       "  'answer': 'You can find a list of your Narratives in the Narratives tab.'},\n",
       " {'question': 'How are Narratives listed in the Narratives tab?',\n",
       "  'answer': 'Narratives are listed in reverse chronological order, with the most recently updated Narrative first. After your personal Narratives, more Narratives owned by others and shared with you are listed.'},\n",
       " {'question': 'Can I revert changes to my Narrative in KBase?',\n",
       "  'answer': 'Yes, KBase stores a history of everything you’ve done to your Narrative, so you are able to revert changes and go back to an earlier version.'},\n",
       " {'question': 'How can I revert to an earlier version of my Narrative in KBase?',\n",
       "  'answer': 'When you hover your cursor over one of your Narratives in the Narratives panel, three icons will appear. The first of these icons (counterclockwise encircling arrow) lets you View Narrative history to revert changes or access prior saved versions of the Narrative. Here you can choose an earlier version of a Narrative to revert to from the saved versions.'},\n",
       " {'question': 'How can I add text to my Narrative in KBase?',\n",
       "  'answer': 'The document does not provide specific instructions on how to add text to a Narrative in KBase.'},\n",
       " {'question': 'What is Markdown?',\n",
       "  'answer': 'Markdown is a markup language (one example is HTML). It’s used to add formatted text to your Narrative in “Markdown Cells”.'},\n",
       " {'question': 'How to add Markdown cells to a Narrative?',\n",
       "  'answer': 'You can add Markdown cells to a Narrative using the translucent blue buttons at the bottom right of the main Narrative panel. The button on the right is the “Add Markdown Cell” button and will add a Markdown cell under the currently selected Narrative cell.'},\n",
       " {'question': 'How to edit a Markdown cell?',\n",
       "  'answer': 'To edit a Markdown cell, double-click it to enter edit mode. When you’re done, press Shift-Return to exit the cell.'},\n",
       " {'question': 'What is the placeholder text in a new Markdown cell?',\n",
       "  'answer': 'New Markdown cells contain the placeholder text “[Empty Markdown / LaTeX cell]”.'},\n",
       " {'question': 'How can you add a Code Cell in KBase?',\n",
       "  'answer': 'You can add a Code Cell in KBase by clicking the \">_\" button.'},\n",
       " {'question': 'How can you execute the code in a code cell?',\n",
       "  'answer': 'To execute the code in a code cell, press Shift-Return within the cell.'},\n",
       " {'question': 'How can you rename a Narrative in KBase?',\n",
       "  'answer': 'To rename a Narrative in KBase, click on its name at the top of the Narrative window. A window will pop up to let you change the name.'},\n",
       " {'question': 'How can you delete a Narrative in KBase?',\n",
       "  'answer': 'To delete a Narrative in KBase, click the trashcan icon. However, delete with caution, because currently there is no way to get your deleted Narrative back.'},\n",
       " {'question': 'How can you save your work in KBase?',\n",
       "  'answer': 'Although your Narrative is periodically autosaved, it is always a good idea to save your work frequently, using the Save button at the top right of the screen.'},\n",
       " {'question': 'How can I share my Narrative in KBase?',\n",
       "  'answer': 'You can share your Narrative in KBase by following the instructions in the \"Share Narratives\" section.'},\n",
       " {'question': 'Is it possible to share my Narrative in KBase with others?',\n",
       "  'answer': 'Yes, it is possible to share your Narrative in KBase with others.'},\n",
       " {'question': 'Where can I find information on how to share my Narrative in KBase?',\n",
       "  'answer': 'You can find information on how to share your Narrative in KBase in the \"Share Narratives\" section.'},\n",
       " {'question': 'What are Markdown cells used for in KBase?',\n",
       "  'answer': 'Markdown cells are used to edit the appearance for publication, teaching, and to add navigation within the Narrative. They are also used to explain your process, interpret results, and communicate to others for sharing and publication and more.'},\n",
       " {'question': 'What languages do Markdown cells use?',\n",
       "  'answer': 'Markdown cells use markup languages of Markdown and HTML.'},\n",
       " {'question': 'Where can I find resources for Markdown syntax?',\n",
       "  'answer': 'Resources for Markdown syntax can be found at the following links: Markdown Cheat Sheet and Markdown Guide.'},\n",
       " {'question': 'Where can I find resources for HTML syntax?',\n",
       "  'answer': 'Resources for HTML syntax can be found at the following links: HTML Cheat Sheet, HTML Guide, and HTML Color Codes.'},\n",
       " {'question': 'What is a KBase Narrative?',\n",
       "  'answer': 'A KBase Narrative is a workspace that can be worked on by multiple people on a project, shared with specific people or everyone, cited as a URL, and can be repeated or even changed by another scientist if they publish it as a Narrative.'},\n",
       " {'question': 'How can you share a KBase Narrative?',\n",
       "  'answer': 'You can share a Narrative that you own with any set of people you choose, or with all KBase Users. To share a Narrative, click the “share” button near the top right and select the \"Manage Sharing\" option in the menu.'},\n",
       " {'question': 'Can a KBase Narrative be cited?',\n",
       "  'answer': 'Yes, a finished Narrative is similar to a research paper, and can be cited as a URL.'},\n",
       " {'question': 'How can I share my Narrative with everyone on KBase?',\n",
       "  'answer': 'You can click the “make public” link near the top of the sharing panel.'},\n",
       " {'question': 'How can I share my Narrative with specific people on KBase?',\n",
       "  'answer': 'You can type their name or username into the “share with” box. Once you have typed something that matches a name in the KBase User Database, the person’s full name will appear. Click the down arrow to the right of the “share with” box to assign permissions to that user.'},\n",
       " {'question': 'What permissions can I assign to a user when sharing my Narrative on KBase?',\n",
       "  'answer': 'You can allow them only to view your Narrative, to edit it, or to edit and share it.'},\n",
       " {'question': 'How can I share my Narrative with KBase staff for troubleshooting?',\n",
       "  'answer': \"You can share your Narrative with 'kbasehelp' so that KBase staff can help you troubleshoot it. You can start typing any part of the username or real name and a list of users that match will appear.\"},\n",
       " {'question': 'What is the default permission level for sharing in KBase?',\n",
       "  'answer': 'The default permission level for sharing is “View only”.'},\n",
       " {'question': 'How can I share my Narrative with a collaborator and give them more privileges?',\n",
       "  'answer': 'Click the down-arrow to see the permission levels and possibly choose a different one.'},\n",
       " {'question': 'How can I share my Narrative with multiple users?',\n",
       "  'answer': 'You can add multiple users to share with by clicking the “Apply” button to add the chosen user(s) to the sharing list for your Narrative.'},\n",
       " {'question': 'How can I share my Narrative with a specific Organization?',\n",
       "  'answer': 'To share your Narrative with a specific Organization, select the Orgs tab. Here you can add a single or multiple Organizations to share with a group of users.'},\n",
       " {'question': 'How can I see the users I have shared my Narrative with?',\n",
       "  'answer': 'The usernames and permission levels for the users you shared with will appear at the bottom of the Share panel.'},\n",
       " {'question': 'How can you close the Share panel in KBase?',\n",
       "  'answer': 'You can close the Share panel by clicking the \"x\" icon.'},\n",
       " {'question': 'How can you change the access permissions in KBase?',\n",
       "  'answer': 'You can change the access permissions at any time by clicking the Share button again. The up/down arrows next to each user’s permissions let you change the permissions.'},\n",
       " {'question': 'What happens if you and a collaborator are working on a shared Narrative at the same time?',\n",
       "  'answer': 'If you and a collaborator are working on a shared Narrative at the same time, a red box will appear at the top of the screen that says “Narrative updated in another session.” You can update to the latest version by refreshing the page, but any changes that you made to the Narrative will be lost, and if you try to save your changes, you might overwrite your collaborator’s changes.'},\n",
       " {'question': 'How will a user know if a Narrative has been shared with them?',\n",
       "  'answer': 'Unless you alert another user that you have shared a Narrative with them, they will not notice a newly shared Narrative until they see your shared Narrative in their “Narratives that have been shared with you” tab or click on your name in her collaborator list.'},\n",
       " {'question': 'How can I share my Narrative in a publication ready, uneditable format?',\n",
       "  'answer': 'You can share your Narrative in a publication ready, uneditable format by creating a static Narrative.'},\n",
       " {'question': 'What are the steps to create a static Narrative?',\n",
       "  'answer': \"There are three steps to create a static Narrative:\\n    1. Click on the share icon in the menu and click on Manage Sharing in the dropdown menu. In the pop-up window click 'make public?' to change the privacy settings. Click the 'x' icon to close the pop-up window.\\n    2. Return and click on the 'share' icon, go to 'Manage Static Narratives in the   dropdown menu. The 'Static Narratives' pop-up window will open and show the status if any versions of this Narrative have been made static. To create a static Narrative of the current version, click on the blue 'Create static narrative' button.\\n    3. Access the static Narrative and share the link by clicking on the 'Existing static Narrative' link within the pop-up window.\"},\n",
       " {'question': 'How can I share the link of the static Narrative?',\n",
       "  'answer': \"You can share the link of the static Narrative by clicking on the 'Existing static Narrative' link within the pop-up window.\"},\n",
       " {'question': 'What is a KBase user profile?',\n",
       "  'answer': 'A KBase user profile is a profile that every KBase user has. Filling in your KBase user profile helps you become part of the social web that KBase is building, making it easier to find and collaborate with other scientists and share data and Narratives.'},\n",
       " {'question': 'Is it mandatory to complete your user profile in order to use KBase?',\n",
       "  'answer': 'No, it is not required to complete your user profile in order to use KBase, but it is encouraged.'},\n",
       " {'question': 'How can you edit your user profile on KBase?',\n",
       "  'answer': 'To see and edit your user profile, go to the Menu sidebar on the left and click the “Account” button on the left side. Your profile page will open, and you will be able to enter or change information including your organization, department, job title, research interests, avatar, and more.'},\n",
       " {'question': 'What fields are required to be filled in order to save your profile on KBase?',\n",
       "  'answer': 'Required fields are marked with a red \"*.\" You must fill in all required fields in order to save your profile.'},\n",
       " {'question': 'What can you see in the small preview below the \"Save\" button on your KBase profile page?',\n",
       "  'answer': 'The small preview shows what your profile page looks like to others.'},\n",
       " {'question': 'What can you do with the “Open Your Profile Page” link?',\n",
       "  'answer': 'You can see a bigger version of your profile, along with a list of your Narratives and collaborators and a search box for searching for other users by name or username.'},\n",
       " {'question': 'What can be edited in the Account tab of your KBase account?',\n",
       "  'answer': 'The only field that can be edited there is your preferred email address.'},\n",
       " {'question': 'Who can see your email address on KBase?',\n",
       "  'answer': 'Other users cannot see your email address; it will be used only by KBase staff to contact you occasionally with important information.'},\n",
       " {'question': 'What does the Linked Sign-In Accounts tab allow you to do?',\n",
       "  'answer': 'The Linked Sign-In Accounts tab lets you see and manage your linked sign-in accounts.'},\n",
       " {'question': 'What can you edit in the User Profile editor on KBase?',\n",
       "  'answer': 'In the User Profile editor, you can edit your position, organization, and location. You can also input your Research Interests, Research or Personal Statement.'},\n",
       " {'question': 'What information does the User Profile editor show?',\n",
       "  'answer': 'The User Profile editor shows which KBase Organizations you belong to and any Affiliations.'},\n",
       " {'question': 'Can you manage external accounts linked to your KBase account?',\n",
       "  'answer': 'Yes, you can manage which external (Globus or Google or ORCiD) accounts are linked to your KBase account.'},\n",
       " {'question': 'How to add KBase static Narrative DOIs to your ORCID account?',\n",
       "  'answer': \"1. Sign in to ORCID\\n    2. Scroll down to the Works section of your ORCID record and hover over the '+Add works' in the menu and select 'Search & link'\\n    3. Under 'Link Works', select DOE / OSTI from the list of member organizations.\\n    4. While using the DOE / OSTI Search & link wizard for the first time, you will need to Authorize access to approve DOE / OSTI to make changes to your ORCID Works.\"},\n",
       " {'question': 'What is ORCID?',\n",
       "  'answer': 'ORCID is a non-profit organization that provides a persistent digital identifier (an ORCID iD) that you own and control, and that distinguishes you from every other researcher.'},\n",
       " {'question': 'What is DOE / OSTI?',\n",
       "  'answer': \"DOE / OSTI is one of the member organizations in ORCID, which you can select under 'Link Works' to add KBase static Narrative DOIs to your ORCID account.\"},\n",
       " {'question': 'What happens after DOE / OSTI is authorized in ORCID?',\n",
       "  'answer': 'Once DOE / OSTI is authorized, the OSTI.GOV page will appear with the green ORCID iD icon next to the user’s ORCID iD name. The ORCID Account Integration page will display.'},\n",
       " {'question': 'How can I add works to my ORCID record using OSTI.GOV’s search?',\n",
       "  'answer': \"You can add works to your ORCID record using OSTI.GOV’s search by following these steps:\\n    1. Browse DOE / OSTI search results to identify records authored, coauthored, or contributed by you.\\n    2. Once you find records in the search results to add to your ORCID record, select 'Add to ORCID Works.'\\n    3. A pop-up text box will appear and confirm that you want to add this record to ORCID Works. Select 'Add to my ORCID Works'.\\n    4. In the search results, works added by you to your ORCID record will now display 'In your ORCID Works' at the bottom right of the popup.\\n    5. To view your list of works added to your ORCID record, navigate back to the 'ORCID Account Integration' page. Select the user name next to the green ORCID iD icon at the top of the page.\\n    6. Click on the ORCID Works in OSTI.GOV to see the works integrated with your account.\"},\n",
       " {'question': \"What does 'In your ORCID Works' mean?\",\n",
       "  'answer': \"'In your ORCID Works' means that the work has been added to your ORCID record.\"},\n",
       " {'question': \"How can I view the works I've added to my ORCID record?\",\n",
       "  'answer': \"To view your list of works added to your ORCID record, navigate back to the 'ORCID Account Integration' page. Select the user name next to the green ORCID iD icon at the top of the page. Click on the ORCID Works in OSTI.GOV to see the works integrated with your account.\"},\n",
       " {'question': 'Where can I find the Narratives that I have created in KBase?',\n",
       "  'answer': 'The Narratives that you have created can be found in the \"My Narratives\" section under the Narratives tab.'},\n",
       " {'question': 'Where can I find the Narratives that have been shared with me in KBase?',\n",
       "  'answer': 'The Narratives that have been shared with you can be found in the \"Shared With Me\" section under the Narratives tab.'},\n",
       " {'question': 'Can I edit or run a Narrative that someone else owns and shared with me without write permission?',\n",
       "  'answer': \"No, you won't be able to edit or run a Narrative that someone else owns and shared with you without write permission. It will open in View-only mode.\"},\n",
       " {'question': 'How can I access public Narratives in KBase?',\n",
       "  'answer': 'The public Narratives can be accessed from the Narratives dash.'},\n",
       " {'question': 'What happens when I click on one of the Narratives in my Narratives panel?',\n",
       "  'answer': 'If you click on one of the Narratives in your Narratives panel, it will open in a new web browser window. If it’s a Narrative that you own, you will be able to run it or edit it. However, if it’s one that someone else owns and shared with you without write permission, you won’t be able to edit or run it–it will open in View-only mode.'},\n",
       " {'question': 'What is the key difference in View-only mode in KBase?',\n",
       "  'answer': 'The key difference in View-only mode is that there is no Analyze panel on the left side–you won’t see any data objects or apps. This is because you can’t modify or run a Narrative that was shared with you with only View permission.'},\n",
       " {'question': \"Why can't you run a Narrative owned by someone else in KBase?\",\n",
       "  'answer': \"You can't run a Narrative owned by someone else because running an app usually creates new data objects in your Narrative, which requires write permission. If you want to run a Narrative owned by someone else who has not given you write permission, you will need to copy it to your account.\"},\n",
       " {'question': 'How can you switch a Narrative that you own to view-only mode in KBase?',\n",
       "  'answer': 'You can switch a Narrative that you own to view-only mode by clicking the pencil icon under the title of the Narrative.'},\n",
       " {'question': 'How can I print a Narrative in KBase?',\n",
       "  'answer': 'To print a Narrative in KBase, switch it to view-only mode first, so it doesn’t waste the left side of the page with empty space.'},\n",
       " {'question': 'How can I copy a Narrative in KBase?',\n",
       "  'answer': 'To copy a Narrative that has been shared with you, you can use the blue “Copy This Narrative” button at the top of the Narratives panel (or, in View-only mode, at the top right of the screen). This will copy the Narrative that you are currently viewing. You can choose a name for the copy or leave the default name.'},\n",
       " {'question': 'What happens when I copy a Narrative in KBase?',\n",
       "  'answer': 'When you click Copy, the Narrative and its associated data objects are copied to your account, and you should see the new copy appear at the top of “My Narratives”.'},\n",
       " {'question': 'What should I be aware of when copying a Narrative in KBase?',\n",
       "  'answer': 'Please be aware that if you copy a copy of a Narrative, some things may not work right in the copy of a copy. KBase is working on resolving this problem.'},\n",
       " {'question': 'How can I copy a Narrative in KBase?',\n",
       "  'answer': 'You can copy any Narrative by hovering your cursor over its name in your Narratives panel to reveal the icons and clicking the Copy icon (the one that looks like two sheets of paper).'},\n",
       " {'question': 'What can I do with a copied Narrative in KBase?',\n",
       "  'answer': 'After you’ve copied a Narrative, you can open it, run it, edit it, or even share it with others (if the owner of the original Narrative gave you sharing privileges). You can run all the steps to reproduce the computational experiment, or try changing the parameters or input data to alter or perhaps even improve the results.'},\n",
       " {'question': 'What is the benefit of copying Narratives in KBase?',\n",
       "  'answer': 'Copying Narratives is the key to leveraging KBase’s commitment to reproducibility and reusability. By making it easy to build on previous work and go through rapid cycles of analysis, KBase aims to accelerate the pace of systems biology research.'},\n",
       " {'question': 'What is a KBase Org?',\n",
       "  'answer': 'In KBase, organizations are a way for teams of scientists to share their data and associated analyses that are in the Narratives they create with each other as a group. Organization members can see information about the team and a list of the Narratives associated with the organization to which can request access. View-only access is granted upon request while all other access to the Narrative is granted by the Narrative owner. KBase users can be members of more than one organization and Narratives with the associated data might also be added to more than one organization.'},\n",
       " {'question': 'Can a KBase user be a member of more than one organization?',\n",
       "  'answer': 'Yes, KBase users can be members of more than one organization.'},\n",
       " {'question': 'Can Narratives with the associated data be added to more than one organization?',\n",
       "  'answer': 'Yes, Narratives with the associated data might also be added to more than one organization.'},\n",
       " {'question': 'Who grants access to the Narrative in a KBase Org?',\n",
       "  'answer': 'View-only access is granted upon request while all other access to the Narrative is granted by the Narrative owner.'},\n",
       " {'question': 'How can I create an organization in KBase?',\n",
       "  'answer': 'On the Orgs page, click \"Create Organization\" button on the right side. This will open the Create Your Organization page. Input the information and Save.'},\n",
       " {'question': 'What information is needed to create an organization in KBase?',\n",
       "  'answer': 'To create an Organization, you will need: Name, ID, Logo URL (optional), Home Page URL (optional), Hidden?, Research Interests, Description.'},\n",
       " {'question': \"What does the 'Hidden?' option mean when creating an organization in KBase?\",\n",
       "  'answer': \"'Hidden?' is an option to check for the Organization to be Hidden - will be visible only for members of this organization or leave unchecked to remain Visible - will be visible to all KBase users.\"},\n",
       " {'question': 'Can I edit an organization after it has been created in KBase?',\n",
       "  'answer': 'Yes, there are options to edit and share an organization after it has been created.'},\n",
       " {'question': 'How can I join an organization in KBase?',\n",
       "  'answer': 'The document does not provide specific steps on how to join an organization in KBase.'},\n",
       " {'question': 'How can I accept an invitation to join an organization in KBase?',\n",
       "  'answer': 'If an invitation appears in your feeds, click on the ID of the organization. A window will pop up with options to Accept or Reject the invitation. Click on your preference.'},\n",
       " {'question': 'How can I join an organization in KBase without an invitation?',\n",
       "  'answer': 'Search for the organization and click on its name. Click on the blue button (upper right) to ‘Join Organization’. A request to join will be sent to the owner and group administrators. An acceptance or rejection will be added to your Feeds.'},\n",
       " {'question': 'How can I view Narratives in an organization in KBase?',\n",
       "  'answer': 'If you haven\\'t opened the Narrative before, there will be a \"Click for Access/View Access\" button. Click the \"Click for View Access\" button. You will be able to open and view the Narrative when the \"No Access\" line disappears. To open a Narrative, click on the Narrative name.'},\n",
       " {'question': 'How can I join an organization in KBase if I know the organization URL?',\n",
       "  'answer': 'Go to the webpage of the organization (URL starts with https://Narrative.kbase.us/#org/) and request membership.'},\n",
       " {'question': 'Can I cancel my request to join an organization in KBase?',\n",
       "  'answer': 'Yes, you can cancel this request if needed.'},\n",
       " {'question': 'How can I access a Narrative in KBase?',\n",
       "  'answer': 'You can access a Narrative by clicking the \"Click for View Access\" button.'},\n",
       " {'question': 'What does the globe icon mean in the Narrative cell?',\n",
       "  'answer': 'The globe icon means that the Narrative is shared publicly with all KBase users.'},\n",
       " {'question': 'How can I add Narratives to an Organization in KBase?',\n",
       "  'answer': 'You can add Narratives by either clicking the \"+ Add a Narrative\" button or by sharing Narratives with an Organization.'},\n",
       " {'question': 'How can I invite members to an Organization in KBase?',\n",
       "  'answer': 'You can invite members by clicking the invite button on the right above the list of members. In the Search box ‘Select User to Invite’, type in the user’s name or KBase ID. A list of users will appear below the search box. Click on the user to invite and the name will appear on the right with more information that can be used to confirm it is the correct user. If correct, click on the blue ‘Send Invitation’ button.'},\n",
       " {'question': 'What does the crown icon mean in the Narrative cell?',\n",
       "  'answer': 'The crown icon means that you are the owner of the Narrative.'},\n",
       " {'question': 'How can a user request to join an organization in KBase?',\n",
       "  'answer': 'Users can request to join an organization in KBase by sending a request which will appear in the \"Inbox\" of the \\'Requests\\' tab on the organization page.'},\n",
       " {'question': 'How can an owner or administrator accept or reject requests to join an organization?',\n",
       "  'answer': \"Requests to join an organization can be accepted or rejected by the owner or administrator by clicking the respective buttons in the 'Requests' tab.\"},\n",
       " {'question': 'How can an owner or administrator promote a member to administrator?',\n",
       "  'answer': 'Owners and administrators can promote members to administrator by clicking on the ‘...’ to the right of the member’s name in the membership list and then clicking on ‘Promote to Admin’.'},\n",
       " {'question': 'How can a member be removed from an organization?',\n",
       "  'answer': 'A member can be removed from an organization by clicking on the ‘...’ to the right of the member’s name in the membership list and then clicking on ‘Remove Member’.'},\n",
       " {'question': 'Can an organization be deleted in KBase?',\n",
       "  'answer': 'No, there is no mechanism to delete an organization in KBase. However, you can un-associate all Narratives, remove all members and set the group to “Private.”'},\n",
       " {'question': 'What was the last modification date of the document?',\n",
       "  'answer': '1 year ago'},\n",
       " {'question': 'What is the next topic after \"Access and Copy Narratives\"?',\n",
       "  'answer': 'FAQs'},\n",
       " {'question': 'What is the previous topic before \"FAQs\"?',\n",
       "  'answer': 'Access and Copy Narratives'},\n",
       " {'question': 'Is KBase free to use?',\n",
       "  'answer': 'Yes, KBase is completely free to use and is supported by the Department of Energy.'},\n",
       " {'question': \"What should I do if I'm getting an error in KBase?\",\n",
       "  'answer': 'You can check the job log and refer to the troubleshooting guide. If you do not see a reason for the error there, please make a ticket on the Help Board.'},\n",
       " {'question': 'Can I obtain publication-quality figures by using KBase tools?',\n",
       "  'answer': 'Some of the Apps can produce publication-quality images that can be views by opening in a separate window. More will be added in the future.'},\n",
       " {'question': 'How can I handle large datasets in KBase?',\n",
       "  'answer': 'For large files or collections of files, it is recommended to use Globus. There is also a guide for importing using Globus.'},\n",
       " {'question': 'Can I import more than one file into a Narrative at the same time in KBase?',\n",
       "  'answer': 'Yes, bulk import allows you to check multiple files in the staging area and import them all by clicking \"Import Selected.\" More instructions on this feature are available in the bulk import guide.'},\n",
       " {'question': 'How many users can view a Narrative in KBase?',\n",
       "  'answer': 'There is no limit to the number of users that can be invited to view or edit Narratives.'},\n",
       " {'question': 'How can I organize many inputs in KBase?',\n",
       "  'answer': 'The best way to organize many inputs is to create sets. There are Apps for building these sets, such as Build AssemblySet that allow you to group those outputs.'},\n",
       " {'question': 'What languages are supported in code cells in KBase?',\n",
       "  'answer': 'Currently, KBase only supports code cells in Python.'},\n",
       " {'question': 'How does KBase compare with the standalone tools available with command line interface?',\n",
       "  'answer': 'The document does not provide information on how KBase compares with standalone tools available with command line interface.'},\n",
       " {'question': 'Does KBase always use the most recent version?',\n",
       "  'answer': 'The document does not provide information on whether KBase always uses the most recent version.'},\n",
       " {'question': 'Are the results consistent if the same data, parameters, and tool version are used in KBase?',\n",
       "  'answer': 'Yes, provided the same data, parameters, and tool version are consistent, the results are the same.'},\n",
       " {'question': 'How can I access advanced parameters in KBase?',\n",
       "  'answer': 'You can view some advanced parameters depending on the app by clicking the Advanced Parameters button. For some apps there are few or no advanced parameters.'},\n",
       " {'question': 'How can I get a tool I like added to KBase?',\n",
       "  'answer': \"There are two ways to get a tool added to KBase. One is to submit a Help Desk New Feature Request. The other is to add as a 3rd-party developer if you are a developer of an app. You can find more information on the developer's page and you can read the documentation for our Software Development Kit.\"},\n",
       " {'question': 'How can I switch to beta Apps in KBase?',\n",
       "  'answer': 'In the Apps panel, click on the \"R\" button to switch from release to beta. When a \"B\" appears in place of the \"R\", you are viewing beta Apps.'},\n",
       " {'question': 'Does KBase provide pipelines?',\n",
       "  'answer': 'KBase does not provide pipelines per se; they try to be as flexible as possible. They do have example Narratives that replicate the paths many users follow, but they encourage everyone to plot their own course.'},\n",
       " {'question': 'Will my data be made public in KBase?',\n",
       "  'answer': 'All user data in KBase is private by default until you share it. You can keep all your data private if you wish. KBase encourages users to make data public in the interest of FAIR - Findable, Accessible, Interoperable, and Reusable data practices.'},\n",
       " {'question': 'What organisms are available in KBase?',\n",
       "  'answer': 'The text does not provide information on what organisms are available in KBase.'},\n",
       " {'question': 'What data can be uploaded to KBase?',\n",
       "  'answer': 'Anything can be uploaded to KBase, the only exception being that human data is not permitted. There are connections to some bacterial, plant, and fungal species in the Public Data tab, but any user data or data found from public sources other than those integrated into KBase can be uploaded.'},\n",
       " {'question': 'Can I access KBase methods from the command line?',\n",
       "  'answer': 'Command line is not supported for the narratives. However, you may find more flexibility with the KBase Software Development Kit (SDK).'},\n",
       " {'question': 'How can I arrange an in-person KBase workshop at my institution?',\n",
       "  'answer': 'If you would like an in-person workshop, contact KBase by email at [email\\xa0protected]'},\n",
       " {'question': 'Can I see the background code for KBase?',\n",
       "  'answer': 'Yes, KBase is open source. You can find the code on GitHub at https://github.com/kbase.'},\n",
       " {'question': 'How can I find the Github repo for each App in KBase?',\n",
       "  'answer': 'You can find the Github repo for each App by following the links in the App Catalog.'},\n",
       " {'question': 'How can I access the version and parameters used during an App run in KBase?',\n",
       "  'answer': 'You can access the version and parameters that were used during an App run by clicking on the ellipsis button and clicking \"Show code.\"'},\n",
       " {'question': 'What is the difference between Public and Private organizations in KBase?',\n",
       "  'answer': 'The existence and contents of Private organizations are only visible to members of the organization. The organization will not be findable by non-members. Any KBase user can see a list of all Public organizations plus a description of the organization. Non-members are given the option to request membership. Members of the organization can access more information than non-members.'},\n",
       " {'question': 'How can I open a Narrative in an Org in KBase?',\n",
       "  'answer': 'Click on ‘Click for Access’ and you are automatically granted View access to the Narrative. You can then click on the linked Narrative to open.'},\n",
       " {'question': 'What should I do if I lose access to my account in KBase?',\n",
       "  'answer': 'The text does not provide information on what to do if you lose access to your account in KBase.'},\n",
       " {'question': 'How can I log into KBase?',\n",
       "  'answer': 'You can log into KBase through Google, Globus, and ORCiD.'},\n",
       " {'question': 'What should I do if I lose access to my KBase account?',\n",
       "  'answer': 'If you lose access completely, please submit a ticket through the Help Board.'},\n",
       " {'question': 'Why is it recommended to have at least one personal account linked to your KBase account?',\n",
       "  'answer': 'It is recommended to always have at least one personal account linked to your KBase account, so that in the event you lose access to an identity supplier i.e., Google, Globus, or ORCiD because of an email provider change, then you will still have another method to login.'},\n",
       " {'question': 'What are the guides for working with data in KBase?',\n",
       "  'answer': 'The guides for working with data in KBase are:\\n    1. Upload and Download Guide\\n    2. How to search, add and upload data in KBase video tutorial\\n    3. How to filter, manage, and view data in KBase video tutorial\\n    4. Linking data and metadata with Samples\\n    5. Using public data within KBase\\n    6. How to transfer data using Globus\\n    7. How to transfer data from JGI'},\n",
       " {'question': 'How can I transfer data from JGI to KBase?',\n",
       "  'answer': 'KBase provides a guide on how to transfer data from JGI.'},\n",
       " {'question': 'How can I use public data within KBase?',\n",
       "  'answer': 'KBase provides a guide on how to use public data within KBase.'},\n",
       " {'question': 'What does the Data Upload and Download Guide provide?',\n",
       "  'answer': 'This guide provides instructions for uploading different types of data to your KBase account and downloading output data from your analysis results.'},\n",
       " {'question': 'What are the contents of the Data Upload and Download Guide?',\n",
       "  'answer': 'The contents of the guide include Data Types, Importing Data, Assemblies, Genomes, FASTQ/SRA Reads, FBA Models, Media, Expression Matrix, Phenotype Set, Amplicon Matrix, Chemical Abundance Matrix, SampleSet, Compressed Files, and Downloading Data.'},\n",
       " {'question': 'How can I upload data to my KBase account?',\n",
       "  'answer': 'You can follow the instructions provided in the Data Upload and Download Guide on the KBase website.'},\n",
       " {'question': 'What are the data types within KBase?',\n",
       "  'answer': 'Data within KBase covers a wide range of data types relevant to systems biology research, including genomes and their annotations, metagenomes, expression, and protein-protein interactions, inferred models of organismal and community metabolism and gene regulation, even geographical information about populations.'},\n",
       " {'question': 'How does KBase handle data?',\n",
       "  'answer': 'KBase handles data as objects versus files for interoperability with KBase Apps.'},\n",
       " {'question': 'What are the file types for the Assembly data type in KBase?',\n",
       "  'answer': 'The file types for the Assembly data type in KBase are FASTA files with extensions .fasta, .fna, .fa, .fas.'},\n",
       " {'question': 'What are the file types for the Genome data type in KBase?',\n",
       "  'answer': 'The file types for the Genome data type in KBase are GenBank or GFF3 with a FASTA file; extensions .genbank, .gb, .gbk, or .gbff, .gff, and .fasta, .fna, .fa, .fas.'},\n",
       " {'question': 'What are the file types for the Metagenome data type in KBase?',\n",
       "  'answer': 'The file types for the Metagenome data type in KBase are GFF Metagenome with a FASTA file; extensions .gbff, .gff, and .fasta, .fna, .fa, .fas.'},\n",
       " {'question': 'What file formats are supported for FBA Model in KBase?',\n",
       "  'answer': 'SBML, Excel, or TSV; extensions .sbml, .xml, .tsv, .xls, .xlsx'},\n",
       " {'question': 'What file formats are supported for Media in KBase?',\n",
       "  'answer': 'Excel or TSV; .xls, .xlsx, .tsv'},\n",
       " {'question': 'What file formats are supported for Amplicon Matrix in KBase?',\n",
       "  'answer': 'Excel and FASTA; .xls, .xlsx, and .fasta, .fna, .fa, .fas'},\n",
       " {'question': 'What file formats are supported for Chemical Abundance Matrix in KBase?',\n",
       "  'answer': 'Excel, CSV, or TSV; .xls, .xlsx, .csv, .tsv'},\n",
       " {'question': 'What is the Data Type Catalog in KBase?',\n",
       "  'answer': 'The Data Type Catalog lists and describes the key data types representing different classifications of biological data within KBase.'},\n",
       " {'question': 'How can I import data into KBase?',\n",
       "  'answer': 'You can import data into KBase by using the Data Browser in the KBase Narrative Interface. You can drag & drop data files from your computer into the Staging Area to upload data into your Narrative.'},\n",
       " {'question': 'What is the first step in uploading data to KBase?',\n",
       "  'answer': 'The first step in uploading your data is to locate the Data Panel along the left side of the Narrative Interface window and click the “Add Data” button, the circular “+” icon, or the arrow at the upper right of the panel to access the slide-out Data Browser.'},\n",
       " {'question': 'How can I close the Import panel in KBase?',\n",
       "  'answer': 'To close the Import panel and return to the Narrative Interface, simply click the “Close” button on the bottom right of the import panel.'},\n",
       " {'question': 'How can I close the Data Browser in KBase?',\n",
       "  'answer': 'You can close the Data Browser by clicking the arrow icon on the top right of your Data Panel.'},\n",
       " {'question': 'How does the import type selection dropdown work in KBase?',\n",
       "  'answer': 'The import type selection dropdown can detect the data type of the file based on its extension where possible. Some types will not automatically select, such as .fasta or .fastq. In these cases more than one importer can use the FASTA file extension. However, suggested importers will be listed at the top of the drop down under “Suggested Types.”'},\n",
       " {'question': 'How can I import multiple files in KBase?',\n",
       "  'answer': 'You can learn how to import multiple files under the Bulk Import Guide.'},\n",
       " {'question': 'How can I import files with the extension FASTQ in KBase?',\n",
       "  'answer': 'Files with the extension FASTQ can import as a single interleaved file, as a pair of forward and reverse non-interleaved files to create a PairedEndLibrary, or as a single non-interleaved file to create a SingleEndLibrary.'},\n",
       " {'question': 'How can I import a file in KBase?',\n",
       "  'answer': 'To import a file, make sure you have an import type selected and the check box is active, then click \"Import Selected.\"'},\n",
       " {'question': 'What is the size limit for drag & drop file upload in KBase?',\n",
       "  'answer': 'The drag & drop from your local computer works for many files, but there is a size limit (about 5GB) dependent on your computer and browser.'},\n",
       " {'question': 'How can I upload larger files to KBase?',\n",
       "  'answer': 'For larger files around 20GB, use the Globus Online transfer or a direct upload with a URL.'},\n",
       " {'question': 'How can I share my data in KBase?',\n",
       "  'answer': 'You can share any of your Narratives (including their associated data) with one or more specific users, or make it publicly available to all KBase users. Please see the Share Narratives page for more information and how to share data and Narratives.'},\n",
       " {'question': \"What is KBase's data privacy policy?\",\n",
       "  'answer': 'Any data you upload to KBase is private unless you choose to share it. KBase policies on data are described on the Terms and Conditions and Data Policy pages.'},\n",
       " {'question': 'How can JGI users transfer data to KBase?',\n",
       "  'answer': 'JGI users can transfer public genome reads and assemblies (including their private data and annotated genomes) from JGI to their KBase account.'},\n",
       " {'question': 'What is the Bulk Import Guide in KBase?',\n",
       "  'answer': 'The Bulk Import Guide in KBase is a feature that allows users to import multiple files at once. All the files to be imported should have an import type selected and the check box should be active, then click \"Import Selected.\"'},\n",
       " {'question': 'What does bulk import currently support in KBase?',\n",
       "  'answer': 'Bulk import currently supports one parameter set per data type. If you need to upload files of a given data type with different parameters, perform a separate set of bulk imports for each parameter set.'},\n",
       " {'question': 'What is the new import cell used for in KBase?',\n",
       "  'answer': 'The new import cell is used for bulk imports in KBase.'},\n",
       " {'question': 'How can you view the data types you selected from the Staging Area in the new import cell?',\n",
       "  'answer': 'You can view the data types you selected from the Staging Area by clicking on the available types within the Data type column on the left in the new import cell.'},\n",
       " {'question': 'How can you view the logs and status details while the import is running?',\n",
       "  'answer': 'While the import is running, you can see all the logs and status details in the Job Status tab, formatted for the bulk import.'},\n",
       " {'question': 'How can you view the details of each of the child jobs?',\n",
       "  'answer': 'You can expand each of the child jobs by clicking on the corresponding line to view details, then expand further to see individual logs.'},\n",
       " {'question': 'What happens when the bulk import cell is set to run?',\n",
       "  'answer': 'When the bulk import cell is set to run, each child job for the individual data object will have a status.'},\n",
       " {'question': 'How can queued and running jobs be cancelled in KBase?',\n",
       "  'answer': 'Queued and running jobs can be cancelled using the bulk action dropdown.'},\n",
       " {'question': 'What happens when the jobs are completed successfully in KBase?',\n",
       "  'answer': 'When the jobs are completed successfully, you’ll be able to see all the information on the data object, repackaged for viewability in bulk. The first new feature is a table of all successful jobs. Clicking on object names adds a viewer widget to the Narrative. All the reports are available under the Reports tree. The full list is collapsed by default to conserve memory. The list can be expanded to show each report, and each report can be viewed in a separate window.'},\n",
       " {'question': 'What data types does the bulk import feature in KBase currently support?',\n",
       "  'answer': 'The bulk import feature in KBase currently supports Assembly - FASTA, SRA Reads, and FASTQ Reads Interleaved - FASTQ Interleaved reads.'},\n",
       " {'question': 'What are the supported data types for Bulk Import within the Staging Area in KBase?',\n",
       "  'answer': 'The supported data types for Bulk Import within the Staging Area in KBase are FASTQ Reads Noninterleaved, GenBank Genome, and GFF Metagenome.'},\n",
       " {'question': 'What happens when selected files are not supported data types for Bulk Import within the Staging Area in KBase?',\n",
       "  'answer': 'When selected files are not supported data types for Bulk Import within the Staging Area, separate, single import app cells will open in the Narrative for each type and file.'},\n",
       " {'question': 'What are the limitations of the new feature in KBase?',\n",
       "  'answer': 'The new feature in KBase is in its beta version and future development is still to come. Some of the most notable limitations are listed in the common bugs and limitations section. This is not a comprehensive list, but does contain the known bugs, issues, and limitations that are the highest priority for future releases.'},\n",
       " {'question': 'What is a known issue with the bulk import process in KBase?',\n",
       "  'answer': 'Only a single set of parameters can be used for a given type in a given bulk import.'},\n",
       " {'question': 'What is the workaround for the known issue with the bulk import process in KBase?',\n",
       "  'answer': 'Subset your imports and run separate imports for each subset of data with a single set of parameters.'},\n",
       " {'question': 'Does the limitation of using a single set of parameters apply to all data types in an import?',\n",
       "  'answer': 'No, it only applies to parameters within an import type. For example, you can import one set of Illumina reads and the assemblies in one import, and another type of reads in a separate import.'},\n",
       " {'question': 'What is the issue with the Staging Area in KBase?',\n",
       "  'answer': 'The Staging Area allows uploading incomplete files, which can lead to data corruption.'},\n",
       " {'question': 'How can I ensure that my files have fully uploaded in the Staging Area?',\n",
       "  'answer': 'You can ensure files have fully uploaded by checking that the file size in the Staging Area matches the size of the file from your machine. You can also use MD5 checksums to verify the file uploaded correctly.'},\n",
       " {'question': 'How can I view the MD5 inside the Staging Area?',\n",
       "  'answer': 'You can view the MD5 inside the Staging Area by viewing the info for the file.'},\n",
       " {'question': 'How can I verify that the MD5 in the Staging Area matches the MD5 on my local machine?',\n",
       "  'answer': 'You can verify that the MD5 in the Staging Area matches the MD5 on your local machine using the following command depending on your operating system (replacing \"assembly_1.fasta\" with your file name): MacOS: md5 assembly_1.fasta  Windows: certutil -hashfile assembly_1.fasta md5 Linux: md5sum assembly_1.fasta.'},\n",
       " {'question': 'What is the issue with file paths in KBase?',\n",
       "  'answer': 'File paths are not autofilled. During import pairs of files that go together, such as pairs of non-interleaved FASTQ files or FASTA and GFF Metagenome files, will need to be manually assigned to a forward or reverse read.'},\n",
       " {'question': 'What is the workaround for the file path issue in KBase?',\n",
       "  'answer': 'The workaround is to manually assign the files to a forward or reverse read during import.'},\n",
       " {'question': 'What happens to the Scientific Name input in KBase?',\n",
       "  'answer': 'The Scientific Name input disappears and becomes undefined within the input parameters when switching between tabs.'},\n",
       " {'question': 'What is the workaround for the Scientific Name input issue in KBase?',\n",
       "  'answer': 'This is a visual issue and the selected Scientific Name will still be included when running the import - even though it is not searchable in the undefined state. To verify the selected name is correct when the parameter does switch to the undefined state, 1) search for and select a different scientific name, 2) click between Data type tabs, 3) return to Data type tab and search for and select the original, preferred scientific name, 4) then Run the import.'},\n",
       " {'question': 'What is a known issue with KBase?',\n",
       "  'answer': 'There have been infrequent bugs found in which logs failed to appear, jobs reported as successful but no report could be generated, and similar issues.'},\n",
       " {'question': 'What is the workaround for the issue where job status/logs/results do not appear or do not update in KBase?',\n",
       "  'answer': 'Switching between tabs or reloading the page has been found to fix these issues.'},\n",
       " {'question': 'What should I do if my job reports as successful but no report could be generated in KBase?',\n",
       "  'answer': 'You can try switching between tabs or reloading the page.'},\n",
       " {'question': 'What is an assembly file in KBase?',\n",
       "  'answer': 'An assembly file is a single file containing one or more contiguous DNA sequences in FASTA format. It can be uploaded to KBase from your local computer or directly from a publicly accessible FTP or HTTP URL.'},\n",
       " {'question': 'How to upload FASTA files into KBase?',\n",
       "  'answer': 'Using a file on your computer, open the Import tab within the Data Browser. Then drag & drop the assembly file into the Staging Area box or select from your computer files. Open the Import As pulldown menu to the right of the filename in your Staging Area and select “Assembly.”'},\n",
       " {'question': 'What is the KBase data type for assembled, unannotated DNA sequence contigs or predicted coding sequences?',\n",
       "  'answer': 'Assembly'},\n",
       " {'question': 'How to upload annotated sequences in GenBank or GFF format?',\n",
       "  'answer': 'Please see the Genome page.'},\n",
       " {'question': 'How to import a FASTA file as Assembly from Staging Area in KBase?',\n",
       "  'answer': 'Make sure the correct file type is selected and the checkbox is active, then click the “Import Selected” button. The data slide-out will close and an app called “Import FASTA File as Assembly from Staging Area” will be added to your Narrative. Adjust the minimum contig length if needed and then click the green \"Run\" button to start the import.'},\n",
       " {'question': 'What happens after the import of a FASTA file is finished in KBase?',\n",
       "  'answer': 'When the import is finished, your Data Panel will update to show the new Assembly object, and a report will appear in the Import App.'},\n",
       " {'question': 'How to upload an Assembly from other sources in KBase?',\n",
       "  'answer': 'You can upload data into your KBase Staging Area using Globus.'},\n",
       " {'question': 'What are the options to upload and import data into KBase?',\n",
       "  'answer': 'There are several options to upload and import data into KBase, they are Drag and Drop, Upload with Globus, Upload with URL.'},\n",
       " {'question': 'What is the limitation of Drag & Drop option?',\n",
       "  'answer': 'The drag & drop option from your local computer works for many files, but there is a size limit that depends on your computer and browser. For larger files (around 20GB), use Globus Online transfer.'},\n",
       " {'question': 'How to import assemblies in bulk?',\n",
       "  'answer': 'Assemblies can be imported as one of the supported bulk import types. You can select multiple assemblies simultaneously from the staging area to import them at once.'},\n",
       " {'question': 'How to transfer assemblies from JGI to KBase?',\n",
       "  'answer': 'If you are a JGI user, you can transfer public genome reads and assemblies (as well as your private data and annotated genomes) from JGI to your KBase account.'},\n",
       " {'question': 'What is the previous topic before this page?',\n",
       "  'answer': 'The previous topic before this page is Bulk Import Limitations.'},\n",
       " {'question': 'What is the next topic after this page?',\n",
       "  'answer': 'The next topic after this page is Genome.'},\n",
       " {'question': 'When was this page last modified?',\n",
       "  'answer': 'This page was last modified 1 year ago.'},\n",
       " {'question': 'What is a Genome object in KBase?',\n",
       "  'answer': 'In KBase, a Genome object is the annotated version of an Assembly or annotated predicted coding sequences and can encompass several types of feature calls.'},\n",
       " {'question': 'What file formats does the Genome importer in KBase support?',\n",
       "  'answer': 'The Genome importer supports only GenBank and GFF formatted files.'},\n",
       " {'question': 'How does KBase handle a GenBank-formatted input file?',\n",
       "  'answer': 'KBase parses the input file into two data objects: an assembly object with the sequence and a genome object containing the original feature calls and annotations.'},\n",
       " {'question': 'Can GenBank-formatted files with no features be uploaded as Genomes in KBase?',\n",
       "  'answer': 'Yes, GenBank-formatted files with no features can be uploaded as Genomes.'},\n",
       " {'question': 'How can a GenBank-formatted file be uploaded into the Staging Area in KBase?',\n",
       "  'answer': 'A GenBank-formatted file can be uploaded into the Staging Area from your local computer (files with.gb, .gbff, or .gbk extensions) or directly from an FTP or HTTP URL.'},\n",
       " {'question': 'What are the two data objects parsed from a GFF-formatted file and a corresponding FASTA file?',\n",
       "  'answer': 'An assembly object with the sequence and a genome object containing the original feature calls and annotations.'},\n",
       " {'question': 'How to import a GenBank-formatted genome in KBase?',\n",
       "  'answer': 'Open the Import tab within the Data Browser. Then drag & drop the genome file into your Staging Area. Open the Import As... pulldown menu to the right of the filename in your Staging Area and select “GenBank Genome.” Make sure the correct file type is selected and the checkbox is active, then click the “Import Selected” button.'},\n",
       " {'question': 'What happens after clicking the “Import Selected” button when importing a GenBank-formatted genome?',\n",
       "  'answer': 'The Data Browser will close and the “Import GenBank File as Genome from Staging Area” App will be added to your Narrative.'},\n",
       " {'question': 'How to start the import in GenBank Genome App?',\n",
       "  'answer': 'Click the green \"Run\" button to start the import.'},\n",
       " {'question': 'What happens when the import is finished in GenBank Genome App?',\n",
       "  'answer': 'When the import is finished, your Data Panel will update to show the new Genome and Assembly objects, and a report will appear in the Import App.'},\n",
       " {'question': 'How to import a GFF-formatted Genome?',\n",
       "  'answer': 'Open the Import tab in the Data Browser and drag and drop the genome corresponding FASTA file into your Staging Area. In your Staging Area, open the pulldown menu to the right of the GFF filename and select “GFF Genome.\"'},\n",
       " {'question': 'How to import GFF/FASTA file as Genome from Staging Area in KBase?',\n",
       "  'answer': 'First, make sure the correct file type is selected and the checkbox is active, then click the \"Import Selected\" button. The data slide-out will close and the “Import GFF/FASTA File as Genome from Staging Area” App will be added to your Narrative. The GFF File Path name will be filled in. You will need to fill in the name of the FASTA file. Using the dropdown for the “FASTA File Path”, select the FASTA file in the Staging Area. Ensure the file type is a FASTA file type. Then, fill in the Scientific Name and the name for the Genome data object that will be created by the import. You can edit the name of the output Genome Object Name, Scientific Name, and any advanced options as needed. Click the green \"Run\" button. When the import is finished, your Data Panel will update to show the new Genome object, and a report will appear in the Import App.'},\n",
       " {'question': 'What happens after clicking the \"Run\" button while importing a GFF/FASTA file?',\n",
       "  'answer': 'When the import is finished, your Data Panel will update to show the new Genome object, and a report will appear in the Import App.'},\n",
       " {'question': 'How to upload a Genome from other sources in KBase?',\n",
       "  'answer': 'You can upload data into your KBase Staging Area.'},\n",
       " {'question': 'What are the options to upload and import data into KBase?',\n",
       "  'answer': 'There are several options to upload and import data into KBase, including Drag and Drop, Upload with Globus, and Upload with URL.'},\n",
       " {'question': 'What are the limitations of the Drag & Drop option?',\n",
       "  'answer': 'The drag & drop option from your local computer works for many files, but there is a size limit that depends on your computer and browser. For larger files (around 20 gigabases), use the Globus Online transfer.'},\n",
       " {'question': 'How can I upload a Genome using a URL in KBase?',\n",
       "  'answer': 'Open the Import tab in the Data Browser and click on the Upload with URL button (below the drag & drop area) to open an Upload App Cell.'},\n",
       " {'question': 'Can I import multiple assemblies simultaneously from the staging area?',\n",
       "  'answer': 'Yes, you can select multiple assemblies simultaneously from the staging area to import them at once.'},\n",
       " {'question': 'How can I upload a file to staging from web app in KBase?',\n",
       "  'answer': 'You can open the “Upload a File to Staging from Web” App in your Narrative or directly from the Apps Panel. From the app, click on the dropdown for the URL Type and select the URL type. Then click the \"+\" button for the URLs and paste in the name of the Genome file (GenBank or GFF). Hit the \"+\" button again and paste in the name of the FASTA file. Then click the green \"Run\" button to start the upload. After the App completes the files will appear in your Staging Area, which you can access via the Import tab in the Data Browser.'},\n",
       " {'question': 'What is the difference between uploading a GenBank Genome and a GFF Genome?',\n",
       "  'answer': 'When uploading a GenBank Genome, you will only need to use one link. When uploading a GFF Genome, you will need to use two links for the GFF file and the FASTA file.'},\n",
       " {'question': 'What should I do after the files appear in my Staging Area?',\n",
       "  'answer': 'After the files appear in your Staging Area, you need to import them to your Narrative to use them in analyses.'},\n",
       " {'question': 'What types of reads library data objects can be created in KBase?',\n",
       "  'answer': 'SingleEndLibrary or a PairedEndLibrary.'},\n",
       " {'question': 'What can the tools in KBase be used for after importing reads from FASTQ and SRA files?',\n",
       "  'answer': 'The tools in KBase can be used to assemble reads into an “Assembly” data object or to align reads to an “Assembly.\"'},\n",
       " {'question': 'What formats can single-end and paired-end reads be uploaded in?',\n",
       "  'answer': 'Single-end and paired-end reads can be uploaded in FASTQ or SRA format.'},\n",
       " {'question': 'What should be the file extension for FASTQ and SRA files?',\n",
       "  'answer': 'For FASTQ files, the filename should end with the .fastq, .fnq, or .fq file extension. SRA files should have the .sra file extension.'},\n",
       " {'question': 'How can files be uploaded into KBase?',\n",
       "  'answer': 'Files can be uploaded into your KBase Staging Area from your local computer or directly from a publicly accessible FTP or HTTP URL.'},\n",
       " {'question': 'How can I import single-end library in FASTQ or SRA format from my computer to KBase?',\n",
       "  'answer': 'Open the Import tab within the Data Browser. Then drag & drop the single-end library into your Staging Area. Open the pulldown menu to the right of the filename in the Staging Area and select “FASTQ Reads NonInterleaved\" or \"SRA Reads.\" Make sure the correct file type is selected and the checkbox is active, then click \"Import Selected\". The slide-out Data Browser will close and an app called “Import FASTQ/SRA File as Reads from Staging Area” will be added to your Narrative.'},\n",
       " {'question': 'What happens after I click \"Import Selected\" in KBase?',\n",
       "  'answer': 'The slide-out Data Browser will close and an app called “Import FASTQ/SRA File as Reads from Staging Area” will be added to your Narrative.'},\n",
       " {'question': 'What file types can I import from my computer to KBase?',\n",
       "  'answer': 'You can import single-end library in FASTQ or SRA format.'},\n",
       " {'question': 'How to import a Reads Library file in KBase?',\n",
       "  'answer': 'The name of the Reads Library file is already filled in, as is a suggested Reads Object Name that will be created by the import. Adjust the Sequencing Technology and any of the advanced options if needed. If it was a metagenomic sample, uncheck the box next to Single Genome. When ready, click the green \"Run\" button to start the import. When the import is finished, your Data Panel will update to show the new SingleEndLibrary object, and a report will appear in the import app cell.'},\n",
       " {'question': 'What is a paired-end library in FASTQ format?',\n",
       "  'answer': 'A paired-end library can be either two files which typically have the same name and are designated as forward and reverse or a single interleaved file. Interleaved files use an 8-line format where forward and reverse reads alternate.'},\n",
       " {'question': 'How to open the Import tab in the Data Browser?',\n",
       "  'answer': 'You can open the Import tab in the Data Browser by clicking on it.'},\n",
       " {'question': 'How to import a file into the Staging Area in KBase?',\n",
       "  'answer': 'Drag either one interleaved file or two paired files into the Staging Area.'},\n",
       " {'question': 'How to import a FASTQ Reads NonInterleaved file in KBase?',\n",
       "  'answer': 'Open the pulldown menu to the right of the filename under the Import As... column in your Staging Area and select “FASTQ Reads NonInterleaved” for the first file in the pair. Make sure the correct file type is selected and the checkbox is active, then click \"Import Selected\". The Data Browser slide-out will close and the “Import FASTQ/SRA File as Reads from Staging Area” App will open.'},\n",
       " {'question': 'What happens after importing a FASTQ/SRA File as Reads from Staging Area in KBase?',\n",
       "  'answer': 'The name of the FASTA/FASTQ file is filled in. A suggested Reads Object Name is also created by the import and can be changed.'},\n",
       " {'question': 'How to import two paired files in KBase?',\n",
       "  'answer': 'If the files are two paired file, you will need to select a file for “Reverse/Right FASTA/FASTQ File Path” from the dropdown.'},\n",
       " {'question': 'How can you adjust the parameters for a single-end library in KBase?',\n",
       "  'answer': 'You can adjust the Sequencing Technology and any of the advanced options if needed. If the file is an interleaved paired-end library, check the box to the right of Interleaved.'},\n",
       " {'question': 'How do you start the import process in KBase?',\n",
       "  'answer': 'Click the green \"Run\" button to start the import.'},\n",
       " {'question': 'What happens when the import is finished in KBase?',\n",
       "  'answer': 'When the import is finished, your Data Panel will update to show the new PairedEndLibrary object, and a report will appear in the Import App.'},\n",
       " {'question': 'What are the other options for adding data to your staging area in KBase?',\n",
       "  'answer': 'You can import reads into KBase using Globus Online, or by supplying a URL for a publicly accessible FTP location, Google Drive, Dropbox, or a direct HTTP link.'},\n",
       " {'question': 'How can you import reads into your Narrative from a publicly accessible URL?',\n",
       "  'answer': 'You can bypass the Staging Area and directly import reads into your Narrative using one of these three apps: Import SRA File as Reads From Web, Import Single-End Reads From Web, or Import Paired-End Reads From Web.'},\n",
       " {'question': 'What is the process of importing SRA reads from NCBI?',\n",
       "  'answer': 'This generally means navigating to the Data access tab of the Run Browser in the NCBI Sequence Read Archive (SRA) for the reads to import. For example, for SRR18272216, navigate to the Data access tab and copy the SRA-download link under Name and paste the full link into the Import from Web App.'},\n",
       " {'question': 'What is the difference between directly importing files into KBase from the web and uploading to the Staging Area then importing to a Narrative?',\n",
       "  'answer': 'The process is essentially the same, except the transfer is carried out by the Importer App when directly importing from the web.'},\n",
       " {'question': 'How can I import SRA sequences into KBase?',\n",
       "  'answer': 'You can copy the SRA-download link located under the Name heading and paste the link into the URL input within the Importer App.'},\n",
       " {'question': 'Where can I find guidance on how to search and download SRA sequences?',\n",
       "  'answer': 'You can refer to the NCBI Search and Download documentation.'},\n",
       " {'question': 'What are the limitations of the drag & drop feature in KBase?',\n",
       "  'answer': 'The drag & drop from your local computer works for many files, but there is a size limit that depends on your computer and browser. Some users have reported problems around 20GB. For larger files, you can use Globus Online transfer.'},\n",
       " {'question': 'How can I import multiple assemblies simultaneously into KBase?',\n",
       "  'answer': 'FASTQ and SRA reads can be imported as one of the supported bulk import types. You can select multiple assemblies simultaneously from the staging area to import them at once.'},\n",
       " {'question': 'How can I transfer reads from JGI to my KBase account?',\n",
       "  'answer': 'If you are a JGI user, you can transfer public genome reads and assemblies (as well as your private data and annotated genomes) from JGI to your KBase account.'},\n",
       " {'question': 'What was the last modified date of the document?',\n",
       "  'answer': 'The document was last modified 1 year ago.'},\n",
       " {'question': 'What is the next topic after Genome?',\n",
       "  'answer': 'The next topic after Genome is Flux Balance Analysis (FBA) Model.'},\n",
       " {'question': 'What is the previous topic before Flux Balance Analysis (FBA) Model?',\n",
       "  'answer': 'The previous topic before Flux Balance Analysis (FBA) Model is Genome.'},\n",
       " {'question': 'What is a Flux Balance Analysis (FBA) Model?',\n",
       "  'answer': 'A Flux Balance Analysis (FBA) Model is a model used to predict biomass yield given a certain amount of input nutrient.'},\n",
       " {'question': 'How can FBA models be uploaded into KBase?',\n",
       "  'answer': 'FBA models can be uploaded into KBase as a Systems Biology Markup Language (SBML) file using the .sbml or .xml file extension, an Excel file using the .xls extension, or a tab-separated values (TSV) file using the .tsv extension.'},\n",
       " {'question': 'What files are required when uploading an FBA model in the TSV format?',\n",
       "  'answer': 'When uploading an FBA model in the TSV format, there will be two files corresponding to chemical compounds and reactions.'},\n",
       " {'question': 'Where can more SBML FBA Models for various organisms be found?',\n",
       "  'answer': 'More SBML FBA Models for various organisms can be found at http://systemsbiology.ucsd.edu/Downloads.'},\n",
       " {'question': 'What is the requirement for the Excel file format in KBase?',\n",
       "  'answer': 'The most important requirement for the Excel file format in KBase is naming the tabs “ModelCompounds” and “ModelReactions.”'},\n",
       " {'question': 'How are the chemical compounds and reactions tables saved in the Tab-Separated Values file format?',\n",
       "  'answer': 'In the Tab-Separated Values file format, the chemical compounds and reactions tables are saved as two separate files named “FBAModelCompounds.tsv” and “FBAModelReactions.tsv”.'},\n",
       " {'question': \"What does the 'id' field represent in the FBAModelCompounds table?\",\n",
       "  'answer': \"The 'id' field in the FBAModelCompounds table represents the Compound identifier.\"},\n",
       " {'question': \"What does the 'aliases' field represent in the FBAModelCompounds table?\",\n",
       "  'answer': \"The 'aliases' field in the FBAModelCompounds table represents the Alternative names for the chemical compound.\"},\n",
       " {'question': \"What does the 'charge' field represent in the FBAModelCompounds table?\",\n",
       "  'answer': \"The 'charge' field in the FBAModelCompounds table represents the Formal charge of the molecule.\"},\n",
       " {'question': 'What is required for importing an FBA model into KBase?',\n",
       "  'answer': 'Importing an FBA model into KBase requires specified reaction IDs for the biomass producing equations in the model.'},\n",
       " {'question': 'How many biomass equations does an individual organism FBA model need?',\n",
       "  'answer': 'An individual organism FBA model needs a single biomass equation.'},\n",
       " {'question': 'How should the biomass producing reactions be denoted in a community model?',\n",
       "  'answer': 'In a community model, the first biomass producing reaction should be denoted as \"bio1\", the second as \"bio2\", the third as \"bio3\", and so on.'},\n",
       " {'question': \"What does the 'id' field represent in the KBase Biochemistry reference?\",\n",
       "  'answer': \"The 'id' field represents the Reaction identifier.\"},\n",
       " {'question': \"What does the 'direction' field indicate in the KBase Biochemistry reference?\",\n",
       "  'answer': \"The 'direction' field indicates the Directionality of chemical reaction specified as forward (->), reversed (<-), or equilibrium (<->).\"},\n",
       " {'question': \"What is the 'compartment' field in the KBase Biochemistry reference?\",\n",
       "  'answer': \"The 'compartment' field represents the Cellular compartment that the reaction takes place in.\"},\n",
       " {'question': \"What does the 'gpr' field represent in the KBase Biochemistry reference?\",\n",
       "  'answer': \"The 'gpr' field represents the Gene-Protein Reaction in PATRIC identifier format.\"},\n",
       " {'question': \"What is the 'name' field in the KBase Biochemistry reference?\",\n",
       "  'answer': \"The 'name' field represents the Protein associated with the reaction.\"},\n",
       " {'question': 'What is an FBA model in KBase?',\n",
       "  'answer': 'An FBA model is a Flux Balance Analysis model that can be uploaded into your KBase workspace.'},\n",
       " {'question': 'What is the first step to upload an FBA model into KBase?',\n",
       "  'answer': 'The first step is to add the Genome that corresponds to the FBA Model you wish to upload.'},\n",
       " {'question': 'How to import an FBA model from a file on your computer to KBase?',\n",
       "  'answer': 'Open the Import tab within the Data Browser, then drag & drop the FBA model file (plus compound file if using TSV format) into your Staging Area. Open the pulldown menu to the right of the filename in the Staging Area and select “FBA Model.\" Then click the import icon (up arrow) to the right of “FBA Model.\"'},\n",
       " {'question': 'What happens after clicking the import icon for the FBA model?',\n",
       "  'answer': 'The Data Browser slide-out will close and an app called “Import TSV/XLS/SBML File as an FBAModel from Staging Area” will be added to your Narrative.'},\n",
       " {'question': 'What is the suggested name for the FBA Model data object that will be created by the import?',\n",
       "  'answer': 'The suggested name for the FBA Model data object that will be created by the import is filled in the Model file path.'},\n",
       " {'question': 'What needs to be linked to the FBA Model?',\n",
       "  'answer': 'The corresponding Genome needs to be linked to the FBA Model.'},\n",
       " {'question': 'What is required to import gene-protein-reactions (GPRs)?',\n",
       "  'answer': 'Adding a Genome with gene IDs that match genes IDs in the model is required to import gene-protein-reactions (GPRs).'},\n",
       " {'question': 'What are the correct Model file types that can be selected?',\n",
       "  'answer': 'The correct Model file types that can be selected are SBML, Excel, Tab-separated values.'},\n",
       " {'question': 'How to add a biomass in KBase?',\n",
       "  'answer': 'Click the \"+\" button to the right of Biomass and type in the name of the first biomass. If there are additional biomass values to enter, click the \"+\" button to open another parameter field for another biomass.'},\n",
       " {'question': 'What should be noted when entering the name of the biomass reaction in KBase?',\n",
       "  'answer': 'If the biomass reaction name starts with “R_”, do NOT include the “R_” when entering the name in this field.'},\n",
       " {'question': 'How to import a TSV-formatted model in KBase?',\n",
       "  'answer': 'Select the corresponding Compounds file path from the dropdown.'},\n",
       " {'question': 'What happens when the import is finished in KBase?',\n",
       "  'answer': 'When the import is finished, your Data Panel will update to show the new FBA Model object, and a report will appear in the Import App.'},\n",
       " {'question': 'What are the limitations of Drag & Drop upload in KBase?',\n",
       "  'answer': 'Drag & drop upload from your local computer works for many files, but there is a size limit  dependent on your computer and browser. For larger files around 20GB or more, use Globus Online transfer.'},\n",
       " {'question': 'Why is it necessary to associate a genome with user uploaded models in KBase Metabolic Modeling Apps?',\n",
       "  'answer': 'It is necessary to ensure fidelity between the genome annotations and gene-protein-reactions in the metabolic model.'},\n",
       " {'question': 'What happens if the gene IDs between the model and genome do not match?',\n",
       "  'answer': 'If the gene IDs do not match, you will need to change either the model or the genome to ensure their gene IDs match.'},\n",
       " {'question': 'What might cause the gene IDs between the model and genome to not match?',\n",
       "  'answer': 'Occasionally, researchers will use locus tags or gene IDs for constructing metabolic models that are different than those found commonly in reference genomes.'},\n",
       " {'question': 'How can I associate a model with a genome in KBase?',\n",
       "  'answer': 'One way to accomplish this is to locate the genome used as the reference for building the model – based on shared gene IDs – and to upload this genome into KBase before uploading the model and associating it with the genome.'},\n",
       " {'question': 'What should I do if I cannot locate the genome with the gene IDs used to build the model?',\n",
       "  'answer': 'The best option for reconciling the gene IDs is to edit the model file (as SBML or TSV) to change the gene IDs contained therein to match the gene IDs of the genome.'},\n",
       " {'question': 'Where can I find a detailed example of how to associate a model with a genome in KBase?',\n",
       "  'answer': 'You can find a detailed example on the public Help Board ticket.'},\n",
       " {'question': 'How to format reaction media files for KBase?',\n",
       "  'answer': 'Media can be uploaded as a TSV (tab-separated values) or Excel file using four columns: Maxflux (maximum allowed uptake/excretion of compound), Minflux (minimum allowed uptake/excretion of compound), Concentration (concentration of compound in mol/L), Compound identifier (e.g., ModelSEED ID, KEGG ID, PubChem ID, or compound name such as “glucose”).'},\n",
       " {'question': 'What are the options for defining specific growth or culture media for metabolic modeling in KBase?',\n",
       "  'answer': 'Within the Narrative Interface, you can use the Edit Media tool to build media based on the hundreds of conditions available in KBase or import your own custom media as a TSV or Excel file.'},\n",
       " {'question': \"What is the purpose of the 'Maxflux' column in the media file for KBase?\",\n",
       "  'answer': \"The 'Maxflux' column represents the maximum allowed uptake/excretion of a compound.\"},\n",
       " {'question': \"What is the purpose of the 'Minflux' column in the media file for KBase?\",\n",
       "  'answer': \"The 'Minflux' column represents the minimum allowed uptake/excretion of a compound.\"},\n",
       " {'question': \"What is the purpose of the 'Concentration' column in the media file for KBase?\",\n",
       "  'answer': \"The 'Concentration' column represents the concentration of a compound in mol/L.\"},\n",
       " {'question': \"What is the purpose of the 'Compound identifier' column in the media file for KBase?\",\n",
       "  'answer': \"The 'Compound identifier' column is for the ModelSEED ID, KEGG ID, PubChem ID, or compound name such as “glucose”.\"},\n",
       " {'question': 'Where can I find a list of all the biochemical compounds in KBase?',\n",
       "  'answer': 'The list of all the biochemical compounds is available in the KBase ModelSEED Biochemistry Database.'},\n",
       " {'question': 'What is the purpose of the concentration field in KBase?',\n",
       "  'answer': 'The concentration field is not used by any of the metabolic modeling tools in KBase. It is best to account for these values accurately so that your media files will retain their utility pending updates.'},\n",
       " {'question': 'What do the minflux and maxflux columns represent in KBase?',\n",
       "  'answer': 'The minflux and maxflux columns connect media to a flux balance analysis (FBA) metabolic model. The values in these columns are absolute units representing the range of possible fluxes for reactions that transport the media compounds into and out of the cell. A negative flux value corresponds to excretion (transport out of the cell), and a positive flux value corresponds to uptake (transport into the cell).'},\n",
       " {'question': 'What does assigning a positive number to minflux for a particular compound do?',\n",
       "  'answer': 'It essentially forces uptake of that compound.'},\n",
       " {'question': 'What does assigning a negative number to maxflux for a particular compound do?',\n",
       "  'answer': 'It will force excretion of the compound.'},\n",
       " {'question': 'How are minimum and maximum flux measured?',\n",
       "  'answer': 'They are measured in mmol per gram cell dry weight per hour.'},\n",
       " {'question': 'What should be the name of the worksheet that contains your media conditions in a media Excel file?',\n",
       "  'answer': 'The name of the worksheet that contains your media conditions must be named “MediaCompounds”.'},\n",
       " {'question': 'What happens if the worksheet that contains your media conditions is not named “MediaCompounds”?',\n",
       "  'answer': 'It will not upload.'},\n",
       " {'question': 'How to import media from a TSV or Excel file in KBase?',\n",
       "  'answer': 'Open the Import tab within the Data Browser. Then drag & drop the media file to upload it to your Staging Area. Choose \"Media\" from the Import As... data type dropdown menu in your Staging Area. Click the upload icon next to the field that now says “Media.”'},\n",
       " {'question': 'What is the next step after uploading the media file in the Staging Area?',\n",
       "  'answer': 'The “Import Media file (TSV/Excel) from Staging Area” App will be added to your Narrative.'},\n",
       " {'question': 'What types of files can be used to import media in KBase?',\n",
       "  'answer': 'TSV or Excel files can be used to import media in KBase.'},\n",
       " {'question': 'How can I change the name of the Media Object in KBase?',\n",
       "  'answer': 'You can change the name of the Media Object by typing in the new name in the appropriate field.'},\n",
       " {'question': 'What happens after I click the \"Run\" button in KBase?',\n",
       "  'answer': 'After clicking the \"Run\" button, the import process will start. Once completed, the Media data object will appear in your Data Panel.'},\n",
       " {'question': 'Where can I find the Media data object after the import process in KBase?',\n",
       "  'answer': 'The Media data object will appear in your Data Panel after the import process.'},\n",
       " {'question': 'What is the Expression Matrix data type in KBase?',\n",
       "  'answer': 'The Expression Matrix data type contains gene expression values taken under given sampling conditions.'},\n",
       " {'question': 'How can Expression Matrices be uploaded and imported in KBase?',\n",
       "  'answer': 'Expression Matrices can be uploaded and imported using the Tab-separated values (TSV) file extension .tsv or the Excel file extension .xls.'},\n",
       " {'question': 'How should a TSV file be formatted for use with KBase?',\n",
       "  'answer': 'A tab-separated values (TSV) file is a tab delimited text file that has genes across the rows and sample observations across the columns. The first label in the first column should be “feature_ids” followed by tab-delimited labels for samples.'},\n",
       " {'question': 'How should an Excel file be formatted for use with KBase?',\n",
       "  'answer': 'In Excel, the first column matches Feature IDs from the genome. This column could also be a gene alias. Gene aliases supported by KBase include NCBI, EMBL, UniProt, BioCyc, and ASAP. The column headings are sample conditions.'},\n",
       " {'question': 'What should be excluded from the gene features?',\n",
       "  'answer': 'Gene features that are missing all expressions or are composed of non-changing expressions across the samples should be excluded.'},\n",
       " {'question': 'What is the format of the expression data file?',\n",
       "  'answer': 'The expression data file is in TSV format.'},\n",
       " {'question': 'Where can I find an example of Expression Matrix format?',\n",
       "  'answer': 'You can download the \"Shewanella_MR-1_M3D_ExpData\" ExpressionMatrix data set from the Example tab in the Data Panel. After downloading, unzip the file and examine the \"matrix.tsv\" file.'},\n",
       " {'question': 'What does the first column in the TSV file represent?',\n",
       "  'answer': 'The first column in the TSV file corresponds to gene identifiers.'},\n",
       " {'question': 'What do the remaining columns in the TSV file represent?',\n",
       "  'answer': 'The remaining columns in the TSV file contain expression values for the appropriate gene and sample.'},\n",
       " {'question': 'What is the data structure for gene IDs in KBase plant genomes?',\n",
       "  'answer': 'The gene IDs retain the data structure from the external source databases (Ensembl or Phytozome) and do not have aliases.'},\n",
       " {'question': 'How to construct an expression dataset in KBase?',\n",
       "  'answer': 'Append gene IDs with the transcript IDs followed by “.CDS.\"'},\n",
       " {'question': 'How to upload an Expression Matrix into KBase?',\n",
       "  'answer': 'In order to successfully upload an Expression Matrix into KBase, you first need to add the Genome that corresponds to references in the Expression Matrix. Using a file on your computer, open the Import tab within the Data Browser. Then drag & drop the expression matrix file into the Staging Area.'},\n",
       " {'question': 'Can Expression Matrices be uploaded into KBase without a Genome?',\n",
       "  'answer': 'Yes, Expression Matrices can be uploaded into KBase without a Genome, but they will have limited use.'},\n",
       " {'question': 'How can I import an expression matrix into my Narrative from the Staging Area in KBase?',\n",
       "  'answer': 'Open the pulldown menu to the right of the filename under the Import As... column in your Staging Area and select “Expression Matrix.\" Then, click the import icon (up arrow) to the right of “Expression Matrix.” The Data Browser will close and the “Import TSV File as Expression Matrix From Staging Area” App will be added to your Narrative.'},\n",
       " {'question': 'What is the next step after the “Import TSV File as Expression Matrix From Staging Area” App is added to the Narrative?',\n",
       "  'answer': 'The name of the Tab-delimited (TSV) file is already filled in, as is a suggested name for the Expression Matrix data object that will be created by the import, which can be edited or changed.'},\n",
       " {'question': 'How can I link a genome to the expression matrix in KBase?',\n",
       "  'answer': 'To add the name of the Genome, click on the ‘show advanced’ link to the right of ‘Input Objects’ in the Import App.'},\n",
       " {'question': 'How to start the import in KBase?',\n",
       "  'answer': 'Use the Genome dropdown to select the corresponding genome. Adjust any of the other advanced options if needed, then click the green \"Run\" button to start the import.'},\n",
       " {'question': 'What happens when the import is finished in KBase?',\n",
       "  'answer': 'When the import is finished, your Data Panel will update to show the new Expression Matrix object, and a report will appear in the Import App.'},\n",
       " {'question': 'How to ensure that the gene identifiers listed in your dataset correspond to the aliases in KBase?',\n",
       "  'answer': 'Click the name of the Genome object to open up the genome viewer. Click on the tab labeled Browse Features and locate a gene of interest by searching for the name of the function or protein associated with the gene. Click the Feature ID of the gene of interest to open a new tab with additional information about the gene.'},\n",
       " {'question': 'What is an Expression Matrix object in KBase?',\n",
       "  'answer': 'An Expression Matrix object is created when the import of a gene expression dataset is finished in KBase. It is shown in the Data Panel and a report appears in the Import App.'},\n",
       " {'question': 'What should you crosscheck the gene labels in your expression dataset with?',\n",
       "  'answer': 'You should crosscheck the gene labels in your expression dataset with either the Feature ID or one of the aliases in the section titled Aliases.'},\n",
       " {'question': 'What is a limitation of the drag & drop feature in KBase?',\n",
       "  'answer': 'The drag & drop feature in KBase has a size limit that depends on your computer and browser. For larger files around 20GB, it is recommended to use the Globus Online transfer.'},\n",
       " {'question': 'What should you use for transferring larger files around 20GB in KBase?',\n",
       "  'answer': 'For transferring larger files around 20GB in KBase, you should use the Globus Online transfer.'},\n",
       " {'question': 'What is the Phenotype Set data type in KBase?',\n",
       "  'answer': 'The Phenotype Set data type represents experimental data about an organism’s ability to grow in specific media conditions, recorded as growth or no growth.'},\n",
       " {'question': 'How can a Phenotype Set be uploaded in KBase?',\n",
       "  'answer': 'A Phenotype Set can be uploaded from a TSV (tab-separated values) file with a .tsv or .tab file extension, or from Excel spreadsheet with a .xls extension.'},\n",
       " {'question': 'What are the required columns in the spreadsheet for uploading a Phenotype Set in KBase?',\n",
       "  'answer': 'The spreadsheet needs to have exactly these five columns: Gene knockout (geneko), Workspace information (mediaws), and Media.'},\n",
       " {'question': \"What does the 'Gene knockout' column represent in the Phenotype Set spreadsheet?\",\n",
       "  'answer': \"'Gene knockout' represents the list of genes knocked out in the phenotype; 'none' is used for wild-type phenotypes. Gene IDs should be in the same format that appears in your metabolic model.\"},\n",
       " {'question': \"What does the 'Workspace information' column represent in the Phenotype Set spreadsheet?\",\n",
       "  'answer': \"'Workspace information' represents the workspace where the media for the phenotype data was loaded into KBase. The workspace information can be found in the Narratives tab by clicking Show Advanced Controls at the end of the list of Narratives.\"},\n",
       " {'question': \"What does 'addtlCpd' stand for in KBase?\",\n",
       "  'answer': \"'addtlCpd' stands for Additional Compounds, which are additional media compound IDs to be added alongside the primary media formulation.\"},\n",
       " {'question': 'How is growth indicated in KBase?',\n",
       "  'answer': 'Growth is indicated by a 1 or 0, with 1 meaning growth and 0 meaning no growth.'},\n",
       " {'question': 'How can I download an example phenotype set data table in KBase?',\n",
       "  'answer': 'To download an example phenotype set data table, go to the \"Example\" tab in the data Data Browser slide-out, add a Phenotype Data Object to the Narrative and download the TSV file.'},\n",
       " {'question': 'What should I do before importing a Phenotype Set file?',\n",
       "  'answer': 'Before importing a Phenotype Set file, you should add the Genome that corresponds to referenced in the Phenotype Set. Add the Genome object to the Narrative.'},\n",
       " {'question': 'How to import a Phenotype Set from the Staging Area in KBase?',\n",
       "  'answer': 'Click the import icon (up arrow) to the right of “Phenotype Set”. The data slide-out will close and the “Import TSV File as Phenotype Set From Staging Area” App will be added to your Narrative. The name of the Phenotype TSV file is already filled in, as is a suggested name for the Phenotype Set data object that will be created by the import, which you can edit. Add the name of the Genome, click on the dropdown to select the corresponding genome that you added to your Narrative earlier. Click the green \"Run\" button to start the import. When the import is finished, your Data Panel will update to show the new Phenotype Set object, and a report will appear in the Import App.'},\n",
       " {'question': 'What is the first step to import a Phenotype Set in KBase?',\n",
       "  'answer': 'The first step is to drag & drop the phenotype set file into your Staging Area. Then open the pulldown menu to the right of the filename in the Staging Area and select “Phenotype Set.\"'},\n",
       " {'question': 'What happens after clicking the green \"Run\" button during the import of a Phenotype Set in KBase?',\n",
       "  'answer': 'After clicking the green \"Run\" button, the import will start. When the import is finished, your Data Panel will update to show the new Phenotype Set object, and a report will appear in the Import App.'},\n",
       " {'question': 'What is the size limit for drag & drop files in KBase?',\n",
       "  'answer': 'The size limit depends on your computer and browser, but for larger files around 20GB, it is recommended to use the Globus Online transfer.'},\n",
       " {'question': 'What should I use for transferring larger files in KBase?',\n",
       "  'answer': 'For larger files around 20GB, use the Globus Online transfer.'},\n",
       " {'question': 'Can I drag and drop files from my local computer to KBase?',\n",
       "  'answer': 'Yes, the drag & drop from your local computer works for many files in KBase.'},\n",
       " {'question': 'What is an Amplicon Matrix?',\n",
       "  'answer': 'Amplicon matrices or taxonomic abundance matrices include OTU (operational taxonomic units) and ASV (amplicon sequence variants).'},\n",
       " {'question': 'How can Amplicon Matrices be uploaded in KBase?',\n",
       "  'answer': 'Amplicon matrices can be uploaded from a TSV (tab-separated values) file with a .tsv file extension.'},\n",
       " {'question': 'What files are needed for uploading amplicon data?',\n",
       "  'answer': 'Two files are needed for uploading amplicon data: a TSV file for the Amplicon Matrix and a FASTA file with consensus sequences.'},\n",
       " {'question': 'What should be included in the FASTA file for amplicon data?',\n",
       "  'answer': 'Consensus sequences must be included in the FASTA file. Sequence names must match exactly with the taxa names in the amplicon matrix.'},\n",
       " {'question': 'What is the recommended format for uploading raw counts in KBase?',\n",
       "  'answer': 'It is recommended to upload raw counts as analysis (rarefaction, standardization, and normalization) will be tracked and provenance maintained on system.'},\n",
       " {'question': 'What is a taxonomic abundance matrix in KBase?',\n",
       "  'answer': 'A taxonomic abundance matrix in KBase is a matrix where each row represents a taxon and each column represents a sample. There can also be an optional column for taxonomy.'},\n",
       " {'question': 'What additional information is needed for parameters in KBase?',\n",
       "  'answer': 'Additional information for parameters include sample processing metadata such as sequencing technology or platform, target gene or region, and bioinformatic processing metadata like clustering method, quality filtering steps.'},\n",
       " {'question': 'How can Amplicon Matrices be uploaded into KBase?',\n",
       "  'answer': 'Amplicon Matrices can be uploaded into KBase without linking metadata. For full functionality of analysis tools using metadata, first upload and import the SampleSet. Using a file on your computer, open the Import tab within the Data Browser. Then drag & drop the amplicon matrix file and the FASTA file with consensus sequences into the Staging Area. Once the files are in your Staging Area, you can import the data into your Narrative.'},\n",
       " {'question': 'How to locate the \"Import Amplicon Matrix from TSV/FASTA File in Staging Area\" App in KBase?',\n",
       "  'answer': 'Go to the APPS panel and open the Upload category.'},\n",
       " {'question': 'How to add the \"Import Amplicon Matrix from TSV/FASTA File in Staging Area\" App to your Narrative in KBase?',\n",
       "  'answer': 'Click on the Import Amplicon App.'},\n",
       " {'question': 'What should be selected in the first section for Input Objects in KBase?',\n",
       "  'answer': 'The previously imported SamplesSet file should be selected from the dropdown menu for linking metadata.'},\n",
       " {'question': 'What parameters require inputs in KBase?',\n",
       "  'answer': 'The parameters that require inputs are the target gene, target subfragment, taxon calling method, i.e. denoising or clustering, and sequence error cutoff.'},\n",
       " {'question': 'How to input additional processing metadata in KBase?',\n",
       "  'answer': 'Additional processing metadata, i.e. primer sequences, library kit, sequencing center, the denoise or clustering methods can be input by clicking \"show advanced\" parameters.'},\n",
       " {'question': 'How to import an amplicon matrix into the Narrative in KBase?',\n",
       "  'answer': 'Fill in the Amplicon Matrix Object Name and click the \"Run\" button.'},\n",
       " {'question': 'How to view the imported AmpliconMatrix object in KBase?',\n",
       "  'answer': 'You can use the \"View Matrix as Table\" App to view your new AmpliconMatrix object.'},\n",
       " {'question': 'What is the purpose of applying the amplicon matrix to an existing SampleSet and Attribute Mapping on-system in KBase?',\n",
       "  'answer': 'It facilitates the interoperability of objects in KBase.'},\n",
       " {'question': 'What is \"chemical abundance\" in KBase?',\n",
       "  'answer': 'The term “chemical abundance” is used to represent a wide array of measurements associated with chemicals. This data type can be used to upload and store diverse types of chemical data in the system such as metabolomics (intracellular and/or extracellular) that is derived based on microbiomes/isolate organism growth experiments etc., computationally predicted compounds, or data collected on the concentration or from elemental analysis.'},\n",
       " {'question': 'What types of data can be uploaded and stored in the system under \"chemical abundance\"?',\n",
       "  'answer': 'Metabolomics (intracellular and/or extracellular) that is derived based on microbiomes/isolate organism growth experiments etc., computationally predicted compounds, or data collected on the concentration or from elemental analysis.'},\n",
       " {'question': 'What are the most popular data that is uploaded and stored in the system under \"chemical abundance\"?',\n",
       "  'answer': 'The metabolomics data derived from the samples are the most popular data that is uploaded and stored in the system.'},\n",
       " {'question': 'What are KBase Apps for metabolomics?',\n",
       "  'answer': 'KBase Apps for metabolomics include Escher mapping and additional statistical analysis of the chemical abundance attribute maps, such as PCA and clustering.'},\n",
       " {'question': 'How can a Chemical Abundance Matrix be uploaded in KBase?',\n",
       "  'answer': 'A Chemical Abundance Matrix can be uploaded from a TSV (tab-separated values) file with a .tsv or .tab file extension, or from Excel spreadsheet with a .xls extension.'},\n",
       " {'question': 'What can each Chemical Type be in KBase?',\n",
       "  'answer': 'Each Chemical Type can be either a specific compound or element, aggregate (totals), exometabolites (measurements of compounds or elements that are consumed or excreted into the medium).'},\n",
       " {'question': 'What does the system create for direct download?',\n",
       "  'answer': 'The system creates an Excel spreadsheet that can be populated with chemical abundance data.'},\n",
       " {'question': 'Is linking a SampleSet required for Chemical abundance data?',\n",
       "  'answer': 'No, linking a SampleSet is not required for Chemical abundance data.'},\n",
       " {'question': 'When does Chemical abundance data work best?',\n",
       "  'answer': 'Chemical abundance data works best and is more meaningful when linked with an existing SampleSet in the system.'},\n",
       " {'question': 'What is the minimal set of metadata in a chemical abundance matrix?',\n",
       "  'answer': 'The minimal set of metadata in a chemical abundance matrix includes an ID (unique value) field, a chemical type (aggregate, exometabolite, specific), and one or more of the following:  Compound ID (e.g; ModelSEED, KEGG, ChEBI), mass, formula, inchikey, inchi, smiles, or compound name.'},\n",
       " {'question': 'What additional metadata is encouraged to provide in a chemical abundance matrix?',\n",
       "  'answer': 'Additional metadata such as units are strongly encouraged to provide with proper information that fits your scientific use cases or be kept as ‘unknown’.'},\n",
       " {'question': 'What can be applied to the chemical abundance data if a SampleSet exists?',\n",
       "  'answer': 'If a SampleSet exists, it can be applied to the chemical abundance data. Chemical abundance data needs to be formatted to ensure Samples are correctly linked.'},\n",
       " {'question': 'Is linking to Samples required in KBase?',\n",
       "  'answer': 'No, linking to Samples is not required, but highly recommended.'},\n",
       " {'question': 'What does the Chemical Abundance Matrix Template App do?',\n",
       "  'answer': 'This App generates a spreadsheet onto which you can copy your data to ensure it links to the SampleSet when uploaded.'},\n",
       " {'question': 'How many columns are there in the default sheet when creating the chemical abundance template?',\n",
       "  'answer': 'There are 10 columns in the default sheet when creating the chemical abundance template.'},\n",
       " {'question': 'What happens when linking to using this app?',\n",
       "  'answer': 'The template will be automatically populated with Sample IDs to ensure the chemical abundance data is properly linked to corresponding Samples in the system.'},\n",
       " {'question': 'What is the ID in KBase?',\n",
       "  'answer': 'The ID is the identifier for the chemical/element/compound or peak value. This can be user-defined but must be unique for each compound.'},\n",
       " {'question': 'What is the Chemical Type in KBase?',\n",
       "  'answer': 'The Chemical Type can be Specific, Aggregate, or Exometabolite. Specific refers to specific compounds in the sample (e.g; TiO2, Tyrosine), Aggregate refers to the measurement of total concentration of individual elements (e.g; Total C, N), and Exometabolite refers to measurements of compounds that are consumed or excreted into the medium by an organism or a microbial community.'},\n",
       " {'question': 'What is the Measurement Type in KBase?',\n",
       "  'answer': 'The Measurement Type can be unknown, FTICR, Orbitrap, Quadrapole.'},\n",
       " {'question': 'What are the possible Units in KBase?',\n",
       "  'answer': 'The possible Units can be mg/Kg, g/Kg, mg/L, mg/g DW, mM (millimolar), M (molar), % (percentage), Total Weight  %, unknown.'},\n",
       " {'question': 'What is the Chemical Ontology Class in KBase?',\n",
       "  'answer': 'The Chemical Ontology Class is used for grouping chemicals by their functional groups (e.g; aromatics) or could use according to ontologies defined in public databases (e.g; ChEBI).'},\n",
       " {'question': 'What is an identifier in the context of KBase?',\n",
       "  'answer': 'An identifier is an optional abbreviation that you use for the compound and or element.'},\n",
       " {'question': 'Can you give an example of an identifier?',\n",
       "  'answer': 'Yes, for example, Menaquinone can be abbreviated as MK-8, and Fatty acid as Iso-C16:1.'},\n",
       " {'question': 'Is it mandatory to use an identifier in KBase?',\n",
       "  'answer': \"No, it's optional.\"},\n",
       " {'question': 'What type of chemical data can be included in KBase?',\n",
       "  'answer': 'The chemical data that can be included in KBase are aggregate M/Z, compound name, predicted formula, and more.'},\n",
       " {'question': 'What should I do if I want to upload data to KBase?',\n",
       "  'answer': 'You should select the chemical data to include, such as aggregate M/Z, compound name, predicted formula, and more, depending on what data you have for upload.'},\n",
       " {'question': 'Can I upload predicted formula data to KBase?',\n",
       "  'answer': 'Yes, you can upload predicted formula data to KBase.'},\n",
       " {'question': 'What does M/Z represent?',\n",
       "  'answer': 'M/Z represents the ratio of mass (M) divided by net charge (Z) of the chemical.'},\n",
       " {'question': 'What is Predicted Formula?',\n",
       "  'answer': 'Formula of the compound (if available). The formula could be a predicted formula based on techniques such as Fourier-transform ion cyclotron resonance (FTICR) mass spectrometry, computationally predicted formula based on molecular weight, derived from instruments based on known protocols, or based on the chemical structure determination from NMR.'},\n",
       " {'question': 'What is Predicted Structure (smiles)?',\n",
       "  'answer': 'Predicted structure using Simplified Molecular Input Line Entry System (SMILES) notation'},\n",
       " {'question': 'What is Predicted Structure (inchi-key)?',\n",
       "  'answer': 'Predicted structure using International Chemical Identifier (InChi) key notation.'},\n",
       " {'question': 'What is Theoretical Mass?',\n",
       "  'answer': 'Theoretical molecular mass of the compound'},\n",
       " {'question': 'What is Retention Time?',\n",
       "  'answer': 'The retention time of the chemical; not validated, so users can include units or not depending on needs'},\n",
       " {'question': 'What is Polarity?', 'answer': 'Polarity of the chemical.'},\n",
       " {'question': 'What are the standard Chemical IDs that can be included in the template?',\n",
       "  'answer': 'The standard Chemical IDs can be selected from the KEGG, ChEBI, and ModelSEED databases.'},\n",
       " {'question': 'Is it necessary to include a chemical ID when creating a chemical abundance sheet from scratch?',\n",
       "  'answer': \"It's not necessary, but it is recommended to include at least one form of standard Chemical IDs in order to compare similar chemical once uploaded.\"},\n",
       " {'question': 'How can I upload and import my metadata to KBase?',\n",
       "  'answer': 'You can upload your metadata to KBase using the Upload and Import button in the Data Browser. After uploading, you can import your metadata into a Narrative by clicking on the Import button next to the uploaded file in the Data Browser.'},\n",
       " {'question': 'What is the benefit of including a chemical ID in my data?',\n",
       "  'answer': 'If you plan to use your data with metabolic modeling analysis pipelines, having at least one type of identifier is highly encouraged as the compound identifiers will be used to map compounds in metabolic models.'},\n",
       " {'question': 'How can I upload my data to the Chemical Abundance Matrix template?',\n",
       "  'answer': 'You can upload it using the Import Chemical Abundance Matrix from CSV/Excel/TSV File in Staging Area App. Open the Import tab within the Data Browser and then drag & drop the chemical abundance matrix file into the Staging Area.'},\n",
       " {'question': 'How can I import the data into my Narrative?',\n",
       "  'answer': 'Once the matrix is in your Staging Area, you can import the data into your Narrative. In the first section for Input Objects, select the previously imported SamplesSet file from the dropdown menu for linking metadata. Under Parameters, select the Chemical Abundance Matrix using the FilePath dropdown. Fill in the Matrix Object Name and click the \"Run\" button to add the metabolomics data to the Narrative.'},\n",
       " {'question': 'What can I do after uploading my chemical abundance data?',\n",
       "  'answer': \"Once you've uploaded your chemical abundance data, you can explore apps using that type of data or use one of our use case demonstration Narratives.\"},\n",
       " {'question': 'What is the last modification date of the document?',\n",
       "  'answer': '1 year ago'},\n",
       " {'question': 'What is the next topic after the chemical abundance section?',\n",
       "  'answer': 'SampleSet'},\n",
       " {'question': 'What is the previous topic before the chemical abundance section?',\n",
       "  'answer': 'Amplicon Matrix'},\n",
       " {'question': 'What is a Sample in KBase?',\n",
       "  'answer': 'A Sample in KBase represents physical material from an experiment with attributes or metadata (environment, location, temperature, etc). It has a unique ID internal to KBase, aliases, related Samples, and structured metadata.'},\n",
       " {'question': 'What is a SampleSet in KBase?',\n",
       "  'answer': 'Samples occur within the Narrative as a SampleSet, and multiple SampleSets may share the same Sample.'},\n",
       " {'question': 'How can I format SampleSets in KBase?',\n",
       "  'answer': 'The SESAR (System for Earth SAmple Registration) format is one format that can be used in KBase for uploading Samples. Their samples use a controlled vocabulary for fields of data and metadata that can be used in KBase.'},\n",
       " {'question': 'How can I participate in beta testing for KBase Apps?',\n",
       "  'answer': 'You can contact KBase at [email\\xa0protected] to express your interest in beta testing their Apps.'},\n",
       " {'question': 'How can I import samples from IGSN SESAR to KBase?',\n",
       "  'answer': 'You can use the \"Import Samples from IGSN\" App in KBase, which takes SESAR number(s) separated by commas as input and imports the sample information directly from SESAR.'},\n",
       " {'question': 'How can I import samples from NCBI BioSamples to KBase?',\n",
       "  'answer': 'You can use the \"Import Samples from NCBI\" App in KBase, which takes the NCBI BioSample accession number(s), again separated by commas when importing multiple samples into the same SampleSet.'},\n",
       " {'question': 'What are the formats of the downloadable SESAR template files?',\n",
       "  'answer': 'The downloadable SESAR template files are available in .xlsx and .csv formats.'},\n",
       " {'question': 'What is ENIGMA?',\n",
       "  'answer': 'The Ecosystems & Networks Integrated with Genes & Molecular Assemblies (ENIGMA) Science Focus Area (SFA) partners with KBase for new functionality development.'},\n",
       " {'question': 'What is the ENIGMA metadata template used for?',\n",
       "  'answer': 'The ENIGMA metadata template is available as a template for the \"Import Samples\" App in KBase.'},\n",
       " {'question': 'How can new Samples be added in KBase?',\n",
       "  'answer': 'New Samples can either be added to an existing SampleSet or used to create a new SampleSet. To add Samples to an existing SampleSet, choose the appropriate SampleSet from the drop-down in the importer. To create a new set, leave it blank and a new SampleSet will be created from the newly imported Samples.'},\n",
       " {'question': 'How can samples be imported from IGSN or NCBI?',\n",
       "  'answer': 'Samples can be directly imported from IGSN or NCBI using the \"Import Samples from IGSN\" or \"Import from NCBI\" Apps, respectively, searchable in the Upload section of the Apps panel.'},\n",
       " {'question': 'What happens when samples are imported from IGSN or NCBI?',\n",
       "  'answer': 'Samples imported using this method can be added to an existing SampleSet through selecting it from the SampleSet drop-down or can be used to create a new SampleSet.'},\n",
       " {'question': 'How are samples uploaded from a local source?',\n",
       "  'answer': 'Samples data is uploaded to the Staging Area and then imported into a Narrative. When importing new Samples, the default action is to create a new SampleSet.'},\n",
       " {'question': 'What is the requirement for individual samples?',\n",
       "  'answer': 'Individual samples must belong to a SampleSet, even if that SampleSet contains only one Sample.'},\n",
       " {'question': 'In what formats must the samples be?',\n",
       "  'answer': 'Samples must be in SESAR or ENIGMA formats.'},\n",
       " {'question': 'What happens if the data uploaded to KBase does not match format requirements?',\n",
       "  'answer': 'If data does not match format requirements, the upload may result in an error. Critical errors, such as unrecognized units for a measurement, will prevent upload entirely. Minor errors, such as an unrecognized controlled vocabulary term, can be ignored by selecting the \"Ignore Warnings\" checkbox in the importer.'},\n",
       " {'question': 'How can I view the data that has been imported to KBase?',\n",
       "  'answer': 'Once imported, the SampleSet viewer widget allows you to see the Samples data that has been imported. This includes all metadata, such as collection information, and can be selectively viewed using the search bar.'},\n",
       " {'question': 'What is the recommendation for uploading non-validated data to KBase?',\n",
       "  'answer': 'Non-validated data can be uploaded as user terms, but this will not allow comparisons across Samples. It is recommended to change the column headers to correspond with the relevant SESAR or ENIGMA headers, and ensure the data matches.'},\n",
       " {'question': 'Where can more details for the SampleSets be found?',\n",
       "  'answer': 'More details for the SampleSets can be found on the landing page.'},\n",
       " {'question': 'How are permissions managed in KBase?',\n",
       "  'answer': 'Permissions are managed separately from the Narrative workspace with some exceptions. The permissions for a Sample can be viewed in the Sample landing page under the Access tab.'},\n",
       " {'question': 'What happens when a SampleSet is shared with a user?',\n",
       "  'answer': 'When a SampleSet is shared with a user, that user automatically has read access to all Samples included in the SampleSet. Read access will automatically be given with the first access to the SampleSet by visiting its landing page or viewing it in the Narrative.'},\n",
       " {'question': 'How to grant a user increased permissions?',\n",
       "  'answer': 'To grant a user increased permissions (e.g. to modify Samples), you should use the Update SampleSet Access Controls App in the Narrative.'},\n",
       " {'question': 'How to change user permissions for specific SampleSets?',\n",
       "  'answer': 'Use the Update SampleSet Access Controls App to change user permissions for specific SampleSets.'},\n",
       " {'question': 'Can permissions be granted for individual samples in the SampleSet?',\n",
       "  'answer': 'No, permissions will be granted for all Samples in the SampleSet. There is not currently a mechanism to change the permissions on a single app and there is not a method to easily revoke permissions.'},\n",
       " {'question': 'How can I update the attributes of the Samples in a SampleSet?',\n",
       "  'answer': 'You will need to export a CSV file of the Samples from a SampleSet, modify the columns and/or their values, then re-import the Samples CSV using the Samples Importer App to edit Sample attributes.'},\n",
       " {'question': 'Will updating the attributes of the Samples in a SampleSet modify existing data or analysis?',\n",
       "  'answer': 'No, this will not modify existing data or analysis that has already been run. To update the analysis results, you will need to re-run the analysis apps using the new SampleSet with your recently updated Samples.'},\n",
       " {'question': 'How can subsets and supersets be managed in KBase?',\n",
       "  'answer': 'The most efficient way manage subsets of Samples is to upload separate SampleSets for each subset, and a complete SampleSet to represent the parent set. A superset can be created by downloading SampleSets, combining them, and then uploading the combined superset.'},\n",
       " {'question': 'What is the purpose of the Samples Importer Apps?',\n",
       "  'answer': 'The Samples Importer Apps are designed to encourage the use of validated data.'},\n",
       " {'question': 'How many types of error reporting are there in KBase?',\n",
       "  'answer': 'There are two types of error reporting in KBase, red and yellow.'},\n",
       " {'question': 'What do red errors indicate in KBase?',\n",
       "  'answer': 'Red errors indicate the validation failed to recognize a value as a validated field. This prevents the importer from creating a SampleSet for any cell with values that do not comply with controlled vocabulary.'},\n",
       " {'question': 'What do yellow errors indicate in KBase?',\n",
       "  'answer': 'Yellow errors indicate a column could not be validated as a controlled field. These result in warnings, but the SampleSet can still be created.'},\n",
       " {'question': 'What does a \"successful\" job that does not create an object in the Data panel mean?',\n",
       "  'answer': 'A successful job run indicates that the sample file was read, but the contents may or may not be valid.'},\n",
       " {'question': 'What happens if a cell in a validated column does not comply with the controlled vocabulary?',\n",
       "  'answer': 'The cell will fail to upload.'},\n",
       " {'question': 'What should you do if your value is not part of the controlled vocabulary and there is not a suitable alternative?',\n",
       "  'answer': 'You can change the column header to a user defined field.'},\n",
       " {'question': 'How can you add your term to the controlled vocabulary?',\n",
       "  'answer': 'You can contact KBase to have your term added to the controlled vocabulary.'},\n",
       " {'question': 'What happens if I accept an unvalidated field warning in KBase?',\n",
       "  'answer': 'If you accept an unvalidated field, it will be stored as a user-defined field. This allows you to include the information, but does not allow comparisons across Samples.'},\n",
       " {'question': 'What should I do if the information in the unvalidated field is critical to Samples which I plan to use in analysis?',\n",
       "  'answer': 'If the information is critical to Samples which you plan to use in analysis, it is recommended to manually fix the values to comply with controlled vocabulary terms.'},\n",
       " {'question': 'Can I accept warnings after the import has run in KBase?',\n",
       "  'answer': 'No, there is no mechanism to accept warnings after the import has run. You can run the Importer App with the \"Ignore Warnings\" option unchecked. This allows you to see the errors and warnings and run does not create a data object. Then proceed to edit your spreadsheet and re-run the App with the warnings fixed, or run the App with the option checked and delete the resulting object if you find warnings you want to fix.'},\n",
       " {'question': 'What was the last modification date of the document?',\n",
       "  'answer': 'The document was last modified 1 year ago.'},\n",
       " {'question': 'What is the next topic after \"Chemical Abundance Matrix\"?',\n",
       "  'answer': 'The next topic after \"Chemical Abundance Matrix\" is \"Compressed/Zipped Files\".'},\n",
       " {'question': 'What is the previous topic before \"Compressed/Zipped Files\"?',\n",
       "  'answer': 'The previous topic before \"Compressed/Zipped Files\" is \"Chemical Abundance Matrix\".'},\n",
       " {'question': 'What compressed file formats are accepted by KBase?',\n",
       "  'answer': 'KBase accepts compressed files in these formats: .zip, .gz, .bz, .bzip2, .bz2, .tar, .tgz.'},\n",
       " {'question': 'How to upload and import compressed files to KBase?',\n",
       "  'answer': 'To add compressed data files from your computer to the KBase Narrative, follow the three-step process: 1. Drag & drop the data file(s) from your computer to the new Import tab to upload them to your Staging Area. 2. Choose a format for importing data from your Staging Area into your Narrative. 3. Run the Import App that is created.'},\n",
       " {'question': 'How to uncompress the compressed files in KBase?',\n",
       "  'answer': 'When files have been uploaded to the Staging Area, compressed files have an outward-facing double arrow icon to the left of the file name. Click the uncompress icon to the left of the file name to uncompress the compressed/zip uploaded files before clicking the import icon. To ensure this works, the file extension must match the type of packing/compression.'},\n",
       " {'question': 'Can importers accept compressed files in KBase?',\n",
       "  'answer': 'Yes, importers will accept compressed files, but not archives, as imports.'},\n",
       " {'question': 'Can I import a series of files in an archive in KBase?',\n",
       "  'answer': 'No, if you have a series of files in an archive, the archive will first have to be unpacked and then the files can be imported.'},\n",
       " {'question': 'Does this rule apply to both single and bulk imports in KBase?',\n",
       "  'answer': 'Yes, this rule applies to both single and bulk imports.'},\n",
       " {'question': 'What is the Bulk Import Specification in KBase?',\n",
       "  'answer': 'The Bulk Import Specification in KBase is a feature that allows users to import multiple files at once through bulk upload using an import specification file with the extension .csv, .tsv, or .xls.'},\n",
       " {'question': 'What types of files can be used for bulk upload in KBase?',\n",
       "  'answer': 'A TSV (tab-separated values) file with a .tsv or .tab file extension, a CSV (comma-separated values) file with a .csv extension, or an Excel spreadsheet with a .xls extension can be used to upload multiple files at once.'},\n",
       " {'question': 'How does the Import Specification file work in KBase?',\n",
       "  'answer': 'The Import Specification file is treated as a directory or manifest containing the upload information for multiple files to fill out an import cell. The TSV, CSV, or Excel file must be formatted exactly to import correctly.'},\n",
       " {'question': 'What is the file path in the context of the Staging Area in KBase?',\n",
       "  'answer': 'The file path is relative to your home directory of the Staging Area. If the files are at the top level, you only need to include the file name. If they are within a folder, the file input fields should include the folder name and the file name, like \"folder1/file1\", \"folder1/file2\", etc.'},\n",
       " {'question': 'What is the Object name in KBase?',\n",
       "  'answer': 'The Object name is a name designated for the imported file, which will be used as the input for analysis Apps.'},\n",
       " {'question': 'How to use the bulk Importer App in KBase?',\n",
       "  'answer': 'You can use the bulk Importer App with a subset of files to generate an Import Specification template. You can then download and modify the template locally to fill out file paths and parameters. After that, you can upload the Import Specification file to import all files at once.'},\n",
       " {'question': 'How to create an import specification template from the Narrative in KBase?',\n",
       "  'answer': 'To create an import specification template CSV, TSV or Excel file for many files from the Narrative, you can begin with the Bulk Import directions using a subset of the files or file types to import.'},\n",
       " {'question': 'How can I upload files to KBase?',\n",
       "  'answer': 'You can drag and drop all files or a folder containing the files to upload to Staging.'},\n",
       " {'question': 'How can I import files from the Staging Area in KBase?',\n",
       "  'answer': 'Once the files appear in the Staging Area, select at least one that represents each file type you wish to import, select the Import As file type and make sure the check box is active, then click \"Import Selected.\" This will create a bulk Importer App in the Narrative. Fill out the parameters for the import. Then click the \"Create Import Template\" button.'},\n",
       " {'question': 'How can I create an import template in KBase?',\n",
       "  'answer': 'After clicking the \"Create Import Template\" button, a pop-up file will prompt the template creation. Select the import type(s), if multiple use key controls (such as command-hold and select with Mac or control-shift and select on Microsoft). Choose the output type as either Comma-separated (CSV), Tab-separated (TSV), or Excel (XLS). Then select or rename the output destination within the Staging Area. Click \"Generate template!\" to create the templates selected.'},\n",
       " {'question': 'What happens if you have an Import Specification Template within the destination that you designate in the pop-up?',\n",
       "  'answer': 'It will be overwritten with the newly created template.'},\n",
       " {'question': 'How can you import multiple data types at once?',\n",
       "  'answer': 'Either create a CSV/TSV for each type, or a tab for each type in a single Excel file.'},\n",
       " {'question': 'What should you do if you do not see the folder or template in the destination designated in the previous step?',\n",
       "  'answer': 'Navigate back to the Staging Area and click \"Refresh\".'},\n",
       " {'question': 'How can you download the Import Specification template?',\n",
       "  'answer': 'Download the Import Specification template using the download button on the left of the trash icon.'},\n",
       " {'question': 'How to start with the Import Specification template in KBase?',\n",
       "  'answer': 'Open the Import Specification template. In the template files 3 rows are already filled; these are required for the staging service to parse the file. The third row displays the column headers for inputs, outputs, and parameters included in the Importer App. The following rows are for each of the files to import, the files selected before will be included here.'},\n",
       " {'question': 'What does each row in the Import Specification template represent?',\n",
       "  'answer': 'Each row corresponds to one data object to import. Some data objects will include more than one file path.'},\n",
       " {'question': 'What should be done if a GFF Genome is to be imported?',\n",
       "  'answer': 'A GFF Genome requires both GFF and FASTA files. All text fields should match exactly to the display text in the Importer. Use the pre-filled parameters to fill out the rest of the rows with designated file paths.'},\n",
       " {'question': 'How to continue filling out the spreadsheet in the Import Specification template?',\n",
       "  'answer': 'Continue filling out the spreadsheet as if you would the Importer App with the remaining files to import. Use the pre-filled parameters to fill out the rest of the rows with designated file paths.'},\n",
       " {'question': 'How to manually fill in the Scientific name into a template in KBase?',\n",
       "  'answer': \"Navigate to an Import cell with a GenBank Genome Data type. Under Parameters, go to the 'Scientific name', search for the genus/species name you want and select it. Click the copy button on the right. You can now paste the name into the template.\"},\n",
       " {'question': 'How to import a bulk Import Specification file in KBase?',\n",
       "  'answer': 'Open the Import tab within the Data Browser. Then drag & drop the Import Specification file and the files to be imported if they are not already in the Staging Area.'},\n",
       " {'question': 'How to import a CSV file in KBase?',\n",
       "  'answer': 'Once the file appears in the Staging Area, select \"\" from the dropdown in the column. Make sure with a filled checkbox (do not select the independent files). Click at the bottom right corner.'},\n",
       " {'question': 'How to import files in KBase?',\n",
       "  'answer': 'Once the Import Specification file has been read, it will create a bulk Importer App in your Narrative and you can proceed as usual following the bulk import directions. Verify the file names and import settings within the App. Click \"Run\" on the upper lefthand corner of the App to import the files from the Import Specification into the Narrative.'},\n",
       " {'question': 'What happens after the Importer App has run?',\n",
       "  'answer': 'Once the Importer App has run, the App will signal that the files have been imported or any errors that occurred across individual jobs.'},\n",
       " {'question': 'What is the current limit of rows or data objects per import in KBase?',\n",
       "  'answer': 'Currently, 500 rows or data objects per import is supported with good performance. The current limit is 10000 rows.'},\n",
       " {'question': 'What happens if you upload an invalid file in KBase?',\n",
       "  'answer': 'If you upload an invalid file, you will receive an error that lists the problem encountered. For errors that can be fixed through the interface, the Importer cell will be created but you must correct the error before running the import.'},\n",
       " {'question': 'Can multiple sets of parameters be used per import job in the current stage of KBase development?',\n",
       "  'answer': 'No, at this stage of development only a single set of parameters can be used per import job.'},\n",
       " {'question': 'What should not be changed in the TSV or CSV file?',\n",
       "  'answer': 'The first 3 rows of the TSV or CSV file should not be changed.'},\n",
       " {'question': 'What should be done with stray cells in Excel?',\n",
       "  'answer': 'Ensure that all the stray cells outside your data block are empty. Consider deleting all the rows and columns outside your data block.'},\n",
       " {'question': 'What should be done with CSV and TSV files?',\n",
       "  'answer': 'Be careful to always include the correct number of separator characters for every row, even if some values are blank. The importer is picky about this to help the user avoid off-by-one (or more) errors in their data.'},\n",
       " {'question': 'How many files or Excel tabs are allowed per data type in KBase?',\n",
       "  'answer': 'Only one file or Excel tab is allowed per data type.'},\n",
       " {'question': 'How is whitespace treated in CSV and TSV files in KBase?',\n",
       "  'answer': 'In CSV and TSV files, white space around non-numerical data is ignored, and whitespace prior to numerical data is ignored.'},\n",
       " {'question': 'How can I include text that contains the separator character in CSV and TSV files in KBase?',\n",
       "  'answer': 'In CSV and TSV files, you can use quotes (\") to surround text that contains the separator character (respectively a comma or a tab).'},\n",
       " {'question': 'What issue might occur if individual files in an import specification are selected along with the import specification in KBase?',\n",
       "  'answer': 'A bug will occur where extra rows that cannot be filled in are added to the bulk import cell for the individual files.'},\n",
       " {'question': 'What is the workaround if the scientific name lookup ignores spaces in KBase?',\n",
       "  'answer': 'The workaround is to create (and optionally delete) a new bulk import cell for the scientific name input.'},\n",
       " {'question': 'What is the last modified date for the \"Downloading Data\" section?',\n",
       "  'answer': '1 month ago.'},\n",
       " {'question': 'How can I download data from KBase?',\n",
       "  'answer': 'You can download data from KBase by following these steps: \\n    1. Locate the data you wish to download in the Data Panel under the Analyze tab\\n    2. Hover your mouse over the data object until a “…” icon appears to the right of its name\\n    3. Click the “…” icon to reveal advanced options\\n    4. Click the seventh icon from the left \"Export / Download Data\"\\n    5. Select a format for exporting and downloading the data\\n    The data will download to the default downloads folder on your computer.'},\n",
       " {'question': 'Where will the downloaded data from KBase be stored on my computer?',\n",
       "  'answer': 'The data will download to the default downloads folder on your computer.'},\n",
       " {'question': 'Can I select a format for exporting and downloading the data from KBase?',\n",
       "  'answer': 'Yes, you can select a format for exporting and downloading the data from KBase.'},\n",
       " {'question': 'How is data downloaded from KBase?',\n",
       "  'answer': 'The data will be compressed into a Zip file (.zip) containing files (or a directory containing files) in the format you selected and a metadata file (in JSON format).'},\n",
       " {'question': 'Can I directly import the Zip file I downloaded from KBase?',\n",
       "  'answer': 'No, you cannot directly import the Zip file – you must first extract the file and then re-upload using the specified uploader for the data type.'},\n",
       " {'question': 'How can I download files from App Cells in KBase?',\n",
       "  'answer': 'A few KBase Apps have file links as part of the output at the bottom of the App Cell. Right click on the file name to download and choose “Save Link As\" to download and save to your local computer.'},\n",
       " {'question': 'How can I download large data objects from KBase?',\n",
       "  'answer': 'Large data objects (over 10 gigabases) include many Reads, Assemblies, Genomes, and Alignment (BAM) objects. They encounter many of the same obstacles as uploading large data sets. Just like uploading, this is a 2-stage process; copy the file to the Staging Area and download from Staging to your local computer with Globus.'},\n",
       " {'question': 'How to copy a file to the Staging Area in KBase?',\n",
       "  'answer': '1. Locate the data you wish to download in the Data Panel under the Analyze tab\\n    2. Hover your mouse over the data object until a “…” icon appears to the right of its name\\n    3. Click the “…” icon to reveal advanced options\\n    4. Click the seventh icon from the left \"Export/Download Data\"\\n    5. Select \"Staging\"\\n    6. This will open “Export Data Object to Staging Area” App.\\n    7. The name of the Input Object is filled in. A default Destination Directory is filled in, but can be edited and changed. A directory will be created in the Staging Area, and all created files will be copied to the directory. The \"show advanced\" Parameters may have more options relevant to your data object.\\n    8. Once in the Staging Area, large files (>10GB) should be transferred to your local machine using Globus. Instead of “To” the “KBase Bulk Share” endpoint, this will be a transfer “From” the “KBase Bulk Share” endpoint.'},\n",
       " {'question': 'What is the Staging Area in KBase?',\n",
       "  'answer': 'The Staging Area in KBase is a place where data files are stored temporarily before they are transferred to your local machine.'},\n",
       " {'question': 'How to transfer large files from the Staging Area in KBase?',\n",
       "  'answer': 'Large files (>10GB) should be transferred to your local machine using Globus. Instead of “To” the “KBase Bulk Share” endpoint, this will be a transfer “From” the “KBase Bulk Share” endpoint.'},\n",
       " {'question': 'How to export data object to staging area in KBase?',\n",
       "  'answer': 'In the App Panel, search for the app “Export Data Object to Staging Area,” put in the name of your Input Object for the data object to download, and follow the prior instructions.'},\n",
       " {'question': 'What is the name of the app to export data object to staging area in KBase?',\n",
       "  'answer': 'The name of the app is “Export Data Object to Staging Area”.'},\n",
       " {'question': 'What do you need to put in for the data object to download in KBase?',\n",
       "  'answer': 'You need to put in the name of your Input Object for the data object to download in KBase.'},\n",
       " {'question': 'How can I add data to a Narrative in KBase?',\n",
       "  'answer': 'You can add data to a Narrative through the Data Browser which allows you to search data in KBase or import data from your computer or a database. You can also add data by navigating to the Data Browser tab and hover over the data object to reveal the import button. Click on the import button to import the data object into the Narrative.'},\n",
       " {'question': \"What does 'My Data' show in KBase?\",\n",
       "  'answer': \"'My Data' shows your data objects available within Narratives.\"},\n",
       " {'question': \"What does 'Shared With Me' include in KBase?\",\n",
       "  'answer': \"'Shared With Me' includes data associated to Narratives shared with you.\"},\n",
       " {'question': \"What does 'Public' display in KBase?\",\n",
       "  'answer': \"'Public' displays Public datasets.\"},\n",
       " {'question': \"What does 'Example' show in KBase?\",\n",
       "  'answer': \"'Example' shows datasets that have been pre-loaded by the KBase team.\"},\n",
       " {'question': \"What does 'Import' allow in KBase?\",\n",
       "  'answer': \"'Import' allows you to upload your own datasets or external datasets for analysis.\"},\n",
       " {'question': 'What can you learn from the video tutorial in KBase?',\n",
       "  'answer': 'You can learn about filtering, managing, and viewing data in KBase.'},\n",
       " {'question': 'What was the previous topic before \"Filtering, Managing, and Viewing Data\"?',\n",
       "  'answer': 'The previous topic was \"Searching, Adding, and Uploading Data\".'},\n",
       " {'question': 'What is the next topic after \"Filtering, Managing, and Viewing Data\"?',\n",
       "  'answer': 'The next topic is \"Linking Metadata\".'},\n",
       " {'question': 'What is a Sample in KBase?',\n",
       "  'answer': 'A Sample in KBase represents measured material from an experiment. It has a unique ID internal to KBase, aliases, related Samples, and structured metadata. Samples occur within a Narrative as a SampleSet, but the same Sample may be shared by multiple SampleSets.'},\n",
       " {'question': 'What is the purpose of Sample aliases in KBase?',\n",
       "  'answer': 'Sample aliases allow the same data uploaded by different users to be linked and connect different analyses of the same data. These aliases can also be used to link Samples in KBase with other methods of sample registration, such as ESS-DIVE or SESAR.'},\n",
       " {'question': 'How can I beta test the Apps for linking environmental sampling metadata with experimental data in KBase?',\n",
       "  'answer': 'You can contact KBase at [email\\xa0protected] to express your interest in beta testing these Apps.'},\n",
       " {'question': 'How can a SampleSet be used in KBase?',\n",
       "  'answer': 'Once a SampleSet is in system, it can be used when uploading amplicon or chemical abundance matrices into KBase. This allows the amplicon and chemical abundance analysis to remain linked to the sample metadata and enrich analysis.'},\n",
       " {'question': 'How to link additional existing data to samples in KBase?',\n",
       "  'answer': 'Additional existing data can be linked to samples. The app \"Link Workspace Objects to Samples\" provides a graphical interface for creating the links. In this app, you can select the SampleSet to create links from, then choose the Sample from the set and object from the Narrative. You can add multiple Samples and objects by clicking the + button.'},\n",
       " {'question': 'What should be selected when uploading and importing either an amplicon matrix or chemical abundance matrix?',\n",
       "  'answer': 'When uploading and importing either an amplicon matrix or chemical abundance matrix, select the corresponding SampleSet from the dropdown menu for the input before running the Import App.'},\n",
       " {'question': 'What should I use if I have a large number of links to create in KBase?',\n",
       "  'answer': 'You should use the \"Batch Link Workspace Objects to Samples\" app.'},\n",
       " {'question': 'What is the \"Batch Link Workspace Objects to Samples\" app used for?',\n",
       "  'answer': 'The \"Batch Link Workspace Objects to Samples\" app is used for creating a large number of links in KBase.'},\n",
       " {'question': 'Is it advisable to use the \"Batch Link Workspace Objects to Samples\" app for creating a large number of links?',\n",
       "  'answer': 'Yes, it is advisable to use the \"Batch Link Workspace Objects to Samples\" app for creating a large number of links in KBase.'},\n",
       " {'question': 'What is the benefit of using validated metadata terms in KBase?',\n",
       "  'answer': 'Using validated metadata terms allows your data to be more easily compared with other data and allows KBase to integrate data when different sources may use different conventions.'},\n",
       " {'question': 'What is the challenge in integrating metadata from different sources?',\n",
       "  'answer': 'The challenge is that many standard terms, such as NCBI BioSamples and IGSN SESAR, aim to describe sample metadata. These standards often overlap and may differ significantly depending on the focus of the organization or institution that manages the samples. At the same time, many labs and smaller institutions have their own templates and organization of their data.'},\n",
       " {'question': 'What is the solution to integrate similar metadata from different sources?',\n",
       "  'answer': 'The solution is to use a controlled vocabulary that can be identified in KBase so your metadata is functional. This allows for the integration of similar metadata from different sources and accounts for the differences between the sources.'},\n",
       " {'question': 'What happens when samples are uploaded into KBase via a spreadsheet?',\n",
       "  'answer': 'They are transformed into an internal representation. Columns within the spreadsheet are mapped and transformed based on the template used. Columns that correspond to recognized terms are validated to ensure they properly formatted.'},\n",
       " {'question': 'What is checked during the validation of recognized terms in KBase?',\n",
       "  'answer': 'The validation checks that the value is the proper type (e.g. string versus a number), in the correct range, match an enumerated list, or appear in an ontology.'},\n",
       " {'question': 'Why are controlled terms useful in KBase?',\n",
       "  'answer': 'Controlled terms are useful both because they undergo this validation and they provide a more precise meaning for the value and comparisons accounting for units.'},\n",
       " {'question': 'What happens when the uploader encounters terms it doesn’t recognize?',\n",
       "  'answer': 'Those terms and values will be stored in a user section of the samples. These values can be used in analysis within that data set and they can provide contextual information that the original uploader understands. However, unrecognized terms can not reliably be compared across SampleSets and other samples in the system.'},\n",
       " {'question': 'What is the benefit of using templates and controlled terms?',\n",
       "  'answer': 'Using templates and controlled terms clarifies the exact meaning of a term and enforces additional validation.'},\n",
       " {'question': 'How can I suggest new terms to be added to KBase?',\n",
       "  'answer': 'You can contact KBase at [email\\xa0protected] to suggest additions.'},\n",
       " {'question': 'How can I add new Sample Attribute columns to KBase?',\n",
       "  'answer': \"You can request the terms to be added to KBase's vocabulary if they are represented as validated terms in SESAR, ENIGMA, or KBase formats.\"},\n",
       " {'question': 'What does KBase review when new attributes are proposed?',\n",
       "  'answer': 'KBase reviews the overlap with current supported attributes, existing or relevant ontological frameworks, and the presence of and interoperability of units for measured data.'},\n",
       " {'question': 'How can I determine if my term is valid?',\n",
       "  'answer': 'The easiest way to determine if your term is valid is to search the ENVO Database.'},\n",
       " {'question': 'What are the commonly used units for length in KBase?',\n",
       "  'answer': 'The commonly used units for length are meter (m), micron/micrometer (µm or um), millimeter (mm), centimeter (cm), and kilometer (km).'},\n",
       " {'question': 'What are the commonly used units for time in KBase?',\n",
       "  'answer': 'The commonly used units for time are second (s), minute (min), hour (hr), day (d), and year (yr or a).'},\n",
       " {'question': 'What are the commonly used units for mass in KBase?',\n",
       "  'answer': 'The commonly used units for mass are gram (g), microgram (µg), milligram (mg), and kilogram (kg).'},\n",
       " {'question': 'What is the last modification date of the \"Public Data in KBase\" section?',\n",
       "  'answer': '1 year ago.'},\n",
       " {'question': 'What type of data does KBase provide?',\n",
       "  'answer': 'KBase provides an extensive set of prokaryotic annotated genomes; a selected set of eukaryotic genomes; and thousands of biochemical compounds and reactions.'},\n",
       " {'question': 'How can I search public data in KBase?',\n",
       "  'answer': 'You can search public data in KBase using the Public tab in the Data Browser within the Narrative Interface.'},\n",
       " {'question': 'How can a JGI user transfer data to KBase?',\n",
       "  'answer': 'A JGI user can transfer public genome reads and assemblies (as well as your private data and annotated genomes) from JGI to your KBase account using specific instructions provided by KBase.'},\n",
       " {'question': 'What is Globus?',\n",
       "  'answer': 'Globus is a data management and file transfer system that can facilitate bulk transfer of data (large data files or a large number of files) between two endpoints.'},\n",
       " {'question': 'How to transfer large files to KBase?',\n",
       "  'answer': 'You can transfer large files to KBase using Globus. You will need a Globus account and your local computer can become a Globus endpoint with a Globus Connect Personal Endpoint. Once you have a Globus account, link to your KBase account to facilitate file transfers.'},\n",
       " {'question': 'What is the KBase endpoint in Globus?',\n",
       "  'answer': 'The KBase endpoint in Globus is called “KBase Bulk Share.”'},\n",
       " {'question': 'How to link my Globus account to my KBase account?',\n",
       "  'answer': 'Once you have a Globus account, you can link it to your KBase account to facilitate file transfers. You can find further guidance in the Linking Accounts section.'},\n",
       " {'question': 'How to start a Globus Data Transfer in KBase?',\n",
       "  'answer': \"To initiate a transfer using Globus, click 'Or upload to this staging area by using Globus Online' in the tab of the below the Staging Area. This brings you to a page where you can set the right panel to “KBase Bulk Share,” which directs to the Staging Area. Set the left panel to the location of the data you want to transfer.\"},\n",
       " {'question': 'What is the function of the \"KBase Bulk Share\" option in Globus Data Transfer?',\n",
       "  'answer': 'The \"KBase Bulk Share\" option directs to the Staging Area in KBase.'},\n",
       " {'question': \"What should I do if I don't see a directory with my name on the KBase endpoint during a Globus Data Transfer?\",\n",
       "  'answer': 'If you don’t see a directory with your name as in the example above, add /username to the Path for the KBase endpoint (/user/ is used in the example above).'},\n",
       " {'question': 'Can I create new directories in my Staging Area for my data transfer in KBase?',\n",
       "  'answer': 'Yes, Globus can be used to create new directories in your Staging Area for your data transfer and to delete staging files when you are done importing into KBase.'},\n",
       " {'question': 'How can I transfer data files or folders in KBase?',\n",
       "  'answer': 'You can drag data files or folders from the left endpoint to the right one, inside the Globus interface, and the files will appear in your KBase Staging Area.'},\n",
       " {'question': 'What happens after I submit my transfer request to Globus?',\n",
       "  'answer': 'Your transfer request will be submitted to Globus. If there are network or other problems, Globus will retry several times before giving up. Globus will send an email when the transfer is complete. In the KBase Staging Area, you may need to click on the refresh icon (two encircling arrows to the left of the user name) to update the Import tab.'},\n",
       " {'question': 'What could be the cause if I see a message that ends with “Details: 500 Command failed…”?',\n",
       "  'answer': 'A firewall may be the cause. Some government institutions have a firewall between users and the outside world. In an effort to increase security, many web links can be blocked, including Globus traffic.'},\n",
       " {'question': 'What is the common cause for \"No ACL rules\" error on the KBase Bulk Share endpoint in the Globus UI?',\n",
       "  'answer': 'The most common cause for this error is multiple Globus accounts where the wrong account is linked to KBase.'},\n",
       " {'question': 'How to fix \"No ACL rules\" error on the KBase Bulk Share endpoint in the Globus UI?',\n",
       "  'answer': 'Be sure to log out of all Globus accounts, then link the correct account at https://narrative.kbase.us/#/auth2/account. If you still get this error, contact the help desk.'},\n",
       " {'question': 'Can I transfer data from JGI to my KBase account?',\n",
       "  'answer': 'Yes, if you are a JGI user, you can transfer public genome reads and assemblies (and your private data and annotated genomes) from JGI to your KBase account.'},\n",
       " {'question': 'How can JGI users transfer data to KBase?',\n",
       "  'answer': 'JGI users can transfer genome reads, assemblies, and annotated genomes to KBase’s Narrative Interface for assembly, annotation, metabolic modeling, and more.'},\n",
       " {'question': 'What are the prerequisites for transferring data from JGI to KBase?',\n",
       "  'answer': 'Before transferring data from JGI to KBase, you must:\\n1. Get a JGI account\\n2. Get a Globus account\\n3. Get a KBase account and link it to Globus\\n4. Use either Chrome, Firefox, or Safari browser'},\n",
       " {'question': \"What is the function of JGI's Globus service?\",\n",
       "  'answer': 'JGI’s Globus service lets users select data from the JGI Genome Portal’s sequencing projects and make them transferable to KBase, all in a few simple steps.'},\n",
       " {'question': 'How to transfer data from JGI to KBase?',\n",
       "  'answer': 'Starting at the JGI Genome Portal, search for the genome or project of interest and click on the download link. Click on the green \"Agree\" button on the next page to accept the data usage policy. Near the top of the download page is a link to JGI’s Globus v.2 service. This version works by 1) transferring the entire project to a JGI/Globus staging area and 2) sending you an email when it is ready. Click on the blue \"Download\" button and one of two windows will pop up: 1. A window to enter a Globus ID or register for a Globus ID OR 2. Or a window saying “There is not enough disk space in the staging area for this data.”'},\n",
       " {'question': 'What to do if there is not enough disk space in the staging area for the data?',\n",
       "  'answer': 'If you do get the message about disk space, this is an issue between JGI and Globus (it has nothing to do with KBase), and it is dependent on other users staging data for transfer. The issue should eventually resolve itself if you try again later.'},\n",
       " {'question': 'What is Globus v.2 service?',\n",
       "  'answer': 'Globus v.2 service is a service provided by JGI that works by 1) transferring the entire project to a JGI/Globus staging area and 2) sending you an email when it is ready.'},\n",
       " {'question': 'What happens after you successfully enter your Globus account name and hit the blue \"Submit\" button?',\n",
       "  'answer': 'Your data will be transferred to the JGI staging area.'},\n",
       " {'question': 'How long does JGI keep the link active for transferring data with Globus?',\n",
       "  'answer': 'JGI keeps this link active for 14 days.'},\n",
       " {'question': 'What happens after the link for transferring data with Globus expires?',\n",
       "  'answer': 'JGI will remove the data from their staging area to make room for other users.'},\n",
       " {'question': 'How to transfer the file to KBase Bulk Share?',\n",
       "  'answer': 'Click on the file name to highlight it and then click on the \"Transfer or Sync to...\" link on the right. Set the “To” endpoint (right) to “KBase Bulk Share”, which is the KBase Staging Area. If you do not see a directory with your name, add /username (i.e. /landml for the user below) to the Path for the KBase endpoint.'},\n",
       " {'question': 'What is the email address that sends the link for transferring data with Globus?',\n",
       "  'answer': 'The email is sent from [email\\xa0protected]'},\n",
       " {'question': 'How can I transfer files to my KBase Staging Area?',\n",
       "  'answer': 'You can drag data files or folders from the left endpoint to the right one, inside the Globus interface, and the files will be copied to your KBase Staging Area.'},\n",
       " {'question': 'What happens if the connection to the server is broken during transfer?',\n",
       "  'answer': 'If the connection to the server is broken during transfer, Globus will start up again where it left off once the connection is reestablished.'},\n",
       " {'question': 'What types of files can be imported into KBase?',\n",
       "  'answer': 'Reads files (.fastq or .fastq.gz), assemblies (.fna or .fasta), and annotated genomes (.gbk or .gff) can be imported into KBase.'},\n",
       " {'question': 'What happens to the files in the Staging Area after 90 days?',\n",
       "  'answer': 'The files in the Staging Area will be removed after 90 days.'},\n",
       " {'question': 'Can I import data from my Data Browser Staging Area to any of my KBase Narratives?',\n",
       "  'answer': 'Yes, you can import data from your Data Browser Staging Area to any of your KBase Narratives.'},\n",
       " {'question': 'How long does the JGI staging area keep file links?',\n",
       "  'answer': 'The JGI staging area keeps file links for 14 days.'},\n",
       " {'question': 'How long does the KBase staging area keep files?',\n",
       "  'answer': 'The KBase staging area keeps files for 90 days.'},\n",
       " {'question': 'How much space does the KBase staging area have available to the user?',\n",
       "  'answer': 'The KBase staging area has 2.8 terabytes available to the user.'},\n",
       " {'question': 'How does the JGI notify users when files have been added to the staging area?',\n",
       "  'answer': 'The JGI notifies users by email when files have been added to the staging area.'},\n",
       " {'question': 'How does KBase stage files?',\n",
       "  'answer': 'KBase stages specific files selected by the user.'},\n",
       " {'question': 'How can I access and run analysis Apps in KBase?',\n",
       "  'answer': 'The specific steps are not provided in the text, but you can find the guidance in the KBase documentation.'},\n",
       " {'question': 'What are Beta Apps in KBase?',\n",
       "  'answer': 'The text does not provide a definition for Beta Apps in KBase, but they are likely to be new or experimental apps that are being tested.'},\n",
       " {'question': 'Where can I find the KBase App Catalog?',\n",
       "  'answer': 'The KBase App Catalog can be visited online, but the specific URL is not provided in the text.'},\n",
       " {'question': 'What are KBase Apps?',\n",
       "  'answer': 'KBase Apps are analysis tools that you can use in KBase. Apps interoperate seamlessly to enable a range of scientific workflows. Some of the apps were written by KBase scientists and developers; others are third-party tools that were integrated into KBase with our Software Developer Kit (SDK).'},\n",
       " {'question': 'How can I find the available Apps in KBase?',\n",
       "  'answer': 'The App Catalog lists currently available Apps. You can access it within KBase by clicking on the Catalog icon in the Menu or in a Narrative by clicking the right arrow at the top of the Apps Panel.'},\n",
       " {'question': 'What are the categories of KBase Apps?',\n",
       "  'answer': 'The majority of KBase Apps fall into the following categories: Assembly, Annotation, Sequence Analysis, Comparative Genomics, Metabolic Modeling, Expression, Microbial Communities.'},\n",
       " {'question': 'What do you need to use KBase tools?',\n",
       "  'answer': 'You will need a KBase user account to use our tools.'},\n",
       " {'question': 'How can you access the App Catalog from inside the Narrative Interface?',\n",
       "  'answer': 'You can access the App Catalog from inside the Narrative Interface by clicking the small arrow in the upper right corner of the Apps Panel.'},\n",
       " {'question': 'How can you add an app to your favorites in KBase?',\n",
       "  'answer': 'You can click the star at the lower left of any app to add it to your “favorites.” The gray star will turn yellow to indicate that you have favorited the app.'},\n",
       " {'question': 'What does the number to the right of the star in an app indicate?',\n",
       "  'answer': 'The number to the right of the star shows how many people have favorited that app.'},\n",
       " {'question': 'How are the apps sorted by default in KBase?',\n",
       "  'answer': 'By default, the apps are sorted by category.'},\n",
       " {'question': 'What do you need to do after adding an app to the Narrative in KBase?',\n",
       "  'answer': 'You will need to select a data object as the Input Object and set Parameters.'},\n",
       " {'question': 'How will you know if there are empty fields or errors when working within an application in KBase?',\n",
       "  'answer': 'A red bar or banner will highlight empty fields and errors.'},\n",
       " {'question': 'Where can you find more information about using the App Catalog in KBase?',\n",
       "  'answer': 'You can find more information in the Narrative Interface User Guide.'},\n",
       " {'question': 'What are some of the tools available in KBase for Assembly and Annotation?',\n",
       "  'answer': 'KBase provides multiple Apps for de novo assembly of prokaryotic Next-Generation Sequencing (NGS) reads from various sequencing platforms. These assemblies can then be annotated to explore structural and functional features of a Genome or use it in other analyses.'},\n",
       " {'question': 'What is the purpose of the interactive tutorials in KBase?',\n",
       "  'answer': 'The interactive tutorials are a good way to learn about these workflows.'},\n",
       " {'question': 'What are some of the tools available for Read Processing in KBase?',\n",
       "  'answer': 'Some of the tools available for Read Processing in KBase are Trim Reads with Trimmomatic, Filter Out Low-Complexity Reads with PRINSEQ, Assess Read Quality with FastQC, and Cutadapt.'},\n",
       " {'question': 'What types of next-generation sequencing reads does KBase support for de novo assembly?',\n",
       "  'answer': 'KBase supports the de novo assembly of Illumina and Ion Torrent next-generation sequencing reads. It supports single-end and paired-end read libraries.'},\n",
       " {'question': 'What is HipMer?',\n",
       "  'answer': 'HipMer is a highly-parallelized port of JGI’s Meraculous assembler. It is a de Bruijn graph-based which increases speed by not performing error correction. Instead, it bases contigs on already high-quality scores and fills the gaps based on localized assemblies from the reads.'},\n",
       " {'question': 'What is IDBA-UD?',\n",
       "  'answer': 'IDBA-UD is an iterative graph-based assembler for single-cell and standard short read data and is good for data of highly uneven sequencing depth. This assembler uses an iterative approach for selecting k-mer size that compensates for the information loss associated with single k-mer based de Bruijn graphs, making IDBA-UD one of the more accurate microbial assemblers.'},\n",
       " {'question': 'How does HipMer enhance the speed of Meraculous?',\n",
       "  'answer': 'HipMer enhances the speed of Meraculous by not performing error correction and basing contigs on already high-quality scores and fills the gaps based on localized assemblies from the reads.'},\n",
       " {'question': 'Why is IDBA-UD considered one of the more accurate microbial assemblers?',\n",
       "  'answer': 'IDBA-UD is considered one of the more accurate microbial assemblers because it uses an iterative approach for selecting k-mer size that compensates for the information loss associated with single k-mer based de Bruijn graphs.'},\n",
       " {'question': 'What is MaSuRCA?',\n",
       "  'answer': 'MaSuRCA is a short read assembler that combines the benefits of de Bruijn graph and overlap layout consensus assembly approaches. It is used for eukaryotic assembly.'},\n",
       " {'question': 'What is the main concept of MaSuRCA?',\n",
       "  'answer': 'The main concept of MaSuRCA is the creation of super-reads that contain sequence information present in the original reads, which super-reads are then extended in both directions using an efficient k-mer lookup table.'},\n",
       " {'question': 'What is MEGAHIT?',\n",
       "  'answer': 'MEGAHIT is a single node assembler for large and complex metagenomics NGS reads. It makes use of succinct de Bruijn graph (SdBG) to achieve low memory assembly, making it fast and especially suitable for assembly of small metagenomes, metatranscriptomes or low-coverage data in general.'},\n",
       " {'question': 'What makes MEGAHIT suitable for assembly of small metagenomes, metatranscriptomes or low-coverage data?',\n",
       "  'answer': 'MEGAHIT makes use of succinct de Bruijn graph (SdBG) to achieve low memory assembly, which makes it fast and especially suitable for these types of assemblies.'},\n",
       " {'question': 'What is SPAdes?',\n",
       "  'answer': 'SPAdes is a single-cell and standard assembler based on paired de Bruijn graphs, considered to be one of the most accurate microbial assemblers.'},\n",
       " {'question': 'How does SPAdes work?',\n",
       "  'answer': 'SPAdes employs a multisized de Bruijn graph which detects and removes bubble and chimeric reads, estimates insert distance from paired kmers, and computes contigs based on paired assembly graph.'},\n",
       " {'question': 'What is Velvet?',\n",
       "  'answer': 'Velvet is a classic de Bruijn graph based assembler that works by efficiently manipulating de Bruijn graphs through simplification and compression.'},\n",
       " {'question': 'How does Velvet work?',\n",
       "  'answer': 'It eliminates errors and resolves repeats by first using an error correction algorithm that merges sequences together. Repeats are then removed from the sequence via the repeat solver that separates paths which share local overlaps.'},\n",
       " {'question': 'What is the purpose of QUAST?',\n",
       "  'answer': 'QUAST is used to assess the output assemblies from different configurations of the same assembler, or compare assemblies from multiple assemblers to determine which one is optimal for downstream analysis.'},\n",
       " {'question': 'How can genomes be annotated in KBase?',\n",
       "  'answer': 'Genomes can be annotated with Prokka or RAST in KBase.'},\n",
       " {'question': \"What is the function of the 'Annotate Domains in a Genome' tool in KBase?\",\n",
       "  'answer': 'It identifies protein domains from widely used domain libraries (COGs, TIGRfams, Pfam).'},\n",
       " {'question': \"What does the 'Annotate Assembly with Prokka' tool do in KBase?\",\n",
       "  'answer': 'It combines multiple open-source annotation tools in a quick and thorough annotation pipeline for prokaryotic sequences for genomes, plasmids, and metagenomes.'},\n",
       " {'question': \"What is the purpose of the 'Annotate Microbial Assembly' tool in KBase?\",\n",
       "  'answer': 'It uses components from the RAST (Rapid Annotations using Subsystems Technology) toolkit to annotate an assembled bacterial or archaeal genome.'},\n",
       " {'question': \"What does the 'Annotate Plant Coding Sequences with Metabolic Functions' tool do in KBase?\",\n",
       "  'answer': 'It performs functional annotation of plant cDNA or protein sequences.'},\n",
       " {'question': 'What does the Bulk Annotate Genomes/Assemblies app do in KBase?',\n",
       "  'answer': 'It uses components from the RAST (Rapid Annotations using Subsystems Technology) toolkit to annotate a set of genomes or assemblies.'},\n",
       " {'question': 'What is the output of the annotation apps in KBase?',\n",
       "  'answer': 'The output of the annotation apps is a Genome, which is displayed in a tabular genome viewer that shows information about the Genome as well as a list of contigs and the genes that were called on each contig.'},\n",
       " {'question': 'What is RAST in KBase?',\n",
       "  'answer': 'RAST (Rapid Annotations using Subsystems Technology) is a toolkit used in KBase to annotate a set of genomes or assemblies.'},\n",
       " {'question': 'What does KBase provide for comparative genomics and phylogenetic analysis?',\n",
       "  'answer': 'KBase provides multiple comparative genomics and phylogenetic analysis Apps to enable researchers to understand evolutionary relationships between organisms and explore structural and functional variance across genomes.'},\n",
       " {'question': 'What are some of the tools available in KBase for phylogenetic and taxonomic analysis?',\n",
       "  'answer': 'Some of the tools are: \\n    1. Insert Genome Into Species Tree – Build a phylogenetic tree with your genomes and RefSeq microbial genomes based on conserved orthologous groups\\n    2. Build Microbial Species Tree (in beta) – Build a phylogenetic tree with your genome and phylum exemplars from RefSeq\\n    3. Build Gene Tree (in beta) – Build an evolutionary reconstruction tree for a collection of sequence homology related genes\\n    4. Build Phylogenetic Tree with FastTree2  – Build a tree with FastTree2 directly on a user-created Multiple Sequence Alignment (MSA)'},\n",
       " {'question': 'What is the purpose of the \"Insert Genome Into Species Tree\" tool in KBase?',\n",
       "  'answer': 'The \"Insert Genome Into Species Tree\" tool in KBase is used to build a phylogenetic tree with your genomes and RefSeq microbial genomes based on conserved orthologous groups.'},\n",
       " {'question': 'What is BLASTx used for in KBase?',\n",
       "  'answer': 'BLASTx is used to search genes in genomes and Annotated Metagenome Assemblies (AMAs) and compare translated protein sequences.'},\n",
       " {'question': 'What is the function of BLASTn in KBase?',\n",
       "  'answer': 'BLASTn is used to search genes and AMAs and compare nucleotide sequences.'},\n",
       " {'question': 'What does MUSCLE do in KBase?',\n",
       "  'answer': 'MUSCLE is used for building and analyzing multiple sequence alignments (MSA) for nucleotide or protein sequences.'},\n",
       " {'question': 'What is the purpose of BLASTp and psiBLAST in KBase?',\n",
       "  'answer': 'BLASTp and psiBLAST are used to search genes and AMAs and compare protein sequences.'},\n",
       " {'question': 'What is HMMER MSA Collection used for in KBase?',\n",
       "  'answer': 'It is used to search and profile genes in genomes and Annotated Metagenome Assemblies (AMAs) with a collection of user-defined MSAs.'},\n",
       " {'question': 'What is the purpose of HMMER Search with dbCAN2 of CAZy families in KBase?',\n",
       "  'answer': 'It is used to search and profile genes in genomes and AMAs with the suite of CAZy dbCAN2 hidden markov models (HMMs).'},\n",
       " {'question': 'What does the Annotate Domains in a GenomeSet function do in KBase?',\n",
       "  'answer': 'It computes canonical gene family membership using COG, Pfam, and TIGRFAMs.'},\n",
       " {'question': 'What can you do with the View Function Profile for Genomes and View Function Profile for a Phylogenetic Tree in KBase?',\n",
       "  'answer': 'You can examine functional content from canonical domain family annotation.'},\n",
       " {'question': 'What is the purpose of HMMER Search for Env BioElement families in KBase?',\n",
       "  'answer': 'It is used to search and profile genes in genomes and AMAs with the suite of Environmental BioElement metabolic families.'},\n",
       " {'question': 'What does the \"Compare Two Proteomes\" function do in KBase?',\n",
       "  'answer': 'It compares proteomes and produces a synteny map dot plot matrix and table of gene differences.'},\n",
       " {'question': 'What is the purpose of the \"Build Pangenome with OrthoMCL\" function in KBase?',\n",
       "  'answer': 'It groups orthologous protein sequences and identifies core and non-core genes across multiple species.'},\n",
       " {'question': 'What does the \"Compute Pangenome\" function do in KBase?',\n",
       "  'answer': 'It builds a pangenome and evaluates protein family conservation.'},\n",
       " {'question': 'What is the \"Pangenome Circle Plot\" function used for in KBase?',\n",
       "  'answer': 'It is used to view a microbial pangenome as a circle plot.'},\n",
       " {'question': 'What does the \"Compare Genomes from Pangenome\" function do in KBase?',\n",
       "  'answer': 'It compares isofunctional and homologous gene families within a pangenome.'},\n",
       " {'question': 'What is the purpose of the \"Phylogenetic Pangenome Accumulation\" function in KBase?',\n",
       "  'answer': 'It is used to view the pangenome in a phylogenetic context.'},\n",
       " {'question': 'What are some of the tools available in KBase for metabolic modeling?',\n",
       "  'answer': 'KBase has a suite of Apps supporting the reconstruction, prediction, and design of metabolic models in microbes and plants. Some of the tools include Build Metabolic Model, Gapfill Metabolic Model, Run Flux Balance Analysis, Compare FBA solutions, Check Model Mass Balance, and Compare Models.'},\n",
       " {'question': 'What can genome-scale metabolic models be used for?',\n",
       "  'answer': 'Genome-scale metabolic models can be used to explore an organism’s growth in specific media conditions, determine which biochemical pathways are present, optimize production of an important metabolite, identify high flux pathways, and more.'},\n",
       " {'question': 'What does the Build Metabolic Model tool do?',\n",
       "  'answer': 'The Build Metabolic Model tool constructs genome-scale metabolic models based on a bacterial genome and media condition.'},\n",
       " {'question': 'What does the Gapfill Metabolic Model tool do?',\n",
       "  'answer': 'The Gapfill Metabolic Model tool fills in missing reactions based on stoichiometry.'},\n",
       " {'question': 'What does the Run Flux Balance Analysis tool do?',\n",
       "  'answer': 'The Run Flux Balance Analysis tool predicts metabolic fluxes.'},\n",
       " {'question': 'What can you do with the Edit Metabolic Model feature in KBase?',\n",
       "  'answer': 'You can build unique models suited to specific experiments.'},\n",
       " {'question': 'What does the Create or Edit Media feature allow you to do in KBase?',\n",
       "  'answer': 'It allows you to create specialized growth conditions.'},\n",
       " {'question': 'What is the purpose of the Bulk Download Modeling Objects feature in KBase?',\n",
       "  'answer': 'It allows you to save modeling data for future analysis.'},\n",
       " {'question': 'What does the Propagate Model to New Genome feature do in KBase?',\n",
       "  'answer': 'It allows you to translate metabolic models from one organism to another.'},\n",
       " {'question': 'What can you do with the Compare Flux with Expression feature in KBase?',\n",
       "  'answer': 'You can compare reaction fluxes with gene expression values to identify metabolic pathways where expression and flux data agree or conflict and compare with chemical abundance data.'},\n",
       " {'question': 'What is the PickAxe App?',\n",
       "  'answer': 'The PickAxe App is a tool to generate novel compounds from an FBA Model and a Chemical Abundance Matrix. It is part of the MINE Databases.'},\n",
       " {'question': 'What does the Fit Model to Exometabolite Data app do?',\n",
       "  'answer': 'The Fit Model to Exometabolite Data app identifies biochemical reactions to add to a draft metabolic model for production and consumption of exometabolites.'},\n",
       " {'question': 'What is the purpose of the Merge Metabolic Models into Community Model app?',\n",
       "  'answer': 'The Merge Metabolic Models into Community Model app is used to investigate community metabolism.'},\n",
       " {'question': 'What does the Build Metagenome Metabolic Model app do?',\n",
       "  'answer': 'The Build Metagenome Metabolic Model app builds a metagenome metabolic model from an annotated assembly or bins.'},\n",
       " {'question': 'What are some of the tools in KBase available for analyzing microbial communities and metagenomics?',\n",
       "  'answer': 'Some of the tools include FastQC for raw read quality control and analysis, Trim Reads with Trimmomatic for read trimming and adaptor removal, Kaiju for taxonomy assignment of raw metagenomic reads, MEGAHIT for assembling large and complex metagenomic reads, MetaSPAdes for assembling shotgun metagenomic reads, and IDBA-UD for assembling short metagenomic reads.'},\n",
       " {'question': 'What is the purpose of the Apps designed to analyze microbial communities and metagenomic reads data in KBase?',\n",
       "  'answer': 'They provide a way to understand the functional interactions between species in microbial communities. Annotated genomes extracted from metagenomic sequence libraries can be used for metabolic modeling, comparative phylogenomics, functional profiling, and more.'},\n",
       " {'question': 'What is FastQC used for in KBase?',\n",
       "  'answer': 'FastQC is used for raw read quality control and analysis in KBase.'},\n",
       " {'question': 'What is the function of Trim Reads with Trimmomatic in KBase?',\n",
       "  'answer': 'Trim Reads with Trimmomatic is used for read trimming and adaptor removal in KBase.'},\n",
       " {'question': 'What is the role of Kaiju in KBase?',\n",
       "  'answer': 'Kaiju is used for taxonomy assignment of raw metagenomic reads in KBase.'},\n",
       " {'question': 'What is MaxBin2 used for in KBase?',\n",
       "  'answer': 'MaxBin2 is used to group metagenomic contigs using depth-of-coverage, nucleotide composition, and marker genes in KBase.'},\n",
       " {'question': 'What is the function of MetaBAT2 in KBase?',\n",
       "  'answer': 'MetaBAT2 is used to group metagenomic contigs using strain abundance and nucleotide composition in KBase.'},\n",
       " {'question': 'What does DAS-Tool do in KBase?',\n",
       "  'answer': 'DAS-Tool optimizes bacterial or archaeal genome bins using a de-replication, aggregation and scoring strategy in KBase.'},\n",
       " {'question': 'What is the purpose of CheckM in KBase?',\n",
       "  'answer': 'CheckM is used to assess genome and Metagenome Assembled Genomes (MAG) quality in KBase.'},\n",
       " {'question': 'What does RAST for Multiple Microbial Genomes and Multiple Microbial Assemblies do in KBase?',\n",
       "  'answer': 'It annotates multiple bacterial or archaeal genomes and assemblies or sets using RASTtk in KBase.'},\n",
       " {'question': 'What is the function of dRep (in beta) in KBase?',\n",
       "  'answer': 'dRep (in beta) is used to dereplicate binned contigs, assemblies, and genomes using nucleotide similarity in KBase.'},\n",
       " {'question': 'What is VirSorter used for in KBase?',\n",
       "  'answer': 'VirSorter is used to identify viral sequences from viral and microbial metagenomes.'},\n",
       " {'question': 'What is the function of Classify Microbes with GTDB-Tk in KBase?',\n",
       "  'answer': 'Classify Microbes with GTDB-Tk is used to assign taxonomy to isolate and classify Metagenome Assembled Genomes (MAGs) using Genome Taxonomy Database (GTDB) genome taxonomy.'},\n",
       " {'question': 'What does ModelSEED do in KBase?',\n",
       "  'answer': 'ModelSEED is used to generate and merge metabolic models from annotated genomes.'},\n",
       " {'question': 'What is the purpose of Phylogenetic Pangenome Accumulation in KBase?',\n",
       "  'answer': 'Phylogenetic Pangenome Accumulation is used to produce phylogenetic trees from annotated genomes.'},\n",
       " {'question': 'What are Metabolic modeling tools for microbial communities used for in KBase?',\n",
       "  'answer': 'Metabolic modeling tools for microbial communities are used to analyze community metabolism.'},\n",
       " {'question': 'What tools are available in KBase for analyzing amplicon matrices and OTU tables?',\n",
       "  'answer': 'KBase provides tools for rarefaction, standardization, and data cleaning.'},\n",
       " {'question': 'What is the purpose of the Transform Matrix App in KBase?',\n",
       "  'answer': 'The Transform Matrix App in KBase is used for data cleaning. It uses an AmpliconMatrix object as input and provides abundance filtering, standardization, and ratio transformation.'},\n",
       " {'question': 'What transformations can be performed using the Transform Matrix App?',\n",
       "  'answer': 'The Transform Matrix App can perform abundance filtering, standardization, and ratio transformation. It can remove rows or columns if all values fall below a user-specified threshold and/or if the sum of all values falls below a user-specified threshold. It can also perform standardization, with options for centering the data before scaling, scaling the data to unit variance, and whether to perform the standardization on columns or rows. It can perform ratio transformation, by either Centre or Isometric Log Ratio Transformation methods, and be applied to columns or rows.'},\n",
       " {'question': 'What does the Rarefy Matrix App do?',\n",
       "  'answer': 'The Rarefy Matrix App allows manual selection of a seed value to aid reproducibility to columns or rows.'},\n",
       " {'question': 'What does the Taxonomic Classification feature do?',\n",
       "  'answer': 'It classifies rRNA with taxonomy using naïve Bayes with RDP Classifier App. It assigns taxonomies to amplicon matrices based on several gene databases including RDP Classifier 2.13 - 16srna, RDP Classifier - fungallsu, and SILVA 138 - SSU, Full Length.'},\n",
       " {'question': 'What does the Taxonomy Abundance Barplot App do?',\n",
       "  'answer': 'The Taxonomy Abundance Barplot App visualizes the taxonomic abundance.'},\n",
       " {'question': 'What types of statistical comparisons can be calculated in KBase?',\n",
       "  'answer': 'Statistical comparisons such as clustering (hierarchical or k-means) and PCA analysis can be calculated in KBase.'},\n",
       " {'question': 'What does the Perform NMDS Analysis App do?',\n",
       "  'answer': 'The Perform NMDS Analysis App performs Non-metric Multidimensional Scaling Analysis on amplicon matrix data to visualize it in two dimensions.'},\n",
       " ...]"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = []\n",
    "for doc in documents_html:\n",
    "    answers = generateQA(instruction_prompt, doc.page_content)\n",
    "    try:\n",
    "        data.extend(extractQA(answers))\n",
    "    except IndexError:\n",
    "        print(answers)\n",
    "        continue\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "e8f40f59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results have been saved to kbase_formatted_new_more_prompt.json.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "output_filepath = 'kbase_formatted_new_more_prompt.json'\n",
    "with open(output_filepath, 'w') as output_file:\n",
    "    json.dump(data, output_file, indent=4)\n",
    "    print(f\"Results have been saved to {output_filepath}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd73f81d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
