{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8045b8f9-6629-4429-aa9a-c32e0edf17da",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import openai\n",
    "import config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b68a2090-9a13-4ba7-a313-62df35e495b6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os \n",
    "import json\n",
    "#import streamlit as st \n",
    "from langchain.llms import OpenAI\n",
    "from langchain.prompts import PromptTemplate\n",
    "from langchain.chains import LLMChain, SequentialChain \n",
    "from langchain.memory import ConversationBufferMemory \n",
    "from langchain.document_loaders import JSONLoader, DirectoryLoader\n",
    "from langchain.document_loaders import BSHTMLLoader\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from bs4 import BeautifulSoup\n",
    "from urllib.parse import urlparse, urljoin\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain.vectorstores import Chroma\n",
    "from langchain.embeddings.openai import OpenAIEmbeddings\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "from langchain.chains import RetrievalQA\n",
    "from langchain.document_loaders import UnstructuredHTMLLoader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7871f892",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from langchain.schema import HumanMessage, SystemMessage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "895c3ce2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8415e0d2-157d-42f2-8279-bb7f85dfaea4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os \n",
    "\n",
    "#Key from Paramvir\n",
    "os.environ[\"OPENAI_API_KEY\"] = \"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\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "09baec3a-247f-4f1a-a19a-50ed07a494f9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from langchain.prompts.few_shot import FewShotPromptTemplate\n",
    "from langchain.prompts.prompt import PromptTemplate\n",
    "\n",
    "examples = [\n",
    "  {\n",
    "    \"question\": \"I have a fastq file of reads from paired end illumina sequencing. I would like your help to analyze this data. Please give me step by step instructions of which KBase apps to use.\",\n",
    "    \"answer\":\n",
    "\"\"\"\n",
    "1. Trim the reads using the \"Trim Reads with Trimmomatic - v0.36\" app.\n",
    "2. Perform quality control on the sequencing data using the \"FastQC-v0.11.8\" app.\n",
    "3. Assemble the metagenomic contigs using a MetaSPAdes metagenome assembly app\n",
    "4. Separate individual bins using binning tools like MaxBin.\n",
    "5. Annotate the assembled contigs using the \"Annotate Metagenome Assembly with Prokka - v1.14.5\" app or a RAST Annotation App.\n",
    "6. KBase has the \"Classify Microbes with GTDB-Tk - v1.7.0\" app for microbial taxonomy determination. \n",
    "\n",
    "\"\"\"\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"I have run QUAST and the metaspades assembly seems clearly better. What is the next step?\",\n",
    "    \"answer\":\n",
    "\"\"\"\n",
    "The next step after determining that the metaspades assembly is better would be to perform metagenome annotation using the \n",
    "Genome Taxonomy Database (GTDB) version R06-RS202 to \n",
    "identify the genes and functional elements present in the microbe.\n",
    "\"\"\"\n",
    "  }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "67be12b1-f8c2-420f-8231-65ff55414d2d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Question: I have a fastq file of reads from paired end illumina sequencing. I would like your help to analyze this data. Please give me step by step instructions of which KBase apps to use.\n",
      "\n",
      "1. Trim the reads using the \"Trim Reads with Trimmomatic - v0.36\" app.\n",
      "2. Perform quality control on the sequencing data using the \"FastQC-v0.11.8\" app.\n",
      "3. Assemble the metagenomic contigs using a MetaSPAdes metagenome assembly app\n",
      "4. Separate individual bins using binning tools like MaxBin.\n",
      "5. Annotate the assembled contigs using the \"Annotate Metagenome Assembly with Prokka - v1.14.5\" app or a RAST Annotation App.\n",
      "6. KBase has the \"Classify Microbes with GTDB-Tk - v1.7.0\" app for microbial taxonomy determination. \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "example_prompt = PromptTemplate(input_variables=[\"question\", \"answer\"], template=\"Question: {question}\\n{answer}\")\n",
    "\n",
    "print(example_prompt.format(**examples[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "26f5955c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# prompt used to extract questions\n",
    "extraction_system_prompt=\"You are an expert user extracting information to quiz people on documentation. Generate questions similar to the examples provided. You will be passed a page extracted from the documentation, write a numbered list of questions that can be answered based *solely* on the given text.\"\n",
    "def create_extraction_conversation_messages(text):\n",
    "    \"\"\"\n",
    "    Takes a piece of text and returns a list of messages designed to extract questions from the text.\n",
    "    \n",
    "    Args:\n",
    "        text (str): The input text for which questions are to be extracted.\n",
    "    \n",
    "    Returns:\n",
    "        list: A list of messages that set up the context for extracting questions.\n",
    "    \"\"\"\n",
    "    # Create a system message setting the context for the extraction task\n",
    "    context_message = SystemMessage(content=prompt.format(input=extraction_system_prompt))\n",
    "    \n",
    "    # Create a human message containing the input text\n",
    "    input_text_message = HumanMessage(content=text)\n",
    "    print('messages',context_message,input_text_message )\n",
    "    # Return the list of messages to be used in the extraction conversation\n",
    "    return [context_message, input_text_message]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b868a1c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Question: I have a fastq file of reads from paired end illumina sequencing. I would like your help to analyze this data. Please give me step by step instructions of which KBase apps to use.\n",
      "\n",
      "1. Trim the reads using the \"Trim Reads with Trimmomatic - v0.36\" app.\n",
      "2. Perform quality control on the sequencing data using the \"FastQC-v0.11.8\" app.\n",
      "3. Assemble the metagenomic contigs using a MetaSPAdes metagenome assembly app\n",
      "4. Separate individual bins using binning tools like MaxBin.\n",
      "5. Annotate the assembled contigs using the \"Annotate Metagenome Assembly with Prokka - v1.14.5\" app or a RAST Annotation App.\n",
      "6. KBase has the \"Classify Microbes with GTDB-Tk - v1.7.0\" app for microbial taxonomy determination. \n",
      "\n",
      "\n",
      "\n",
      "Question: I have run QUAST and the metaspades assembly seems clearly better. What is the next step?\n",
      "\n",
      "The next step after determining that the metaspades assembly is better would be to perform metagenome annotation using the \n",
      "Genome Taxonomy Database (GTDB) version R06-RS202 to \n",
      "identify the genes and functional elements present in the microbe.\n",
      "\n",
      "\n",
      "Question: You are an expert user extracting information to quiz people on documentation. Generate questions similar to the examples provided. You will be passed a page extracted from the documentation, write a numbered list of questions that can be answered based *solely* on the given text.\n"
     ]
    }
   ],
   "source": [
    "prompt = FewShotPromptTemplate(\n",
    "    examples=examples,\n",
    "    example_prompt=example_prompt,\n",
    "    suffix=\"Question: {input}\",\n",
    "    input_variables=[\"input\"]\n",
    ")\n",
    "\n",
    "print(prompt.format(input=extraction_system_prompt))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7601ff42",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "llm = ChatOpenAI(temperature=0, model=\"gpt-4\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b43aa91c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# prompt used to answer a question\n",
    "answering_system_prompt=\"You are an expert user answering questions. You will be passed a page extracted from a documentation and a question. Generate a comprehensive and informative answer to the question based *solely* on the given text.\"\n",
    "\n",
    "\n",
    "def create_answering_conversation_messages(question, text):\n",
    "    \"\"\"\n",
    "    Takes a question and a text and returns a list of messages designed to answer the question based on the text.\n",
    "    \n",
    "    Args:\n",
    "        question (str): The question to be answered.\n",
    "        text (str): The text containing information for answering the question.\n",
    "    \n",
    "    Returns:\n",
    "        list: A list of messages that set up the context for answering the question.\n",
    "    \"\"\"\n",
    "    # Create a system message setting the context for the answering task\n",
    "    context_message = SystemMessage(content=answering_system_prompt)\n",
    "    \n",
    "    # Create a human message containing the input text\n",
    "    input_text_message = HumanMessage(content=text)\n",
    "    \n",
    "    # Create a human message containing the question to be answered\n",
    "    input_question_message = HumanMessage(content=question)\n",
    "    \n",
    "    # Return the list of messages to be used in the answering conversation\n",
    "    \n",
    "    return [context_message, input_text_message, input_question_message]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ce0aa7f9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def extract_answers(questions,ref_text):\n",
    "    # Create the input messages for the chat model\n",
    "\n",
    "    answers=[]\n",
    "    for item in questions:\n",
    "        message = create_answering_conversation_messages(item, ref_text)\n",
    "        answer = llm(message)\n",
    "        answers.append(answer.content)\n",
    "        # run the chat model with the input messages\n",
    "    return answers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "254a4d17",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "text_splitter = RecursiveCharacterTextSplitter(\n",
    "    chunk_size = 2000,\n",
    "    chunk_overlap  = 20,\n",
    "    length_function = len,\n",
    ")\n",
    "dir_loader = DirectoryLoader('/scratch/ac.pgupta/convLLM/convLL_addMem/convLLM_addNarratives/convLLM_addNarratives_add embeddings/kbasedocs/', glob=\"**/[!.]*.html\", loader_cls=BSHTMLLoader)\n",
    "data = dir_loader.load()\n",
    "documents_html = text_splitter.split_documents(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "c7308fc6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "messages content='Question: I have a fastq file of reads from paired end illumina sequencing. I would like your help to analyze this data. Please give me step by step instructions of which KBase apps to use.\\n\\n1. Trim the reads using the \"Trim Reads with Trimmomatic - v0.36\" app.\\n2. Perform quality control on the sequencing data using the \"FastQC-v0.11.8\" app.\\n3. Assemble the metagenomic contigs using a MetaSPAdes metagenome assembly app\\n4. Separate individual bins using binning tools like MaxBin.\\n5. Annotate the assembled contigs using the \"Annotate Metagenome Assembly with Prokka - v1.14.5\" app or a RAST Annotation App.\\n6. KBase has the \"Classify Microbes with GTDB-Tk - v1.7.0\" app for microbial taxonomy determination. \\n\\n\\n\\nQuestion: I have run QUAST and the metaspades assembly seems clearly better. What is the next step?\\n\\nThe next step after determining that the metaspades assembly is better would be to perform metagenome annotation using the \\nGenome Taxonomy Database (GTDB) version R06-RS202 to \\nidentify the genes and functional elements present in the microbe.\\n\\n\\nQuestion: You are an expert user extracting information to quiz people on documentation. Generate questions similar to the examples provided. You will be passed a page extracted from the documentation, write a numbered list of questions that can be answered based *solely* on the given text.' content='KBase DocumentationSearch⌃KKBase DocumentationKBase Terms & ConditionsGetting StartedWorking with DataData Upload and Download GuideData TypesImporting DataAssemblyGenomeFASTQ/SRA ReadsFlux Balance Analysis (FBA) ModelMediaExpression MatrixPhenotype SetAmplicon MatrixChemical Abundance MatrixSampleSetCompressed/Zipped FilesBulk Import SpecificationDownloading DataSearching, Adding, and Uploading DataFiltering, Managing, and Viewing DataLinking MetadataPublic Data in KBaseTransferring Data with GlobusTransferring Data from JGIUsing AppsRunning Common WorkflowsTroubleshootingDeveloping AppsExternal LinksKBase Narrative InterfaceKBase web siteKBase App Catalogkbase.usPowered By GitBookChemical Abundance MatrixMetabolomics, exometabolite, and chemical abundance data can be integrated with metabolic modeling and flux balance analysis tools in KBase.What is \"chemical abundance\" in KBase?The name of this data type “chemical abundance” is a broad term that we use to represent a wide array of measurements associated with chemicals. This data type can be used to upload and store diverse types of chemical data in the system such as metabolomics (intracellular and/or extracellular) that is derived based on microbiomes/isolate organism growth experiments etc., computationally predicted compounds, or data collected on the concentration or from elemental analysis. These data could be collected on environmental samples, such as soil, sediment, or water. Currently, the metabolomics data derived from the samples are the most popular data that is uploaded and stored in the system.Once chemical abundance data matrices are uploaded, they can be analyzed using KBase Apps for metabolomics, such as Escher mapping. Additional statistical analysis of the chemical abundance attribute maps, such as PCA and clustering, can also be performed.A Chemical Abundance Matrix can be uploaded from a TSV (tab-separated values) file with a .tsv or .tab file extension, or from Excel spreadsheet with a'\n",
      "messages content='Question: I have a fastq file of reads from paired end illumina sequencing. I would like your help to analyze this data. Please give me step by step instructions of which KBase apps to use.\\n\\n1. Trim the reads using the \"Trim Reads with Trimmomatic - v0.36\" app.\\n2. Perform quality control on the sequencing data using the \"FastQC-v0.11.8\" app.\\n3. Assemble the metagenomic contigs using a MetaSPAdes metagenome assembly app\\n4. Separate individual bins using binning tools like MaxBin.\\n5. Annotate the assembled contigs using the \"Annotate Metagenome Assembly with Prokka - v1.14.5\" app or a RAST Annotation App.\\n6. KBase has the \"Classify Microbes with GTDB-Tk - v1.7.0\" app for microbial taxonomy determination. \\n\\n\\n\\nQuestion: I have run QUAST and the metaspades assembly seems clearly better. What is the next step?\\n\\nThe next step after determining that the metaspades assembly is better would be to perform metagenome annotation using the \\nGenome Taxonomy Database (GTDB) version R06-RS202 to \\nidentify the genes and functional elements present in the microbe.\\n\\n\\nQuestion: You are an expert user extracting information to quiz people on documentation. Generate questions similar to the examples provided. You will be passed a page extracted from the documentation, write a numbered list of questions that can be answered based *solely* on the given text.' content='spreadsheet with a .xls extension.Each Chemical Type can be either a specific compound or element, aggregate (totals), exometabolites (measurements of compounds or elements that are consumed or excreted into the medium). Chemical Abundance Upload WebinarFormatting chemical abundance matricesThe Create Chemical Abundance Matrix Template App creates an Excel spreadsheet for direct download that can be populated with chemical abundance data. While Chemical abundance data works best and more meaningful when linked with an existing SampleSet in the system, linking a SampleSet is not required (See section Linking SampleSet). While Chemical abundance data works best and more meaningful when linked with an existing SampleSet in the system, linking a SampleSet is not required.The minimal set of metadata in a chemical abundance matrix includes an ID (unique value) field, a chemical type (aggregate, exometabolite, specific), and one or more of the following:  Compound ID (e.g; ModelSEED, KEGG, ChEBI), mass, formula, inchikey, inchi, smiles, or compound name. Additional metadata such as units are strongly encouraged to provide with proper information that fits your scientific use cases or be kept as ‘unknown’.  (see section \"Template Fields Descriptions\" for an explanation of each field)  Providing additional metadata may enhance the downstream analysis of use cases for you and other readers.If a SampleSet exists, it can be applied to the chemical abundance data. Chemical abundance data needs to be formatted to ensure Samples are correctly linked.Note that linking to Samples is not required, but highly recommended. When linking to using this app, the template will be automatically populated with Sample IDs to ensure the chemical abundance data is properly linked to corresponding Samples in the system.Create Chemical Abundance Matrix TemplateThis App generates a spreadsheet onto which you can copy your data to ensure it links to the SampleSet when uploaded. When creating the'\n"
     ]
    }
   ],
   "source": [
    "result=[]\n",
    "for item in documents_html[150:152]:\n",
    "    result.append(create_extraction_conversation_messages(item.page_content))\n",
    "output=[]\n",
    "for item in result:\n",
    "    output.append(llm(item))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "9557f42b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[AIMessage(content='1. What is the term \"chemical abundance\" used to represent in KBase?\\n2. What types of chemical data can be stored in KBase?\\n3. What are some examples of environmental samples from which data could be collected?\\n4. What is the most popular type of data that is uploaded and stored in the system?\\n5. What can be done once chemical abundance data matrices are uploaded?\\n6. What are some of the KBase Apps that can be used to analyze chemical abundance data?\\n7. What types of statistical analysis can be performed on the chemical abundance attribute maps?\\n8. What file formats can a Chemical Abundance Matrix be uploaded from?'),\n",
       " AIMessage(content='1. What is the extension of the spreadsheet created by the \"Create Chemical Abundance Matrix Template\" App?\\n2. What are the different types of Chemical Types that can be included in the chemical abundance matrix?\\n3. Is it mandatory to link a SampleSet with the chemical abundance data?\\n4. What is the minimal set of metadata required in a chemical abundance matrix?\\n5. What are some examples of additional metadata that can be included in the chemical abundance matrix?\\n6. What is the benefit of linking the chemical abundance data to Samples?\\n7. What does the \"Create Chemical Abundance Matrix Template\" App do?\\n8. What happens when a SampleSet is linked to the chemical abundance data using the app?')]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "76671c62",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'extract_questions' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[75], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m#questions=[]\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m item \u001b[38;5;129;01min\u001b[39;00m output:\n\u001b[0;32m----> 3\u001b[0m     questions \u001b[38;5;241m=\u001b[39m extract_questions(item)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'extract_questions' is not defined"
     ]
    }
   ],
   "source": [
    "#questions=[]\n",
    "for item in output:\n",
    "    questions = extract_questions(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f2a01f93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In the Excel format, where are the tables containing information about chemical compounds and reactions stored?\n",
      "What are the required names for the tabs in the Excel file?\n",
      "In the TSV format, how are the chemical compounds and reactions tables saved?\n",
      "What are the names of the files for the chemical compounds and reactions in the TSV format?\n",
      "What are the columns in the \"FBAModelCompounds\" table?\n",
      "What is the purpose of the \"id\" column in the \"FBAModelCompounds\" table?\n",
      "What is the purpose of the \"name\" column in the \"FBAModelCompounds\" table?\n",
      "What is the purpose of the \"formula\" column in the \"FBAModelCompounds\" table?\n",
      "What is the purpose of the \"charge\" column in the \"FBAModelCompounds\" table?\n",
      "What is the purpose of the \"aliases\" column in the \"FBAModelCompounds\" table?\n"
     ]
    }
   ],
   "source": [
    "for item in questions:\n",
    "    print(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "02ba37d1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'questions' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[76], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m answers \u001b[38;5;241m=\u001b[39m extract_answers(questions,item\u001b[38;5;241m.\u001b[39mpage_content)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'questions' is not defined"
     ]
    }
   ],
   "source": [
    "def extract_answers(questions,ref_text):\n",
    "    # Create the input messages for the chat model\n",
    "\n",
    "    answers=[]\n",
    "    for item in questions:\n",
    "        message = create_answering_conversation_messages(item, ref_text)\n",
    "        answer = llm(message)\n",
    "        answers.append(answer.content)\n",
    "        # run the chat model with the input messages\n",
    "    return answers\n",
    "answers = extract_answers(questions,item.page_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "bab1e456",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The given text does not provide any information about the storage location of tables containing information about chemical compounds and reactions in the Excel format.\n",
      "The given text does not provide any information about the required names for the tabs in an Excel file. It solely discusses Newton's laws of motion in classical mechanics. To determine the required names for the tabs in an Excel file, you would need to refer to the specific instructions or guidelines provided by the document or the context in which the Excel file is being used.\n",
      "The given text does not provide any information about the TSV format or how chemical compounds and reactions tables are saved in it. Therefore, it is not possible to generate a comprehensive and informative answer to the question based solely on the given text.\n",
      "The given text does not provide any information about the names of files for chemical compounds and reactions in the TSV format. It solely discusses Newton's laws of motion in classical mechanics. To find the names of files for chemical compounds and reactions in the TSV format, you may need to refer to a different source or consult specific documentation related to chemical compounds and reactions in the TSV format.\n",
      "The given text does not provide any information about the \"FBAModelCompounds\" table or its columns. Therefore, without additional information, it is not possible to determine the specific columns in the \"FBAModelCompounds\" table.\n",
      "The given text does not provide any information about the \"FBAModelCompounds\" table or the purpose of the \"id\" column within it. Therefore, without additional information, it is not possible to determine the purpose of the \"id\" column in the \"FBAModelCompounds\" table.\n",
      "The given text does not provide any information about a specific table called \"FBAModelCompounds\" or a column named \"name\" within that table. Therefore, it is not possible to determine the purpose of the \"name\" column in the \"FBAModelCompounds\" table based solely on the given text.\n",
      "The given text does not provide any information about a specific table called \"FBAModelCompounds\" or a \"formula\" column within it. Therefore, it is not possible to determine the purpose of the \"formula\" column based solely on the given text.\n",
      "The given text does not provide any information about a \"charge\" column in the \"FBAModelCompounds\" table. Therefore, it is not possible to determine the purpose of the \"charge\" column based solely on the given text. Additional information or context is needed to answer this question accurately.\n",
      "The given text does not provide any information about the \"aliases\" column in the \"FBAModelCompounds\" table. Therefore, it is not possible to generate a comprehensive and informative answer to the question based solely on the given text. Additional information or context is needed to provide an accurate answer.\n"
     ]
    }
   ],
   "source": [
    "for item in answers:\n",
    "    print(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dd3039f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
