{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "035a8d8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import openai\n",
    "import config\n",
    "import os \n",
    "import json\n",
    "from langchain import FewShotPromptTemplate\n",
    "from langchain.llms import OpenAI\n",
    "from langchain.prompts import PromptTemplate\n",
    "from langchain.document_loaders import JSONLoader, DirectoryLoader, UnstructuredHTMLLoader\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "from langchain.schema import HumanMessage, SystemMessage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "71c4a9c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = \"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\" #gpt 4\n",
    "# key = \"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\" # gpt 3.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cebf6779",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"OPENAI_API_KEY\"] = key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4a1c4a90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "92"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "html_dir_loader = DirectoryLoader('/scratch/ac.pgupta/convLLM/kbasedocs', glob=\"**/[!.]*.html\", loader_cls=UnstructuredHTMLLoader)\n",
    "data_html = html_dir_loader.load()\n",
    "len(data_html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b3d1ed72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'KBase Documentation\\n\\nHere you will find guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.\\n\\nHow to cite KBase:\\n\\nArkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163\\n\\nFunding:\\n\\nThis work is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase (KBase) funded by the\\n\\nU.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\\n\\nunder Award Numbers DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.\\n\\nNextKBase Terms & Conditions\\n\\nLast modified \\n\\n2yr ago'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text1 = data_html[0].page_content\n",
    "text1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7532b67a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/index.html'}\n",
      "1 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started.html'}\n",
      "2 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up.html'}\n",
      "3 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up_step-by-step.html'}\n",
      "4 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up_linking-orcid.html'}\n",
      "5 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up_linking-accounts.html'}\n",
      "6 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_sign-up_auth-update.html'}\n",
      "7 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_browsers.html'}\n",
      "8 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_quick-start.html'}\n",
      "9 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative.html'}\n",
      "10 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_about.html'}\n",
      "11 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_access.html'}\n",
      "12 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_tour.html'}\n",
      "13 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_narratives.html'}\n",
      "14 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_create.html'}\n",
      "15 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_explore-data.html'}\n",
      "16 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_add-data.html'}\n",
      "17 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_add-apps.html'}\n",
      "18 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_analyze-data.html'}\n",
      "19 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_job-browser.html'}\n",
      "20 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_revise.html'}\n",
      "21 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_markdown.html'}\n",
      "22 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_share.html'}\n",
      "23 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_link-doi.html'}\n",
      "24 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_access-and-copy.html'}\n",
      "25 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_narrative_orgs.html'}\n",
      "26 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/getting-started_faq.html'}\n",
      "27 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data.html'}\n",
      "28 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide.html'}\n",
      "29 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_data-types.html'}\n",
      "30 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_uploads.html'}\n",
      "31 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_uploads_bulk-limitations.html'}\n",
      "32 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_assembly.html'}\n",
      "33 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_genome.html'}\n",
      "34 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_reads.html'}\n",
      "35 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_fba-model.html'}\n",
      "36 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_media.html'}\n",
      "37 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_expression-matrix.html'}\n",
      "38 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_phenotype-set.html'}\n",
      "39 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_amplicon-matrix.html'}\n",
      "40 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_chemical-abundance-matrix.html'}\n",
      "41 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_sampleset.html'}\n",
      "42 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_compressed-files.html'}\n",
      "43 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_bulk.html'}\n",
      "44 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_upload-download-guide_downloads.html'}\n",
      "45 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_search-add-upload.html'}\n",
      "46 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_filter-manage-view.html'}\n",
      "47 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_samples.html'}\n",
      "48 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_samples_ontology.html'}\n",
      "49 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_public.html'}\n",
      "50 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_globus.html'}\n",
      "51 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/data_jgi-transfer.html'}\n",
      "52 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps.html'}\n",
      "53 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis.html'}\n",
      "54 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_assembly-and-annotation.html'}\n",
      "55 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_comparative-genomics.html'}\n",
      "56 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_metabolic-modeling.html'}\n",
      "57 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_metagenomics.html'}\n",
      "58 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_matrix.html'}\n",
      "59 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_chemical-abundance.html'}\n",
      "60 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_analysis_expression.html'}\n",
      "61 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/apps_beta.html'}\n",
      "62 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows.html'}\n",
      "63 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_assembly-annotation.html'}\n",
      "64 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_assembly-annotation_faq-assembly-annotation.html'}\n",
      "65 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_comparative-genomics.html'}\n",
      "66 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_comparative-genomics_faq-comparative-genomics.html'}\n",
      "67 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metagenomic-analysis.html'}\n",
      "68 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metagenomic-analysis_faq-metagenomics.html'}\n",
      "69 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_rnaseq.html'}\n",
      "70 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_rnaseq_faq-rnaseq.html'}\n",
      "71 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metabolic-models.html'}\n",
      "72 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metabolic-models_metabolic-flux-models.html'}\n",
      "73 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/workflows_metabolic-models_faq-metabolic-modeling.html'}\n",
      "74 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting.html'}\n",
      "75 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_narrative.html'}\n",
      "76 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_support.html'}\n",
      "77 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_report.html'}\n",
      "78 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors.html'}\n",
      "79 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_common.html'}\n",
      "80 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_common_job-log.html'}\n",
      "81 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_import.html'}\n",
      "82 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_assembly.html'}\n",
      "83 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_annotation.html'}\n",
      "84 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_functional-genomics.html'}\n",
      "85 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/troubleshooting_job-errors_modeling.html'}\n",
      "86 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/development.html'}\n",
      "87 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/development_kbase-sdk.html'}\n",
      "88 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/development_create-a-kbase-developer-account.html'}\n",
      "89 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/development_github-repository.html'}\n",
      "90 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/kbase.us.html'}\n",
      "91 {'source': '/scratch/ac.pgupta/convLLM/kbasedocs/cdn-cgi_l_email-protection.html'}\n"
     ]
    }
   ],
   "source": [
    "for i, d in enumerate(data_html):\n",
    "    print(i, d.metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a1698632",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm = ChatOpenAI(temperature=0, model=\"gpt-4\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "27f3a2f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_splitter = RecursiveCharacterTextSplitter(\n",
    "    chunk_size = 8000,\n",
    "    chunk_overlap  = 20,\n",
    "    length_function = len,\n",
    ")\n",
    "documents_html = text_splitter.split_documents(data_html)\n",
    "len(documents_html)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3f3654d",
   "metadata": {},
   "source": [
    "### Few shot example prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9197810f",
   "metadata": {},
   "outputs": [],
   "source": [
    "few_shot_examples = [\n",
    "  {\n",
    "    \"Instruction\": \n",
    "                \"\"\"I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
    "                for each document provided, I need you to create 1-3 quiz questions. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
    "                The questions can be True or False questions or muliselection, generate all choices and the correct answers.\n",
    "                text is:\n",
    "                Analysis Apps in KBase\\n\\nKBase Apps are analysis tools that you can use in KBase. Apps interoperate seamlessly to enable a range of scientific workflows (see figure below). Some of the apps were written by KBase scientists and developers; others are third-party tools that were integrated into KBase with our\\n\\nSoftware Developer Kit (SDK)\\n\\n. The number of apps available in KBase increases as members of the community use our SDK to integrate their analysis tools into the KBase platform.\\n\\nThe App Catalog lists currently available Apps.\\n\\nExternal App Catalog (no user account or sign-in required for browsing).\\n\\nWithin KBase by clicking on the Catalog icon in the Menu\\n\\nIn a Narrative by clicking the right arrow at the top of the Apps Panel.\\n\\nThe majority of KBase Apps fall into the following categories:\\n\\n\\u200bAssembly\\u200b\\n\\n\\u200bAnnotation\\u200b\\n\\n\\u200bSequence Analysis\\u200b\\n\\n\\u200bComparative Genomics\\u200b\\n\\n\\u200bMetabolic Modeling\\u200b\\n\\n\\u200bExpression\\u200b\\n\\n\\u200bMicrobial Communities\\u200b\\n\\nShow me!\\nGo straight to the\\n\\nApp Catalog\\n\\n\\u200b\\u200b\\n\\nNote: you will need a\\n\\nKBase user account\\n\\nto use our tools.\\n\\nTips on using the App Catalog\\u200b\\n\\nEach app links to a reference page (which includes technical details about the inputs and outputs) called an App Details Page.\\n\\nTo run apps, you will need to sign in to the Narrative Interface.\\n\\nYou can access the App Catalog from inside the Narrative Interface by clicking the small arrow in the upper right corner of the Apps Panel.\\n\\nYou can click the star at the lower left of any app to add it to your “favorites.” The gray star will turn yellow to indicate that you have favorited the app. The number to the right of the star shows how many people have favorited that app.\\n\\nBy default, the apps are sorted by category. Try the options in the “Organize by” menu to sort the apps by My Favorites, Run Count and other options.\\n\\nRunning Analysis Apps\\n\\nOnce you have added an app to the Narrative, you will need to select a data object as the Input Object and set Parameters. When working within an application, a red bar or banner will highlight empty fields and errors.\\n\\nMore info\\n\\nFor more information about using the App Catalog, see the\\n\\nNarrative Interface User Guide\\n\\nPreviousUsing Apps\\n\\nNextAssembly & Annotation\\n\\nLast modified \\n\\n1yr ago\n",
    "                \"\"\",\n",
    "    \"Generated content\":\n",
    "    \"\"\"\n",
    "    ### Question: Which is Not in the majority of KBase Apps categories:\n",
    "    ## Choice: A. Assembly B.Annotation C. Expression D. Genome Reading\n",
    "    # Answer: D\n",
    "    ### Question: Is Comparison in the majority of KBase Apps categories?\n",
    "    ## Choice: A. True B False\n",
    "    # Answer: B\n",
    "    \"\"\"\n",
    "  },\n",
    "  {\n",
    "    \"Instruction\": \n",
    "                \"\"\"I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
    "                for each document provided, I need you to create 1-3 quiz questions. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
    "                The questions can be True or False questions or muliselection, generate all choices and the correct answers.\n",
    "                text is:\n",
    "                KBase Documentation\\n\\nHere you will find guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.\\n\\nHow to cite KBase:\\n\\nArkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163\\n\\nFunding:\\n\\nThis work is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase (KBase) funded by the\\n\\nU.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\\n\\nunder Award Numbers DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.\\n\\nNextKBase Terms & Conditions\\n\\nLast modified \\n\\n2yr ago\n",
    "                \"\"\",\n",
    "    \"Generated content\":\n",
    "    \"\"\"\n",
    "    ### Question: How is KBase funded?\n",
    "    ## Choice: A. KBase is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase\n",
    "               B. KBase is supported as part of the DOE Joint Genome Institute(JGI)\n",
    "               C. KBase is funded by Lawrence Berkeley National Laboratory\n",
    "               D. KBase is funded by University of California San Diego\n",
    "    # Answer: A\n",
    "    \"\"\"\n",
    "  }\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0261550c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Instruction: I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
      "                for each document provided, I need you to create 1-3 quiz questions. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
      "                The questions can be True or False questions or muliselection, generate all choices and the correct answers.\n",
      "                text is:\n",
      "                Analysis Apps in KBase\n",
      "\n",
      "KBase Apps are analysis tools that you can use in KBase. Apps interoperate seamlessly to enable a range of scientific workflows (see figure below). Some of the apps were written by KBase scientists and developers; others are third-party tools that were integrated into KBase with our\n",
      "\n",
      "Software Developer Kit (SDK)\n",
      "\n",
      ". The number of apps available in KBase increases as members of the community use our SDK to integrate their analysis tools into the KBase platform.\n",
      "\n",
      "The App Catalog lists currently available Apps.\n",
      "\n",
      "External App Catalog (no user account or sign-in required for browsing).\n",
      "\n",
      "Within KBase by clicking on the Catalog icon in the Menu\n",
      "\n",
      "In a Narrative by clicking the right arrow at the top of the Apps Panel.\n",
      "\n",
      "The majority of KBase Apps fall into the following categories:\n",
      "\n",
      "​Assembly​\n",
      "\n",
      "​Annotation​\n",
      "\n",
      "​Sequence Analysis​\n",
      "\n",
      "​Comparative Genomics​\n",
      "\n",
      "​Metabolic Modeling​\n",
      "\n",
      "​Expression​\n",
      "\n",
      "​Microbial Communities​\n",
      "\n",
      "Show me!\n",
      "Go straight to the\n",
      "\n",
      "App Catalog\n",
      "\n",
      "​​\n",
      "\n",
      "Note: you will need a\n",
      "\n",
      "KBase user account\n",
      "\n",
      "to use our tools.\n",
      "\n",
      "Tips on using the App Catalog​\n",
      "\n",
      "Each app links to a reference page (which includes technical details about the inputs and outputs) called an App Details Page.\n",
      "\n",
      "To run apps, you will need to sign in to the Narrative Interface.\n",
      "\n",
      "You can access the App Catalog from inside the Narrative Interface by clicking the small arrow in the upper right corner of the Apps Panel.\n",
      "\n",
      "You can click the star at the lower left of any app to add it to your “favorites.” The gray star will turn yellow to indicate that you have favorited the app. The number to the right of the star shows how many people have favorited that app.\n",
      "\n",
      "By default, the apps are sorted by category. Try the options in the “Organize by” menu to sort the apps by My Favorites, Run Count and other options.\n",
      "\n",
      "Running Analysis Apps\n",
      "\n",
      "Once you have added an app to the Narrative, you will need to select a data object as the Input Object and set Parameters. When working within an application, a red bar or banner will highlight empty fields and errors.\n",
      "\n",
      "More info\n",
      "\n",
      "For more information about using the App Catalog, see the\n",
      "\n",
      "Narrative Interface User Guide\n",
      "\n",
      "PreviousUsing Apps\n",
      "\n",
      "NextAssembly & Annotation\n",
      "\n",
      "Last modified \n",
      "\n",
      "1yr ago\n",
      "                \n",
      "Generated content: \n",
      "    ### Question: Which is Not in the majority of KBase Apps categories:\n",
      "    ## Choice: A. Assembly B.Annotation C. Expression D. Genome Reading\n",
      "    # Answer: D\n",
      "    ### Question: Is Comparison in the majority of KBase Apps categories?\n",
      "    ## Choice: A. True B False\n",
      "    # Answer: B\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "example_prompt = PromptTemplate(input_variables=[\"Instruction\", \"Generated content\"], template=\"Instruction: {Instruction}\\nGenerated content: {Generated content}\")\n",
    "\n",
    "print(example_prompt.format(**few_shot_examples[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7c6438d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Instruction: I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
      "                for each document provided, I need you to create 1-3 quiz questions. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
      "                The questions can be True or False questions or muliselection, generate all choices and the correct answers.\n",
      "                text is:\n",
      "                Analysis Apps in KBase\n",
      "\n",
      "KBase Apps are analysis tools that you can use in KBase. Apps interoperate seamlessly to enable a range of scientific workflows (see figure below). Some of the apps were written by KBase scientists and developers; others are third-party tools that were integrated into KBase with our\n",
      "\n",
      "Software Developer Kit (SDK)\n",
      "\n",
      ". The number of apps available in KBase increases as members of the community use our SDK to integrate their analysis tools into the KBase platform.\n",
      "\n",
      "The App Catalog lists currently available Apps.\n",
      "\n",
      "External App Catalog (no user account or sign-in required for browsing).\n",
      "\n",
      "Within KBase by clicking on the Catalog icon in the Menu\n",
      "\n",
      "In a Narrative by clicking the right arrow at the top of the Apps Panel.\n",
      "\n",
      "The majority of KBase Apps fall into the following categories:\n",
      "\n",
      "​Assembly​\n",
      "\n",
      "​Annotation​\n",
      "\n",
      "​Sequence Analysis​\n",
      "\n",
      "​Comparative Genomics​\n",
      "\n",
      "​Metabolic Modeling​\n",
      "\n",
      "​Expression​\n",
      "\n",
      "​Microbial Communities​\n",
      "\n",
      "Show me!\n",
      "Go straight to the\n",
      "\n",
      "App Catalog\n",
      "\n",
      "​​\n",
      "\n",
      "Note: you will need a\n",
      "\n",
      "KBase user account\n",
      "\n",
      "to use our tools.\n",
      "\n",
      "Tips on using the App Catalog​\n",
      "\n",
      "Each app links to a reference page (which includes technical details about the inputs and outputs) called an App Details Page.\n",
      "\n",
      "To run apps, you will need to sign in to the Narrative Interface.\n",
      "\n",
      "You can access the App Catalog from inside the Narrative Interface by clicking the small arrow in the upper right corner of the Apps Panel.\n",
      "\n",
      "You can click the star at the lower left of any app to add it to your “favorites.” The gray star will turn yellow to indicate that you have favorited the app. The number to the right of the star shows how many people have favorited that app.\n",
      "\n",
      "By default, the apps are sorted by category. Try the options in the “Organize by” menu to sort the apps by My Favorites, Run Count and other options.\n",
      "\n",
      "Running Analysis Apps\n",
      "\n",
      "Once you have added an app to the Narrative, you will need to select a data object as the Input Object and set Parameters. When working within an application, a red bar or banner will highlight empty fields and errors.\n",
      "\n",
      "More info\n",
      "\n",
      "For more information about using the App Catalog, see the\n",
      "\n",
      "Narrative Interface User Guide\n",
      "\n",
      "PreviousUsing Apps\n",
      "\n",
      "NextAssembly & Annotation\n",
      "\n",
      "Last modified \n",
      "\n",
      "1yr ago\n",
      "                \n",
      "Generated content: \n",
      "    ### Question: Which is Not in the majority of KBase Apps categories:\n",
      "    ## Choice: A. Assembly B.Annotation C. Expression D. Genome Reading\n",
      "    # Answer: D\n",
      "    ### Question: Is Comparison in the majority of KBase Apps categories?\n",
      "    ## Choice: A. True B False\n",
      "    # Answer: B\n",
      "    \n",
      "\n",
      "Instruction: I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
      "                for each document provided, I need you to create 1-3 quiz questions. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
      "                The questions can be True or False questions or muliselection, generate all choices and the correct answers.\n",
      "                text is:\n",
      "                KBase Documentation\n",
      "\n",
      "Here you will find guidance for getting started with KBase, full documentation for users and App developers, and help with troubleshooting.\n",
      "\n",
      "How to cite KBase:\n",
      "\n",
      "Arkin AP, Cottingham RW, Henry CS, Harris NL, Stevens RL, Maslov S, et al. KBase: The United States Department of Energy Systems Biology Knowledgebase. Nature Biotechnology. 2018;36: 566. doi: 10.1038/nbt.4163\n",
      "\n",
      "Funding:\n",
      "\n",
      "This work is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase (KBase) funded by the\n",
      "\n",
      "U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\n",
      "\n",
      "under Award Numbers DE-AC02-05CH11231, DE-AC02-06CH11357, DE-AC05-00OR22725, and DE-AC02-98CH10886.\n",
      "\n",
      "NextKBase Terms & Conditions\n",
      "\n",
      "Last modified \n",
      "\n",
      "2yr ago\n",
      "                \n",
      "Generated content: \n",
      "    ### Question: How is KBase funded?\n",
      "    ## Choice: A. KBase is supported as part of the Genomic Sciences Program DOE Systems Biology Knowledgebase\n",
      "               B. KBase is supported as part of the DOE Joint Genome Institute(JGI)\n",
      "               C. KBase is funded by Lawrence Berkeley National Laboratory\n",
      "               D. KBase is funded by University of California San Diego\n",
      "    # Answer: A\n",
      "    \n",
      "\n",
      "Instruction:  I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
      "                for each document provided, I need you to create 1-3 quiz questions. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
      "                The questions can be True or False questions or muliselection, generate all choices and the correct answers.\n",
      "                text is:\n"
     ]
    }
   ],
   "source": [
    "instruction_prompt=\"\"\"I have some KBase related docs retrieved from KBase Webpage by langchain html loader. The Department of Energy Systems Biology Knowledgebase (KBase) is a software and data science platform designed to meet the grand challenge of systems biology: predicting and designing biological function.\n",
    "                for each document provided, I need you to create 1-3 quiz questions. Be noted, the texts are converted from html file, so some content like the tables are messed up in one line. please try the best to figure this out.\n",
    "                The questions can be True or False questions or muliselection, generate all choices and the correct answers.\n",
    "                text is:\"\"\"\n",
    "prompt = FewShotPromptTemplate(\n",
    "    examples=few_shot_examples,\n",
    "    example_prompt=example_prompt,\n",
    "    suffix=\"Instruction:  {input}\",\n",
    "    input_variables=[\"input\"]\n",
    ")\n",
    "prompt\n",
    "print(prompt.format(input=instruction_prompt))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fd314096",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generateQA(instruction_prompt, text):\n",
    "    system_message = SystemMessage(content=prompt.format(input=instruction_prompt))\n",
    "    human_message = HumanMessage(content=text)\n",
    "    res = llm([system_message, human_message])\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "069bab0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "answers = generateQA(instruction_prompt, data_html[0].page_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ac13055f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated content: \n",
      "    ### Question: What can you find in the KBase Documentation?\n",
      "    ## Choice: A. Guidance for getting started with KBase\n",
      "               B. Full documentation for users and App developers\n",
      "               C. Help with troubleshooting\n",
      "               D. All of the above\n",
      "    # Answer: D\n",
      "    ### Question: Who funds the work of KBase?\n",
      "    ## Choice: A. U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\n",
      "               B. National Institutes of Health\n",
      "               C. National Science Foundation\n",
      "               D. Department of Defense\n",
      "    # Answer: A\n",
      "    ### Question: Was the KBase documentation last modified 2 years ago?\n",
      "    ## Choice: A. True B. False\n",
      "    # Answer: A\n"
     ]
    }
   ],
   "source": [
    "print(answers.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9943f0ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'What can you find in the KBase Documentation?'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "answers.content.split(\"###\")[1].split(\"##\")[0].strip()[10:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bbc42eec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A. Guidance for getting started with KBase\\n               B. Full documentation for users and App developers\\n               C. Help with troubleshooting\\n               D. All of the above'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "answers.content.split(\"###\")[1].split(\"##\")[1].split(\"#\")[0].strip()[8:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3c405a41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'D'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "answers.content.split(\"###\")[1].split(\"##\")[1].split(\"#\")[1].strip()[8:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "eebdc025",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'question': 'What can you find in the KBase Documentation?',\n",
       "  'choices': 'A. Guidance for getting started with KBase\\n               B. Full documentation for users and App developers\\n               C. Help with troubleshooting\\n               D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Who funds the work of KBase?',\n",
       "  'choices': 'A. U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\\n               B. National Institutes of Health\\n               C. National Science Foundation\\n               D. Department of Defense',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Was the KBase documentation last modified 2 years ago?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'}]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extractQA(answers):\n",
    "    answers = answers.content.split(\"###\")\n",
    "    res = []\n",
    "    for i in range(1, len(answers)):\n",
    "        q = answers[i].split(\"##\")[0].strip()[10:]\n",
    "        c = answers[i].split(\"##\")[1].split(\"#\")[0].strip()[8:]\n",
    "        a = answers[i].split(\"##\")[1].split(\"#\")[1].strip()[8:]\n",
    "        res.append({'question': q, \"choices\": c, \"answer\": a})\n",
    "    return res\n",
    "extractQA(answers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "dcb9cf3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'question': 'What can you find in the KBase Documentation?',\n",
       "  'choices': 'A. Guidance for getting started with KBase\\n               B. Full documentation for users and App developers\\n               C. Help with troubleshooting\\n               D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Who funds the work of KBase?',\n",
       "  'choices': 'A. U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research\\n               B. National Institutes of Health\\n               C. National Science Foundation\\n               D. Department of Defense',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Was the KBase Documentation last modified 2 years ago?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the first thing you can find in the \"Getting Started\" section of KBase?',\n",
       "  'choices': 'A. KBase Supported Browsers B. Signing Up and Signing In C. A Quick Start Guide to KBase D. KBase Narrative Interface User Guide',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Is there a FAQ section in the \"Getting Started\" section of KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is not included in the \"Getting Started\" section of KBase?',\n",
       "  'choices': 'A. KBase Supported Browsers B. Signing Up and Signing In C. A Quick Start Guide to KBase D. KBase App Developer Guide',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Which of the following is NOT a supported identity provider for creating a KBase account?',\n",
       "  'choices': 'A. Google B. ORCiD C. Globus D. Facebook',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Does KBase manage its own passwords?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Is KBase free for all to use, even if you are not in the United States?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the steps to sign up for KBase?',\n",
       "  'choices': 'A. Click the “Sign Up” button on the KBase Narrative login page, use your existing Google, ORCiD, or Globus account to sign up for KBase, enter some required information, agree to the KBase User Agreement and KBase Data Policy, and then create your KBase account.\\n               B. Click the “Sign Up” button on the KBase Narrative login page, use your existing Facebook or Twitter account to sign up for KBase, enter some required information, agree to the KBase User Agreement and KBase Data Policy, and then create your KBase account.\\n               C. Click the “Sign Up” button on the KBase Narrative login page, use your existing email account to sign up for KBase, enter some required information, agree to the KBase User Agreement and KBase Data Policy, and then create your KBase account.\\n               D. Click the “Sign Up” button on the KBase Narrative login page, use your existing LinkedIn account to sign up for KBase, enter some required information, agree to the KBase User Agreement and KBase Data Policy, and then create your KBase account.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What do you need to do before creating a KBase account?',\n",
       "  'choices': 'A. Agree to the KBase User Agreement and KBase Data Policy\\n               B. Agree to the KBase Privacy Policy\\n               C. Agree to the KBase Terms and Conditions\\n               D. Agree to the KBase Data Sharing Policy',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you sign up for KBase with a Facebook account?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can you sign into KBase using your ORCiD account?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the first step to link your ORCiD account to KBase if you are already a KBase User?',\n",
       "  'choices': 'A. Sign into KBase through your Globus or Google accounts\\n               B. Go to the sign-in page and click the Sign In button\\n               C. On the left sidebar menu, click the “Account” icon\\n               D. Choose \"ORCiD” from the pulldown list of identity providers',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What should you do after choosing \"ORCiD” from the pulldown list of identity providers?',\n",
       "  'choices': 'A. Sign in using your ORCiD account to link\\n               B. Click the “Link” button\\n               C. You will be returned to the Linked Sign-in Accounts screen in your Account Manager\\n               D. You will see your newly linked ORCiD account',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Which accounts can be linked to KBase?',\n",
       "  'choices': 'A. Google B. Globus C. ORCiD D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Can you link more than one Google account to KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'After linking your Google account to KBase, you can sign in to KBase with your Google credentials.',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the three ways you can sign into KBase?',\n",
       "  'choices': 'A. Google, Globus, ORCiD B. Google, Facebook, Twitter C. Google, Microsoft, Apple D. Google, Yahoo, Bing',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Do you need to sign into Globus every time you use KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What do you need to do the first time you sign into KBase after the authentication change?',\n",
       "  'choices': 'A. You need to create a new account B. You need to convert your existing KBase account to the new system C. You need to link your Google account D. You need to sign in with your email address',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Which browser is primarily supported by KBase?',\n",
       "  'choices': 'A. Chrome B. Firefox C. Safari D. Microsoft Edge',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is KBase supported on mobile browsers?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Which browser is not supported by KBase?',\n",
       "  'choices': 'A. Chrome B. Internet Explorer C. Safari D. Microsoft Edge',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is a Narrative in KBase?',\n",
       "  'choices': 'A. A Narrative in KBase is a shareable, reproducible workflow that includes data, analysis steps, results, visualizations, and commentary.\\n               B. A Narrative in KBase is a type of data analysis tool.\\n               C. A Narrative in KBase is a type of data storage system.\\n               D. A Narrative in KBase is a type of user account.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you add data to your Narrative in KBase?',\n",
       "  'choices': 'A. By clicking the “Add Data” button in the Data Panel.\\n               B. By clicking the “New Narrative” button in Narratives.\\n               C. By clicking the “Run” button near the top left of the app cell.\\n               D. By clicking the “share” button near the top right of your Narrative.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Are all Narratives that you create in KBase private by default?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the first topic in the Narrative Interface User Guide?',\n",
       "  'choices': 'A. About B. Accessing the Narrative C. Tour the Narrative D. Narratives Navigator',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is \"Create an Org\" a topic in the Narrative Interface User Guide?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the last topic in the Narrative Interface User Guide?',\n",
       "  'choices': 'A. Share a Narrative B. Access and copy Narratives C. Create an Org D. Formatting Markdown in the Narrative',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What is the purpose of the Narrative Interface in KBase?',\n",
       "  'choices': 'A. It enables researchers to design and carry out computational experiments while creating interactive, reproducible records of the data, computational steps, and thought processes underpinning their results.\\n               B. It is a tool for data visualization only.\\n               C. It is a tool for data storage only.\\n               D. It is a tool for data analysis only.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is the Narrative Interface built on top of the Jupyter Notebook framework?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can the records created by the Narrative Interface be shared with collaborators and published as “active papers,” called Narratives?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you access the Narrative Interface from KBase?',\n",
       "  'choices': 'A. Use the menu at the top left of the landing page and select “Narrative Interface”\\n               B. Open a public Narrative or one that has been shared with you\\n               C. Start a new Narrative by clicking the “+ New Narrative” button\\n               D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What should you do if a new version of the Narrative Interface is available?',\n",
       "  'choices': 'A. Ignore it\\n               B. Click the green button near the top right of your Narrative that says “New Version Available”\\n               C. Contact KBase support\\n               D. Uninstall and reinstall KBase',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Is the Narrative Interface automatically updated to the latest version in most cases?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the Share button in the Narrative controls do?',\n",
       "  'choices': 'A. It allows you to make your Narrative public or give specific KBase users the ability to view, edit, or share it further.\\n               B. It allows you to save your work.\\n               C. It allows you to sign in/out or access your user profile.\\n               D. It allows you to link to documentation as well as to the Narrative Tour.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the Data Panel in the Analyze tab do?',\n",
       "  'choices': 'A. It allows you to create a new Narrative, copy an existing Narrative, and see a list of other Narratives created by or shared with you.\\n               B. It allows you to add data to your Narrative by searching KBase’s reference collection, importing datasets from external sources, or generating new data from KBase analyses.\\n               C. It allows you to add rich text commentary and graphics to your Narrative.\\n               D. It allows you to list KBase analysis functions.',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What does the App Controls do?',\n",
       "  'choices': 'A. It allows you to search and refresh the list of apps, as well as expose apps that are available but still in beta.\\n               B. It allows you to add data to your Narrative by searching KBase’s reference collection, importing datasets from external sources, or generating new data from KBase analyses.\\n               C. It allows you to create a new Narrative, copy an existing Narrative, and see a list of other Narratives created by or shared with you.\\n               D. It allows you to add rich text commentary and graphics to your Narrative.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the five tabs for exploring and locating existing Narratives in KBase?',\n",
       "  'choices': 'A. My Narratives, Shared with Me, Tutorials, Public, and Private\\n               B. My Narratives, Shared with Me, Tutorials, Public, and Data\\n               C. My Narratives, Shared with Me, Tutorials, Public, and Jobs\\n               D. My Narratives, Shared with Me, Tutorials, Public, and Account',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the Job Browser allow you to do?',\n",
       "  'choices': 'A. View apps that you have run recently, see their progress, search for specific runs, and more\\n               B. Edit your profile information\\n               C. View notifications from KBase\\n               D. Search for data present in your own Narratives',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What can you do in the Account Manager?',\n",
       "  'choices': 'A. View apps that you have run recently\\n               B. Edit your profile information\\n               C. View notifications from KBase\\n               D. Search for data present in your own Narratives',\n",
       "  'answer': 'B'},\n",
       " {'question': 'How can you create a new Narrative in KBase?',\n",
       "  'choices': 'A. By clicking the “+ New Narrative” button in your Narratives\\n               B. By selecting the Narratives tab in the Narrative Interface and clicking the button labeled “+ New Narrative.”\\n               C. By clicking the \"Create New\" button on the homepage\\n               D. By selecting the \"New Narrative\" option in the \"File\" menu',\n",
       "  'answer': 'A, B'},\n",
       " {'question': 'When will the new Narrative appear in your Narratives list?',\n",
       "  'choices': \"A. Immediately after you create it\\n               B. After you name it something other than 'Untitled' and save it\\n               C. After you add data to it\\n               D. After you share it with other users\",\n",
       "  'answer': 'B'},\n",
       " {'question': 'Are the Narratives you create private by default?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the different ways to explore data available in KBase?',\n",
       "  'choices': 'A. Using the Data Browser slide-out B. Using the Data Panel C. Using Data Landing and Provenance pages D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What does the binoculars icon represent in the Data Browser panel?',\n",
       "  'choices': 'A. It represents the \"Explore data\" button B. It represents the \"View associated report\" button C. It represents the \"View data provenance and relationships\" button D. It represents the \"Export/Download data\" button',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the trash can icon represent in the Data Panel?',\n",
       "  'choices': 'A. It represents the \"Delete data\" button B. It represents the \"Rename data\" button C. It represents the \"Export/Download data\" button D. It represents the \"View data provenance and relationships\" button',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of Provenance pages in KBase?',\n",
       "  'choices': 'A. To facilitate reproducibility and transparency of scientific results\\n               B. To provide a platform for data analysis\\n               C. To store all the data in KBase\\n               D. To provide a platform for data visualization',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you identify the original “owners” of the data from the Provenance page?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can older versions of a data object be accessed from the Provenance page?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of the Staging Area in KBase?',\n",
       "  'choices': 'A. It is a temporary holding place for your uploaded data files. B. It is a permanent storage for your data files. C. It is a place to store your favorite apps. D. It is a place to store your narratives.',\n",
       "  'answer': 'A'},\n",
       " {'question': \"What is the difference between 'Upload' and 'Import' in KBase?\",\n",
       "  'choices': 'A. Upload refers to getting data from your computer into your Staging Area and Import refers to the process of converting a data file from your Staging Area into a data object in your Narrative. B. Upload refers to the process of converting a data file from your Staging Area into a data object in your Narrative and Import refers to getting data from your computer into your Staging Area. C. There is no difference between Upload and Import in KBase. D. Upload refers to adding apps to your narrative and Import refers to removing apps from your narrative.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How long do files stay in your Staging Area in KBase?',\n",
       "  'choices': 'A. Indefinitely B. 30 days C. 60 days D. 90 days',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What is one of the most common causes of import failure in KBase?',\n",
       "  'choices': 'A. Attempting to import a file that’s the wrong data type or not the expected format for that data type\\n               B. Network issues during the import process\\n               C. The file size is too large for import\\n               D. The file is corrupted or damaged',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can multiple files be imported together in a single Import App in KBase?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What happens when you delete a file from your Staging Area in KBase?',\n",
       "  'choices': 'A. The file is permanently deleted and cannot be recovered\\n               B. The file is moved to the trash bin and can be restored\\n               C. The file is deleted from the Staging Area, but any data object imported from the file in your Narrative remains\\n               D. All data objects imported from the file in your Narrative are also deleted',\n",
       "  'answer': 'C'},\n",
       " {'question': 'How can you browse KBase Apps?',\n",
       "  'choices': 'A. Scrolling through the list in the Apps Panel\\n               B. Searching for an app by name\\n               C. Using the App Catalog\\n               D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What does the number beside the star icon in the App Catalog indicate?',\n",
       "  'choices': 'A. The number of times the app has been run\\n               B. The number of users who have selected the app as a favorite\\n               C. The number of users who have used the app\\n               D. The number of times the app has been updated',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What does the \"B\" in the Apps Panel indicate?',\n",
       "  'choices': 'A. The app is in Beta\\n               B. The app is in the Best category\\n               C. The app is in the Basic category\\n               D. The app is in the Business category',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the three tabs in an App Cell?',\n",
       "  'choices': 'A. Configure, Info, Job Status B. Configure, Info, Result C. Configure, Job Status, Result D. Info, Job Status, Result',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What does the green \"Run\" button indicate?',\n",
       "  'choices': 'A. The app is currently running B. All required fields are filled in C. The app has finished running D. The app has encountered an error',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What happens when you rerun an App with the same output object name?',\n",
       "  'choices': 'A. The App will not run B. The information in the App cells will be overwritten C. The data objects in your Narrative will be overwritten D. Both B and C',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What is the default timeframe for jobs shown in the Job Browser?',\n",
       "  'choices': 'A. Last 24 hours B. Last 48 hours C. Last week D. Last month',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What happens when you click on the Narrative name in the Job Browser?',\n",
       "  'choices': 'A. It opens a new tab with the Narrative B. It opens the catalog page for the app C. It expands the table to show the contents of the log D. It prompts you to cancel the job',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can Job Logs be downloaded in CSV, TSV, JSON, and TEXT formats?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the function of the first icon (counterclockwise encircling arrow) that appears when you hover your cursor over one of your Narratives in the Narratives panel?',\n",
       "  'choices': 'A. It allows you to delete the Narrative B. It allows you to rename the Narrative C. It allows you to view the Narrative history and revert changes D. It allows you to share the Narrative',\n",
       "  'answer': 'C'},\n",
       " {'question': 'How can you add text to your Narrative?',\n",
       "  'choices': 'A. By adding text cells in the \"Markdown Cells\" B. By adding text cells in the \"LaTeX Cells\" C. By adding text cells in the \"HTML Cells\" D. By adding text cells in the \"CSS Cells\"',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you rename a Narrative?',\n",
       "  'choices': 'A. By clicking on its name at the top of the Narrative window B. By clicking the trashcan icon C. By clicking the \"Add Markdown Cell\" button D. By clicking the \"…\" icon in the upper right corner of a cell',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the uses of Markdown cells in KBase?',\n",
       "  'choices': 'A. To edit the appearance in Markdown cells for publication, teaching, and to add navigation within the Narrative\\n               B. To explain your process, interpret results, and communicate to others for sharing and publication and more\\n               C. To run apps and analysis tools\\n               D. Both A and B',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Does every formatting option work in the Narrative Interface?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What languages do Markdown cells use?',\n",
       "  'choices': 'A. Markdown and HTML B. Python and HTML C. Markdown and Python D. HTML and CSS',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you share your KBase Narrative with everyone?',\n",
       "  'choices': 'A. Click the “share” button near the top right and select the \"Manage Sharing\" option in the menu, then click the “make public” link near the top of the sharing panel.\\n               B. Click the “share” button near the top right and select the \"Manage Sharing\" option in the menu, then click the “make private” link near the top of the sharing panel.\\n               C. Click the “share” button near the top right and select the \"Manage Sharing\" option in the menu, then click the “make public” link near the bottom of the sharing panel.\\n               D. Click the “share” button near the top right and select the \"Manage Sharing\" option in the menu, then click the “make private” link near the bottom of the sharing panel.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the default permission level for sharing a Narrative?',\n",
       "  'choices': 'A. Edit only B. View and Edit C. View only D. Edit and Share',\n",
       "  'answer': 'C'},\n",
       " {'question': 'Can you change the access permissions after sharing a Narrative?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the first step to add KBase static Narrative DOIs to your ORCID account?',\n",
       "  'choices': \"A. Scroll down to the Works section of your ORCID record and hover over the '+Add works' in the menu and select 'Search & link'\\n               B. Sign in to ORCID\\n               C. Under 'Link Works', select DOE / OSTI from the list of member organizations.\\n               D. Authorize access to approve DOE / OSTI to make changes to your ORCID Works.\",\n",
       "  'answer': 'B'},\n",
       " {'question': 'What will appear once DOE / OSTI is authorized?',\n",
       "  'choices': 'A. The OSTI.GOV page will appear with the green ORCID iD icon next to the user’s ORCID iD name\\n               B. A pop-up text box will appear and confirm that the user wants to add this record to ORCID Works.\\n               C. The user name next to the green ORCID iD icon at the top of the page.\\n               D. The ORCID Works in OSTI.GOV to see the works integrated with your account.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What will display in the search results, once works are added by the user to their ORCID record?',\n",
       "  'choices': \"A. 'In your ORCID Works' at the bottom right of the popup.\\n               B. The green ORCID iD icon next to the user’s ORCID iD name\\n               C. The user name next to the green ORCID iD icon at the top of the page.\\n               D. The ORCID Works in OSTI.GOV to see the works integrated with your account.\",\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the key difference when a Narrative opens in View-only mode?',\n",
       "  'choices': 'A. There is no Analyze panel on the left side B. There is no Data Panel on the left side C. There is no Narratives list on the left side D. There is no Apps Panel on the left side',\n",
       "  'answer': 'A'},\n",
       " {'question': \"Why can't you run a Narrative owned by someone else?\",\n",
       "  'choices': \"A. Because running an app usually creates new data objects in your Narrative, which requires write permission B. Because you don't have the access to the data objects in the Narrative C. Because you don't have the access to the apps in the Narrative D. Because you don't have the access to the Analyze panel in the Narrative\",\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you run or change a Narrative that has been shared with you?',\n",
       "  'choices': 'A. You can request the owner to give you write permission B. You can copy the Narrative to your account C. You can request the owner to give you edit permission D. You can request the owner to give you run permission',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is a KBase Org?',\n",
       "  'choices': 'A. KBase Org is a way for teams of scientists to share their data and associated analyses that are in the Narratives they create with each other as a group.\\n               B. KBase Org is a platform for individual scientists to share their data and associated analyses.\\n               C. KBase Org is a tool for data analysis in KBase.\\n               D. KBase Org is a database for storing scientific data.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can a KBase user be a member of more than one organization?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can a Narrative with the associated data be added to more than one organization?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is there a mechanism to delete an organization in KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Is KBase free to use?',\n",
       "  'choices': 'A. Yes B. No',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What languages are supported in code cells in KBase?',\n",
       "  'choices': 'A. Python B. Java C. C++ D. Ruby',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you get a tool added to KBase?',\n",
       "  'choices': 'A. Submit a Help Desk New Feature Request B. Add as a 3rd-party developer C. Both A and B D. None of the above',\n",
       "  'answer': 'C'},\n",
       " {'question': 'Is user data in KBase public by default?',\n",
       "  'choices': 'A. Yes B. No',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can human data be uploaded in KBase?',\n",
       "  'choices': 'A. Yes B. No',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can you access KBase methods from the command line?',\n",
       "  'choices': 'A. Yes B. No',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can you see the background code for KBase?',\n",
       "  'choices': 'A. Yes B. No',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the difference between Public and Private organizations in KBase?',\n",
       "  'choices': 'A. The existence and contents of Private organizations are only visible to members of the organization. The organization will not be findable by non-members. Any KBase user can see a list of all Public organizations plus a description of the organization. Non-members are given the option to request membership. Members of the organization can access more information than non-members.\\n               B. There is no difference between Public and Private organizations in KBase.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What should you do if you lose access to your KBase account?',\n",
       "  'choices': 'A. Submit a ticket through the Help Board B. Contact KBase by email C. Both A and B D. None of the above',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is not a guide for working with data in KBase?',\n",
       "  'choices': 'A. How to search, add and upload data in KBase video tutorial\\n               B. Using public data within KBase\\n               C. How to transfer data using Globus\\n               D. How to analyze data in KBase',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Is there a guide on how to transfer data from JGI in KBase?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the guide about?',\n",
       "  'choices': 'A. Data Upload and Download Guide B. Data Analysis Guide C. Data Visualization Guide D. Data Management Guide',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Which of the following is not a content in the guide?',\n",
       "  'choices': 'A. Data Types B. Importing Data C. Assemblies D. Data Cleaning',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Is there a section about Compressed Files in the guide?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the file extension for Assembly data type in KBase?',\n",
       "  'choices': 'A. .fasta, .fna, .fa, .fas B. .fastq, .fq, .sra C. .genbank, .gb, .gbk, or .gbff, .gff, and .fasta, .fna, .fa, .fas D. .sbml, .xml, .tsv, .xls, .xlsx',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the file extension for FBA Model data type in KBase?',\n",
       "  'choices': 'A. .fasta, .fna, .fa, .fas B. .fastq, .fq, .sra C. .genbank, .gb, .gbk, or .gbff, .gff, and .fasta, .fna, .fa, .fas D. .sbml, .xml, .tsv, .xls, .xlsx',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Does KBase handle data as files for interoperability with KBase Apps?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the size limit for drag & drop data files from your computer into the Staging Area in KBase?',\n",
       "  'choices': 'A. 1GB B. 5GB C. 10GB D. 20GB',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the privacy policy for data uploaded to KBase?',\n",
       "  'choices': 'A. All data uploaded to KBase is public B. All data uploaded to KBase is private unless you choose to share it C. All data uploaded to KBase is shared with specific users D. All data uploaded to KBase is shared with all KBase users',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What types of data does the bulk import currently support in KBase?',\n",
       "  'choices': 'A. Assembly - FASTA, SRA Reads, FASTQ Reads Interleaved - FASTQ Interleaved reads, FASTQ Reads Noninterleaved - FASTQ paired-end reads or FASTQ single-end reads library, GenBank Genome - GenBank, GFF Metagenome - FASTA and GFF3\\n               B. Assembly - FASTA, SRA Reads, FASTQ Reads Interleaved - FASTQ Interleaved reads, FASTQ Reads Noninterleaved - FASTQ paired-end reads or FASTQ single-end reads library, GenBank Genome - GenBank\\n               C. Assembly - FASTA, SRA Reads, FASTQ Reads Interleaved - FASTQ Interleaved reads, FASTQ Reads Noninterleaved - FASTQ paired-end reads or FASTQ single-end reads library\\n               D. Assembly - FASTA, SRA Reads',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is a known issue with the Staging Area in KBase?',\n",
       "  'choices': 'A. It does not allow uploading of incomplete files\\n               B. It allows uploading of incomplete files, leading to data corruption\\n               C. It does not allow uploading of files larger than 1GB\\n               D. It does not support uploading of multiple files at once',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is a workaround for the issue of only a single set of parameters being used for a given type in a given bulk import?',\n",
       "  'choices': 'A. Run separate imports for each subset of data with a single set of parameters\\n               B. Use different parameters for each data type\\n               C. Use the same parameters for all data types\\n               D. There is no workaround for this issue',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is a known issue with Job Status/Logs/Results in KBase?',\n",
       "  'choices': 'A. They do not appear or do not update\\n               B. They appear but are not accurate\\n               C. They update but do not save\\n               D. They appear but cannot be downloaded',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the KBase data type for assembled, unannotated DNA sequence contigs or predicted coding sequences?',\n",
       "  'choices': 'A. Assembly B. Genome C. FASTA D. GenBank',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the file extensions that can be uploaded to KBase as an assembly file?',\n",
       "  'choices': 'A. .fasta, .fna, .fa, .fas B. .txt, .doc, .xls, .ppt C. .jpg, .png, .gif, .bmp D. .mp3, .mp4, .avi, .mkv',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you upload data into your KBase Staging Area using Globus?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you import multiple assemblies simultaneously from the staging area?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you transfer public genome reads and assemblies from JGI to your KBase account?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What types of files does the Genome importer in KBase support?',\n",
       "  'choices': 'A. GenBank and GFF formatted files B. FASTA files C. CSV files D. TXT files',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the function of the Genome object in KBase?',\n",
       "  'choices': 'A. It is the annotated version of an Assembly or annotated predicted coding sequences B. It is used to store the DNA sequence from a FASTA file C. It is used to store the taxonomy information for the organism D. It is used to store the sequence contig(s)',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can GenBank-formatted files with no features be uploaded as Genomes in KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the two types of reads library data objects in KBase?',\n",
       "  'choices': 'A. SingleEndLibrary and DoubleEndLibrary B. SingleEndLibrary and PairedEndLibrary C. PairedEndLibrary and DoubleEndLibrary D. SingleEndLibrary and TripleEndLibrary',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What are the two ways that KBase and GenBank SRA recognize a paired-end library?',\n",
       "  'choices': 'A. A paired-end library can be either two files which typically have the same name and are designated as forward and reverse or a single interleaved file. B. A paired-end library can be either two files which typically have different names and are designated as forward and reverse or a single interleaved file. C. A paired-end library can be either two files which typically have the same name and are designated as forward and reverse or a double interleaved file. D. A paired-end library can be either three files which typically have the same name and are designated as forward and reverse or a single interleaved file.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you import reads into KBase using Globus Online?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the file formats that can be used to upload an FBA model into KBase?',\n",
       "  'choices': 'A. SBML B. Excel C. TSV D. PDF',\n",
       "  'answer': 'A, B, C'},\n",
       " {'question': 'What is the requirement for the Excel file when uploading an FBA model in the Excel format?',\n",
       "  'choices': 'A. The file should be named as \"ModelCompounds\" and \"ModelReactions\" B. The tables containing the information about the chemical compounds and reactions in an FBA model are stored in two separate tabs respectively, named as “ModelCompounds” and “ModelReactions” C. The file should be saved as .xls extension D. All of the above',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the requirement for the TSV file when uploading an FBA model in the TSV format?',\n",
       "  'choices': 'A. The file should be named as \"FBAModelCompounds.tsv\" and “FBAModelReactions.tsv” B. The chemical compounds and reactions tables are saved as two separate files C. The file should be saved as .tsv extension D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What is the requirement for importing an FBA model into KBase?',\n",
       "  'choices': 'A. The corresponding Genome needs to be linked to the FBA Model B. The gene IDs between the model and genome need to match C. Both A and B D. None of the above',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What is the size limit for drag & drop upload from your local computer?',\n",
       "  'choices': 'A. 10GB B. 20GB C. 30GB D. 40GB',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What are the four columns required when uploading a custom media file in KBase?',\n",
       "  'choices': 'A. Maxflux, Minflux, Concentration, Compound identifier B. Maxflux, Minflux, Compound name, Compound identifier C. Maxflux, Minflux, Concentration, Compound name D. Maxflux, Minflux, Concentration, Compound ID',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does a negative flux value correspond to in KBase?',\n",
       "  'choices': 'A. Uptake B. Excretion C. Neither D. Both',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What should be the name of the worksheet that contains your media conditions in an Excel file for KBase?',\n",
       "  'choices': 'A. MediaCompounds B. MediaConditions C. MediaSheet D. MediaData',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What file extensions can be used to upload and import Expression Matrices in KBase?',\n",
       "  'choices': 'A. .tsv and .xls B. .csv and .xlsx C. .txt and .docx D. .pdf and .doc',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can Expression Matrices be uploaded into KBase without a Genome?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the first label in the first column of a TSV file for Expression Matrix in KBase?',\n",
       "  'choices': 'A. feature_ids B. gene_ids C. sample_ids D. expression_ids',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does a Phenotype Set data type represent?',\n",
       "  'choices': 'A. Experimental data about an organism’s ability to grow in specific media conditions\\n               B. Genetic information of an organism\\n               C. Metabolic pathways of an organism\\n               D. Genetic mutations in an organism',\n",
       "  'answer': 'A'},\n",
       " {'question': \"What does the 'Growth' column in a Phenotype Set indicate?\",\n",
       "  'choices': 'A. The rate of growth of the organism\\n               B. The size of the organism\\n               C. Whether or not the organism grew in the specified media with the specified knockouts\\n               D. The age of the organism',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What is the first step in importing a Phenotype Set?',\n",
       "  'choices': 'A. Add the Genome that corresponds to referenced in the Phenotype Set\\n               B. Open the Import tab within the Data Browser\\n               C. Drag & drop the phenotype set file into your Staging Area\\n               D. Click the import icon (up arrow) to the right of “Phenotype Set”',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the two files needed for uploading amplicon data in KBase?',\n",
       "  'choices': 'A. FASTA file with consensus sequences and Taxonomic abundance matrix\\n               B. FASTA file with consensus sequences and Phenotype Set\\n               C. Chemical Abundance Matrix and Taxonomic abundance matrix\\n               D. Chemical Abundance Matrix and Phenotype Set',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can Amplicon Matrices be uploaded into KBase without linking metadata?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the recommended format for uploading Amplicon matrices in KBase?',\n",
       "  'choices': 'A. CSV B. TSV C. XLSX D. DOCX',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the \"chemical abundance\" in KBase?',\n",
       "  'choices': 'A. It is a term used to represent a wide array of measurements associated with chemicals.\\n               B. It is a term used to represent the total number of chemicals in a sample.\\n               C. It is a term used to represent the concentration of a specific chemical in a sample.\\n               D. It is a term used to represent the diversity of chemicals in a sample.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What types of files can be used to upload a Chemical Abundance Matrix in KBase?',\n",
       "  'choices': 'A. TSV and Excel files B. CSV and Excel files C. TSV and CSV files D. Only Excel files',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is it mandatory to link a SampleSet when uploading chemical abundance data in KBase?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is a Sample in KBase?',\n",
       "  'choices': 'A. A Sample in KBase represents physical material from an experiment with attributes or metadata\\n               B. A Sample in KBase represents a set of data from an experiment\\n               C. A Sample in KBase represents a set of tools for data analysis\\n               D. A Sample in KBase represents a set of apps for data analysis',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the two types of errors in KBase?',\n",
       "  'choices': 'A. Red and Blue B. Red and Yellow C. Blue and Yellow D. Green and Red',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can Samples be directly imported from IGSN SESAR and NCBI BioSamples?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What happens if a cell in a validated column does not comply with the controlled vocabulary?',\n",
       "  'choices': 'A. The cell will be highlighted in red B. The upload will fail C. The cell will be automatically corrected D. The cell will be deleted',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What happens if you accept an unvalidated field?',\n",
       "  'choices': 'A. It will be stored as a user-defined field B. It will be deleted C. It will be automatically corrected D. It will be highlighted in red',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you accept warnings after the import has run?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What types of compressed files are accepted by KBase data importers?',\n",
       "  'choices': 'A. .zip, .gz, .bz, .bzip2, .bz2, .tar, .tgz\\n               B. .rar, .7z, .tar, .tgz\\n               C. .zip, .rar, .7z, .tar\\n               D. .zip, .gz, .rar, .7z',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you import archives directly into KBase?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the icon for compressed files in the Staging Area?',\n",
       "  'choices': 'A. A star icon B. An outward-facing double arrow icon C. A cloud icon D. A folder icon',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the maximum number of rows or data objects per import that is currently supported with good performance in KBase?',\n",
       "  'choices': 'A. 100 B. 500 C. 1000 D. 10000',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What file formats can be used for the Import Specification file in KBase?',\n",
       "  'choices': 'A. TSV, CSV, Excel B. TXT, DOC, PDF C. JPG, PNG, GIF D. MP3, MP4, WAV',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of an Import Specification file in KBase?',\n",
       "  'choices': 'A. To upload multiple files at once B. To download multiple files at once C. To delete multiple files at once D. To rename multiple files at once',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the format of the data when downloading from KBase?',\n",
       "  'choices': 'A. The data will be compressed into a Zip file B. The data will be in JSON format C. The data will be in CSV format D. The data will be in TXT format',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you directly import the Zip file you have downloaded from the system?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'How can you transfer large files (>10GB) to your local machine from the Staging Area?',\n",
       "  'choices': 'A. Using Globus B. Using FTP C. Using HTTP D. Using SCP',\n",
       "  'answer': 'A'},\n",
       " {'question': \"What does the 'Shared With Me' section in the Data Browser include?\",\n",
       "  'choices': 'A. Data associated to Narratives shared with you B. Public datasets C. Datasets pre-loaded by the KBase team D. Your own datasets',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you add data to a Narrative in KBase?',\n",
       "  'choices': 'A. By navigating to the Data Browser tab and clicking on the import button B. By emailing the data to KBase C. By uploading the data to a third-party website D. By sending the data via postal mail',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you upload your own datasets for analysis in KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the tutorial about in this section?',\n",
       "  'choices': 'A. Filtering, managing, and viewing data in KBase\\n               B. Searching, adding, and uploading data in KBase\\n               C. Linking Metadata in KBase\\n               D. Analyzing data in KBase',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is a Sample in KBase?',\n",
       "  'choices': 'A. A Sample in KBase represents measured material from an experiment. \\n               B. A Sample in KBase represents a set of data from an experiment.\\n               C. A Sample in KBase represents a unique ID internal to KBase.\\n               D. A Sample in KBase represents a unique user in the system.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of Sample aliases in KBase?',\n",
       "  'choices': 'A. Sample aliases allow the same data uploaded by different users to be linked and connect different analyses of the same data.\\n               B. Sample aliases allow users to change their username.\\n               C. Sample aliases allow users to hide their identity.\\n               D. Sample aliases allow users to create multiple accounts.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the function of the app \"Link Workspace Objects to Samples\"?',\n",
       "  'choices': 'A. The app provides a graphical interface for creating links between samples and workspace objects.\\n               B. The app provides a graphical interface for uploading new samples.\\n               C. The app provides a graphical interface for deleting samples.\\n               D. The app provides a graphical interface for editing samples.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the currently supported ontology systems in KBase?',\n",
       "  'choices': 'A. Gene Ontology (GO) and Environmental Ontology (ENVO) B. Gene Ontology (GO) and Biological Ontology (BIO) C. Environmental Ontology (ENVO) and Biological Ontology (BIO) D. Gene Ontology (GO) and Chemical Ontology (CO)',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What happens when the uploader encounters terms it doesn’t recognize in KBase?',\n",
       "  'choices': 'A. The terms and values will be deleted B. The terms and values will be stored in a user section of the samples C. The terms and values will be replaced with default terms D. The terms and values will be flagged for review',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the base unit for mass in KBase?',\n",
       "  'choices': 'A. Kilogram B. Gram C. Milligram D. Microgram',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What types of data does KBase provide?',\n",
       "  'choices': 'A. An extensive set of prokaryotic annotated genomes\\n               B. A selected set of eukaryotic genomes\\n               C. Thousands of biochemical compounds and reactions\\n               D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Can JGI users transfer public genome reads and assemblies from JGI to their KBase account?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Where can you search public data in KBase?',\n",
       "  'choices': 'A. Public tab in the Data Browser within the Narrative Interface\\n               B. Private tab in the Data Browser within the Narrative Interface\\n               C. Public tab in the Data Browser outside the Narrative Interface\\n               D. Private tab in the Data Browser outside the Narrative Interface',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of Globus in KBase?',\n",
       "  'choices': 'A. It is used for bulk transfer of data between two endpoints\\n               B. It is used for data analysis\\n               C. It is used for data visualization\\n               D. It is used for data storage',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the name of the KBase endpoint in Globus?',\n",
       "  'choices': 'A. KBase Data Share B. KBase Bulk Share C. KBase File Share D. KBase Data Transfer',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can you use Globus to create new directories in your Staging Area for your data transfer and to delete staging files when you are done importing into KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the requirements before transferring data from JGI to KBase?',\n",
       "  'choices': 'A. Get a JGI account B. Get a Globus account C. Get a KBase account and link it to Globus D. Use either Chrome, Firefox, or Safari browser E. All of the above',\n",
       "  'answer': 'E'},\n",
       " {'question': 'How long does the JGI staging area keep file links?',\n",
       "  'choices': 'A. 14 days B. 30 days C. 60 days D. 90 days',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How long does the KBase staging area keep files?',\n",
       "  'choices': 'A. 14 days B. 30 days C. 60 days D. 90 days',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What is the capacity of the KBase staging area?',\n",
       "  'choices': 'A. 1 terabyte B. 2.8 terabytes C. 5 terabytes D. 10 terabytes',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What can you learn from the \"Using Apps\" section in KBase?',\n",
       "  'choices': 'A. How to access and run analysis Apps in KBase\\n               B. How to access Beta Apps in KBase\\n               C. How to transfer data from JGI to KBase\\n               D. Both A and B',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Can you learn how to transfer data from JGI to KBase in the \"Using Apps\" section?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the purpose of the Software Developer Kit (SDK) in KBase?',\n",
       "  'choices': 'A. It is used to integrate third-party tools into KBase. B. It is used to write new apps by KBase scientists and developers. C. It is used to increase the number of apps available in KBase. D. All of the above.',\n",
       "  'answer': 'D'},\n",
       " {'question': 'How can you access the App Catalog from inside the Narrative Interface?',\n",
       "  'choices': 'A. By clicking the small arrow in the upper right corner of the Apps Panel. B. By clicking the Catalog icon in the Menu. C. By clicking the right arrow at the top of the Apps Panel. D. By clicking the star at the lower left of any app.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What happens when you click the star at the lower left of any app?',\n",
       "  'choices': 'A. The app is added to your “favorites.” B. The app is removed from your “favorites.” C. The app is sorted by category. D. The app is sorted by My Favorites, Run Count and other options.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the output of the annotation apps in KBase?',\n",
       "  'choices': 'A. A Genome B. A Protein C. A Cell D. A Gene',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Which tool is not used for Assembly in KBase?',\n",
       "  'choices': 'A. Assemble with HipMer B. Assemble with IDBA-UD C. Assemble with MaSuRCA D. Assemble with QUAST',\n",
       "  'answer': 'D'},\n",
       " {'question': \"What does the tool 'Annotate Assembly with Prokka' do in KBase?\",\n",
       "  'choices': 'A. It identifies protein domains from widely used domain libraries B. It combines multiple open-source annotation tools in a quick and thorough annotation pipeline for prokaryotic sequences for genomes, plasmids, and metagenomes C. It uses components from the RAST (Rapid Annotations using Subsystems Technology) toolkit to annotate an assembled bacterial or archaeal genome D. It performs functional annotation of plant cDNA or protein sequences',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the purpose of the comparative genomics and phylogenetic analysis Apps in KBase?',\n",
       "  'choices': 'A. To enable researchers to understand evolutionary relationships between organisms and explore structural and functional variance across genomes.\\n               B. To enable researchers to understand the genetic makeup of organisms.\\n               C. To enable researchers to understand the metabolic pathways of organisms.\\n               D. To enable researchers to understand the gene expression of organisms.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the \"Build Microbial Species Tree\" app do?',\n",
       "  'choices': 'A. It builds a phylogenetic tree with your genome and phylum exemplars from RefSeq.\\n               B. It builds a phylogenetic tree with your genomes and RefSeq microbial genomes based on conserved orthologous groups.\\n               C. It builds an evolutionary reconstruction tree for a collection of sequence homology related genes.\\n               D. It builds a tree with FastTree2 directly on a user-created Multiple Sequence Alignment (MSA).',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the \"Compare Two Proteomes\" app do?',\n",
       "  'choices': 'A. It compares proteomes and produces a synteny map dot plot matrix and table of gene differences.\\n               B. It groups orthologous protein sequences and identifies core and non-core genes across multiple species.\\n               C. It builds a pangenome and evaluates protein family conservation.\\n               D. It compares isofunctional and homologous gene families within a pangenome.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of the \"Build Metabolic Model\" tool in KBase?',\n",
       "  'choices': 'A. To construct genome-scale metabolic models based on a bacterial genome and media condition\\n               B. To fill in missing reactions based on stoichiometry\\n               C. To predict metabolic fluxes\\n               D. To determine optimal conditions of flux',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the \"Compare Flux with Expression\" tool do?',\n",
       "  'choices': 'A. It compares reaction fluxes with gene expression values to identify metabolic pathways where expression and flux data agree or conflict and compare with chemical abundance data\\n               B. It creates specialized growth conditions\\n               C. It saves modeling data for future analysis\\n               D. It translates metabolic models from one organism to another',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the \"Merge Metabolic Models into Community Model\" tool do?',\n",
       "  'choices': 'A. It investigates community metabolism\\n               B. It builds a metagenome metabolic model from an annotated assembly or bins\\n               C. It identifies biochemical reactions to add to a draft metabolic model for production and consumption of exometabolites\\n               D. It generates novel compounds from an FBA Model and a Chemical Abundance Matrix',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of the FastQC tool in KBase?',\n",
       "  'choices': 'A. Raw read quality control and analysis B. Read trimming and adaptor removal C. Taxonomy assignment of raw metagenomic reads D. Assemble large and complex metagenomic reads',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of the Trim Reads with Trimmomatic tool in KBase?',\n",
       "  'choices': 'A. Raw read quality control and analysis B. Read trimming and adaptor removal C. Taxonomy assignment of raw metagenomic reads D. Assemble large and complex metagenomic reads',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the purpose of the Kaiju tool in KBase?',\n",
       "  'choices': 'A. Raw read quality control and analysis B. Read trimming and adaptor removal C. Taxonomy assignment of raw metagenomic reads D. Assemble large and complex metagenomic reads',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What is the purpose of the Transform Matrix App in KBase?',\n",
       "  'choices': 'A. It is used for taxonomic classification B. It is used for data cleaning, including abundance filtering and standardization C. It is used for statistical analysis D. It is used for functional prediction for amplicon matrices',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What does the Rarefy Matrix App in KBase allow?',\n",
       "  'choices': 'A. It allows manual selection of a seed value to aid reproducibility to columns or rows B. It allows taxonomic classification C. It allows statistical analysis D. It allows functional prediction for amplicon matrices',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the Perform NMDS Analysis App do?',\n",
       "  'choices': 'A. It performs Non-metric Multidimensional Scaling Analysis on amplicon matrix data to visualize it in two dimensions B. It performs taxonomic classification C. It performs data cleaning D. It performs functional prediction for amplicon matrices',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What can be done with both chemical abundance and amplicon matrices in KBase?',\n",
       "  'choices': 'A. Cleaning B. Normalizing C. Performing statistical comparisons D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What is one of the popular use cases for chemical abundance tables in KBase?',\n",
       "  'choices': 'A. Metabolomics data representation in Escher pathway maps B. Genomic data representation in Escher pathway maps C. Proteomics data representation in Escher pathway maps D. Transcriptomics data representation in Escher pathway maps',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you reconcile metabolic models to exo-metabolite data in KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': \"What is the purpose of the 'Align Reads using Bowtie2' tool in KBase?\",\n",
       "  'choices': 'A. It aligns the sequencing reads for a set of two or more samples to long reference sequences of a prokaryotic genome.\\n               B. It aligns the sequencing reads for a set of two or more samples to an eukaryotic genome.\\n               C. It aligns the sequencing reads for a set of two or more samples to long reference sequences of a genome.\\n               D. It aligns the sequencing reads of a single or a set of two (paired end) reads to long reference sequences of a prokaryotic genome.',\n",
       "  'answer': 'A'},\n",
       " {'question': \"What does the 'Identify Differential Expression using Cuffdiff' tool do?\",\n",
       "  'choices': 'A. It uses the Cufflinks transcripts for two or more samples to calculate gene and transcript levels in more than one condition and finds significant changes in the expression levels.\\n               B. It uses the transcripts for two or more samples obtained from either Cufflinks or StringTie to calculate gene and transcript levels in more than one condition and finds significant changes in the expression levels.\\n               C. It filters an expression matrix using either Log Odds Ratio (LOR) or ANalysis of VAriance (ANOVA) algorithms.\\n               D. It performs hierarchical clustering to group gene expression data into a dendrogram.',\n",
       "  'answer': 'A'},\n",
       " {'question': \"What is the function of the 'View Interactive Heatmap' tool?\",\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you access the apps that are still in development in KBase?',\n",
       "  'choices': 'A. Click the “R” in the Apps Panel B. Click the “B” in the Apps Panel C. Click the \"OK\" button D. Click the \"B\" icon',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the \"B\" icon represent in the Apps menu?',\n",
       "  'choices': 'A. Beta version of available apps B. Released version of available apps C. Both A and B D. None of the above',\n",
       "  'answer': 'A'},\n",
       " {'question': 'True or False: Choosing Beta will show both the apps that are still being tested and the released apps.',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What does the green \"G\" icon represent in KBase App?',\n",
       "  'choices': 'A. The app is in beta version B. The app requires a genome as input C. The app is related to Genomics D. The app is developed by KBase',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Which of the following is NOT an example workflow in KBase?',\n",
       "  'choices': 'A. Assembling & Annotating Microbial Genomes B. Comparative Genomics & Phylogenetic Analysis C. Constructing Metabolic Models D. DNA Sequencing',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What do the numbers in parentheses indicate in KBase App?',\n",
       "  'choices': 'A. The number of users who have used the app B. The number of alternative apps that implement each function C. The rating of the app D. The number of updates the app has had',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What are the steps for assembly and annotation in KBase?',\n",
       "  'choices': 'A. Upload reads to assemble, Add your Reads data object to the Narrative, Search the App Catalog and insert an Assembly App, Search the App Catalog and insert an Annotation App, Examine assembly job statistics and genome annotation\\n               B. Upload reads to assemble, Add your Reads data object to the Narrative, Search the App Catalog and insert an Assembly App, Examine assembly job statistics and genome annotation\\n               C. Upload reads to assemble, Add your Reads data object to the Narrative, Search the App Catalog and insert an Annotation App, Examine assembly job statistics and genome annotation\\n               D. Upload reads to assemble, Search the App Catalog and insert an Assembly App, Search the App Catalog and insert an Annotation App, Examine assembly job statistics and genome annotation',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the Microbial Genome Assembly and Annotation Tutorial guide through?',\n",
       "  'choices': 'A. Assembling a set of NGS short reads into contigs and annotate the assembled contigs, assigning biological functions derived from RAST\\n               B. Quality control and assessment of NGS reads, de novo assembly, and genome annotation\\n               C. Drafting isolate genomes from quality assessment to assembly and annotation\\n               D. Building a workflow for searching and identifying features within your genomes',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the Genome Analysis 1: Drafting Isolate Genomes tutorial demonstrate?',\n",
       "  'choices': 'A. Assembling a set of NGS short reads into contigs and annotate the assembled contigs, assigning biological functions derived from RAST\\n               B. Quality control and assessment of NGS reads, de novo assembly, and genome annotation\\n               C. Drafting isolate genomes from quality assessment to assembly and annotation\\n               D. Building a workflow for searching and identifying features within your genomes',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What is the typical threshold to determine whether our assembled genome is contaminated?',\n",
       "  'choices': 'A. >90% complete and <5% contaminated B. >80% complete and <10% contaminated C. >70% complete and <15% contaminated D. >60% complete and <20% contaminated',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is there a limit on the size of the FASTQ files I can upload to KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Does KBase support co-assembly?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the difference between RAST and Prokka annotations?',\n",
       "  'choices': 'A. They have different options in the app, their method for assigning the annotation is different. B. They have the same options in the app, their method for assigning the annotation is the same. C. They have different options in the app, their method for assigning the annotation is the same. D. They have the same options in the app, their method for assigning the annotation is different.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can I annotate plants or fungi?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Are there tools specialized for fungal data?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': \"What does KBase's comparative genomics and phylogenetic analysis tools enable researchers to do?\",\n",
       "  'choices': 'A. Understand evolutionary relationships between organisms and explore structural and functional variance across genomes\\n               B. Analyze the chemical composition of organisms\\n               C. Study the behavior of organisms in different environments\\n               D. Understand the genetic makeup of a single organism',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is one of the tutorials available for comparative genomics and phylogenetic analysis in KBase?',\n",
       "  'choices': 'A. Build a Gene Tree\\n               B. Analyzing the Behavior of Organisms\\n               C. Studying the Chemical Composition of Organisms\\n               D. Understanding the Genetic Makeup of a Single Organism',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is there a video tutorial available for comparative genomics and phylogenetic analysis in KBase?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can the GTDB database be used within Kaiju?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Does the Kaiju analysis provide a text output that can be used in R?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the tail <1% each taxon mean in the Kaiju output?',\n",
       "  'choices': 'A. It means that the analysis is affected and the data cannot be replotted.\\n               B. It means that the analysis is not affected and the data can be replotted.\\n               C. It means that the analysis is affected and the data can be replotted.\\n               D. It means that the analysis is not affected and the data cannot be replotted.',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the purpose of the tools designed to analyze microbial communities and metagenomic reads data in KBase?',\n",
       "  'choices': 'A. To understand the functional interactions between species in microbial communities\\n               B. To analyze the genetic makeup of individual organisms\\n               C. To study the effects of environmental factors on microbial communities\\n               D. To study the effects of microbial communities on human health',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What can annotated genomes extracted from metagenomic samples be used for in KBase?',\n",
       "  'choices': 'A. Metabolic modeling B. Comparative phylogenomics C. Functional profiling D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What does the \"Genome Extraction from Shotgun Metagenome Sequence Data\" tutorial demonstrate?',\n",
       "  'choices': 'A. The application of bioinformatic tools for metagenome assembly and genome binning\\n               B. The process of annotating plant genome sequences\\n               C. The process of understanding a metagenomic pipeline\\n               D. The process of extracting and analyzing metagenome-assembled genomes',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What is the JGI Metagenome Assembly App in KBase?',\n",
       "  'choices': 'A. It is an app for annotating the complete metagenome assembly\\n               B. It is a beta app with a complete workflow, optimized by the JGI, that goes from raw reads to an assembly using BFC, BBTools for read QC and metaSPAdes\\n               C. It is an app for metabolic modeling of an entire community\\n               D. It is an app for extracting genome sequences from metagenomes',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is considered a high-quality genome in KBase?',\n",
       "  'choices': 'A. A genome that is 80% complete and <10% contaminated\\n               B. A genome that is 90% complete and <5% contaminated\\n               C. A genome that is 100% complete and <1% contaminated\\n               D. A genome that is 70% complete and <15% contaminated',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can you exclude eukaryotes and viruses from a marine metagenome in KBase?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the first step in the RNA-seq pipeline in KBase?',\n",
       "  'choices': 'A. Read Alignment B. Transcriptome Assembly and Quantification C. Differential Gene Expression D. Filtering',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of the Create RNA-seq Sample Set App in KBase?',\n",
       "  'choices': 'A. To align reads to map short reads to the reference genome B. To generate full-length transcripts and quantify transcripts and genes C. To group together your reads into an RNA-seq sample set with associated experimental metadata to run RNA-seq Apps in batch mode D. To generate gene- or transcript-level differential expression based on the quantification',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What can you do with the Hierarchical, K-Means or WGCNA clustering App in KBase?',\n",
       "  'choices': 'A. Create a filtered expression matrix and associated feature set based on fold-change or adjusted p-value B. Group features into clusters based on gene expression C. Assess the functional enrichment in plant genomes for a set of features using associated GO terms D. Assimilate the expression data from RNA-seq into the metabolic models to compare reaction fluxes with gene expression',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can you assemble a transcriptome for a species that does not have a reference genome using KBase?',\n",
       "  'choices': 'A. Yes B. No',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Which tools are recommended only for prokaryotes in KBase?',\n",
       "  'choices': 'A. Bowtie2 B. HISAT2 C. StringTie D. DEseq2',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you compare RNA sequencing data with genome-scale metabolic modes in KBase?',\n",
       "  'choices': 'A. Yes B. No',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Does KBase support metatranscriptomic assembly?',\n",
       "  'choices': 'A. Yes B. No',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the uses of genome-scale metabolic models in KBase?',\n",
       "  'choices': 'A. To explore an organism’s growth in specific media conditions\\n               B. To determine which biochemical pathways are present\\n               C. To optimize production of an important metabolite\\n               D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Can metabolic models be used to evaluate an organism’s metabolic capability by simulating growth under different conditions?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What databases does KBase use for the reconstruction, prediction, and design of metabolic models?',\n",
       "  'choices': 'A. ModelSEED and PlantSEED\\n               B. GenBank and Swiss-Prot\\n               C. PDB and UniProt\\n               D. EMBL and RefSeq',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the main focus of the research in modeling the growth and behavior of microbial communities?',\n",
       "  'choices': 'A. Metabolic flux modeling techniques B. Genetic modification techniques C. Protein synthesis techniques D. Cell division techniques',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the three types of models discussed in the article?',\n",
       "  'choices': 'A. Individual species models, mixed-bag models, multi-species models B. Single cell models, multi cell models, organ models C. Animal models, plant models, fungi models D. Bacterial models, viral models, prion models',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is it true that the article concludes that no single community modeling approach is better than the others?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of the KBase Gapfill Metabolic Models App?',\n",
       "  'choices': 'A. To find a minimal set of reactions that, when added to your model, will allow it to grow\\n               B. To annotate the metabolic models\\n               C. To predict soil amendments that improve the productivity of plant bioenergy feedstocks\\n               D. To optimize microbial production of a certain biofuel',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the default media used in the Gapfilling App if the Media field is left blank?',\n",
       "  'choices': 'A. Complete media B. Minimal media C. Custom media D. No media',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is Flux Balance Analysis?',\n",
       "  'choices': 'A. A method to annotate metabolic models\\n               B. A method to optimize microbial production of a certain biofuel\\n               C. A method to understand how FBA works in general\\n               D. A method to find the minimal media conditions under which that fuel is generated',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What is the primary function of gapfilling in KBase?',\n",
       "  'choices': 'A. To ensure that the model can produce the biomass of the organism\\n               B. To add new reactions to the model\\n               C. To make the model reversible\\n               D. To predict the enzymes that would catalyze each of the reactions in the primary metabolic network',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the directionality of a reaction in the “Equation” column of the “Reactions” tab of the output table indicate?',\n",
       "  'choices': 'A. The reaction is reversible in the gapfilled model\\n               B. The reaction is irreversible in the gapfilled model\\n               C. The reaction was already in the draft model, albeit previously irreversible, and gap-filling made it reversible\\n               D. The reaction is a new reaction that wasn’t previously present in the draft model, and has been added to the model by the gapfilling algorithm',\n",
       "  'answer': 'C, D'},\n",
       " {'question': 'What do the compartment IDs mean when viewing a model?',\n",
       "  'choices': 'A. They represent the subcellular localization of compounds, enzymes, and reactions\\n               B. They represent the index number of the model\\n               C. They represent the type of the model\\n               D. They represent the number of reactions in the model',\n",
       "  'answer': 'A, B'},\n",
       " {'question': 'What is the unit of measurement for the estimated biomass from a metabolic model?',\n",
       "  'choices': 'A. The biomass components are normalized so that a result of 1 is estimated to be 1 gram of dry weight of biomass.\\n               B. The biomass components are normalized so that a result of 1 is estimated to be 1 gram of wet weight of biomass.\\n               C. The biomass components are normalized so that a result of 1 is estimated to be 1 mole of biomass.\\n               D. The biomass components are normalized so that a result of 1 is estimated to be 1 milligram of dry weight of biomass.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does a positive flux value indicate in a reaction flux?',\n",
       "  'choices': 'A. A reaction is proceeding from left to right as written in the equation\\n               B. A reaction is proceeding from right to left as written in the equation\\n               C. A reaction is not proceeding\\n               D. A reaction is proceeding in both directions',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you constrain the flux of a reaction in your model?',\n",
       "  'choices': 'A. By exporting your model, editing it, and re-importing it\\n               B. By changing the uptake of a media compound by editing the media object itself\\n               C. By changing the “bounds” on a specific reaction using the “Custom flux bounds” field under advanced options in the Run Flux Balance Analysis App\\n               D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What does the Gapfill Metabolic Model App do in KBase?',\n",
       "  'choices': 'A. It fills in the gaps of necessary reactions for completing the pathway for a set of specified reactions\\n               B. It helps to edit the media or metabolic model\\n               C. It helps to find gene candidates for gapfilled reactions\\n               D. It helps to simulate growth on phenotype data',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the Find Candidate Genes for a Reaction tool do in KBase?',\n",
       "  'choices': 'A. It helps to find gene candidates for gapfilled reactions\\n               B. It helps to edit the media or metabolic model\\n               C. It helps to simulate growth on phenotype data\\n               D. It fills in the gaps of necessary reactions for completing the pathway for a set of specified reactions',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the Simulate Growth on Phenotype Data App do in KBase?',\n",
       "  'choices': 'A. It performs multiple flux balance analyses (FBA) using each media condition in a phenotype dataset and compares the results\\n               B. It helps to find gene candidates for gapfilled reactions\\n               C. It helps to edit the media or metabolic model\\n               D. It fills in the gaps of necessary reactions for completing the pathway for a set of specified reactions',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the main goal of the gap filling algorithm in KBase?',\n",
       "  'choices': 'A. To ensure that given the media provided by the user, all compounds in the biomass reaction can be produced by the model\\n               B. To ensure that all reactions in the model are maximized\\n               C. To ensure that all reactions in the model are minimized\\n               D. To ensure that all reactions in the model are balanced',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is one of the reasons for gap-filling reactions in KBase?',\n",
       "  'choices': 'A. The genome used to build the model has missing or incorrect annotations for a given gene(s)\\n               B. The model has too many reactions\\n               C. The model has too few reactions\\n               D. The model has incorrect reactions',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is there an app in KBase that can be used to propagate a model to a new genome?',\n",
       "  'choices': 'A. True B False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of the Troubleshooting section in KBase?',\n",
       "  'choices': 'A. To provide guidance on troubleshooting and issue reporting\\n               B. To provide guidance on how to use KBase\\n               C. To provide guidance on how to develop apps for KBase\\n               D. To provide guidance on how to fund KBase',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you report issues to the Help Board in KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is there a section in KBase that explains common error messages and their meanings?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': \"What should you do if your Narrative Interface freezes or won't load properly?\",\n",
       "  'choices': 'A. Try a hard reload of your Narrative B. Log out of KBase and log in again C. Clear your browser cache D. All of the above',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Is it recommended to use any web browser for KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What does the \"Shutdown and Restart\" option do in the Narrative Interface?',\n",
       "  'choices': 'A. It closes the current session and opens a new one B. It deletes all your data C. It saves your current session for later use D. It restarts your computer',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the purpose of the KBase Help Board?',\n",
       "  'choices': 'A. To report bugs, ask questions, and suggest new features\\n               B. To provide tutorials on how to use KBase\\n               C. To share research findings and data\\n               D. To discuss the latest developments in systems biology',\n",
       "  'answer': 'A'},\n",
       " {'question': 'How can you access the KBase Help Board?',\n",
       "  'choices': 'A. By emailing the KBase team\\n               B. By visiting https://kbase-jira.atlassian.net/\\n               C. By calling the KBase customer service line\\n               D. By sending a letter to the KBase office',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can you engage in a two-way dialog with KBase staff using the Help Board?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Where can you report issues or suggest new features for KBase?',\n",
       "  'choices': 'A. KBase Help Board B. KBase User Forum C. KBase Feedback Form D. KBase Email Support',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the required fields when creating a ticket on the KBase Help Board?',\n",
       "  'choices': 'A. Issue type, short summary, long description, URL of affected Narrative, Job ID, Category, KBase Username, Environment, Attachments, Incident Start and End Times\\n               B. Issue type, short summary, long description\\n               C. Issue type, short summary, long description, URL of affected Narrative, Job ID\\n               D. All fields are required',\n",
       "  'answer': 'B'},\n",
       " {'question': 'Can you attach files when reporting an issue on the KBase Help Board?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What are the categories of job errors in KBase?',\n",
       "  'choices': 'A. Common Job Errors B. Import Job Errors C. Assembly App Errors D. Annotation App Errors E. Functional Genomics App Errors F. Modeling App Errors G. Export Job Errors',\n",
       "  'answer': 'A, B, C, D, E, F'},\n",
       " {'question': 'Are there specific error messages for Annotation apps in KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Are there specific error messages for Exporting data in KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': \"What does the error message 'listener timeout after waiting for [600000] ms' indicate?\",\n",
       "  'choices': 'A. The job ran for more than 7 days and crashed. B. The utility ElasticSearch went down. C. The job requires data owned by another user and you do not have access. D. A component in KBase needed to be rebooted and will recover in a few minutes.',\n",
       "  'answer': 'B'},\n",
       " {'question': \"What does the error message 'User XXXX may not read workspace nnnnn' indicate?\",\n",
       "  'choices': 'A. The job ran for more than 7 days and crashed. B. The utility ElasticSearch went down. C. The job requires data owned by another user and you do not have access. D. A component in KBase needed to be rebooted and will recover in a few minutes.',\n",
       "  'answer': 'C'},\n",
       " {'question': \"What does the error message '502 Server Error: Bad Gateway….' indicate?\",\n",
       "  'choices': 'A. The job ran for more than 7 days and crashed. B. The utility ElasticSearch went down. C. The job requires data owned by another user and you do not have access. D. A known temporary error.',\n",
       "  'answer': 'D'},\n",
       " {'question': 'Where can the Job Log be found?',\n",
       "  'choices': \"A. An app's Job Status tab and the Job Browser B. The Job Status tab C. The Job Browser D. The Job Log cannot be found\",\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the maximum run time for jobs in KBase?',\n",
       "  'choices': 'A. 7 days B. 14 days C. 30 days D. There is no limit',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is a job ID in KBase?',\n",
       "  'choices': 'A. A 24-digit hexadecimal number B. A 16-digit hexadecimal number C. A 32-digit hexadecimal number D. A 12-digit hexadecimal number',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does UE stand for in the context of KBase Import Job Errors?',\n",
       "  'choices': 'A. User Error B. Unidentified Error C. Unresolved Error D. Unusual Error',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does KE stand for in the context of KBase Import Job Errors?',\n",
       "  'choices': 'A. KBase Error B. Known Error C. Key Error D. Kernel Error',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What should you do if you encounter the error \"Cannot connect to URL: ftp://ftp.imicrobe.us. ...\"?',\n",
       "  'choices': 'A. Recheck the URL and permissions, then try resubmitting the job. B. Submit a ticket to KBase Help Board. C. Change the file format. D. Rename the output file.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What should you do if you encounter an error while unzipping a file in KBase?',\n",
       "  'choices': 'A. Ignore the error and continue with your work\\n               B. Verify the file can be unzipped locally\\n               C. Delete the file and upload a new one\\n               D. Contact KBase support immediately',\n",
       "  'answer': 'B'},\n",
       " {'question': 'True or False: If a file cannot be unzipped by KBase, it most likely can be unzipped by the user.',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What does the error message \"There are no contigs to save, thus there is no valid assembly\" mean?',\n",
       "  'choices': 'A. The coverage of your input is so uneven that everything is disconnected\\n               B. The reads contain too many k-mers to fit into available memory\\n               C. There are no contigs that passed the minimum contig size\\n               D. The interleaved file does not appear to be the correct format',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What does the error message \"Invalid type for object 27459/32/1\" mean?',\n",
       "  'choices': 'A. The user is allowed to enter an Assembly for the long reads (like in the MaSuRCA app) but this results in an error\\n               B. The input object and the selected parameters disagree\\n               C. There is more than one input library\\n               D. The interleaved file does not appear to be the correct format',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the error message \"[Errno 28] No space left on device\" mean?',\n",
       "  'choices': 'A. The combined size of the libraries exceeds the current KBase hardware\\n               B. The user is allowed to enter an Assembly for the long reads (like in the MaSuRCA app) but this results in an error\\n               C. The input object and the selected parameters disagree\\n               D. There is more than one input library',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the error message \"too many contigs\" mean in the Annotation Apps?',\n",
       "  'choices': 'A. The input genome does not have the needed features\\n               B. RAST has a limit of 10,000 contigs. Prokka has a limit of 30,000 contigs\\n               C. The wrong delimiter was used in the free text of the Assembly list or Genome list\\n               D. The app requires data which is owned by another user and you do not have access',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What does the error message \"name assembly_info is not defined\" mean in the Annotation Apps?',\n",
       "  'choices': 'A. The input genome does not have the needed features\\n               B. RAST has a limit of 10,000 contigs. Prokka has a limit of 30,000 contigs\\n               C. The wrong delimiter was used in the free text of the Assembly list or Genome list\\n               D. You are running the beta version of the app',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What does the error message \"Object NNN cannot be accessed: User user_name may not read workspace XXXXX\" mean in the Annotation Apps?',\n",
       "  'choices': 'A. The input genome does not have the needed features\\n               B. RAST has a limit of 10,000 contigs. Prokka has a limit of 30,000 contigs\\n               C. The wrong delimiter was used in the free text of the Assembly list or Genome list\\n               D. The app requires data which is owned by another user and you do not have access',\n",
       "  'answer': 'D'},\n",
       " {'question': 'What does the error type \"KE\" stand for in KBase?',\n",
       "  'choices': 'A. KBase Error B. Known Error C. Key Error D. Kernel Error',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the error type \"UE\" stand for in KBase?',\n",
       "  'choices': 'A. User Error B. Unknown Error C. Unhandled Error D. Unresolved Error',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the suggested solution when the error \"cannot have both input_one_sequence and input_one_ref parameter\" occurs in BLASTp prot-prot Search?',\n",
       "  'choices': 'A. The app requires either a single Query Sequence or a Query Object with single sequence. Change the inputs and try resubmitting the job.\\n               B. The app requires both a single Query Sequence and a Query Object with single sequence. Change the inputs and try resubmitting the job.\\n               C. The app requires multiple Query Sequences. Change the inputs and try resubmitting the job.\\n               D. The app requires a Query Object with multiple sequences. Change the inputs and try resubmitting the job.',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does UE stand for in the context of KBase error messages?',\n",
       "  'choices': 'A. User Error B. Unidentified Error C. Unresolved Error D. Unusual Error',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does KE stand for in the context of KBase error messages?',\n",
       "  'choices': 'A. KBase Error B. Known Error C. Key Error D. Kernel Error',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What should a user do when they encounter the error message \"Must select at least two models to compare\"?',\n",
       "  'choices': 'A. Submit a ticket to KBase Help Board B. Wait for KBase to fix the problem C. Add at least one more model and try resubmitting the job D. Delete all models and start over',\n",
       "  'answer': 'C'},\n",
       " {'question': 'What is the cornerstone of KBase as a scientific platform?',\n",
       "  'choices': 'A. Reproducibility B. Provenance C. Sharing and data privacy D. Credit where credit is due',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the principle of Provenance in the context of KBase?',\n",
       "  'choices': 'A. It explains how data comes to exist B. It guarantees data and activities are private unless shared C. It ensures users receive credit for their work D. It supports cross-references',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Is KBase software publicly available?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can third parties extend the functionality of KBase?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is required to develop Apps using the SDK in KBase?',\n",
       "  'choices': 'A. A KBase user account B. A KBase developer account C. A KBase admin account D. A KBase guest account',\n",
       "  'answer': 'B'},\n",
       " {'question': 'What is the purpose of the KBase Software Development Kit (SDK)?',\n",
       "  'choices': 'A. It offers members of the KBase community a mechanism to add open-source, open-license analysis tools to KBase\\n               B. It is used for creating user accounts in KBase\\n               C. It is used for accessing KBase’s user interfaces, APIs, workspaces, and more\\n               D. It is used for creating a KBase developer account',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the first step to get started with KBase SDK?',\n",
       "  'choices': 'A. Create a KBase developer account\\n               B. Read the SDK Documentation\\n               C. Sign up for a user account\\n               D. Check which tools have already been made available in the App Catalog',\n",
       "  'answer': 'C'},\n",
       " {'question': 'Can community developers write new tools that take advantage of the KBase Structured Data Types and large KBase Reference Data Resources?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the first step to create a KBase Developer Account?',\n",
       "  'choices': 'A. Apply for an Argonne Collaborator Account B. Join the KBase-Developers project C. Email [email\\xa0protected] D. Login to https://accounts.cels.anl.gov',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the correct email to contact for KBase Developer Account creation?',\n",
       "  'choices': 'A. [email\\xa0protected] B. [email\\xa0protected] C. [email\\xa0protected] D. [email\\xa0protected]',\n",
       "  'answer': 'B'},\n",
       " {'question': 'True or False: You need to provide full documentation if you are not a US citizen when filling out the Argonne Collaborator Registration form.',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does the KBase architecture overview outline?',\n",
       "  'choices': \"A. The components and relationships between KBase’s user interfaces, services and databases\\n               B. The structure of KBase's GitHub Repository\\n               C. The process of creating a KBase Developer Account\\n               D. The details of kbase.us\",\n",
       "  'answer': 'A'},\n",
       " {'question': 'What is the maximum width for images viewed on a larger screen like a computer in WordPress?',\n",
       "  'choices': 'A. 500px B. 600px C. 700px D. 800px',\n",
       "  'answer': 'D'},\n",
       " {'question': 'How to create boxes around text in WordPress?',\n",
       "  'choices': 'A. <p style=\"border: Npx; border-style:solid; border-color:',\n",
       "  'answer': ' padding: 1em;\"> text here </p>\\n               B. <div style=\"border: Npx; border-style:solid; border-color:'},\n",
       " {'question': 'What is the color code for Teal boxes for developer highlights in WordPress?',\n",
       "  'choices': 'A.',\n",
       "  'answer': '.'},\n",
       " {'question': \"What is the purpose of Cloudflare's email protection?\",\n",
       "  'choices': 'A. To hide email addresses from malicious bots B. To filter spam emails C. To encrypt email content D. To block unwanted emails',\n",
       "  'answer': 'A'},\n",
       " {'question': 'Can you sign up for Cloudflare?',\n",
       "  'choices': 'A. True B. False',\n",
       "  'answer': 'A'},\n",
       " {'question': 'What does Cloudflare provide?',\n",
       "  'choices': 'A. Performance & security B. Email service C. Web hosting D. Domain registration',\n",
       "  'answer': 'A'}]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = []\n",
    "for doc in documents_html:\n",
    "    answers = generateQA(instruction_prompt, doc.page_content)\n",
    "    try:\n",
    "        data.extend(extractQA(answers))\n",
    "    except IndexError:\n",
    "        print(answers)\n",
    "        continue\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e8f40f59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results have been saved to kbase_quiz.json.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "output_filepath = 'kbase_quiz.json'\n",
    "with open(output_filepath, 'w') as output_file:\n",
    "    json.dump(data, output_file, indent=4)\n",
    "    print(f\"Results have been saved to {output_filepath}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3347c372",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
